---
title: "RNA_seq_DESeq2"
author: "echille"
edited by: "daniellembecker"
date: "4/12/2021"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Differential Expression Analysis 
## Follow RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow before statistical analysis to make gene counts and transcript counts matrices

### Set up workspace

Load libraries
```{r, message=FALSE, warning=FALSE}
library("genefilter")
library("car")
library("DESeq2")
library("factoextra")
library("MuMIn")
library("NbClust")
library("ComplexHeatmap")
library("tidyverse")
library("tidyr")
library("lme4")
library("RColorBrewer")
library("magrittr")
library("ggplot2")
library("goseq")
library("pheatmap")
library("stats")
library("gridExtra")
library("VennDiagram")
library("here")
library("plotrix")
```

Import the data files 
```{r}
#treatment information
treatmentinfo <- read.csv("../../Data/RNA-seq/metadata.RNAseq.csv", header = TRUE, sep = ",")
rownames(treatmentinfo) <- treatmentinfo$sample_id
str(treatmentinfo)
head(treatmentinfo)

#gene count matrix
gcount <- as.data.frame(read.csv("../../Data/RNA-seq/Poc_gene_count_matrix.csv", row.names="gene_id"))
dim(gcount)
head(gcount)

#remove extra characters from multiple column header names for sample ids, skip first column labeled gene counts so specify 1:ncol
for ( col in 1:ncol(gcount)){
  colnames(gcount)[col] <-  sub("_R1_001.fastq.gz.sam.sorted.bam.merge.gtf", "", colnames(gcount)[col])
}

#subset the gcount matrix and reorder it so it matches the treatment info
gcount <- gcount[ ,treatmentinfo$sample_id]
head(gcount)

# Make sure treatment info ='s the rows of count data
all(rownames(treatmentinfo) %in% colnames(gcount)) # must come out TRUE

```


##### Pre-filter gene counts
## Pre-filtering our dataset to reduce the memory size dataframe, increase the speed of the transformation and testing functions, and improve quality of statistical analysis by removing low-coverage counts. Removed counts could represent outliers in the data and removing these improves sensitivity of statistical tests. We will filter out low coverage samples. Here, we will keep P=87.5% percent of the samples have counts over A=5, allowing only 1 of 8 samples to have a value less than 5 per gene.

```{r}
#Erin Set filter values for PoverA, P=87.5% percent of the samples have counts over A=5. They chose their value allowing only 1 of 8 samples to have a value less than 5 per gene. 
filt <- filterfun(pOverA(0.875,5))

#create filter for the counts data
gfilt <- genefilter(gcount, filt)

#identify genes to keep by count filter
gkeep <- gcount[gfilt,]

#identify gene lists
gn.keep <- rownames(gkeep)

#gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt <- as.data.frame(gcount[which(rownames(gcount) %in% gn.keep),])
head(gcount_filt)
dim(gcount_filt)


```

#testing PoverA cut off values and influence on differential expression number and change in enrichment

```{r}
#make data frame to test the change in the PoverA results to see how much our differential expression number and enrichment changes for the PoverA
#set PoverA at 0.50, 0.60, 0.60, 0.70, 0.80, 0.90, and 1.0

PoverA_cut_off <- c(0.50, 0.60, 0.70, 0.80, 0.85, 0.86, 0.87, 0.88, 0.89, 0.90, 1.0)
diff.expres.num <- c(20306, 19685, 19163, 18601, 18108, 18108, 18108, 17804, 17804, 17804, 16088)

povera_test <- data_frame(PoverA_cut_off, diff.expres.num)

povera_ggplot <- ggplot(povera_test, aes(x = PoverA_cut_off, y = diff.expres.num)) +
  geom_point() 

ggsave(path = "../../Output/Final_Figures/", filename = "povera_ggplot.pdf")

```


##Merge the treatment columns into a new column , group. Set group as a factor.
```{r}
treatmentinfo$treatment <- factor(treatmentinfo$treatment, levels = c("control","enriched"))
```

#Create a DESeqDataSet design from gene count matrix and labels. Here we set the design to look at treatment to test for any differences in gene expression across timepoints attributed to treatment.
#not using DESeq due to random factors, but can still use this to visualize clusters
```{r}
#Set DESeq2 design
gdds <- DESeqDataSetFromMatrix(countData = gcount_filt,
                              colData = treatmentinfo,
                              design = ~treatment)
```

#### Visualize gene count data

We're looking to see if the samples of the same treatments cluster

##### Log-transform the count data
First we are going to log-transform the data using a variance stabilizing transforamtion (vst). This is only for visualization purposes. Essentially, this is roughly similar to putting the data on the log2 scale. It will deal with the sampling variability of low counts by calculating within-group variability (if blind=FALSE). Importantly, it does not use the design to remove variation in the data, and so can be used to examine if there may be any variability do to technical factors such as extraction batch effects.

To do this we first need to calculate the size factors of our samples. This is a rough estimate of how many reads each sample contains compared to the others. In order to use VST (the faster log2 transforming process) to log-transform our data, the size factors need to be less than 4. Otherwise, there could be artefacts in our results.
```{r}
SF.gdds <- estimateSizeFactors(gdds) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than four to use vst
print(sizeFactors(SF.gdds)) #View size factors
```

Our size factors are all less than 4, so we can use VST!
```{r}
gvst <- vst(gdds, blind=FALSE) #apply a variance stabilizing transformation to minimize effects of small counts and normalize library size
```

#### Principal component plot of samples
```{r}
gPCAdata <- plotPCA(gvst, intgroup = c("treatment"), returnData=TRUE)
percentVar <- round(100*attr(gPCAdata, "percentVar")) #plot PCA of samples with all data
ggplot(gPCAdata, aes(PC1, PC2, color=treatment)) + 
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  scale_color_manual(values = c(control="slateblue", enriched="indianred3")) +
  coord_fixed() +
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
                     #panel.grid.major = element_blank(), #Set major gridlines
                     #panel.grid.minor = element_blank(), #Set minor gridlines
                     axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) + #Set the plot background
  theme(legend.position = ("none")) #set title attributes
```

##### Run DE analysis for overall example 

Run differential expression test using a Wald model. 
```{r, message = FALSE}
DEG_enriched <- DESeq(gdds) #run differential expression test by group using the Wald model
```

Explore significant p-values for treatment
```{r, message = FALSE}
DEG.results <- results(DEG_enriched, contrast=c("treatment","control","enriched")) #see code underlying function is to apply to glmer DEG results, FDR correction
head(DEG.results)
sum(DEG.results$padj < 0.05, na.rm=TRUE) #How many adjusted p-values were less than 0.05?

#create data frame with only DESeq2 DEG's with < 0.05 significance 

DEGSeq2.sig.results <- subset(DEG.results, padj < 0.05)
DEGSeq2.sig.results <- as.data.frame(DEGSeq2.sig.results)
DEGSeq2.sig.results$gene_id <-rownames(DEGSeq2.sig.results)
write.csv(DEGSeq2.sig.results, '../../Output/RNA-seq/DEG/DEGSeq2.sig.results.csv')  

```

#### Differential Gene Expression Analysis
# For data with random effects

```{r}
#reshape count data and join data frames for treatment and gene counts
head(gcount_filt)

#preserve rownames in data frame while converting data frame from wide to long format
gcount_filt_long <- gcount_filt %>% 
    as_tibble(rownames = "gene_id") %>% 
    pivot_longer(cols = -gene_id, names_to = "sample_id", values_to = "count")

head(gcount_filt_long)

#join treatment information with gene count data so gene_ids can be assigned to samples and treatments
combined.data <- left_join(gcount_filt_long, treatmentinfo)

head(combined.data)
```

#### Fit glmer position model

```{r}
# create data frame to store results
results <- data.frame()
gs <- unique(combined.data$gene_id)

#make block a factor 
combined.data$block <- as.factor(combined.data$block)

```

#### Fit glmer position model

```{r}

#Notes on 0615:
#Found a way to resolve the error, for the for loop, length needed to be amended to just look through the gs data frame, the length for combined.data:gene_id was going through each observation and not just looking at the unique genes

#Notes on 0608: 
#last thing I did was run random effect (block) as fixed factor but when I do summary, block 8 has NA's, maybe an issue with formatting/subsetting section? TBD
#the "Error: Invalid grouping factor specification, 1 = block" is only occurring for certain gene ID and sample pairs.
#when I selected i after the error was reported in the for loop, it said line 18109 which was gene ID g16996 was causing this error. I thought maybe E6's block had a typo or something associated with it did, but tried other gene ID's with the E6 line (which was taken from metadata sheet so all would be the same) and it ran through the model fine. I then thought maybe it was this gene ID, but other combinations I tried with this gene ID ran through the model fine. I then thought maybe it was gene ID's with the labeling gXXX but tried random gene ID's around this labeling and numbering and they still worked. So I do not think it is a typo but something about the specific combinations at this point.

#Notes on 0605:
#confident in model selection from below when checking with i = 1 and i = 30
#in for loop, receiving "Error: Invalid grouping factor specification, 1 = 1block" which others have run into and said it is due to NA values which are not present throughout our data frames
#when run individual i = 1 or i = 30 for glmer, will run fine without an error
#I also get the warning for boundary (singular) fit: see ?isSingular as the for loop runs which is saying that the random effects are not accounting for much more if any more variation in the model,
# ?isSingular explanation: complex mixed-effect models (i.e., those with a large number of variance-covariance parameters) frequently result in singular fits, i.e. estimated variance-covariance matrices with less than full rank. Less technically, this means that some "dimensions" of the variance-covariance matrix have been estimated as exactly zero. For scalar random effects such as intercept-only models, or 2-dimensional random effects such as intercept+slope models, singularity is relatively easy to detect because it leads to random-effect variance estimates of (nearly) zero, or estimates of correlations that are (almost) exactly -1 or 1.
#when run with DESeq2, 223 significant DEGs, when run with glmer 1,953

#Notes on 0526:
#think about this as just a logistic regression, just adding a random effect for block. P-values based off of maximum likelihood approach, not gonna be like an anova table.
# for loop is sub-setting the combined data sheet to go through gene by gene and compare treatment by the count data 
# we found that we needed to use a summary table for glmer models in R and pull the coefficients to observed the p values for this type of model. We picked coefficient number 8 because that was the column and row that had our p value data between treatments
# to be confident with our for loop, we first used ggplot to visualize one gene with treatment and count data and saw that the data aligned how we expected for a gene that was insignificant

combined.data <- na.omit(combined.data)

for(i in 1:length((gs))){
  
  #subset the dataframe gene by gene
  combined.data1 <- subset(combined.data, gene_id == gs[i])
  
  # SHOULD THIS BE THE NEGATIVE BINOMIAL DISTRIBUTION USING GLMER.NB?
  fit <- glmer(count ~ treatment + (1|block), family="poisson", data=combined.data1)
  
  a <- summary(fit)
  
  # capture summary stats to data frame
  df <- data.frame(gene = combined.data1[1,1],  #prints gene id into output table
                   pval.treatment = a$coefficients[8])   #prints pval from coefficients table
  
  # bind rows of temporary data frame to the results data frame
  results <- rbind(results, df) 
  
}

```


# Explore significant p values between treatments
```{r}
#How many significant p-values were less than 0.05 without p adjust?

sum(results$pval.treatment < 0.05, na.rm=TRUE) 
#1953

#create data frame with only DEG's with < 0.05 significance 

DEG.results.all <- results

#use p.adjust to adjust p values from glmer using multiple methods: https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/p.adjust; https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6099145/

DEG.results.all$padj <- p.adjust(DEG.results.all$pval.treatment, method = "bonferroni", n = length(DEG.results.all$pval.treatment))

#How many significant p-values were less than 0.05 after p adjust?

DEG.sig.results.sum <- sum(DEG.results.all$padj < 0.05, na.rm=TRUE) 

#create data frame with only DEG's with < 0.05 significance 

DEG.sig.results <- subset(DEG.results.all, padj < 0.05)

write.csv(DEG.sig.results, '../../Output/RNA-seq/DEG/DEG.sig.results.csv')  

read.csv('../../Output/RNA-seq/DEG/DEG.sig.results.csv')  

view(DEG.sig.results)

```


## comparison of degs (padj = 0.05) between deseq2 and glmer 
```{r}

mixed.model.results <- DEG.sig.results
trt.results <- as.data.frame(DEGSeq2.sig.results)
trt.results$gene_id <- rownames(trt.results)
  
compare.df <- full_join(trt.results, mixed.model.results, by="gene_id")

compare.list <- compare.df$gene_id 
compare.counts <- gcount_filt[rownames(gcount_filt) %in% compare.list,]

#preserve rownames in data frame while converting data frame from wide to long format
compare.counts.long <- compare.counts %>% 
    as_tibble(rownames = "gene_id") %>% 
    pivot_longer(cols = -gene_id, names_to = "sample_id", values_to = "count")


#join treatment information with gene count data so gene_ids can be assigned to samples and treatments
compare.counts.long <- left_join(compare.counts.long, treatmentinfo)

high <- compare.counts.long %>%	
  filter(gene_id== unique(compare.counts.long$gene_id)[1])
means <- aggregate(count ~ treatment, data=high, FUN=mean)
ses <- aggregate(count ~ treatment, data=high, FUN=std.error)
means$ses <- ses$count
ggplot(means, aes(x=treatment, y=count)) +
  geom_bar(stat="identity", color="black",
             position=position_dodge()) +
    geom_errorbar(aes(ymin=count-ses, ymax=count+ses), width=.2,
                  position=position_dodge(.9)) +
    xlab("treatment") + #plot x axis label
    ylab("Expression") + #plot y axis label
    theme_bw()



```


## Visualize differentially expressed genes from DESeq2
```{r}

# Subset list of genes by those which padj>0.
DEGSeq2.output <- as.data.frame(DEGSeq2.sig.results)
dim(DEGSeq2.output)
DEG.results.all <- DEGSeq2.sig.results
DEGs <- rownames(DEG.results.all) # list all gene names 
DEGs <- unique(DEGs) # select only unique gene names 
DEGs_list <- gdds[which(rownames(gdds) %in% DEGs)] # filter gdds DESeq2 object by unique gene names
dim(DEGs_list) # 15 x 15
print(counts(DEGs_list))

# As determined above, size factors all less than 4, so proceed with VST
#apply a variance stabilizing transformation to minimize effects of small counts and normalize wrt library size

DEGvst <- vst(DEGs_list, blind = FALSE, nsub = nrow(counts(DEGs_list)))

# -- note: fitType='parametric', but the dispersion trend was not well captured by the
# function: y = a/x + b, and a local regression fit was automatically substituted.
# specify fitType='local' or 'mean' to avoid this message next time.
dim(DEGvst) # 15 x 15
print(assay(DEGvst)) # look at vst-transformed gene count data 

# Plot heat map with diff expressed genes
# Testing if the first two command in the heatmap is necessary given that we are clustering the columns anyways.
#topVarGenes <- head(order(rowVars(assay(DEGvst)),decreasing=TRUE), DEG.sig.results.sum) #sort by decreasing sig ?
mat <- assay(DEGvst)[DEG.results.all, ] #make an expression object
mat <- mat - rowMeans(mat) #diff in expression compared to average across all samples
dim(mat)
ann_colors <- list(treatment= c(control="black", enriched="cadetblue3"))
df_DEG <- as.data.frame(colData(DEGvst)[c("treatment")]) #make dataframe for column naming and associated treatment
DEG_heatmap <- pheatmap(mat, scale= "row", legend=TRUE, annotation_legend=TRUE, annotation_col=df_DEG, annotation_colors = ann_colors,
                            clustering_distance_rows="euclidean", clustering_method = "average",
                            show_rownames =FALSE,
                            show_colnames =TRUE,
                            cluster_cols = TRUE)
pdf("../../Output/Final_Figures/DEG_heatmap.pdf")
DEG_heatmap
draw(DEG_heatmap)
dev.off()

# PCA plot of diff-expressed genes 
DEGPCAdata <- plotPCA(DEGvst, intgroup = c("treatment"), returnData=TRUE)
percentVar_pca <- round(100*attr(DEGPCAdata, "percentVar")) #plot PCA of samples with all data
DEG_PCA_plot <- ggplot(DEGPCAdata, aes(PC1, PC2, color=treatment)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar_pca[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_pca[2],"% variance")) +
  scale_color_manual(values = c(control="black", enriched="cadetblue3")) +
  coord_fixed() +
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
        #panel.grid.major = element_blank(), #Set major gridlines
        #panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background

DEG_PCA_plot

ggsave(path = "../../Output/Final_Figures/", filename = "DEG_PCA_plot.pdf")

# PCA plot is of deferentially expressed genes only

# Save results

write.csv(counts(DEGs_list), file="../../Output/RNA-seq/DEG/DEG_list_unique.csv")

```


