---
title: "Fix GFF format"
author: "Ariana S Huffmyer"
date: "2022"
output: html_document
editor_options: 
  chunk_output_type: console
---
This script adds transcript and gene id into GFF file for alignment.  

Here, I'll be adding transcript_id= and gene_id= to 'gene' column because we needs that label to map our RNAseq data  

Load libraries and data. 
```{r}
#Load libraries
library(tidyverse)
library(R.utils)
```

Load  gene gff file

```{r}
gff <- read.csv(file="~/Desktop/Pver_genome_assembly_v1.0.gff3", header=FALSE, sep="\t", skip=1) 
#gff <- na.omit(gff)

#gff.DB <- read.csv(file="data/Tagseq_zymo_genohub/Pver_genome_assembly_v1.0_modified.gff3", header=FALSE, sep="\t", skip=1) 
#gff.DB <- na.omit(gff.DB)
```

Rename columns 
```{r}
colnames(gff) <- c("scaffold", "Gene.Predict", "id", "gene.start","gene.stop", "pos1", "pos2","pos3", "gene")
head(gff)

unique(gff$scaffold)
#shows that #PROT and # PASA_UPDATE # ORIGINAL: lines have been included but only have information in column 1 scaffold
```

Remove extra #PROT and # PASA_UPDATE and # ORIGINAL:
check the number of genes against the number in the genome
```{r}
#genes in the Pverrucosa genome 27,439

# remove any rows with NA
gff <- na.omit(gff)

# check the number of genes against the number in the genome
# Pverrucosa should return 27439
nrow(gff %>%filter(id=="gene"))

#genes <- gff %>%filter(id=="mRNA")
#genes$gene <- gsub("Pver_gene_split_gene_g", "Pver_split_gene_g", genes$gene) #remove extra "_gene_" from name

```

Remove cds. and duplication of _gene in the names
```{r}
#gff$ID <- sub(";.*", "", gff$gene) #move everything before the first ";" to a new column
#gff$ID <- gsub("cds.", "", gff$gene) #remove "cds." from the new column
gff$gene <- gsub("Pver_gene_g", "Pver_g", gff$gene) #remove "_gene" from the new column
gff$gene <- gsub("Pver_gene_split_gene_g", "Pver_split_gene_g", gff$gene) #remove extra "_gene_" from name
gff$gene <- gsub("Pver_gene_novel_model_", "Pver_novel_model_", gff$gene) #remove extra "_gene_" from name

head(gff)
```

Create a Parent column. 
```{r}
gff$parent_id <- sub(".*Parent=", "", gff$gene)
gff$parent_id <- sub(";.*", "", gff$parent_id)
gff$parent_id <- gsub("ID=", "", gff$parent_id) #remove ID= 
head(gff)
```

Create a transcript ID column
```{r}
gff$transcript_id <- sub(".*Name=", "", gff$gene)
gff$transcript_id <- gsub("ID=", "", gff$transcript_id) # remove ID
gff$transcript_id <- gsub("Parent=", "", gff$transcript_id) # remove Parent
gff$transcript_id <- sub(".*;", "", gff$transcript_id) #keep everything after semi colon
head(gff)
```

Now add these values into the gene column separated by semicolons.  

```{r}
gff <- gff %>% 
  mutate(test = ifelse(id != "gene", paste0("ID=", gff$transcript_id, ";parent_id=", gff$parent_id),  paste0(gene)))
head(gff)
```

Check which things are not matching between the gff and annotation file. 
```{r}
#Read in annotation file to check gene names are the same in the gff and the "Query" column of the annotation
Annot <- readxl::read_xlsx("~/Desktop/FileS2_Pver_gene_annot_May28.xlsx", skip=4) 

gff_list<-gff$transcript_id
head(gff_list)

annot_list<-Annot$Query
head(annot_list)

missing1 <- setdiff(gff_list, annot_list) #things in gff that are not in annotation
missing1
length(missing1)
#the gff has 2543 extra things 

missing2 <- setdiff(annot_list, gff_list) #things in annotation that are not in gff
missing2
length(missing2)

```

Now remove the transcript and parent ID separate columns.  

```{r}
gff<-gff %>%
  dplyr::select(!transcript_id)%>%
  dplyr::select(!parent_id)%>%
  dplyr::select(!gene)

head(gff)

gff<-gff%>%
  dplyr::rename(gene=test)

head(gff)
```

Save file. Gzip and then upload this to the server for use in bioinformatic steps.  
```{r}
write.table(gff, file="RAnalysis/Output/RNA-seq/Pver_genome_assembly_v1.0_fixed.gff3", sep="\t", col.names = FALSE, row.names=FALSE, quote=FALSE)

#gzip the file 
#gzip("RAnalysis/Output/RNA-seq/Pver_genome_assembly_v1.0_fixed.gff3")
```
