---
title: "GO_Term_Enrichment_Analysis"
author: "daniellembecker"
edited by: "daniellembecker"
date: "9/20/2021"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Gene Ontology (GO) Enrichment Analysis 
## Previous steps include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DEG statistical data sheet

# This code is using the subsetted pver data. This code is also using the GO terms generated by Diamond (B2G), Swiss Prot (B2G), trembl (B2G), and InterProScan [here](https://github.com/hputnam/Becker_E5/tree/master/Functional_Annotation/Final_Annotations) in Dec. 2021. The workflow to make the functional annotation was [here](https://github.com/hputnam/Becker_E5/blob/master/Functional_Annotation/2021-10-15-P-verrucosa-functional-annotation-pipeline.md) and how is was compiled is [here](https://github.com/hputnam/Becker_E5/blob/master/Functional_Annotation/Scripts/pver_annot_compile.Rmd) 



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
library("DESeq2")
library("tidyverse")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
library("genefilter")
library("ggplot2")
library("gplots")
library("limma")
library("spdep") 
library("adegenet") 
library("goseq")
library("gridExtra")
library("clusterProfiler")
library("DataCombine")
library("VennDiagram")
```


# Import the data files 

# Structural annotation GFF3 file
```{r, echo=FALSE}
#structural annotation file needed to calculate the length by stop position - start position
#bring in the gene specific gff3 and pull out gene id 
GFF3 <- read.csv("RAnalysis/Genome/Pver_genome_assembly_v1.0.gene.gff3", header = FALSE, sep="", skip = 1)
length(unique(GFF3$V9)) # 27439
colnames(GFF3) <- c("prot", "Predict", "id", "start","stop", "pos1", "pos2","pos3", "gene_id") #name columns
GFF3$length <- GFF3$stop - GFF3$start #calculate gene length

```



# Identifying treatment and all expressed pver genes (poverA = 0.90,10)
```{r}
#treatment information
treatmentinfo <- read.csv("RAnalysis/Data/RNA-seq/metadata.RNAseq.csv", header = TRUE, sep = ",")
str(treatmentinfo)
head(treatmentinfo)

#DEG significant results
DEG.res <- read.csv("RAnalysis/Output/RNA-seq/DEG/DEGSeq2.sig.results.csv")[,-1]
nrow(DEG.res)
DEG.res$gene_id <-  gsub("_gene","",DEG.res$gene_id) #remove extra characters

```

# Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for each set of DEGs.
#Build GOSEQ vector 
#GOseq requires a vector of all genes, all differentially expressed genes, and gene lengths
```{r}
#Make ID and length vectors
DEG.vector <- DEG.res$gene_id #the enrichment test list
ALL.vector <- GFF3$gene_id #the full gene list
ID.vector <- GFF3$gene_id #id obtained from gff3 file
length.vector <- GFF3$length #length obtained from gff3 file

gene.vector=as.integer(ALL.vector%in%DEG.vector) #Construct new vector with 1 for DEG and 0 for others
names(gene.vector)=ALL.vector #set names
```

#Calculate Probability Weighting Function
```{r}
pwf<-nullp(gene.vector, ID.vector, bias.data=length.vector) #weight vector by length of gene
```


# GOSeq w/ my annotations from Dec. 2021 (GO terms generated by Diamond (B2G), Swiss Prot (B2G), trembl (B2G), and IPS)

## Load GO terms 
```{r}
annot_GO <- read.csv("Functional_Annotation/Final_Annotations/pver_GOterms_interprot_swissprot_blast_trembl_20211203.csv", header = TRUE)[,2:3]

# GO terms already split into single GO term per row/sequence 
colnames(annot_GO) <- c("gene_id", "GO.ID")
annot_GO[annot_GO == "NA"] <- NA
annot_GO$GO.ID<- as.character(annot_GO$GO.ID)
annot_GO$GO.ID <- replace_na(annot_GO$GO.ID, "unknown")
annot_GO$GO.ID <- as.factor(annot_GO$GO.ID)
annot_GO$gene_id <- as.factor(annot_GO$gene_id)
annot_GO <- unique(annot_GO)
dim(annot_GO) # 282963 x 2
length(unique(annot_GO$GO.ID)) # 17023 unique GO.IDs
head(annot_GO)
```

# Perform GOseq

## Find enriched GO terms, "selection-unbiased testing for category enrichment amongst differentially expressed (DE) genes for RNA-seq data"
```{r}
### Perform GOseq
# Find enriched GO terms, "selection-unbiased testing for category enrichment amongst differentially expressed (DE) genes for RNA-seq data"
GO.wall<-goseq(pwf, ID.vector, gene2cat=annot_GO, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)

# Using manually entered categories.
# Calculating the p-values...
# 'select()' returned 1:1 mapping between keys and columns

write.csv(GO.wall, file = "Output/GOSeq/acerv/pver_GO_ALL_20211203.csv")

# Check GO.wall df
class(GO.wall)
head(GO.wall)
tail(GO.wall)
nrow(GO.wall)

# Find significantly enriched GO terms 
enriched.GO.05.a<-GO.wall$category[GO.wall$over_represented_pvalue<.05]
enriched.GO.05<-data.frame(enriched.GO.05.a)
colnames(enriched.GO.05) <- c("category")
enriched.GO.05 <- merge(enriched.GO.05, GO.wall, by="category")
enriched.GO.05 <- enriched.GO.05[order(-enriched.GO.05$numDEInCat),]
enriched.GO.05$term <- as.factor(enriched.GO.05$term)
head(enriched.GO.05)

# Subset enriched GO terms by ontology (BP, CC, MF) and save as csv. 
MF <- subset(enriched.GO.05, ontology=="MF")
MF <- MF[order(-MF$numDEInCat),]
CC <- subset(enriched.GO.05, ontology=="CC") 
CC <- CC[order(-CC$numDEInCat),] 
BP <- subset(enriched.GO.05, ontology=="BP")
BP <- BP[order(-BP$numDEInCat),]
write.csv(MF, file = "Output/GOSeq/acerv/acerv_sub_MF_Sig_Enriched_GO.05_20211104.csv")
write.csv(CC, file = "Output/GOSeq/acerv/acerv_CC_Sig_Enriched_GO.05_20211104.csv") 
write.csv(BP, file = "Output/GOSeq/acerv/acerv_sub_BP_Sig_Enriched_GO.05_20211104.csv")
write.csv(enriched.GO.05, file = "Output/GOSeq/acerv/acerv_sub_Sig_Enriched_GO.05_ALL_20211104.csv")
```

## Merge DEG gene ids and GO info
```{r}
# Merge gene ids and sig enriched GO terms 
colnames(annot_GO) <- c("gene_id", "category")
merge <- merge(enriched.GO.05, annot_GO, by.x = "category") # contains gene ids and GO info 
#merge <- unique(merge)
# Merge DEG with merge 
merge_again <- merge(merge, DEG, by = "gene_id") # contains GO info and gene counts for DEGs
```

## Add treatment info to df and save 
```{r}
# Read in DEG treatment info
acerv_treatment <- read.csv("Output/DESeq2/acerv/acerv_sub_DEGs.all_treatment_20210219.csv", header = TRUE)
acerv_treatment <- select(acerv_treatment, c("baseMean", "log2FoldChange", "lfcSE", "stat", "pvalue", "padj", "Treatment_Compare", "gene_id"))
acerv_treatment$gene_id <- gsub("TU", "model", acerv_treatment$gene_id)
# Merge treatment info with GO info 
merge_all <- merge(merge_again, acerv_treatment, by = "gene_id")
# yay now I have a df with all the info - GO info, gene ids, DESeq2 stats, gene counts
dim(merge_all) # 994 x 29
length(unique(merge_all$gene_id)) # 83 unique genes that are differentially expressed and have GO term info
length(unique(merge_all$category)) # 366 unique GO terms 
write.csv(merge_all, file = "Output/GOSeq/acerv/acerv_sub_ByTreatment_GO.terms_20211104.csv")
# Subset enriched GO terms by ontology (BP, CC, MF) and save as csv. 
MF_ByTreatment <- subset(merge_all, ontology=="MF")
MF_ByTreatment <- MF_ByTreatment[order(-MF_ByTreatment$numDEInCat),]
CC_ByTreatment <- subset(merge_all, ontology=="CC")
CC_ByTreatment <- CC_ByTreatment[order(-CC_ByTreatment$numDEInCat),] # no sig enriched under CC
BP_ByTreatment <- subset(merge_all, ontology=="BP")
BP_ByTreatment <- BP_ByTreatment[order(-BP$numDEInCat),]
write.csv(MF_ByTreatment, file = "Output/GOSeq/acerv/acerv_sub_MF_Sig_Enriched_ByTreatment_GO.05_20211104.csv")
write.csv(CC_ByTreatment, file = "Output/GOSeq/acerv/acerv_sub_CC_Sig_Enriched_ByTreatment_GO.05_20211104.csv")
write.csv(BP_ByTreatment, file = "Output/GOSeq/acerv/acerv_sub_BP_Sig_Enriched_ByTreatment_GO.05_20211104.csv")
```

## Plot terms by ontology numbers 
```{r}
MFplot <- MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background
MFplot
CCplot <- CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold',
                                  size = 12,
                                  hjust = 0)) +
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background
CCplot 
BPplot <- BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background
BPplot # lots of BP, hard to see. Could subset df and plot by BP dfs
# Save MF, CC, and BP plot in grid layout
GOplot <- grid.arrange(MFplot, BPplot, ncol=3, clip="off")
ggsave("Output/Figs/acerv/acerv_sub_GOplot_05_FullAnnot_20211104.pdf", GOplot, width = 21, height = 21, units = c("in"))
```

# Combining MF, CC, and BP into one plot and order by pvalue
```{r}
GOplot2 <- enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  ylim(0,45) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
GOplot2
ggsave("Output/Figs/acerv/acerv_sub_GOplot2_05_FullAnnot_20211104.pdf", GOplot2, width = 28, height = 28, units = c("in"))
# Combining into one plot and order by pvalue 
GOplot_pvalue <- enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, over_represented_pvalue)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=over_represented_pvalue) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=over_represented_pvalue), color="grey") +
  geom_text(aes(label = numDEInCat), hjust = -1, vjust = 0.5, size = 3) +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  ylim(0,0.05) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("p-value") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
GOplot_pvalue
ggsave("Output/Figs/acerv/acerv_sub_GOplot_pvalue_05_FullAnnot_20211104.pdf", GOplot_pvalue, width = 28, height = 28, units = c("in"))
```




