---
title: "GO_Term_Enrichment_Analysis"
author: "echille"
edited by: "daniellembecker"
date: "6/3/2021"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Gene Ontology (GO) Enrichment Analysis 
## Previous steps include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DEG statistical data sheet

### Set up workspace

Load necessary libraries
```{r}
library(goseq)
library(tidyverse)
library(GSEABase)
library(data.table)
library(ggplot2)
library(cowplot)
library(patchwork)
```

Import the data files 
```{r}
#treatment information
treatmentinfo <- read.csv("../Data/RNA-seq/metadata.RNAseq.csv", header = TRUE, sep = ",")
str(treatmentinfo)
head(treatmentinfo)

#gene count matrix
gcount <- as.data.frame(read.csv("../Data/RNA-seq/Poc_gene_count_matrix.csv", row.names="gene_id"))
#remove extra characters from multiple column header names for sample ids, skip first column labeled gene counts so specify 1:ncol
for ( col in 1:ncol(gcount)){
  colnames(gcount)[col] <-  sub("_R1_001.fastq.gz.sam.sorted.bam.merge.gtf", "", colnames(gcount)[col])
}
#subset the gcount matrix and reorder it so it matches the treatment info
gcount <- gcount[ ,treatmentinfo$sample_id]
gcount$gene_id <- rownames(gcount)
rownames(gcount) <- NULL
dim(gcount)
head(gcount)

#DEG significant results
DEG.res <- read.csv("../Data/RNA-seq/DEG.sig.results.csv")[,-1]
nrow(DEG.res)

#transcript annotations
annot <- read_tsv( "../Genome/Pver_gene_annotation_file.txt", col_names = TRUE) #biological annotation information
go.ref <- subset(annot, select= c(Gene, Query_length)) #Select only relevant information
#rename Gene column to gene_id to match gcount data frame and rename query_length to length
go.ref <- data.frame(gene_id=go.ref$Gene, length=go.ref$Query_length) #rename columns
#Filter gcount by available annotations
go.ref <- merge(gcount, go.ref, by = "gene_id")
```

Make a dataframe containing the gene_ids and cluster for each cluster.
#to seperate upreg and downreg genes
```{r}
# #Select only gene_id and cluster from DEseq2 res
# DEGclust <- subset(DEG.res, select=c(gene_id, cluster))
# DEGclust <- unique(DEGclust)
# clust1 <- filter(DEGclust, cluster=="1")
# nrow(clust1) #nrow clust1
# clust2 <- filter(DEGclust, cluster=="2")
# nrow(clust2) #nrow clust2
```

Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for each set of DEGs.
```{r}
#Make ID and length vectors
go.ref <- unique(go.ref)
dim(go.ref)
IDvector <- go.ref$gene_id
lengthVector <- go.ref$length
#Cluster 1
clust1.genes <- as.vector(DEG.res$gene_id)
clust1.genes=as.integer(go.ref$gene_id%in%clust1.genes)
names(clust1.genes)=go.ref$gene_id
length(clust1.genes)
length(names(clust1.genes))
length(unique(names(clust1.genes)))

#uses gene bias 
pwf.C1<-nullp(DEgenes=clust1.genes, id=IDvector, bias.data=lengthVector) #weight vector by length of gene

```

Prepare GO term dataframe
```{r}
#taking GO terms nd making them into a list to match with gene they are associated with, multiple genes for each GO term
GO.annot <- subset(annot, select=c(Gene, Gene_ontology_IDs))
splitted <- strsplit(as.character(GO.annot$Gene_ontology_IDs), ";") #split into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(GO.annot$Gene, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
colnames(GO.terms) <- c("gene_id", "GO.ID")
GO.terms$GO.ID<- as.character(GO.terms$GO.ID)
GO.terms$GO.ID <- replace_na(GO.terms$GO.ID, "unknown")
GO.terms$GO.ID <- as.factor(GO.terms$GO.ID)
GO.terms$gene_id <- as.factor(GO.terms$gene_id)
GO.terms$GO.ID <- gsub(" ", "", GO.terms$GO.ID)
GO.terms <- unique(GO.terms)
dim(GO.terms)
head(GO.terms, 10)
nrow(GO.terms)/length(unique(GO.terms$gene_id)) #avg GO IDs per gene, very high ratio of GO term to gene_id
```

Find enriched GO terms, "selection-unbiased testing for category enrichment amongst significantly expressed genes for RNA-seq data"
```{r, warning=FALSE, message=FALSE}
#calculating which GO terms are significantly up regulated, all differential significant results
#taking bias information, all gene ids in GO reference, handmade database of GO terms and test significance for different categories, cellular component, biological process, and molecular function using wallenius function and output genes that are significant that don't have a function and they will say unknown
#need to add KEGG terms and can use goseq for that as well
GOwall <- goseq(pwf.C1, GOref$gene_id, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
#numDEInCat = number of your DEG in that category
#numInCat = how many of all of your genes fit in that category
#use ratio of numDE/numIncat 

```

Find only enriched GO terms that are statistically significant at cutoff
```{r}
#filter out all that have p value < 0.05 
GO.05<-GOwall$category[GOwall$over_represented_pvalue<.05]
GO.05<-data.frame(GO.05)
colnames(GO.05) <- c("category")
GO.05 <- merge(GO.05, GOwall, by="category")
GO.05 <- GO.05[order(GO.05$ontology, GO.05$over_represented_pvalue, -GO.05$numDEInCat),]
GO.05$term <- as.factor(GO.05$term)
nrow(filter(GO.05, ontology=="BP")) #number sig BP terms
nrow(filter(GO.05, ontology=="MF")) #number sig MF terms
nrow(filter(GO.05, ontology=="CC")) #number sig CC terms #number sig MF terms
nrow(GO.05)
#409 that are biological process
#116 that are molecular function
#65 that are molecular function
#593 total
```

Correct any un-annotated terms/ontologies
```{r}
#look up GO term and what ontology they are, could be due to retired GO term entries 
NAs.ontology <- GO.05 %>% subset(is.na(term))
print(NAs.ontology)
```
*There were three un-annotated functions except*

Save significant terms
```{r, warning=FALSE}
write.csv(GO.05, file = "../Output/GO.05.csv", row.names = FALSE)
```

## Find GOslim terms
## This is where you put GO terms to larger umbrella terms

Read in files if previous steps not run.
```{r}
GO.05 <- read_csv("../Output/GO.05.csv")
#GO.05$dir <- "C1 (Up)"
```

Run GOslim to get broader categories
```{r}
#load in generic GO database, has all of the upperlevel categories for GO terms. Ex: regulation of cell division would be under cellular regulation
slim <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") #get GO database

## filtering all of BP (do MF and CC seperately)
BP_GO <- GO.05 %>%
  filter(ontology=="BP")
BPGO_collection <- GOCollection(BP_GO$category) #Make library of query terms
slims_bp <- data.frame(goSlim(BPGO_collection, slim, "BP")) #Find common parent terms to slim down our list
slims_bp$category <- row.names(slims_bp) #save rownames as category

## filtering all of MF
MF_GO <- GO.05 %>%
  filter(ontology=="MF")
MFGO_collection <- GOCollection(MF_GO$category) #Make library of query terms
slims_mf <- data.frame(goSlim(MFGO_collection, slim, "MF")) #Find common parent terms to slim down our list
slims_mf$category <- row.names(slims_mf) #save rownames as category

## filtering all of CC
CC_GO <- GO.05 %>%
  filter(ontology=="CC")
CCGO_collection <- GOCollection(CC_GO$category) #Make library of query terms
slims_cc <- data.frame(goSlim(CCGO_collection, slim, "CC")) #Find common parent terms to slim down our list
slims_cc$category <- row.names(slims_cc) #save rownames as category
```

Get mapped terms, using functions from Sam White's Biostars [post](https://support.bioconductor.org/p/128407/#128409).
```{r}
#custom function from Sam White's, gets mapped ids for all of your query terms 
#Write function mappedIds to get the query terms that mapped to the slim categories
mappedIds <-
  function(df, collection, OFFSPRING) #the command to run requires a dataframe of slim terms, like slims_MF above, your list of query terms, and the offspring from the GOCollection by goSlim
  {
    map <- as.list(OFFSPRING[rownames(df)]) # Subset GOcollection offspring by the rownames of your dataframe
    mapped <- lapply(map, intersect, ids(collection)) #Find the terms that intersect between the subset made above of your query terms and the GOids from the GO collection
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L)) #Add column "go_terms" with matching terms 
    df #show resulting dataframe
  }
#Run function for MF and BP terms
BPslim <- mappedIds(slims_bp, BPGO_collection, GOBPOFFSPRING)
MFslim <- mappedIds(slims_mf, MFGO_collection, GOMFOFFSPRING)
CCslim <- mappedIds(slims_cc, CCGO_collection, GOCCOFFSPRING)
```

Remove duplicate matches, keeping the broader umbrella term
```{r}
#filtering out duplicates, keep the largest branch that has the most GO terms

#BP
BPslim <- filter(BPslim, Count>0 & Term!="biological_process") #filter out empty slims and term "biological process"
BPsplitted <- strsplit(as.character(BPslim$go_terms), ";") #split into multiple GO ids
BPslimX <- data.frame(Term = rep.int(BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) #list all
BPslimX <- merge(BPslimX, BPslim[,c(1,3:4)], by="Term") #Add back counts, term, and category info
BPslimX <- unique(setDT(BPslimX)[order(go_term, -Count)], by = "go_term") #remove duplicate offspring terms, keeping only those that appear in the larger umbrella term (larger Count number)
BPslim <- data.frame(slim_term=BPslimX$Term, slim_cat=BPslimX$category, category=BPslimX$go_term) #rename columns
head(BPslim)

#MF
MFslim <- filter(MFslim, Count>0 & Term!="molecular_function") #filter out empty slims and term "molecular function"
MFsplitted <- strsplit(as.character(MFslim$go_terms), ";") #split into multiple GO ids
MFslimX <- data.frame(Term = rep.int(MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) #list all
MFslimX <- merge(MFslimX, MFslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
MFslimX <- unique(setDT(MFslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
MFslim <- data.frame(slim_term=MFslimX$Term, slim_cat=MFslimX$category, category=MFslimX$go_term) #rename columns
head(MFslim)

#CC
CCslim <- filter(CCslim, Count>0 & Term!="cellular_component") #filter out empty slims and term "molecular function"
CCsplitted <- strsplit(as.character(CCslim$go_terms), ";") #split into multiple GO ids
CCslimX <- data.frame(Term = rep.int(CCslim$Term, sapply(CCsplitted, length)), go_term = unlist(CCsplitted)) #list all
CCslimX <- merge(CCslimX, CCslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
CCslimX <- unique(setDT(CCslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
CCslim <- data.frame(slim_term=CCslimX$Term, slim_cat=CCslimX$category, category=CCslimX$go_term) #rename columns
head(CCslim)

```

Save slim info with GO enrichment info for heatmap dataframes.
```{r}
GO.BP <- right_join(BPslim, filter(GO.05, ontology=="BP"), by="category") #add back GO enrichment info for each offspring term
GO.MF <- right_join(MFslim, filter(GO.05, ontology=="MF"), by="category") #add back GO enrichment info for each offspring term
GO.CC <- right_join(CCslim, filter(GO.05, ontology=="CC"), by="category") #add back GO enrichment info for each offspring term
```

## Make heatmap
```{r}
#make figure 5 in ph plot
#currently do not have direction for DEGs, replaced x with ontology
BPplot <- GO.BP %>% filter(numInCat>5) %>% mutate(term = fct_reorder(term, -over_represented_pvalue)) %>% ggplot(aes(x = ontology, y = term)) + 
  geom_tile(aes(fill =over_represented_pvalue)) + 
  facet_grid(slim.term ~ ., scales = "free_y", labeller = label_wrap_gen(width = 5, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
  strip.text.y = element_text(angle=0, size = 10),
  strip.text.x = element_text(size = 20),
  axis.text = element_text(size = 8),
  axis.title.x = element_blank(),
  axis.title.y = element_blank())
MFplot <- GO.MF %>% filter(numInCat>5) %>% mutate(term = fct_reorder(term, -over_represented_pvalue)) %>% ggplot(aes(x = ontology, y = term)) + 
  geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
  scale_y_discrete(position = "right") +
    facet_grid(slim_term~ ., scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
  strip.text.y.left = element_text(angle=0, size = 11, face = "bold"),
  strip.text.x = element_text(size = 12, face = "bold"),
  axis.title = element_blank(),
  axis.text = element_text(size = 12), legend.title = element_text(size = 12), legend.text = element_text(size = 11))
fig5 <- BPplot + MFplot
fig5 <- ggdraw(plot = fig5) + draw_plot_label(c("a)", "b)"), c(0, 0.33), c(1, 1), size = 15)
ggsave("../Output/GOslim.pdf", fig5, width = 20, height = 2, units = c("in"))
ggsave("../Output/heatmap_GO_terms.png", fig5, width = 20, height = 20, units = c("in"))
```

## Make supplemental table summarizing GO enrichment
```{r}
DF_GOhm <- bind_rows(GO.BP, GO.MF, GO.CC)
head(DF_GOhm)
```


Make dataframe of GO results for clustering and heatmap. Using R script from Mass Lab. Needs columns "experiment" where we will put cluster information, "term", "percentDEInCat", and "gene" with all the DE genes associated with that GO term.
```{r}
#add gene_IDs. To get gene_IDs we will merge with the GO.terms DF.
GOgenes <- data.frame(gene_id=GO.terms$gene_id, category=GO.terms$GO.ID) #First have to make the "by" column the same for both
GOdf <- left_join(DF_GOhm, GOgenes, by="category" ) #join the DFs
GOdf$gene_id <- as.character(GOdf$gene_id) #make gene ID a character so we can collasp our many near-identical columns
GOdf <- GOdf %>% #collapse and have gene IDs for a particular term in a single row as a comma-sep list. 
  filter(gene_id%in%DEG.res$gene_id) %>% 
  group_by(ontology, slim_term, slim_cat, category, term, over_represented_pvalue, numDEInCat, numInCat) %>%
  summarise(genes = toString(gene_id)) %>% #rename collapsed gene_ID column "gene"
  ungroup()
head(GOdf)
str(GOdf)
write.csv(GOdf, file = "../Output/GOenrichmentsummary.csv", row.names = FALSE)
```

## Kegg enrichment analysis
## I need to use KOALA-KOfamscan on bluewaves to generate KEGG terms 

Load KO -- gene mapping info
```{r}
#Extract from InterProScan search Chille et al., 2021 project
IPS.KO <- na.omit(subset(annot, select=c(gene_id, ko)))
IPS.KO$ko <- gsub("\\;", "", IPS.KO$ko)
#Load in and prep KofamScan search results
KFS.KO.orig <- read_tsv("5-Planula-GO-Enrichment-Analysis/a-Kegg-ontology/Mcap_KO_annot.tsv", col_names = TRUE)
KFS.KO <- KFS.KO.orig[-1,c(1,2,3)]
colnames(KFS.KO) <- c("sig","gene_id", "ko")
KFS.KO <- filter(KFS.KO, sig=="*")
KFS.KO <- KFS.KO[,c(2:3)]
KFS.KO$gene_id <- gsub("augustus.", "", KFS.KO$gene_id)
KFS.KO$gene_id <- gsub(".t1", "", KFS.KO$gene_id)
head(KFS.KO)
tail(KFS.KO)
#bind all KO results
KO.terms <- bind_rows(IPS.KO, KFS.KO)
KO.terms <- unique(KO.terms)
str(KO.terms)
head(KO.terms)
colnames(KO.terms) <- c("gene_id", "GO.ID")
#Bind KO and GO references
GO.terms <- bind_rows(GO.terms, KO.terms)
```

Obtain KO terms for each cluster of DEGs to be used for the KEGG Mapper [(link)](https://www.genome.jp/kegg/mapper.html)
```{r}
koC1 <- filter(KO.terms, gene_id%in%clust1$gene_id) #Filter for gene IDs in cluster1
koC2 <- filter(KO.terms, gene_id%in%clust2$gene_id) #Filter for gene IDs in cluster2
dim(koC1)
dim(koC2)
write.table(koC1, "5-Planula-GO-Enrichment-Analysis/Output/koC1.KEGG_mapper_input.txt", quote=FALSE,col.names=FALSE,row.names=FALSE,sep="\t") #TO BE USED FOR KEGG Mapper https://www.genome.jp/kegg/mapper.html.
write.table(koC2, "5-Planula-GO-Enrichment-Analysis/Output/koC2.KEGG_mapper_input.txt", quote=FALSE,col.names=FALSE,row.names=FALSE,sep="\t") #TO BE USED FOR KEGG Mapper https://www.genome.jp/kegg/mapper.html.
```

Perform Kegg enrichment with goseq package
```{r}
#Perform goseq
KOwall.C1 <- goseq(pwf.C1, GOref$gene_id, gene2cat=GO.terms, test.cats=c("KEGG"), method="Wallenius", use_genes_without_cat=TRUE)
KOwall.C2 <- goseq(pwf.C2, GOref$gene_id, gene2cat=GO.terms, test.cats=c("KEGG"), method="Wallenius", use_genes_without_cat=TRUE)
```

Replace NAs with "KEGG" and extract KO terms from results
```{r}
#Cluster1
C1.KO.05<-KOwall.C1$category[KOwall.C1$over_represented_pvalue<.05]
C1.KO.05<-data.frame(C1.KO.05)
colnames(C1.KO.05) <- c("category")
C1.KO.05 <- merge(C1.KO.05, KOwall.C1, by="category")
C1.KO.05$ontology <- replace_na(C1.KO.05$ontology, "KEGG")
C1.KO.05 <- filter(C1.KO.05, ontology=="KEGG")
C1.KO.05 <- C1.KO.05[order(C1.KO.05$ontology, C1.KO.05$over_represented_pvalue, -C1.KO.05$numDEInCat),]
C1.KO.05$term <- as.factor(C1.KO.05$term)
nrow(C1.KO.05)
#Cluster2
C2.KO.05<-KOwall.C2$category[KOwall.C2$over_represented_pvalue<.05]
C2.KO.05<-data.frame(C2.KO.05)
colnames(C2.KO.05) <- c("category")
C2.KO.05 <- merge(C2.KO.05, KOwall.C2, by="category")
C2.KO.05$ontology <- replace_na(C2.KO.05$ontology, "KEGG")
C2.KO.05 <- filter(C2.KO.05, ontology=="KEGG")
C2.KO.05 <- C2.KO.05[order(C2.KO.05$ontology, C2.KO.05$over_represented_pvalue, -C2.KO.05$numDEInCat),]
C2.KO.05$term <- as.factor(C2.KO.05$term)
nrow(C2.KO.05)
```

Add KO definitions
```{r}
#Prep definition data
KFS.KO.def <- subset(KFS.KO.orig, select=c("#","KO", "KO definition"))
colnames(KFS.KO.def) <- c("sig","category", "term")
KFS.KO.def <- filter(KFS.KO.def, sig=="*")
KFS.KO.def <- KFS.KO.def[,c(2:3)]
#Merge with KEGG output
C1.KO.05 <- unique(left_join(C1.KO.05[,-6], KFS.KO.def, by=c("category")))
C2.KO.05 <- unique(left_join(C2.KO.05[,-6], KFS.KO.def, by=c("category")))
```

Write output KEGG enrichment files
```{r}
write.csv(C1.KO.05, file = "5-Planula-GO-Enrichment-Analysis/Output/KO.05.C1.csv")
write.csv(C2.KO.05, file = "5-Planula-GO-Enrichment-Analysis/Output/KO.05.C2.csv")
```
