---
title: "GO_Term_Enrichment_Analysis"
author: "daniellembecker"
edited by: "daniellembecker"
date: "9/20/2021"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Gene Ontology (GO) Enrichment Analysis 
## Previous steps include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DEG statistical data sheet

# This code is using the subsetted pver data. This code is also using the GO terms generated by Diamond (B2G), Swiss Prot (B2G), trembl (B2G), and InterProScan [here](https://github.com/hputnam/Becker_E5/tree/master/Functional_Annotation/Final_Annotations) in Dec. 2021. The workflow to make the functional annotation was [here](https://github.com/hputnam/Becker_E5/blob/master/Functional_Annotation/2021-10-15-P-verrucosa-functional-annotation-pipeline.md) and how is was compiled is [here](https://github.com/hputnam/Becker_E5/blob/master/Functional_Annotation/Scripts/pver_annot_compile.Rmd) 



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
library("DESeq2")
library("tidyverse")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
library("genefilter")
library("ggplot2")
library("gplots")
library("limma")
library("spdep") 
library("adegenet") 
library("goseq")
library("gridExtra")
library("clusterProfiler")
library("DataCombine")
library("VennDiagram")
```


# Import the data files 

# Structural annotation GFF3 file
```{r, echo=FALSE}
#structural annotation file needed to calculate the length by stop position - start position
#bring in the gene specific gff3 and pull out gene id 
GFF3 <- read.csv("RAnalysis/Genome/Pver_genome_assembly_v1.0.gene.gff3", header = FALSE, sep="", skip = 1)
length(unique(GFF3$V9)) # 27439
colnames(GFF3) <- c("prot", "Predict", "id", "start","stop", "pos1", "pos2","pos3", "gene_id") #name columns
GFF3$length <- GFF3$stop - GFF3$start #calculate gene length

#Rename gene_id column to match DESeq2 results, currently has too many characters and separations
GFF3$gene_id <-  gsub(";.*","",GFF3$gene_id) #remove extra characters after the first ;
#There also appears to be the presence of the characters ID= in each row
GFF3$gene_id <- gsub("ID=","",GFF3$gene_id) #remove ID=
#There is also an extra "-g#" that can be removed
GFF3$gene_id <-  gsub("-.*","",GFF3$gene_id) #remove the extra g number after the dash

#remove the extra "gene" to match the "Pver_gene_g#" format
GFF3$gene_id <-  gsub("Pver_gene_novel_gene_","Pver_novel_gene_",GFF3$gene_id)
GFF3$gene_id <-  gsub("Pver_gene_split_gene_g","Pver_split_gene_g",GFF3$gene_id)

#remove the _split_ from gene_id name
GFF3$gene_id <-  gsub("Pver_split_gene_g","Pver_gene_g",GFF3$gene_id)

#remove extra characters to match DESeq2
GFF3$gene_id <-  gsub("_gene_g","_g",GFF3$gene_id) #sub gene_g with _g
GFF3$gene_id <-  gsub("_gene_","_",GFF3$gene_id) #remove extra _
GFF3$gene_id <-  gsub("_split_","_",GFF3$gene_id) #remove split naming in gene id


```



# Identifying treatment and all expressed pver genes (poverA = 0.90,10)
```{r}
#treatment information
treatmentinfo <- read.csv("RAnalysis/Data/RNA-seq/metadata.RNAseq.csv", header = TRUE, sep = ",")
str(treatmentinfo)
head(treatmentinfo)

#DEG significant results
DEG.res <- read.csv("RAnalysis/Output/RNA-seq/DEG/DEGSeq2.sig.results.csv")[,-1]
nrow(DEG.res)
DEG.res$gene_id <-  gsub("_gene","",DEG.res$gene_id) #remove extra characters

colnames(DEG.res)[7] <-"gene_id" # make colnames a true column called gene_id
head(DEG.res)

```

# Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for each set of DEGs.
#Build GOSEQ vector 
#GOseq requires a vector of all genes, all differentially expressed genes, and gene lengths
```{r}
#Make ID and length vectors

DEG <- GFF3[GFF3$gene_id %in% DEG.res$gene_id, ]
dim(DEG) # 213 x 10
DEG_names <- as.vector(DEG$gene_id)
gene_vector <- as.integer(GFF3$gene_id%in%DEG_names)
names(gene_vector) <- GFF3$gene_id

# Make ID vector 
IDvector <- GFF3$gene_id

# Make length vector
lengthVector <- GFF3$length

```

#Calculate Probability Weighting Function
```{r}
pwf<-nullp(gene_vector, ID.vector, bias.data=lengthVector) #weight vector by length of gene
```


# GOSeq w/ my annotations from Dec. 2021 (GO terms generated by Diamond (B2G), Swiss Prot (B2G), trembl (B2G), and IPS)

## Load GO terms 
```{r}
annot_GO <- read.csv("Functional_Annotation/Final_Annotations/pver_GOterms_interprot_swissprot_blast_trembl_20211203.csv", header = TRUE)[,2:3]

# GO terms already split into single GO term per row/sequence 
colnames(annot_GO) <- c("gene_id", "GO.ID")
annot_GO[annot_GO == "NA"] <- NA
annot_GO$GO.ID<- as.character(annot_GO$GO.ID)
annot_GO$GO.ID <- replace_na(annot_GO$GO.ID, "unknown")
annot_GO$GO.ID <- as.factor(annot_GO$GO.ID)
annot_GO$gene_id <- as.factor(annot_GO$gene_id)
annot_GO <- unique(annot_GO)
annot_GO$gene_id <-  gsub("\\..*","",annot_GO$gene_id) #remove extra characters after the first .


dim(annot_GO) # 282963 x 2
length(unique(annot_GO$GO.ID)) # 17023 unique GO.IDs
head(annot_GO)
```

# Perform GOseq

## Find enriched GO terms, "selection-unbiased testing for category enrichment amongst differentially expressed (DE) genes for RNA-seq data"
```{r}
### Perform GOseq
## Perform GOseq
# Find enriched GO terms, "selection-unbiased testing for category enrichment amongst differentially expressed (DE) genes for RNA-seq data"
# should I include this: test.cats=c("GO:CC", "GO:BP", "GO:MF") ?
GO.wall<-goseq(pwf, ID_vector, gene2cat=annot_GO, method="Wallenius", use_genes_without_cat=TRUE)
# Using manually entered categories.
# Calculating the p-values...
# 'select()' returned 1:1 mapping between keys and columns
write.csv(GO.wall, file = "RAnalysis/Output/RNA-seq/GOSeq/pver_GO_ALL.csv")

# Check GO.wall df
class(GO.wall)
head(GO.wall)
tail(GO.wall)
nrow(GO.wall)

# Find significantly enriched GO terms 
enriched.GO.05.a<-GO.wall$category[GO.wall$over_represented_pvalue<.05]
enriched.GO.05<-data.frame(enriched.GO.05.a)
colnames(enriched.GO.05) <- c("category")
enriched.GO.05 <- merge(enriched.GO.05, GO.wall, by="category")
enriched.GO.05 <- enriched.GO.05[order(-enriched.GO.05$numDEInCat),]
enriched.GO.05$term <- as.factor(enriched.GO.05$term)
head(enriched.GO.05)

# Subset enriched GO terms by ontology (BP, CC, MF) and save as csv. 
MF <- subset(enriched.GO.05, ontology=="MF")
MF <- MF[order(-MF$numDEInCat),]
CC <- subset(enriched.GO.05, ontology=="CC") 
CC <- CC[order(-CC$numDEInCat),] 
BP <- subset(enriched.GO.05, ontology=="BP")
BP <- BP[order(-BP$numDEInCat),]
write.csv(MF, file = "RAnalysis/Output/RNA-seq/GOSeq/pver_MF_Sig_Enriched_GO.05.csv")
write.csv(CC, file = "RAnalysis/Output/RNA-seq/GOSeq/pver_CC_Sig_Enriched_GO.05.csv") 
write.csv(BP, file = "RAnalysis/Output/RNA-seq/GOSeq/pver_BP_Sig_Enriched_GO.05.csv")
write.csv(enriched.GO.05, file = "RAnalysis/Output/RNA-seq/GOSeq/pver_Sig_Enriched_GO.05_ALL.csv")

nrow(enriched.GO.05) # 308
nrow(filter(enriched.GO.05, ontology=="BP")) #number sig BP terms 211
nrow(filter(enriched.GO.05, ontology=="MF")) #number sig MF terms 70
nrow(filter(enriched.GO.05, ontology=="CC")) #number sig CC terms 27
```

## Merge DEG gene ids and GO info
```{r}
# Merge gene ids and sig enriched GO terms 
colnames(annot_GO) <- c("gene_id", "category")
merge <- merge(enriched.GO.05, annot_GO, by.x = "category") # contains gene ids and GO info 
#merge <- unique(merge)
# Merge DEG with merge 
merge_again <- merge(merge, DEG, by = "gene_id") # contains GO info and gene counts for DEGs
```


## Plot terms by ontology numbers 
```{r}
MFplot <- MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background
MFplot
CCplot <- CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold',
                                  size = 12,
                                  hjust = 0)) +
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background
CCplot 
BPplot <- BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background
BPplot # lots of BP, hard to see. Could subset df and plot by BP dfs
# Save MF, CC, and BP plot in grid layout
GOplot <- grid.arrange(MFplot, BPplot, ncol=3, clip="off")
ggsave("RAnalysis/Output/Final_Figures/pver_GOplot_05_FullAnnot.pdf", GOplot, width = 21, height = 21, units = c("in"))
```

# Combining MF, CC, and BP into one plot and order by pvalue
```{r}
GOplot2 <- enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  ylim(0,45) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
GOplot2
ggsave("RAnalysis/Output/Final_Figures/pver_GOplot2_05.pdf", GOplot2, width = 28, height = 28, units = c("in"))
# Combining into one plot and order by pvalue 
GOplot_pvalue <- enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, over_represented_pvalue)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=over_represented_pvalue) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=over_represented_pvalue), color="grey") +
  geom_text(aes(label = numDEInCat), hjust = -1, vjust = 0.5, size = 3) +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  ylim(0,0.05) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("p-value") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
GOplot_pvalue
ggsave("RAnalysis/Output/Final_Figures/pver_GOplot_pvalue_05_FullAnnot.pdf", GOplot_pvalue, width = 28, height = 28, units = c("in"))
```


## Find GOslim terms

# Run GOslim to get broader categories
```{r}
#load in generic GO database, has all of the upperlevel categories for GO terms. Ex: regulation of cell division would be under cellular regulation
slim <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") #get GO database

## filtering all of BP (do MF and CC seperately)
BP_GO <- enriched.GO.05 %>%
  filter(ontology=="BP")
BPGO_collection <- GOCollection(BP_GO$category) #Make library of query terms
slims_bp <- data.frame(goSlim(BPGO_collection, slim, "BP")) #Find common parent terms to slim down our list
slims_bp$category <- row.names(slims_bp) #save rownames as category

## filtering all of MF
MF_GO <- enriched.GO.05 %>%
  filter(ontology=="MF")
MFGO_collection <- GOCollection(MF_GO$category) #Make library of query terms
slims_mf <- data.frame(goSlim(MFGO_collection, slim, "MF")) #Find common parent terms to slim down our list
slims_mf$category <- row.names(slims_mf) #save rownames as category

## filtering all of CC
CC_GO <- enriched.GO.05 %>%
  filter(ontology=="CC")
CCGO_collection <- GOCollection(CC_GO$category) #Make library of query terms
slims_cc <- data.frame(goSlim(CCGO_collection, slim, "CC")) #Find common parent terms to slim down our list
slims_cc$category <- row.names(slims_cc) #save rownames as category
```

# Get mapped terms, using functions from Sam White's Biostars [post](https://support.bioconductor.org/p/128407/#128409).
```{r}
#custom function from Sam White's, gets mapped ids for all of your query terms 
#Write function mappedIds to get the query terms that mapped to the slim categories
mappedIds <-
  function(df, collection, OFFSPRING) #the command to run requires a dataframe of slim terms, like slims_MF above, your list of query terms, and the offspring from the GOCollection by goSlim
  {
    map <- as.list(OFFSPRING[rownames(df)]) # Subset GOcollection offspring by the rownames of your dataframe
    mapped <- lapply(map, intersect, ids(collection)) #Find the terms that intersect between the subset made above of your query terms and the GOids from the GO collection
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L)) #Add column "go_terms" with matching terms 
    df #show resulting dataframe
  }
#Run function for MF and BP terms
BPslim <- mappedIds(slims_bp, BPGO_collection, GOBPOFFSPRING)
MFslim <- mappedIds(slims_mf, MFGO_collection, GOMFOFFSPRING)
CCslim <- mappedIds(slims_cc, CCGO_collection, GOCCOFFSPRING)
```

# Remove duplicate matches, keeping the broader umbrella term
```{r}
#filtering out duplicates

#BP
BPslim <- filter(BPslim, Count>0 & Term!="biological_process") #filter out empty slims and term "biological process"
BPsplitted <- strsplit(as.character(BPslim$go_terms), ";") #split into multiple GO ids
BPslimX <- data.frame(Term = rep.int(BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) #list all
BPslimX <- merge(BPslimX, BPslim[,c(1,3:4)], by="Term") #Add back counts, term, and category info
BPslimX <- unique(setDT(BPslimX)[order(go_term, -Count)], by = "go_term") #remove duplicate offspring terms, keeping only those that appear in the larger umbrella term (larger Count number)
BPslim <- data.frame(slim_term=BPslimX$Term, slim_cat=BPslimX$category, category=BPslimX$go_term) #rename columns
head(BPslim)

#MF
MFslim <- filter(MFslim, Count>0 & Term!="molecular_function") #filter out empty slims and term "molecular function"
MFsplitted <- strsplit(as.character(MFslim$go_terms), ";") #split into multiple GO ids
MFslimX <- data.frame(Term = rep.int(MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) #list all
MFslimX <- merge(MFslimX, MFslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
MFslimX <- unique(setDT(MFslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
MFslim <- data.frame(slim_term=MFslimX$Term, slim_cat=MFslimX$category, category=MFslimX$go_term) #rename columns
head(MFslim)

#CC
CCslim <- filter(CCslim, Count>0 & Term!="cellular_component") #filter out empty slims and term "molecular function"
CCsplitted <- strsplit(as.character(CCslim$go_terms), ";") #split into multiple GO ids
CCslimX <- data.frame(Term = rep.int(CCslim$Term, sapply(CCsplitted, length)), go_term = unlist(CCsplitted)) #list all
CCslimX <- merge(CCslimX, CCslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
CCslimX <- unique(setDT(CCslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
CCslim <- data.frame(slim_term=CCslimX$Term, slim_cat=CCslimX$category, category=CCslimX$go_term) #rename columns
head(CCslim)

```

# Save slim info with GO enrichment info for heatmap dataframes.
```{r}
GO.BP <- right_join(BPslim, filter(enriched.GO.05, ontology=="BP"), by="category") #add back GO enrichment info for each offspring term
GO.MF <- right_join(MFslim, filter(enriched.GO.05, ontology=="MF"), by="category") #add back GO enrichment info for each offspring term
GO.CC <- right_join(CCslim, filter(enriched.GO.05, ontology=="CC"), by="category") #add back GO enrichment info for each offspring term
```


## Make heatmap
```{r}
BPplot <- GO.BP %>% mutate(term = fct_reorder(term, -over_represented_pvalue)) %>%
    ggplot(aes(x = ontology, y = term)) + 
    geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
      scale_y_discrete(position = "right") +
      facet_grid(slim_term~ ., scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
      theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
      strip.text.y.left = element_text(angle=0, size = 11, face = "bold"),
      strip.text.x = element_text(size = 12, face = "bold"),
      axis.title = element_blank(),
      axis.text = element_text(size = 12), legend.title = element_text(size = 12), legend.text =     
      element_text(size = 11))
MFplot <- GO.MF %>% mutate(term = fct_reorder(term, -over_represented_pvalue)) %>% 
    ggplot(aes(x = ontology, y = term)) + 
    geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
      scale_y_discrete(position = "right") +
      facet_grid(slim_term~ ., scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
      theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
      strip.text.y.left = element_text(angle=0, size = 11, face = "bold"),
      strip.text.x = element_text(size = 12, face = "bold"),
      axis.title = element_blank(),
      axis.text = element_text(size = 12), legend.title = element_text(size = 12), legend.text =
      element_text(size = 11))
fig5 <- BPplot + MFplot
ggsave("RAnalysis/Output/Final_Figures/GO_enrichment_DEG_heatmap_.pdf", fig5, width = 30, height = 30, units = c("in"))
```

## Make supplemental table summarizing GO enrichment
```{r}
GO.enrichment.dat <- bind_rows(GO.BP, GO.MF, GO.CC)
head(GO.enrichment.dat)

#Make dataframe of GO results for clustering and heatmap. 
#add gene_IDs. To get gene_IDs we will merge with the GO.terms DF.
GOgenes <- data.frame(gene_id=annot_GO$gene_id, category=annot_GO$category) 
GOgenes$gene_id <- as.character(GOgenes$gene_id) #make gene ID a character so we can collapse our many near-identical columns

GO.enrichment.summary  <- left_join(GO.enrichment.dat, GOgenes, by="category" ) #join the DFs

GO.enrichment.summary <- GO.enrichment.summary %>% #collapse and have gene IDs for a particular term in a single row as a comma-sep list. 
  group_by(slim_term, slim_cat, category, over_represented_pvalue, under_represented_pvalue, numDEInCat, numInCat, term, ontology) %>%
  summarise(genes = toString(gene_id)) %>% #rename collapsed gene_ID column "gene"
  ungroup()


write.csv(GO.enrichment.summary, 'RAnalysis/Output/RNA-seq/GOSeq/GO.enrichment.summary.table.csv') 
```



## Modified biological processes heatmap
```{r}
# 
# GO.BP2 <- GO.BP %>%
#   group_by(slim_term) %>%
#   filter(n() > 4)
# 
# BPplot2 <- GO.BP2 %>% mutate(term = fct_reorder(term, -over_represented_pvalue)) %>%
#     ggplot(aes(x = ontology, y = term)) + 
#     geom_tile(aes(fill=over_represented_pvalue, width = 10)) + 
#       scale_y_discrete(position = "right") +
#       facet_grid(slim_term~ ., scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
#       theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
#       panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
#       strip.text.y.left = element_text(angle=0, size = 20, face = "bold"),
#       strip.text.y.right = element_text(angle=90, size = 30, face = "bold"),
#       strip.text.x = element_text(size = 22, face = "bold"),
#       axis.title = element_blank(),
#       axis.text = element_text(size = 22), legend.title = element_text(size = 22), legend.text =     element_text(size = 22)) 
# 
# 
# 
# ggsave("../../../Downloads/modified.heatmap.png", BPplot2, width = 25, height = 30, units = c("in"))
```








