---
title: "GO_Term_Enrichment_Analysis"
author: "daniellembecker"
edited by: "daniellembecker"
date: "9/20/2021"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Gene Ontology (GO) Enrichment Analysis 
## Previous steps include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DEG statistical data sheet
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
if (!requireNamespace("DESeq2", quietly = TRUE)) { BiocManager::install("DESeq2") }
library("DESeq2")
library("tidyverse")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
if (!requireNamespace("genefilter", quietly = TRUE)) { BiocManager::install("genefilter") }
library("genefilter")
library("ggplot2")
library("gplots")
if (!requireNamespace("limma", quietly = TRUE)) { BiocManager::install("limma") }
library("limma")
library("spdep") 
library("patchwork")
library("adegenet") 
if (!requireNamespace("goseq", quietly = TRUE)) { BiocManager::install("goseq") }
library("goseq")
library("gridExtra")
if (!requireNamespace("clusterProfiler", quietly = TRUE)) { BiocManager::install("clusterProfiler") }
library("clusterProfiler")
library("DataCombine")
library("VennDiagram")
if (!requireNamespace("GSEABase", quietly = TRUE)) { BiocManager::install("GSEABase") }
library("GSEABase")
library("data.table")
if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")
library("rrvgo")
if ("org.Ce.eg.db" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("org.Ce.eg.db")
library(org.Ce.eg.db)
if ("topGO" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("topGO")
library(topGO)
if ("Rgraphviz" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("Rgraphviz")
library(Rgraphviz)
```

# Import the data files 
```{r, echo=FALSE}
# Structural annotation GFF3 file
GFF3 <- read.csv("RAnalysis/Output/RNA-seq/DEG/rename_structure_gff3.csv")

# Load go terms from functional annotation file
annot_GO <- read.csv("RAnalysis/Output/RNA-seq/DEG/annot_GO.terms.host.csv")[-1]

# Check if all gene IDs in GFF3 are in annot_GO
unmatched_genes <- setdiff(GFF3$gene_id, annot_GO$gene_id)
# Print the unmatched gene IDs
print(unmatched_genes)

# Remove rows where 'GO.ID' is 'unknown'
annot_GO <- subset(annot_GO, GO.ID != 'unknown')

# edit annot_GO to list GO terms all in one row per gene
annot_GO <- annot_GO %>%
  group_by(gene_id) %>%
  summarise(GO.ID = paste(GO.ID, collapse = ";")) %>%
  ungroup()

```

# Identifying treatment and all expressed pver genes (poverA = 0.50,10)
```{r}
#treatment information
treatmentinfo <- read.csv("RAnalysis/Data/RNA-seq/metadata.RNAseq.csv", header = TRUE, sep = ",")
str(treatmentinfo)
head(treatmentinfo)

#DEG significant results
DEG.res <- read.csv("RAnalysis/Output/RNA-seq/DEG/DEGSeq2.sig.results.host.csv")[,-1]
nrow(DEG.res)
#DEG.res$gene_id <-  gsub("_gene","",DEG.res$gene_id) #remove extra characters

colnames(DEG.res)[7] <-"gene_id" # make colnames a true column called gene_id
head(DEG.res)

# Load filtered gene count file
gcount_filt <- read.csv("RAnalysis/Output/RNA-seq/DEG/gene_count_matrix_filtered.csv")

# Rename first column X to gene_id
names(gcount_filt)[names(gcount_filt) == "X"] <- "gene_id"

```

Match up genes in gene list file to annotation file
```{r}
names(annot_GO)

gcount_filt2annot = match(gcount_filt$gene_id, annot_GO$gene_id) #match genes in gcount_filt to annot_GO

# The following is the number of probes without annotation 
sum(is.na(gcount_filt2annot))

row_nas<-which(is.na(gcount_filt2annot))

#view the genes that do not have a match in the annotation file
missing<-as.data.frame(gcount_filt2annot[row_nas])
#print(missing)
```
1,663 genes in gcount_filt do not have GO annotations

Reduce annotation file to only contain genes detected in our dataset.  
```{r}
filtered_Pverr.annot <- annot_GO[annot_GO$gene_id %in% gcount_filt$gene_id, ]
dim(filtered_Pverr.annot)

```

The annotation file now only contains genes that were detected in our dataset that have annotation information. This is 18158 out of the 27439 genes in our dataset. 

Add in length to the annotation file.    

```{r}
filtered_Pverr.annot$length <- GFF3$length[match(filtered_Pverr.annot$gene_id, GFF3$gene_id)]

which(is.na(filtered_Pverr.annot$length)) #all genes have lengths 
```

All have length information. 

Conduct at bp,mf, and cc level for upregulated genes. Filter for BH adjusted p-value <0.05.  
```{r}
### Generate vector with names of all genes detected in our dataset 
ALL.vector <- c(filtered_Pverr.annot$gene_id)
  
### Generate length vector for all genes 
LENGTH.vector <- as.integer(filtered_Pverr.annot$length)
    
### Generate vector with names in just the contrast we are analyzing
ID.vector.up <- DEG.res %>%
      filter(direction=="Upregulated") %>%
      pull(gene_id)
    
##Get a list of GO Terms for all genes detected in our dataset 
    GO.terms <- filtered_Pverr.annot%>%
      dplyr::select(gene_id, GO.ID)
    
##Format to have one goterm per row with gene ID repeated
    split <- strsplit(as.character(GO.terms$GO.ID), ";") 
    split2 <- data.frame(v1 = rep.int(GO.terms$gene_id, sapply(split, length)), v2 = unlist(split)) 
    colnames(split2) <- c("gene_id", "GO.ID")
    GO.terms<-split2

##Construct list of genes with 1 for genes in contrast and 0 for genes not in the contrast
    gene.vector.up=as.integer(ALL.vector %in% ID.vector.up) 
    names(gene.vector.up)<-ALL.vector#set names

#weight gene vector by bias for length of gene: load in the gene.vector(a list of all genes with 1 indicating it is a DEG in the group of interest and 0 meaning it is not a DEG) and bias.data (list of lengths for all genes)
    pwf.up<-nullp(gene.vector.up, bias.data=LENGTH.vector) 
    
#run goseq using Wallenius method for all categories of GO terms: load in the pwf object (gene name, a 1 or 0 to indicate DEG, the length, and the pwf function; generated from pwf above), and the list of GO terms for all genes
    GO.wall.up<-goseq(pwf.up, gene2cat=GO.terms, test.cats=c("GO:BP", "GO:MF", "GO:CC"), method="Wallenius", use_genes_without_cat=TRUE)

    GO.up <- GO.wall.up[order(GO.wall.up$over_represented_pvalue),]
    colnames(GO.up)[1] <- "GOterm"
    
    #adjust p-values 
    GO.up$bh_adjust <- p.adjust(GO.up$over_represented_pvalue, method="BH") #add adjusted p-values
  
    
    #Filtering for p < 0.05
    GO.up <- GO.up %>%
        dplyr::filter(bh_adjust<0.05) %>%
        dplyr::arrange(., ontology, bh_adjust)

#Write file of results 
write_csv(GO.up, "RAnalysis/Output/RNA-seq/GOSeq/func.enrich.upregulated.csv")
    

```

Conduct at bp,mf, and cc level for downregulated genes. Filter for BH adjusted p-value <0.05.  
```{r}
### Generate vector with names of all genes detected in our dataset 
ALL.vector <- c(filtered_Pverr.annot$gene_id)
  
### Generate length vector for all genes 
LENGTH.vector <- as.integer(filtered_Pverr.annot$length)
    
### Generate vector with names in just the contrast we are analyzing
ID.vector.down <- DEG.res %>%
      filter(direction=="Downregulated") %>%
      pull(gene_id)
    
##Get a list of GO Terms for all genes detected in our dataset 
    GO.terms <- filtered_Pverr.annot%>%
      dplyr::select(gene_id, GO.ID)
    
##Format to have one goterm per row with gene ID repeated
    split <- strsplit(as.character(GO.terms$GO.ID), ";") 
    split2 <- data.frame(v1 = rep.int(GO.terms$gene_id, sapply(split, length)), v2 = unlist(split)) 
    colnames(split2) <- c("gene_id", "GO.ID")
    GO.terms<-split2

##Construct list of genes with 1 for genes in contrast and 0 for genes not in the contrast
    gene.vector.down=as.integer(ALL.vector %in% ID.vector.down) 
    names(gene.vector.down)<-ALL.vector#set names

#weight gene vector by bias for length of gene: load in the gene.vector(a list of all genes with 1 indicating it is a DEG in the group of interest and 0 meaning it is not a DEG) and bias.data (list of lengths for all genes)
    pwf.down<-nullp(gene.vector.down, bias.data=LENGTH.vector) 

#run goseq using Wallenius method for all categories of GO terms: load in the pwf object (gene name, a 1 or 0 to indicate DEG, the length, and the pwf function; generated from pwf above), and the list of GO terms for all genes
    GO.wall.down<-goseq(pwf.down, gene2cat=GO.terms, test.cats=c("GO:BP", "GO:MF", "GO:CC"), method="Wallenius", use_genes_without_cat=TRUE)

    GO.down <- GO.wall.down[order(GO.wall.down$over_represented_pvalue),]
    colnames(GO.down)[1] <- "GOterm"
    
    #adjust p-values 
    GO.down$bh_adjust <- p.adjust(GO.down$over_represented_pvalue, method="BH") #add adjusted p-values
  
    #Filtering for p < 0.05
    GO.down <- GO.down %>%
        dplyr::filter(bh_adjust<0.05) %>%
        dplyr::arrange(., ontology, bh_adjust)

#Write file of results 
write_csv(GO.down, "RAnalysis/Output/RNA-seq/GOSeq/func.enrich.downregulated.csv")

```


Use topGO for enrichment analysis on upregulated data
```{r}
# Create a gene-to-GO mapping list
geneID2GO <- split(as.character(GO.terms$GO.ID), GO.terms$gene_id)
geneNames <- unique(GO.terms$gene_id)

# Create a named vector for allGenes
allGenes.up <- factor(as.integer(geneNames %in% ID.vector.up))
names(allGenes.up) <- geneNames

# Create a topGOdata object using the custom gene2GO mapping
GOdata.up <- new("topGOdata",
              ontology = "BP",
              allGenes = allGenes.up,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.up
numSigGenes(GOdata.up) #15385
sum(feasible(GOdata.up)) #15385
```
Building most specific GOs .....
	( 10601 GO terms found. )

Build GO DAG topology ..........
	( 14202 GO terms and 31626 relations. )

Annotating nodes ...............
	( 15385 genes annotated to the GO terms. )

------------------------- topGOdata object -------------------------

 Ontology:
   -  BP 

 18158 available genes (all genes from the array):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 158  significant genes. 

 15385 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 137  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 8152 
   - number of edges = 17684 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled upregulated data
```{r}
resultFisher.up <- runTest(GOdata.up, algorithm = "classic", statistic = "fisher")
resultFisher.up
hist(score(resultFisher.up), 50, xlab = "p-values")

resultKS.up <- runTest(GOdata.up, algorithm = "classic", statistic = "ks")
resultKS.up
hist(score(resultKS.up), 50, xlab = "p-values")

resultWeight.up <- runTest(GOdata.up, algorithm = "weight01", statistic = "fisher")
resultWeight.up
hist(score(resultWeight.up), 50, xlab = "p-values")
```

Combine dataframes and adjust p values for upregulated data
```{r}
allRes.up <- GenTable(GOdata.up, classicFisher.up = resultFisher.up, classicKS.up = resultKS.up, weightFisher.up = resultWeight.up, orderBy = "weightFisher.up", ranksOf = "classicFisher.up", topNodes = length(score(resultFisher.up)))

# Convert character columns to numeric
allRes.up$classicFisher.up <- as.numeric(allRes.up$classicFisher.up)
allRes.up$classicKS.up <- as.numeric(allRes.up$classicKS.up)
allRes.up$weightFisher.up <- as.numeric(allRes.up$weightFisher.up)

#adjust p-values 
allRes.up$weight_bh_adjust <- p.adjust(allRes.up$classicFisher.up, method="BH") #add adjusted p-values
allRes.up
```

Create node figrues of top 5-10 sig terms for upregulated genes
```{r}
showSigOfNodes(GOdata.up, score(resultFisher.up), firstSigNodes = 5, useInfo = 'all')
printGraph(GOdata.up, resultFisher.up, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = TRUE)
printGraph(GOdata.up, resultKS.up, firstSigNodes = 10, fn.prefix = "tGOKS", useInfo = "all", pdfSW = TRUE)
printGraph(GOdata.up, resultWeight.up, firstSigNodes = 10, fn.prefix = "tGOKS", useInfo = "all", pdfSW = TRUE)
```



Use topGO for enrichment analysis on downregulated data
```{r}
# Create a gene-to-GO mapping list
geneID2GO <- split(as.character(GO.terms$GO.ID), GO.terms$gene_id)
geneNames <- unique(GO.terms$gene_id)

# Create a named vector for allGenes
allGenes.down <- factor(as.integer(geneNames %in% ID.vector.down))
names(allGenes.down) <- geneNames

# Create a topGOdata object using the custom gene2GO mapping
GOdata.down <- new("topGOdata",
              ontology = "BP",
              allGenes = allGenes.down,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.down
numSigGenes(GOdata.down) 
sum(feasible(GOdata.down)) 
```
Building most specific GOs .....
	( 10601 GO terms found. )

Build GO DAG topology ..........
	( 14202 GO terms and 31626 relations. )

Annotating nodes ...............
	( 15385 genes annotated to the GO terms. )

------------------------- topGOdata object -------------------------

 Ontology:
   -  BP 

 18158 available genes (all genes from the array):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 74  significant genes. 

 15385 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 58  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 8152 
   - number of edges = 17684 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled data
```{r}
resultFisher.down <- runTest(GOdata.down, algorithm = "classic", statistic = "fisher")
resultFisher.down
hist(score(resultFisher.down), 50, xlab = "p-values")

resultKS.down <- runTest(GOdata.down, algorithm = "classic", statistic = "ks")
resultKS.down
hist(score(resultKS.down), 50, xlab = "p-values")

resultWeight.down <- runTest(GOdata.down, algorithm = "weight01", statistic = "fisher")
resultWeight.down
hist(score(resultWeight.down), 50, xlab = "p-values")
```

Combine dataframes and adjust p values for downregulated data
```{r}
allRes.down <- GenTable(GOdata.down, classicFisher.down = resultFisher.down, classicKS.down = resultKS.down, weightFisher.down = resultWeight.down, orderBy = "weightFisher.down", ranksOf = "classicFisher.down", topNodes = length(score(resultFisher.down)))

# Convert character columns to numeric
allRes.down$classicFisher.down <- as.numeric(allRes.down$classicFisher.down)
allRes.down$classicKS.down <- as.numeric(allRes.down$classicKS.down)
allRes.down$weightFisher.down <- as.numeric(allRes.down$weightFisher.down)

#adjust p-values 
allRes.down$weight_bh_adjust <- p.adjust(allRes.down$classicFisher.down, method="BH") #add adjusted p-values
allRes.down
```

Create node figrues of top 5-10 sig terms for downregulated genes
```{r}
showSigOfNodes(GOdata.down, score(resultFisher.down), firstSigNodes = 5, useInfo = 'all')
printGraph(GOdata.down, resultFisher.down, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = TRUE)
printGraph(GOdata.down, resultKS.down, firstSigNodes = 10, fn.prefix = "tGOKS", useInfo = "all", pdfSW = TRUE)
printGraph(GOdata.down, resultWeight.down, firstSigNodes = 10, fn.prefix = "tGOKS", useInfo = "all", pdfSW = TRUE)
```

















