---
title: "GO_Term_Enrichment_Analysis"
author: "daniellembecker"
edited by: "daniellembecker"
date: "9/20/2021"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Gene Ontology (GO) Enrichment Analysis 
## Previous steps include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DEG statistical data sheet
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
if (!requireNamespace("DESeq2", quietly = TRUE)) { BiocManager::install("DESeq2") }
library("DESeq2")
library("tidyverse")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
if (!requireNamespace("genefilter", quietly = TRUE)) { BiocManager::install("genefilter") }
library("genefilter")
library("ggplot2")
library("gplots")
if (!requireNamespace("limma", quietly = TRUE)) { BiocManager::install("limma") }
library("limma")
library("spdep") 
library("patchwork")
library("adegenet") 
if (!requireNamespace("goseq", quietly = TRUE)) { BiocManager::install("goseq") }
library("goseq")
library("gridExtra")
if (!requireNamespace("clusterProfiler", quietly = TRUE)) { BiocManager::install("clusterProfiler") }
library("clusterProfiler")
library("DataCombine")
library("VennDiagram")
if (!requireNamespace("GSEABase", quietly = TRUE)) { BiocManager::install("GSEABase") }
library("GSEABase")
library("data.table")
if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")
library("rrvgo")
if ("org.Ce.eg.db" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("org.Ce.eg.db")
library(org.Ce.eg.db)
if ("topGO" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("topGO")
library(topGO)
if ("Rgraphviz" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("Rgraphviz")
library(Rgraphviz)
```

# Import the data files 
```{r, echo=FALSE}
# Structural annotation GFF3 file
GFF3 <- read.csv("RAnalysis/Output/RNA-seq/DEG/Host/rename_structure_gff3.csv")

# Load go terms from functional annotation file
annot_GO <- read.csv("RAnalysis/Output/RNA-seq/DEG/Host/annot_GO.terms.host.csv")[-1]

# Check if all gene IDs in GFF3 are in annot_GO
unmatched_genes <- setdiff(GFF3$gene_id, annot_GO$gene_id)
# Print the unmatched gene IDs
print(unmatched_genes)

# Remove rows where 'GO.ID' is 'unknown'
annot_GO <- subset(annot_GO, GO.ID != 'unknown')

# edit annot_GO to list GO terms all in one row per gene
annot_GO <- annot_GO %>%
  group_by(gene_id) %>%
  summarise(GO.ID = paste(GO.ID, collapse = ";")) %>%
  ungroup()

```

# Identifying treatment and all expressed pver genes (poverA = 0.50,10)
```{r}
#treatment information
treatmentinfo <- read.csv("RAnalysis/Data/RNA-seq/metadata.RNAseq.csv", header = TRUE, sep = ",")
str(treatmentinfo)
head(treatmentinfo)

#DEG significant results
DEG.res <- read.csv("RAnalysis/Output/RNA-seq/DEG/Host/DEGSeq2.sig.results.host.csv")[,-1]
nrow(DEG.res)
#DEG.res$gene_id <-  gsub("_gene","",DEG.res$gene_id) #remove extra characters

colnames(DEG.res)[7] <-"gene_id" # make colnames a true column called gene_id
head(DEG.res)

# Load filtered gene count file
gcount_filt <- read.csv("RAnalysis/Output/RNA-seq/DEG/gene_count_matrix_filtered.csv")

# Rename first column X to gene_id
names(gcount_filt)[names(gcount_filt) == "X"] <- "gene_id"

```

Match up genes in gene list file to annotation file
```{r}
names(annot_GO)

gcount_filt2annot = match(gcount_filt$gene_id, annot_GO$gene_id) #match genes in gcount_filt to annot_GO

# The following is the number of probes without annotation 
sum(is.na(gcount_filt2annot))

row_nas<-which(is.na(gcount_filt2annot))

#view the genes that do not have a match in the annotation file
missing<-as.data.frame(gcount_filt2annot[row_nas])
#print(missing)
```
1,663 genes in gcount_filt do not have GO annotations

Reduce annotation file to only contain genes detected in our dataset.  
```{r}
filtered_Pverr.annot <- annot_GO[annot_GO$gene_id %in% gcount_filt$gene_id, ]
dim(filtered_Pverr.annot)

```

The annotation file now only contains genes that were detected in our dataset that have annotation information. This is 18158 out of the 27439 genes in our dataset. 

Add in length to the annotation file.    

```{r}
filtered_Pverr.annot$length <- GFF3$length[match(filtered_Pverr.annot$gene_id, GFF3$gene_id)]

which(is.na(filtered_Pverr.annot$length)) #all genes have lengths 
```

All have length information. 

Conduct at bp,mf, and cc level for upregulated genes. Filter for BH adjusted p-value <0.05.  
```{r}
### Generate vector with names of all genes detected in our dataset 
ALL.vector <- c(filtered_Pverr.annot$gene_id)
  
### Generate length vector for all genes 
LENGTH.vector <- as.integer(filtered_Pverr.annot$length)
    
### Generate vector with names in just the contrast we are analyzing
ID.vector.up <- DEG.res %>%
      filter(direction=="Upregulated") %>%
      pull(gene_id)
    
##Get a list of GO Terms for all genes detected in our dataset 
    GO.terms <- filtered_Pverr.annot%>%
      dplyr::select(gene_id, GO.ID)
    
##Format to have one goterm per row with gene ID repeated
    split <- strsplit(as.character(GO.terms$GO.ID), ";") 
    split2 <- data.frame(v1 = rep.int(GO.terms$gene_id, sapply(split, length)), v2 = unlist(split)) 
    colnames(split2) <- c("gene_id", "GO.ID")
    GO.terms<-split2

##Construct list of genes with 1 for genes in contrast and 0 for genes not in the contrast
    gene.vector.up=as.integer(ALL.vector %in% ID.vector.up) 
    names(gene.vector.up)<-ALL.vector#set names

#weight gene vector by bias for length of gene: load in the gene.vector(a list of all genes with 1 indicating it is a DEG in the group of interest and 0 meaning it is not a DEG) and bias.data (list of lengths for all genes)
    pwf.up<-nullp(gene.vector.up, bias.data=LENGTH.vector) 
    
#run goseq using Wallenius method for all categories of GO terms: load in the pwf object (gene name, a 1 or 0 to indicate DEG, the length, and the pwf function; generated from pwf above), and the list of GO terms for all genes
    GO.wall.up<-goseq(pwf.up, gene2cat=GO.terms, test.cats=c("GO:BP", "GO:MF", "GO:CC"), method="Wallenius", use_genes_without_cat=TRUE)

    GO.up <- GO.wall.up[order(GO.wall.up$over_represented_pvalue),]
    colnames(GO.up)[1] <- "GOterm"
    
    #adjust p-values 
    GO.up$bh_adjust <- p.adjust(GO.up$over_represented_pvalue, method="BH") #add adjusted p-values
  
    
    #Filtering for p < 0.05
    GO.up <- GO.up %>%
        dplyr::filter(bh_adjust<0.05) %>%
        dplyr::arrange(., ontology, bh_adjust)

#Write file of results 
write_csv(GO.up, "RAnalysis/Output/RNA-seq/GOSeq/Host/func.enrich.upregulated.csv")
    

```

Conduct at bp,mf, and cc level for downregulated genes. Filter for BH adjusted p-value <0.05.  
```{r}
### Generate vector with names of all genes detected in our dataset 
ALL.vector <- c(filtered_Pverr.annot$gene_id)
  
### Generate length vector for all genes 
LENGTH.vector <- as.integer(filtered_Pverr.annot$length)
    
### Generate vector with names in just the contrast we are analyzing
ID.vector.down <- DEG.res %>%
      filter(direction=="Downregulated") %>%
      pull(gene_id)
    
##Get a list of GO Terms for all genes detected in our dataset 
    GO.terms <- filtered_Pverr.annot%>%
      dplyr::select(gene_id, GO.ID)
    
##Format to have one goterm per row with gene ID repeated
    split <- strsplit(as.character(GO.terms$GO.ID), ";") 
    split2 <- data.frame(v1 = rep.int(GO.terms$gene_id, sapply(split, length)), v2 = unlist(split)) 
    colnames(split2) <- c("gene_id", "GO.ID")
    GO.terms<-split2

##Construct list of genes with 1 for genes in contrast and 0 for genes not in the contrast
    gene.vector.down=as.integer(ALL.vector %in% ID.vector.down) 
    names(gene.vector.down)<-ALL.vector#set names

#weight gene vector by bias for length of gene: load in the gene.vector(a list of all genes with 1 indicating it is a DEG in the group of interest and 0 meaning it is not a DEG) and bias.data (list of lengths for all genes)
    pwf.down<-nullp(gene.vector.down, bias.data=LENGTH.vector) 

#run goseq using Wallenius method for all categories of GO terms: load in the pwf object (gene name, a 1 or 0 to indicate DEG, the length, and the pwf function; generated from pwf above), and the list of GO terms for all genes
    GO.wall.down<-goseq(pwf.down, gene2cat=GO.terms, test.cats=c("GO:BP", "GO:MF", "GO:CC"), method="Wallenius", use_genes_without_cat=TRUE)

    GO.down <- GO.wall.down[order(GO.wall.down$over_represented_pvalue),]
    colnames(GO.down)[1] <- "GOterm"
    
    #adjust p-values 
    GO.down$bh_adjust <- p.adjust(GO.down$over_represented_pvalue, method="BH") #add adjusted p-values
  
    #Filtering for p < 0.05
    GO.down <- GO.down %>%
        dplyr::filter(bh_adjust<0.05) %>%
        dplyr::arrange(., ontology, bh_adjust)

#Write file of results 
write_csv(GO.down, "RAnalysis/Output/RNA-seq/GOSeq/Host/func.enrich.downregulated.csv")

```


Use topGO for enrichment analysis on upregulated data
```{r}
# Create a gene-to-GO mapping list
geneID2GO <- split(as.character(GO.terms$GO.ID), GO.terms$gene_id)
geneNames <- unique(GO.terms$gene_id)

# Create a named vector for allGenes
allGenes.up <- factor(as.integer(geneNames %in% ID.vector.up))
names(allGenes.up) <- geneNames

# Create a topGOdata object using the custom gene2GO mapping
GOdata.up <- new("topGOdata",
              ontology = "BP",
              allGenes = allGenes.up,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.up
numSigGenes(GOdata.up) #15385
sum(feasible(GOdata.up)) #15385
```
Building most specific GOs .....
	( 10601 GO terms found. )

Build GO DAG topology ..........
	( 14202 GO terms and 31626 relations. )

Annotating nodes ...............
	( 15385 genes annotated to the GO terms. )

------------------------- topGOdata object -------------------------

 Ontology:
   -  BP 

 18158 available genes (all genes from the array):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 158  significant genes. 

 15385 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 137  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 8152 
   - number of edges = 17684 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled upregulated data
```{r}
resultFisher.up <- runTest(GOdata.up, algorithm = "classic", statistic = "fisher")
resultFisher.up
hist(score(resultFisher.up), 50, xlab = "p-values")

resultKS.up <- runTest(GOdata.up, algorithm = "classic", statistic = "ks")
resultKS.up
hist(score(resultKS.up), 50, xlab = "p-values")

resultWeight.up <- runTest(GOdata.up, algorithm = "weight01", statistic = "fisher")
resultWeight.up
hist(score(resultWeight.up), 50, xlab = "p-values")
```

Combine dataframes and adjust p values for upregulated data
```{r}
allRes.up <- GenTable(GOdata.up, classicFisher.up = resultFisher.up, classicKS.up = resultKS.up, weightFisher.up = resultWeight.up, orderBy = "weightFisher.up", ranksOf = "classicFisher.up", topNodes = length(score(resultFisher.up)))

# Convert character columns to numeric
allRes.up$classicFisher.up <- as.numeric(allRes.up$classicFisher.up)
allRes.up$classicKS.up <- as.numeric(allRes.up$classicKS.up)
allRes.up$weightFisher.up <- as.numeric(allRes.up$weightFisher.up)

#adjust p-values 
allRes.up$weight_bh_adjust <- p.adjust(allRes.up$classicFisher.up, method="BH") #add adjusted p-values
allRes.up
```

Create node figrues of top 5-10 sig terms for upregulated genes
```{r}
showSigOfNodes(GOdata.up, score(resultFisher.up), firstSigNodes = 5, useInfo = 'all')
printGraph(GOdata.up, resultFisher.up, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = TRUE)
printGraph(GOdata.up, resultKS.up, firstSigNodes = 10, fn.prefix = "tGOKS", useInfo = "all", pdfSW = TRUE)
printGraph(GOdata.up, resultWeight.up, firstSigNodes = 10, fn.prefix = "tGOKS", useInfo = "all", pdfSW = TRUE)
```



Use topGO for enrichment analysis on downregulated data
```{r}
# Create a gene-to-GO mapping list
geneID2GO <- split(as.character(GO.terms$GO.ID), GO.terms$gene_id)
geneNames <- unique(GO.terms$gene_id)

# Create a named vector for allGenes
allGenes.down <- factor(as.integer(geneNames %in% ID.vector.down))
names(allGenes.down) <- geneNames

# Create a topGOdata object using the custom gene2GO mapping
GOdata.down <- new("topGOdata",
              ontology = "BP",
              allGenes = allGenes.down,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.down
numSigGenes(GOdata.down) 
sum(feasible(GOdata.down)) 
```
Building most specific GOs .....
	( 10601 GO terms found. )

Build GO DAG topology ..........
	( 14202 GO terms and 31626 relations. )

Annotating nodes ...............
	( 15385 genes annotated to the GO terms. )

------------------------- topGOdata object -------------------------

 Ontology:
   -  BP 

 18158 available genes (all genes from the array):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 74  significant genes. 

 15385 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 58  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 8152 
   - number of edges = 17684 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled data
```{r}
resultFisher.down <- runTest(GOdata.down, algorithm = "classic", statistic = "fisher")
resultFisher.down
hist(score(resultFisher.down), 50, xlab = "p-values")

resultKS.down <- runTest(GOdata.down, algorithm = "classic", statistic = "ks")
resultKS.down
hist(score(resultKS.down), 50, xlab = "p-values")

resultWeight.down <- runTest(GOdata.down, algorithm = "weight01", statistic = "fisher")
resultWeight.down
hist(score(resultWeight.down), 50, xlab = "p-values")
```

Combine dataframes and adjust p values for downregulated data
```{r}
allRes.down <- GenTable(GOdata.down, classicFisher.down = resultFisher.down, classicKS.down = resultKS.down, weightFisher.down = resultWeight.down, orderBy = "weightFisher.down", ranksOf = "classicFisher.down", topNodes = length(score(resultFisher.down)))

# Convert character columns to numeric
allRes.down$classicFisher.down <- as.numeric(allRes.down$classicFisher.down)
allRes.down$classicKS.down <- as.numeric(allRes.down$classicKS.down)
allRes.down$weightFisher.down <- as.numeric(allRes.down$weightFisher.down)

#adjust p-values 
allRes.down$weight_bh_adjust <- p.adjust(allRes.down$classicFisher.down, method="BH") #add adjusted p-values
allRes.down
```

Create node figrues of top 5-10 sig terms for downregulated genes
```{r}
showSigOfNodes(GOdata.down, score(resultFisher.down), firstSigNodes = 5, useInfo = 'all')
printGraph(GOdata.down, resultFisher.down, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = TRUE)
printGraph(GOdata.down, resultKS.down, firstSigNodes = 10, fn.prefix = "tGOKS", useInfo = "all", pdfSW = TRUE)
printGraph(GOdata.down, resultWeight.down, firstSigNodes = 10, fn.prefix = "tGOKS", useInfo = "all", pdfSW = TRUE)
```

















## Merge DEG gene ids and GO info
```{r}
# #include enrichment information for each category (GO Term)
#split into one gene and go term for annot.GO
annot.GO.split <- annot_GO
split <- strsplit(as.character(annot.GO.split$GO.ID), ";") 
split2 <- data.frame(v1 = rep.int(annot.GO.split$gene_id, sapply(split, length)), v2 = unlist(split)) 
colnames(split2) <- c("gene_id", "GOterm")
annot.GO.split<-split2
merge <- merge(GO.down, annot.GO.split, by.y = "GOterm") # contains gene ids and GO info 

# Merge DEG with merge 
merge_again <- merge(merge, DEG.res, by = "gene_id") # contains GO info and gene counts for DEGs
```


## Find GOslim terms
# Run GOslim to get broader categories
```{r}
#load in generic GO database, has all of the upper level categories for GO terms. Ex: regulation of cell division would be under cellular regulation
slim <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") #get GO database

## filtering all of BP (do MF and CC seperately)
BP_GO <- enriched.GO.05 %>%
  filter(ontology=="BP")
BPGO_collection <- GOCollection(BP_GO$category) #Make library of query terms
slims_bp <- data.frame(goSlim(BPGO_collection, slim, "BP")) #Find common parent terms to slim down our list
slims_bp$category <- row.names(slims_bp) #save rownames as category

## filtering all of MF
MF_GO <- enriched.GO.05 %>%
  filter(ontology=="MF")
MFGO_collection <- GOCollection(MF_GO$category) #Make library of query terms
slims_mf <- data.frame(goSlim(MFGO_collection, slim, "MF")) #Find common parent terms to slim down our list
slims_mf$category <- row.names(slims_mf) #save rownames as category

## filtering all of CC
CC_GO <- enriched.GO.05 %>%
  filter(ontology=="CC")
CCGO_collection <- GOCollection(CC_GO$category) #Make library of query terms
slims_cc <- data.frame(goSlim(CCGO_collection, slim, "CC")) #Find common parent terms to slim down our list
slims_cc$category <- row.names(slims_cc) #save rownames as category
```

# Get mapped terms, using functions from Sam White's Biostars [post](https://support.bioconductor.org/p/128407/#128409).
```{r}
#custom function from Sam White's, gets mapped ids for all of your query terms 
#Write function mappedIds to get the query terms that mapped to the slim categories
mappedIds <-
  function(df, collection, OFFSPRING) #the command to run requires a dataframe of slim terms, like slims_MF above, your list of query terms, and the offspring from the GOCollection by goSlim
  {
    map <- as.list(OFFSPRING[rownames(df)]) # Subset GOcollection offspring by the rownames of your dataframe
    mapped <- lapply(map, intersect, ids(collection)) #Find the terms that intersect between the subset made above of your query terms and the GOids from the GO collection
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L)) #Add column "go_terms" with matching terms 
    df #show resulting dataframe
  }
#Run function for MF and BP terms
BPslim <- mappedIds(slims_bp, BPGO_collection, GOBPOFFSPRING)
MFslim <- mappedIds(slims_mf, MFGO_collection, GOMFOFFSPRING)
CCslim <- mappedIds(slims_cc, CCGO_collection, GOCCOFFSPRING)
```

# Remove duplicate matches, keeping the broader umbrella term
```{r}
#filtering out duplicates
#BP
BPslim <- filter(BPslim, Count>0 & Term!="biological_process") #filter out empty slims and term "biological process"
BPsplitted <- strsplit(as.character(BPslim$go_terms), ";") #split into multiple GO ids
BPslimX <- data.frame(Term = rep.int(BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) #list all
BPslimX <- merge(BPslimX, BPslim[,c(1,3:4)], by="Term") #Add back counts, term, and category info
BPslimX <- unique(setDT(BPslimX)[order(go_term, -Count)], by = "go_term") #remove duplicate offspring terms, keeping only those that appear in the larger umbrella term (larger Count number)
BPslim <- data.frame(slim_term=BPslimX$Term, slim_cat=BPslimX$category, category=BPslimX$go_term) #rename columns
head(BPslim)

#MF
MFslim <- filter(MFslim, Count>0 & Term!="molecular_function") #filter out empty slims and term "molecular function"
MFsplitted <- strsplit(as.character(MFslim$go_terms), ";") #split into multiple GO ids
MFslimX <- data.frame(Term = rep.int(MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) #list all
MFslimX <- merge(MFslimX, MFslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
MFslimX <- unique(setDT(MFslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
MFslim <- data.frame(slim_term=MFslimX$Term, slim_cat=MFslimX$category, category=MFslimX$go_term) #rename columns
head(MFslim)

#CC
CCslim <- filter(CCslim, Count>0 & Term!="cellular_component") #filter out empty slims and term "molecular function"
CCsplitted <- strsplit(as.character(CCslim$go_terms), ";") #split into multiple GO ids
CCslimX <- data.frame(Term = rep.int(CCslim$Term, sapply(CCsplitted, length)), go_term = unlist(CCsplitted)) #list all
CCslimX <- merge(CCslimX, CCslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
CCslimX <- unique(setDT(CCslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
CCslim <- data.frame(slim_term=CCslimX$Term, slim_cat=CCslimX$category, category=CCslimX$go_term) #rename columns
head(CCslim)

```

# Save slim info with GO enrichment info for heatmap dataframes.
```{r}
GO.BP <- right_join(BPslim, filter(enriched.GO.05, ontology=="BP"), by="category") #add back GO enrichment info for each offspring term
GO.MF <- right_join(MFslim, filter(enriched.GO.05, ontology=="MF"), by="category") #add back GO enrichment info for each offspring term
GO.CC <- right_join(CCslim, filter(enriched.GO.05, ontology=="CC"), by="category") #add back GO enrichment info for each offspring term
```



## Make supplemental table summarizing GO enrichment
```{r}
GO.enrichment.dat <- bind_rows(GO.BP, GO.MF, GO.CC)
head(GO.enrichment.dat)

#Make dataframe of GO results for clustering and heatmap. 
#add gene_IDs. To get gene_IDs we will merge with the GO.terms DF.
GOgenes <- data.frame(gene_id=annot_GO$gene_id, category=annot_GO$category) 
GOgenes$gene_id <- as.character(GOgenes$gene_id) #make gene ID a character so we can collapse our many near-identical columns

GO.enrichment.summary  <- left_join(GO.enrichment.dat, GOgenes, by="category" ) #join the DFs

GO.enrichment.summary <- GO.enrichment.summary %>% #collapse and have gene IDs for a particular term in a single row as a comma-sep list. 
  group_by(slim_term, slim_cat, category, over_represented_pvalue, under_represented_pvalue, numDEInCat, numInCat, term, ontology, bh_adjust) %>%
  summarise(genes = toString(gene_id)) %>% #rename collapsed gene_ID column "gene"
  ungroup()


write.csv(GO.enrichment.summary, 'RAnalysis/Output/RNA-seq/GOSeq/Host/GO.enrichment.summary.table.DEG.host.csv') 
```


## Exploring unique go_slim terms
```{r}
#number of unique go slim terms per category
length(unique(GO.enrichment.summary$slim_term)) #44 unique go terms

#view slim terms per ontology category
## filtering go slim category of BP 
BP_slim <- GO.enrichment.summary %>%
  filter(ontology=="BP")
view(BP_slim)
#number of BP go slim terms
length(unique(BP_slim$slim_term)) #32

## filtering go slim category of MF 
MF_slim <- GO.enrichment.summary %>%
  filter(ontology=="MF")
view(MF_slim)
#number of BP go slim terms
length(unique(MF_slim$slim_term)) #8

## filtering go slim category of CC 
CC_slim <- GO.enrichment.summary %>%
  filter(ontology=="CC")
view(CC_slim)

#number of CC go slim terms
length(unique(CC_slim$slim_term)) #4

```

# Make plots of ontology seperated DEGs
```{r}
MF_plot_GO <- GO.MF  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                   #mutate(term = fct_reorder(term, Day)) %>%
                   ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom"
                   ) +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("Molecular Function") +
                   theme_bw() + #Set background color 
   #facet_wrap(~slim_term) +
   theme(axis.text.y = element_text(hjust = 1))   +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes

 MF_plot_GO

 ggsave("RAnalysis/Output/Final_Figures/Host/GO.plot.MF.terms.host.pdf", MF_plot_GO , width = 10, height = 10, units = c("in"))

 BP_plot_GO <- GO.BP  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                   ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom") +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("Biological Processes") +
   facet_grid(slim_term ~ ., scales = "free", space='free') + 
   theme_bw() +
   theme(axis.text.y = element_text(hjust = 1))   +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes



BP_plot_GO

 ggsave("RAnalysis/Output/Final_Figures/Host/GO.plot.BP.terms.host.pdf", BP_plot_GO , width = 10, height = 49, units = c("in"))
 

 CC_plot_GO <- GO.CC  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                   #mutate(term = fct_reorder(term, Day)) %>%
                   ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom"
                   ) +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("Cellular Components") +
                   theme_bw() + #Set background color 
   #facet_wrap(~slim_term) +
   theme(axis.text.y = element_text(hjust = 1))   +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes

 CC_plot_GO

 ggsave("RAnalysis/Output/Final_Figures/Host/GO.plot.CC.terms.host.pdf", CC_plot_GO , width = 10, height = 10, units = c("in"))

```

#trying different approach to heatmaps using Ariana approach
```{r}
GO.plot.BP <-  ggplot(GO.BP, aes(x = ontology, y = term)) + 
    geom_point(aes(size=numInCat, fill= -log10(over_represented_pvalue), colour=-log10(over_represented_pvalue))) + 
    scale_size(name="Number of Genes", range=c(5,12), limits=c(1,500))+
    scale_fill_gradient(name = "Adjusted p-value", low="black", high="gray")+
    scale_colour_gradient(name = "Adjusted p-value", low="black", high="gray")+
    facet_grid(slim_term ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 60, multi_line = TRUE))+
    theme_bw() + 
    theme(panel.border = element_blank(), 
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
          strip.background =element_rect(fill="white"),
          strip.text.y = element_text(angle=0, size = 18, face="bold"),
          strip.text.x = element_text(size = 16),
          axis.text.x = element_text(size = 20, angle=45, hjust=1),
          axis.text.y = element_text(size = 16),
          axis.title.x = element_text(size = 22, face="bold"),
          legend.text=element_text(size=16),
          title=element_text(size=22, face="bold"),
          plot.title = element_text(hjust = 0.1),
          axis.title.y = element_blank())

ggsave(filename=paste0("RAnalysis/Output/RNA-seq/GOSeq/Host/final.BP.plot.host.png"), plot=GO.plot.BP, dpi=300, width=30, height=49, units="in")

```


# Make upregulated and downregulated plots
```{r}
#remove NA from enriched upregulated data frame
enriched.GO.05.up <- na.omit(enriched.GO.05.up)

#select for 30 lowest p values for each group 
BP.filt.u <- enriched.GO.05.up %>% filter(ontology == "BP")

CC.filt.u <- enriched.GO.05.up %>% filter(ontology == "CC")

MF.filt.u <- enriched.GO.05.up %>% filter(ontology == "MF")

#select only 30 lowest pvalues and rejoin 
BP.filt2.u <- BP.filt.u %>% slice_min(order_by = over_represented_pvalue, n = 30)

MF.filt2.u <- MF.filt.u %>% slice_min(order_by = over_represented_pvalue, n = 30)

#rejoin dataframes

enriched.up <- rbind(MF.filt2.u, BP.filt2.u, CC.filt.u)

# rename ontology terms
enriched.up <- enriched.up %>%
    mutate(ontology = recode(ontology, BP = 'Biological Process', MF = 'Molecular Function', CC =  'Cellular Component' ))

#create upreg plot
upreg.DEG.plot <- enriched.up  %>%  
  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>% 
  mutate(ontology = factor(ontology, levels = c("Biological Process", "Molecular Function", "Cellular Component"))) %>% ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom"
                   ) +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("Upregulated Differentially Expressed Genes") +
                   theme_bw() + #Set background color 
   facet_grid(ontology ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 55, multi_line = TRUE)) +
   theme(axis.text.y = element_text(hjust = 1))   +
  theme(plot.title = element_text(size = 25, face = "bold", hjust = 0.95)) +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes


 ggsave("RAnalysis/Output/Final_Figures/Host/upreg.DEG.plot.pdf", upreg.DEG.plot , width = 12, height = 20, units = c("in"))
 
 
##make downregulated plot
#remove NA from enriched downregulated data frame
enriched.GO.05.down <- na.omit(enriched.GO.05.down)

#select for 30 lowest p values for each group, first filter by ontology
BP.filt.d <- enriched.GO.05.down %>% filter(ontology == "BP")

CC.filt.d <- enriched.GO.05.down %>% filter(ontology == "CC")

MF.filt.d <- enriched.GO.05.down %>% filter(ontology == "MF")

#select only 30 lowest pvalues and rejoin 
BP.filt2.d <- BP.filt.d %>% slice_min(order_by = over_represented_pvalue, n = 30)

MF.filt2.d <- MF.filt.d %>% slice_min(order_by = over_represented_pvalue, n = 30)

#rejoin dataframes

enriched.down <- rbind(MF.filt2.d, BP.filt2.d, CC.filt.d)


# rename ontology terms
enriched.down <- enriched.down %>%
    mutate(ontology = recode(ontology, BP = 'Biological Process', MF = 'Molecular Function', CC =  'Cellular Component' ))


#make downreg plot
downreg.DEG.plot <- enriched.down  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>% 
  mutate(ontology = factor(ontology, levels = c("Biological Process", "Molecular Function", "Cellular Component"))) %>% ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom"
                   ) +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("Downregulated Differentially Expressed Genes") +
                   theme_bw() + #Set background color 
   facet_grid(ontology ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 55, multi_line = TRUE))+
   theme(axis.text.y = element_text(hjust = 1))   +
  theme(plot.title = element_text(size = 25, face = "bold", hjust = 0.95)) +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes


ggsave("RAnalysis/Output/Final_Figures/Host/downreg.DEG.plot.pdf", downreg.DEG.plot , width = 12, height = 20, units = c("in"))


#use patchwork to combine figures and add labels
Figure1 <- upreg.DEG.plot + downreg.DEG.plot + plot_annotation(tag_levels = 'A')
 
ggsave(filename=paste0("RAnalysis/Output/Final_Figures/Host/manuscript/down.upreg.DEG.png"), plot=Figure1, dpi=300, width=18, height=20, units="in")
ggsave(filename=paste0("RAnalysis/Output/Final_Figures/Host/manuscript/down.upreg.DEG.pdf"), plot=Figure1, dpi=300, width=18, height=20, units="in")

```


#######################################################################################

## KEGG enrichment analysis

## Obtained KEGG ontology terms by using KofamScan in command-line; see step 3 in E. Chille [lab notebook](https://github.com/echille/E.-Chille-Open-Lab-Notebook/blob/master/_posts/2020-10-08-M-capitata-functional-annotation-pipeline.md) for instructions. 
## If run into issues with downloading KofamScan software, etc.; see https://github.com/Putnam-Lab/Lab_Management/issues/21

# INFORMATION FOR KEGG ENRICHMENT ANALYSIS FOR R FOUND HERE:
# https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html 
#http://yulab-smu.top/clusterProfiler-book/chapter6.html#kegg-over-representation-test

# Load KO -- gene mapping info
```{r}

#Load in KO term functional annotation and edit to match GFF3 and annotation file
KFS.KO.orig <- read_tsv("../../../Desktop/Pver_KO_annot.tsv", col_names = TRUE)

KFS.KO <- KFS.KO.orig[c(1,2,3)]
colnames(KFS.KO) <- c("sig", "gene_id", "ko")
KFS.KO <- filter(KFS.KO, sig=="*") #filter for significant genes 9,300 total
KFS.KO <- KFS.KO[,c(2:3)]

#need to apply the same changes from Pver_annot file to the KFS.KO for gene_id
KFS.KO$gene_id <-  gsub(".t[1-67890900]","", KFS.KO$gene_id) #remove everything after g#
KFS.KO$gene_id <- gsub("Pver_split_", "Pver_", KFS.KO$gene_id)
KFS.KO$gene_id <- gsub("Pver_", "Pver_gene_", KFS.KO$gene_id)
KFS.KO$gene_id <-  gsub("_gene_g","_g", KFS.KO$gene_id) #sub gene_g with _g
KFS.KO$gene_id <-  gsub("_gene_","_", KFS.KO$gene_id) #remove extra _
KFS.KO$gene_id <-  gsub("\\..*","",KFS.KO$gene_id) #remove extra characters after the first .
head(KFS.KO)
tail(KFS.KO)

#remove eval and rename to KO.terms all
KO.terms <- KFS.KO
KO.terms <- unique(KO.terms)
str(KO.terms)
head(KO.terms)
colnames(KO.terms) <- c("gene_id", "KEGG.ID")

```

Reduce annotation file to only contain genes detected in our dataset.  
```{r}
filtered_Pverr.KEGG <- annot_GO[annot_GO$gene_id %in% KO.terms$gene_id, ]
dim(filtered_Pverr.KEGG)

```

The annotation file now only contains genes that were detected in our dataset that have annotation information. This is 8122 out of the 27439 genes in our dataset. 

Add in length to the annotation file.    

```{r}
filtered_Pverr.KEGG$length <- GFF3$length[match(filtered_Pverr.KEGG$gene_id, GFF3$gene_id)]

which(is.na(filtered_Pverr.KEGG$length)) #all genes have lengths 
```

Conduct at KEGG process level for upregulated. Filter for adjusted p-value <0.05.  
```{r}
### Generate vector with names of all genes detected in our dataset 
ALL.vector <- c(filtered_Pverr.KEGG$gene_id)
  
### Generate length vector for all genes 
LENGTH.vector <- as.integer(filtered_Pverr.KEGG$length)
    
### Generate vector with names in just the contrast we are analyzing
ID.vector <- KFS.KO %>%
      #filter(direction=="Downregulated") %>%
      pull(gene_id)
    
##Get a list of GO Terms for all genes detected in our dataset 
    KEGG.terms <- filtered_Pverr.KEGG%>%
      dplyr::select(gene_id, GO.ID)
    
##Format to have one goterm per row with gene ID repeated
    split <- strsplit(as.character(KEGG.terms$GO.ID), ";") 
    split2 <- data.frame(v1 = rep.int(KEGG.terms$gene_id, sapply(split, length)), v2 = unlist(split)) 
    colnames(split2) <- c("gene_id", "KO.ID")
    KEGG.terms<-split2

##Construct list of genes with 1 for genes in contrast and 0 for genes not in the contrast
    gene.vector=as.integer(ALL.vector %in% ID.vector) 
    names(gene.vector)<-ALL.vector#set names

#weight gene vector by bias for length of gene: load in the gene.vector(a list of all genes with 1 indicating it is a DEG in the group of interest and 0 meaning it is not a DEG) and bias.data (list of lengths for all genes)
    pwf.KEGG<-nullp(gene.vector, bias.data=LENGTH.vector) 
    
#run goseq using Wallenius method for all categories of GO terms: load in the pwf object (gene name, a 1 or 0 to indicate DEG, the length, and the pwf function; generated from pwf above), and the list of GO terms for all genes
    KEGG.wall<-goseq(pwf.KEGG, gene2cat=KEGG.terms, test.cats=c("KEGG", "GO:BP", "GO:MF", "GO:CC"), method="Wallenius", use_genes_without_cat=TRUE)

    KEGG <- KEGG.wall[order(KEGG.wall$over_represented_pvalue),]
    colnames(KEGG)[1] <- "KEGG.term"
    
    #adjust p-values 
    KEGG$bh_adjust <- p.adjust(KEGG$over_represented_pvalue, method="BH") #add adjusted p-values
  
    #Filtering for p < 0.05
    GO <- GO %>%
        dplyr::filter(bh_adjust<0.1) %>%
        dplyr::arrange(., ontology, bh_adjust)

    #Write file of results 
    #write_csv(GO, file = paste0("Mcap2020/Output/TagSeq/GenomeV3/DESeq/functional_enrichment/goseq_cluster", cluster_id, ".csv"))
    

```

## Perform Kegg enrichment with goseq package
```{r}
#used str() to see that GOKO.terms was a tibble not a dataframe, convert to dataframe
#GOKO.terms <- as.data.frame(GOKO.terms)

#Perform goseq
KOwall.up <- goseq(pwf.up, gene2cat=filtered_Pverr.KEGG, test.cats=c("KEGG"), method="Wallenius", use_genes_without_cat=TRUE)
KOwall.down <- goseq(pwf.down, GFF3$gene_id, gene2cat=GOKO.terms, test.cats=c("KEGG"), method="Wallenius", use_genes_without_cat=TRUE)
```

Replace NAs with "KEGG" and extract KO terms from results
```{r}
#Upregulated
up.KO.05<-KOwall.up$category[KOwall.up$over_represented_pvalue<.05]
up.KO.05<-data.frame(up.KO.05)
colnames(up.KO.05) <- c("category")
up.KO.05 <- merge(up.KO.05, KOwall.up, by="category")
up.KO.05$ontology <- replace_na(up.KO.05$ontology, "KEGG")
up.KO.05 <- filter(up.KO.05, ontology=="KEGG")
up.KO.05 <- up.KO.05[order(up.KO.05$ontology, up.KO.05$over_represented_pvalue, -up.KO.05$numDEInCat),]
up.KO.05$term <- as.factor(up.KO.05$term)
nrow(up.KO.05)

#Downregulated
down.KO.05<-KOwall.down$category[KOwall.down$over_represented_pvalue<.05]
down.KO.05<-data.frame(down.KO.05)
colnames(down.KO.05) <- c("category")
down.KO.05 <- merge(down.KO.05, KOwall.down, by="category")
down.KO.05$ontology <- replace_na(down.KO.05$ontology, "KEGG")
down.KO.05 <- filter(down.KO.05, ontology=="KEGG")
down.KO.05 <- down.KO.05[order(down.KO.05$ontology, down.KO.05$over_represented_pvalue, -down.KO.05$numDEInCat),]
down.KO.05$term <- as.factor(down.KO.05$term)
nrow(down.KO.05)
```

Add KO definitions
```{r}
#Prep definition data
KFS.KO.def <- subset(KFS.KO.orig, select=c("#","KO", "KO definition"))
colnames(KFS.KO.def) <- c("sig","category", "term")
KFS.KO.def <- filter(KFS.KO.def, sig=="*")
KFS.KO.def <- KFS.KO.def[,c(2:3)]
#Merge with KEGG output
up.KO.05 <- unique(left_join(up.KO.05[,-6], KFS.KO.def, by=c("category")))
down.KO.05 <- unique(left_join(down.KO.05[,-6], KFS.KO.def, by=c("category")))
```

Write output KEGG enrichment files
```{r}
write.csv(up.KO.05, file = "RAnalysis/Output/RNA-seq/KEGG_ontology/KO.05.up.csv")
write.csv(down.KO.05, file = "RAnalysis/Output/RNA-seq/KEGG_ontology/KO.05.down.csv")
```

## Make supplemental table summarizing KO enrichment
## Must do separately for up and down so we don't get duplicate gene entries. Add dir info
```{r}
#Add dir info
up.KO.05$dir <- "up"
down.KO.05$dir <- "down"
```

Make dataframe of GO results for clustering and heatmap. 
```{r}
#add gene_IDs. To get gene_IDs we will merge with the GO.terms DF.
GOKOgenes <- data.frame(gene_id=GOKO.terms$gene_id, category=GOKO.terms$GO.ID) #First have to make the "by" column the same for both
```

Using R script from Mass Lab. Needs columns "experiment" where we will put cluster information, "term", "percentDEInCat", and "gene" with all the DE genes associated with that GO term.
```{r}
#DO UP (up)
KOgenes_U <- filter(GOKOgenes)

KOdf_U <- left_join(up.KO.05, KOgenes_U, by="category" ) #join the DFs
KOdf_U$gene_id <- as.character(KOdf_U$gene_id) #make gene ID a character so we can collapse our many near-identical columns

```

```{r}
#DO DOWN (down)
KOgenes_D <- filter(GOKOgenes)
KOdf_D <- left_join(down.KO.05, KOgenes_D, by="category" ) #join the DFs
KOdf_D$gene_id <- as.character(KOdf_D$gene_id) #make gene ID a character so we can collapse our many near-identical columns

```

Bind rows for up and down and save!
```{r}
KOdf <- bind_rows(KOdf_U, KOdf_D)
KOdf <- na.omit(KOdf)
head(KOdf)
str(KOdf)
write.csv(KOdf, file = "RAnalysis/Output/RNA-seq/KEGG_ontology/KOenrichmentsummary.csv", row.names = FALSE)
```

```{r}
#KEGG plot
KEGGplot <- KOdf  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                   #mutate(term = fct_reorder(term, Day)) %>%
                   ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom"
                   ) +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("KEGG terms") +
                   theme_bw() + #Set background color 
   theme(axis.text.y = element_text(hjust = 1))   +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes

KEGGplot


ggsave("RAnalysis/Output/RNA-seq/KEGG_ontology/KEGGplot.png", KEGGplot, width = 10, height = 10, units = c("in"))
 
```





