---
title: "KEGG_Pathway_Analysis"
author: "HM Putnam"
date: "9/20/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## KEGG enrichment analysis
## Obtained KEGG ontology terms by using KofamScan in command-line; see step 3 in E. Chille lab notebook https://github.com/echille/E.-Chille-Open-Lab-Notebook/blob/master/_posts/2020-10-08-M-capitata-functional-annotation-pipeline.md for instructions. 
## If run into issues with downloading KofamScan software, etc.; see https://github.com/Putnam-Lab/Lab_Management/issues/21

# Load KO -- gene mapping info
```{r}

#Load in and prep KofamScan search results
KFS.KO.orig <- read_tsv("RAnalysis/Genome/KEGG-ontology/Pver_KO_annot.tsv.gz", col_names = TRUE)
KFS.KO <- KFS.KO.orig[-1,-1] #remove the first row and column becasue they  contain no data
KFS.KO <- KFS.KO[, c(1,2,5)] #subset the gene ID, KO term, and evalue
colnames(KFS.KO) <- c("gene_id", "ko", "eval") #name columns
#NEED TO DECIDE KO TERM CUTOFF AND WHY OR WHY NOT TO KEEP MORE THAN ONE PER GENE
#FILTER FOR EVALUE AS NEEDED
#KFS.KO <- KFS.KO[,c(1:2)] #remove evalue from the dataframe
Kegg.name <- as.data.frame(unique(KFS.KO$gene_id)) #identify all of the unique gene names
colnames(Kegg.name) <- c("gene_id") #set column name for consistency across files

#Annot appears to have an extra Pver_g1.t2 and are missing the word gene
KFS.KO$gene_id <-  gsub("\\..*","", KFS.KO$gene_id)
KFS.KO$gene_id <-  gsub("Pver_novel_model_","Pver_novel_gene_",KFS.KO$gene_id)
#g# is duplicated after dash 
KFS.KO$gene_id <-  gsub("-.*","",KFS.KO$gene_id) #remove the extra g number after the dash
KFS.KO$gene_id <-  gsub("_gene_g","_g",KFS.KO$gene_id)
KFS.KO$gene_id <-  gsub("_gene_","_",KFS.KO$gene_id)

Kegg.name <- as.data.frame(unique(KFS.KO$gene_id)) #identify all of the unique gene names
colnames(Kegg.name) <- c("gene_id") #set column name for consistency across files

KFS.KO$gene_id <- gsub("Pver_split_", "Pver_", KFS.KO$gene_id)
KFS.KO$gene_id <- gsub("Pver_", "Pver_gene_", KFS.KO$gene_id)
head(KFS.KO)
tail(KFS.KO)
```


```{r}
#following steps here: https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html to complete KEGG enrichment
#to search KEGG organism list, go to: https://www.genome.jp/kegg/catalog/org_list.html
#Hollie, I choose Pocillopora damicornis from the Cnidarians section on the genome list, think it is closest we will get
#clusterProfiler (Yu et al. 2012). clusterprofiler supports direct online access of the current KEGG database, rather than relying on R annotation packages, it also provides some nice visualisation options.
#need to download package from devtools because it wasn't compatable with new rstudio version
#devtools::install_github(c("GuangchuangYu/DOSE", "GuangchuangYu/clusterProfiler"))

library(clusterProfiler)
search_kegg_organism('pdam', by='kegg_code')

#need list of FDR and log fold change for genes here, that is in the DESeq2 DEG results list
#loading data for enrichment here
#DEG significant results
DEG.sig <- read.csv("RAnalysis/Output/RNA-seq/DEG/DEGSeq2.sig.results.csv")[,-1]
#how many significant DEGs
nrow(DEG.sig)
#add _gene_ to the gene_id name to match other nomenclature
DEG.sig$gene_id <-  gsub("_g","_gene_g",DEG.sig$gene_id) #remove extra characters

#rename columns for FDR and logFC
#colnames(DEG.sig)[2] <- c("logFC") #rename columns
#colnames(DEG.sig)[6] <- c("FDR") #rename columns
#head(DEG.sig)

# #filter significant DEGs for FDR and logFC cutoffs
# sigGenes <- DEG.sig$gene_id[ DEG.sig$FDR < 0.01 & 
#                               !is.na(DEG.sig$FDR) &
#                               abs(DEG.sig$logFC) > 1 ]
# sigGenes <- na.exclude(sigGenes)
# kk <- enrichKEGG(gene = sigGenes, organism = 'pdam')
# head(kk, n=10)

KEGG_vectore_Pver <- as.vector(DEG.sig)
KEGG_pver <- enrichKEGG(gene = DEG.sig, 
                            organism  = 'pdam', # 'pdam' is Pocillopora damicornis
                            pvalueCutoff = 0.05) 
```


# Perform KEGG enrichment with GOSeq package
```{r}
#Perform goseq
KOwall <- goseq(pwf, GOref$gene_id, gene2cat=GOKO.terms, test.cats=c("KEGG"), method="Wallenius", use_genes_without_cat=TRUE)

```

# Replace NAs with "KEGG" and extract KO terms from results
```{r}
#Cluster1
KO.05<-KOwall$category[KOwall$over_represented_pvalue<.05]
KO.05<-data.frame(KO.05)
colnames(KO.05) <- c("category")
KO.05 <- merge(KO.05, KOwall.C1, by="category")
KO.05$ontology <- replace_na(KO.05$ontology, "KEGG")
KO.05 <- filter(KO.05, ontology=="KEGG")
KO.05 <- KO.05[order(KO.05$ontology, KO.05$over_represented_pvalue, -KO.05$numDEInCat),]
KO.05$term <- as.factor(KO.05$term)
nrow(KO.05)

```

# Add KO definitions
```{r}
#Prep definition data
KFS.KO.def <- subset(KFS.KO.orig, select=c("#","KO", "KO definition"))
colnames(KFS.KO.def) <- c("sig","category", "term")
KFS.KO.def <- filter(KFS.KO.def, sig=="*")
KFS.KO.def <- KFS.KO.def[,c(2:3)]
#Merge with KEGG outp
KO.05 <- unique(left_join(C1.KO.05[,-6], KFS.KO.def, by=c("category")))

```

# Write output KEGG enrichment files
```{r}
write.csv(C1.KO.05, file = "RAnalysis/Output/RNA-seq/GOSeq/KO.05.csv")

```
