---
title: "DEG_DMG_pathway_analysis"
author: "daniellembecker"
date: "2023-01-20"
output: html_document
---


# Molecular Underpinnings Chronic Nutrient Enrichment Project

## Analysis of comparisons between gene ontology enrichment from differentially methylated genes and differentially expressed genes.
## Previous steps for gene expression include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DEG statistical data sheet
## Previous steps for DNA methylation include WGBS workflow in Bioinformatics>WGBS>WGBS workflow and Differential Methylated Genes statistical analysis in RAnalysis>Scripts>WGBS_WGBS.rmd to make DMG statistical dataframe

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries

library("tidyverse")
library("RColorBrewer")
library("gridExtra")
library("ComplexHeatmap")
library("dplyr")
library("clusterProfiler")
library("pheatmap")
library("magrittr")
library("rtracklayer")
library("GenomicRanges")
library("GSEABase")
library("GOSim")
library("VennDiagram")
library("stats")
library("GO.db")
library("rrvgo")
library("GOSemSim")
library("AnnotationDbi")
library("janitor")
library(rtracklayer) # on bioconducter
library(dplyr)
library(edgeR)
library(factoextra)
library(cowplot)
library(wesanderson)
library(viridis)
library(adegenet)
library(ggplot2)
library(matrixStats)
library(grid)
library(gridExtra)
library(RColorBrewer)
library(patchwork)
library(stringr)
library(tidyverse)
library(ggpubr)
pal <- brewer.pal(n = 3, name = "PRGn")
```

# Import the data files 
```{r}
#upload go enrichment summary for DEGs
DEG <- read.csv("RAnalysis/Output/RNA-seq/GOSeq/Host/GO.enrichment.summary.table.DEG.host.csv", header = TRUE, row.names = 1)

#rename category to GOterm
colnames(DEG)[3] ="GOterm" #rename columns

#write csv with bh_adjust
write.csv(DEG, file = "RAnalysis/Output/RNA-seq/GOSeq/Host/GO.DEG.parent.terms.host.csv")

#upload go enrichment summary for DMGs
DMG <- read.csv("RAnalysis/Output/WGBS/GO.enrichment.summary.table.ontology.10x.csv", header = TRUE, row.names = 1)

#rename category to GOterm
colnames(DMG)[3] ="GOterm" #rename columns

#add column to specific type of measure
DMG <- DMG %>% add_column(type = "DMG.GO.terms")

#adjust p-values 
DMG$bh_adjust <-  p.adjust(DMG$over_represented_pvalue, method="BH") 

#write csv with bh_adjust
write.csv(DMG, file = "RAnalysis/Output/WGBS/GO.DMG.parent.terms.csv")

```

#make Venn Diagram of overlapping GOslim terms for DMG and DEG comparison

```{r}
venn.plot <- venn.diagram(
  x = list(DEG = DEG$term, DMG = DMG$term),
  filename = 'RAnalysis/Output/meth_express_compare/venn_diagram.png',
  category.names = c('DEG GO terms', 'DMG GO terms'),
  output = TRUE,
  col = "black",
	fill = c("dodgerblue", "goldenrod1"),
  cat.default.pos = "outer",
  cat.fontface = "bold",
  cat.pos = c(-18, 18),
          cat.dist = c(0.055, 0.055)
)
```

#subset data sets to see alike terms
```{r}
#compare these datasets, simply pass them to the comparedf() function
#comparedf(DMG, DEG)

#Use summary() to get a more detailed summary

#summary(comparedf(DEG, DMG, by = "term"))

#these examples show unshared characteristics but don't list the actual shared terms
```

```{r}
#use dplyr inner_join instead

common <- intersect(DMG$term, DEG$term) 

common <- as.data.frame(common)

#found three common terms for DMG and DEG

#leukocyte migration involved in inflammatory response (parent term cell motility, immune system response, response to stress)

#positive regulation of transcription from RNA polymerase II promoter in response to heat stress (DNA-templated transcription; response to stress; nitrogen compound metabolic process)

#fatty acid derivative biosynthetic process (lipid metabolic process)

#main question is why 10x terms have three overlapping but 5x does not
```

#next steps, fill in remaining parent terms for the NAs in these data sets and see the overlap there
#filter datas sets with NAs for their slim terms
```{r}
DMG.NA <- DMG %>%
  filter(is.na(slim_term))

DEG.NA <- DEG %>%
  filter(is.na(slim_term))
```


#Filled in all NAs with parent terms for completed data frames below
#Used QuickGO to search and fill in the remaining NA parent terms https://www.ebi.ac.uk/QuickGO/term/GO:1901363
#DMG go plot
```{r}
#upload complete parent term dataframe DMGs and make figures

#upload complete go enrichment summary for DMGs
DMG.complete <- read.csv("RAnalysis/Data/WGBS/DMG_GO_parent_terms_NAs_completed.10x.csv", header = TRUE)

#adjust p-values 
DMG.complete$bh_adjust <-  p.adjust(DMG.complete$over_represented_pvalue, method="BH") 

GO.plot.DMG.complete <-  ggplot(DMG.complete, aes(x = ontology, y = term)) + 
    geom_point(aes(size=numInCat, fill=bh_adjust, colour=bh_adjust)) + 
    scale_size(name="Number of Genes", range=c(5,12), limits=c(1,550))+
    scale_fill_gradient(name = "Adjusted p-value", low="black", high="gray")+
    scale_colour_gradient(name = "Adjusted p-value", low="black", high="gray")+
    xlab("Ontology Category")+
    facet_grid(slim_term ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 55, multi_line = TRUE))+
    ggtitle("")+
    theme_bw() + 
    theme(panel.border = element_blank(), 
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
          strip.background =element_rect(fill="white"),
          strip.text.y = element_text(angle=0, size = 18, face="bold"),
          strip.text.x = element_text(size = 16),
          axis.text.x = element_text(size = 20, angle=45, hjust=1),
          axis.text.y = element_text(size = 16),
          axis.title.x = element_text(size = 22, face="bold"),
          legend.text=element_text(size=16),
          title=element_text(size=22, face="bold"),
          plot.title = element_text(hjust = 0),
          axis.title.y = element_blank())

 
ggsave("RAnalysis/Output/Final_Figures/DMG.complete.slimterms.pdf", GO.plot.DMG.complete , width = 20, height = 49, units = c("in"))


```

#DEG go plot
```{r}
#upload complete parent term dataframe DMGs and make figures

#upload complete go enrichment summary for DMGs
DEG.complete <- read.csv("RAnalysis/Data/RNA-seq/Host/DEG_GO_parent_terms_NAs.csv", header = TRUE)

#adjust p-values 
DEG.complete$bh_adjust <-  p.adjust(DEG.complete$over_represented_pvalue, method="BH") 

GO.plot.DEG.complete <-  ggplot(DEG.complete, aes(x = ontology, y = term)) + 
    geom_point(aes(size=numInCat, fill=bh_adjust, colour=bh_adjust)) + 
    scale_size(name="Number of Genes", range=c(5,12), limits=c(1,550))+
    scale_fill_gradient(name = "Adjusted p-value", low="black", high="gray")+
    scale_colour_gradient(name = "Adjusted p-value", low="black", high="gray")+
    xlab("Ontology Category")+
    facet_grid(slim_term ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 55, multi_line = TRUE))+
    ggtitle("")+
    theme_bw() + 
    theme(panel.border = element_blank(), 
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
          strip.background =element_rect(fill="white"),
          strip.text.y = element_text(angle=0, size = 18, face="bold"),
          strip.text.x = element_text(size = 16),
          axis.text.x = element_text(size = 20, angle=45, hjust=1),
          axis.text.y = element_text(size = 16),
          axis.title.x = element_text(size = 22, face="bold"),
          legend.text=element_text(size=16),
          title=element_text(size=22, face="bold"),
          plot.title = element_text(hjust = 0),
          axis.title.y = element_blank())

 
ggsave("RAnalysis/Output/Final_Figures/DEG.complete.slimterms.pdf", GO.plot.DEG.complete , width = 25, height = 49, units = c("in"))


```


#Filter dataframes for gene ontology groups that are involved in nitrogen pathways, stress pathways to see if expression and methylation are correlated
```{r}
phenotype.pathways.DEG <- filter(DEG.complete, slim_term %in%  c("nitrogen compound metabolic process", "response to nitrogen compound", "response to nutrient levels", "heat shock protein binding", "response to stress", "cellular nitrogen compound"))


phenotype.pathways.DMG <- filter(DMG.complete, slim_term %in%  c("nitrogen compound metabolic process", "response to nitrogen compound", "response to nutrient levels", "heat shock protein binding", "response to stress", "cellular nitrogen compound"))
```

#change data frames from wide to long format
```{r}
#change data frames from wide to long so each gene has a row 
long.pathway.DEG <- phenotype.pathways.DEG %>% 
    mutate(genes = strsplit(as.character(genes), ",")) %>% #defines what character separates the list
    unnest(genes) #unnests and makes each gene have its own row

#rename column from genes to gene
long.pathway.DEG <- long.pathway.DEG %>% 
       rename("gene_id" = "genes")

#remove type column from data frame to match DMG 
long.pathway.DEG = subset(long.pathway.DEG, select = -c(type) )

#remove extra first space in gene_id names in column 
long.pathway.DEG$gene_id <- gsub(' Pver', 'Pver', long.pathway.DEG$gene_id)

#change data frames from wide to long so each gene has a row
long.pathway.DMG <- phenotype.pathways.DMG %>% 
    mutate(genes = strsplit(as.character(genes), ",")) %>% 
    unnest(genes)

#remove the _gene_ in the DMG gene list to match DEG gene numbers
long.pathway.DMG$genes <-  gsub("_gene_","_", long.pathway.DMG$genes)

#rename column from genes to gene
long.pathway.DMG <- long.pathway.DMG %>% 
       rename("gene_id" = "genes")

#remove extra first space in gene_id names in column 
long.pathway.DMG$gene_id <- gsub(' Pver', 'Pver', long.pathway.DMG$gene_id)

```


#compare gene lists for correlation, DEG, DMG need to bring in statistic data frame
```{r}
#methylation and expression dataframe calculated and complete in meth_GE_compare.Rmd script, found here: https://github.com/hputnam/Becker_E5/blob/master/RAnalysis/Scripts/methyl_expression_compare/meth_GE_compare.Rmd
#this data frame lists only the genes found in meth and express data frames
meth.express <- readRDS("RAnalysis/Data/GE_meth_compare/Multi_geneSummaryComplete.RData")

#remove _gene_ in gene id to match other data frames
meth.express$gene_id <-  gsub("_gene_","_", meth.express$gene_id)

#this data frame lists all methylated genes found 
methyl.annot <- readRDS("RAnalysis/Data/GE_meth_compare/meth_annot_full.RData")

#remove _gene_ in gene id to match other data frames
methyl.annot$gene_id <-  gsub("_gene_","_", methyl.annot$gene_id)

#this data frame lists all expressed genes found
express.annot <- readRDS("RAnalysis/Data/GE_meth_compare/gene_annot_full.RData")

#remove _gene_ in gene id to match other data frames
express.annot$gene_id <-  gsub("_gene_","_", express.annot$gene_id)

#remove gene_ from column 2 to 8
colnames(express.annot)[2:8] <-  sub(".*gene_", "", colnames(express.annot)[2:8])

```


```{r}
#combine DMG and DEG data frames, by binding below so all rows are present
#DMG columns = 12 rows and DEG = 435 rows, total rows should be 447
DMG.DEG.dat <- rbind(long.pathway.DEG, long.pathway.DMG)
#yes rows equal estimate
```

```{r}
#combine all meth express data frame with DMG.DEG data frame
#this will merge only genes found in both DMG and DEG
meth.exp.filt <- merge(meth.express, DMG.DEG.dat, by = 'gene_id')

#this will merge to list all DEG and DMG express and meth statistics
#all DMG rows should stay = 12, only 9
meth.merge <- merge(methyl.annot, long.pathway.DMG)

#all DEG rows should stay = 435, only 410
express.merge <- merge(express.annot, long.pathway.DEG, by = 'gene_id')

#combine express and meth data


```

```{r}
## Mean Methylation vs Gene Expression following Liew et al
df <- meth.exp.filt

df %>% drop_na()

#control log gene expression x % methylation
df$log_gene_C <- log(df$gene_mean_express_C+1)
df$per_meth_C <- df$mean_meth_C * 100

#df$density <- get_density(df$gene_mean_express_C, df$mean_meth_C*100, n = 308)

#correlation for control
cor.C <- cor.test(df$per_meth_C, df$gene_mean_express_C, 
                    method = "pearson")
cor.C

control <- ggscatter(df, x=per_meth_C, y=log_gene_C, cor.coef = TRUE, cor.method = "pearson") + 
    #geom_point() + ylim(0,2) + 
    #geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
    theme_cowplot();control


#enriched log gene expression x % methylation
df$log_gene_E <- log(df$gene_mean_express_E+1)
df$per_meth_E <- df$mean_meth_E * 100

#correlation for control
cor.E <- cor.test(df$per_meth_E, df$gene_mean_express_E, 
                    method = "pearson")
cor.E

enriched <- ggplot(df,aes(x=per_meth_E,y=log_gene_E, cor.coef = TRUE, cor.method = "pearson")) + 
    geom_point() + ylim(0,2) + 
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
    theme_cowplot();enriched


### Figure Mean Methylation vs Gene Expression###

meth_meanexpress <- control + labs(title = "", subtitle = "Control") + enriched + labs(title = "", subtitle = "Enriched")

meth_meanexpress

ggsave(meth_meanexpress, filename="RAnalysis/Output/meth_express_compare/pathway.specific.express.meth.png", height = 7, width = 15)
```


