---
title: "DEG_DMG_pathway_analysis"
author: "daniellembecker"
date: "2023-01-20"
output: html_document
---


# Molecular Underpinnings Chronic Nutrient Enrichment Project

## Analysis of comparisons between gene ontology enrichment from differentially methylated genes and differentially expressed genes.
## Previous steps for gene expression include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DEG statistical data sheet
## Previous steps for DNA methylation include WGBS workflow in Bioinformatics>WGBS>WGBS workflow and Differential Methylated Genes statistical analysis in RAnalysis>Scripts>WGBS_WGBS.rmd to make DMG statistical dataframe

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
library("tidyverse")
library("dplyr")
library("RColorBrewer")
library("ggplot2")
library("gplots")
library("spdep") 
library("patchwork")
library("VennDiagram")
library("arsenal")
library("GSEABase")
```

# Import the data files 

```{r}
#upload go enrichment summary for DEGs
DEG <- read.csv("RAnalysis/Output/RNA-seq/GOSeq/GO.enrichment.summary.table.DEG.csv", header = TRUE, row.names = 1)

#upload go enrichment summary for DMGs
DMG <- read.csv("RAnalysis/Output/WGBS/GO.enrichment.summary.table.ontology.10x.csv", header = TRUE, row.names = 1)

#add column to specific type of measure
DMG <- DMG %>% add_column(type = "DMG.GO.terms")

```

#make Venn Diagram of overlapping GOslim terms for DMG and DEG comparison

```{r}
venn.plot <- venn.diagram(
  x = list(DEG = DEG$term, DMG = DMG$term),
  filename = 'RAnalysis/Output/meth_express_compare/venn_diagram.png',
  category.names = c('DEG GO terms', 'DMG GO terms'),
  col = "black",
	fill = c("dodgerblue", "goldenrod1"),
	cat.col = c("dodgerblue", "goldenrod1"),
)
```

#subset data sets to see alike terms

```{r}
#compare these datasets, simply pass them to the comparedf() function
comparedf(DMG, DEG)

#Use summary() to get a more detailed summary

summary(comparedf(DEG, DMG, by = "term"))

#these examples show unshared characteristics but don't list the actual shared terms
```

```{r}
#use dplyr inner_join instead

common <- intersect(DMG$term, DEG$term) 

common <- as.data.frame(common)

#found three common terms for DMG and DEG

#leukocyte migration involved in inflammatory response (parent term immune system response)

#positive regulation of transcription from RNA polymerase II promoter in response to heat stress (DNA-templated transcription; response to stress)

#fatty acid derivative biosynthetic process (lipid metabolic process)

#main question is why 10x terms have three overlapping but 5x does not
```

#next steps, fill in remaining parent terms for the NAs in these data sets and see the overlap there
#filter datas sets with NAs for their slim terms
```{r}
DMG.NA <- DMG %>%
  filter(is.na(slim_term))

DEG.NA <- DEG %>%
  filter(is.na(slim_term))
```
#use goslim to find slim terms for Go ids
```{r}



```












