---
title: "METH_GE_COMPARE"
author: "Danielle M. Becker"
date: "20220720"
output: html_document
editor_options: 
  chunk_output_type: console
---

#### Gene Expression and DNA Methylation comparison ####
## Used to compare general relationship between DNA methylation and gene expression
## Code developed from Javier A Rodriguez-Casariego paper: https://onlinelibrary.wiley.com/doi/pdf/10.1111/mec.16246?casa_token=DGoNlwKZByAAAAAA:i6JDcylh98txNEMQnr6dUTFxKd1sX7xotbSMZJq-DDOyqZGUPHObc30PhaWOAqpEufHMcJJ_wlQbQvbG
## https://github.com/jarcasariego/ACER_clonal_divergence/blob/main/GE_GbM_association/rmd/GE_DNAm_control.Rmd

```{r load packages}

## load packages
library(rtracklayer) # on bioconducter
library(dplyr)
library(edgeR)
library(factoextra)
library(cowplot)
library(wesanderson)
library(ggpmisc)
library(viridis)
library(adegenet)
library(ggplot2)
library(matrixStats)
library(grid)
library(gridExtra)
library(RColorBrewer)
library(patchwork)
library(stringr)
library(tidyverse)
library(ggpubr)
pal <- brewer.pal(n = 3, name = "PRGn")
```

#### Functions ####
```{r}
# Function for subset 'mat' for the target expression,DNA methylation, and gene attributes modified from @adowneywall
subMat_control <- function(x){
  y <- data.frame(gene_length=log2(x$gene_length),
                  exon=log2(x$exon_count),totalCpG=log2(x$all_count),
                  Methylation=x$mean_Cc,Methylation_CV=x$cv_Cc,
                  Gene_Expression=x$gene_mean_Cc,Gene_Expression_CV=x$gene_cv_Cc)
  return(y)
}
get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}
perCentCpGCoverage <- function(x){sum(c(mat$cov5_count/mat$all_count)*100 > x)}
# Function for generating three panel contribution plot for the different PCs
suppContriFig <- function(x){
  temp <- list()
  for(i in 1:3){
    temp[[i]] <- fviz_contrib(x, choice = "var", axes = i, top = 10,color = "black",fill = "grey",
                              sort.val = "none")
  }
  return(temp)
}
```

Load metadata
```{r}
#### Data ####
# Read in metadata
meta.data <- read.csv("RAnalysis/Data/RNA-seq/P.verrucosa/metadata.RNAseq.csv", header=T, sep=",", na.string="NA", stringsAsFactors = F) #read in info
colnames(meta.data) <- c("colonyID","treatment","block","Sample_ID")
ID <- meta.data$Sample_ID
```

Processing expression data
```{r}
counts <- read.csv("RAnalysis/Data/RNA-seq/P.verrucosa/Host/Poc_gene_count_matrix.csv", header = TRUE, row.names = 1) #read in genes count matrix
colnames(counts) <- str_sub(colnames(counts), 1, 3) #remove everything after the first _ to just have the sample names
# Order columns by sample name
colnames(counts) <- c("C17","C18","C19","C20","C21","C22","C23","C24","C25","C26","C27","C28","C29","C30","C31","C32","E10","E11","E12","E13","E14","E15","E16","E1","E2","E3", "E4", "E5", "E6", "E7", "E8", "E9")
counts <- counts[, colnames(counts) %in% ID] #intersect with ID to maintain only samples a
counts <- counts[, order(colnames(counts))]
dge_gene <- DGEList(counts,genes = rownames(counts))
test <- as.data.frame(dge_gene)
saveRDS(dge_gene, "RAnalysis/Data/GE_meth_compare/RNA_gene_preNormalization_DGEListObj.Rdata")
```


```{r}
# Methylation Summary Table
# load gene reference file - this contains the chromosome and starting gene coordinate (used to uniquely identify genes in the CpG file) and the gene_id which will be used to merge the expression and CpG data. I created this before and saved it as gene_GeneLoc.RData
#ref <- readRDS("../WGBS/data/gene_GeneLoc.RData")

#SAMPLES E2, E16, AND C19 WERE EXCLUDED DUE TO DEDUPLICATION AND LOW COUNTS
# Read in gene methylation data
gene_dnam <- readRDS("RAnalysis/Output/WGBS/sample_all_methylated_data_5x.Rdata") #read in info


#rename sample.ID to include corresponding E and C in front for enriched and control for later analysis
gene_dnam <- gene_dnam %>%
    mutate(Sample.ID = recode(Sample.ID, '1' = 'E1', '3' = 'E3', '4' = 'E4', '5' = 'E5', '6' = 'E6', '7' = 'E7', '8' = 'E8', '9' = 'E9','10' = 'E10', '11' = 'E11', '12' = 'E12','13' = 'E13', '14' = 'E14', '15' = 'E15', '16' = 'E16', '17' = 'C17', '18' = 'C18', '20' = 'C20', '21' = 'C21', '22' = 'C22', '23' = 'C23', '24' = 'C24', '25' = 'C25', '26' = 'C26', '27' = 'C27', '28' = 'C28', '29' = 'C29', '30' = 'C30', '31' = 'C31', '32' = 'C32'))


#check to see all where renamed correctly
unique(gene_dnam$Sample.ID)

#change from long to wide format
meth <- data.frame (gene_dnam %>%
                   group_by(Sample.ID, gene) %>%
                   summarise(mean = mean(per.meth), n = n()) %>%
                   spread(., key = Sample.ID, value = mean))

#filter for a min 3 Cpg per gene
meth <- meth[which(meth$n > 2),]
cpg10xgene <- meth[,1:2]  
colnames(cpg10xgene) <- c("gene_id", "cov10_count")
rownames(meth) <- meth$gene
meth <- meth[,-1]
meth <- as.matrix(meth)
meth <- meth/100
saveRDS(meth, "RAnalysis/Data/GE_meth_compare/meth_dataset.RData")

meth.test <- data_frame(meth)

## Create a summary of methylation for each gene ##
mean_meth <- rowMeans(meth)

meta.data <- na.omit(meta.data)

# create list for sample IDs from metadata and select from that list with filtering, pipe, row means
list_E <- filter(meta.data, treatment == "enriched") #filter only enriched data from metadata

list_C <- filter(meta.data, treatment == "control") #filter only control data from metadata

list_E <- as.list(list_E$Sample_ID) #make list of just enriched IDs

list_C <- as.list(list_C$Sample_ID) #make list of just control IDs
  

# mean gene meth enriched 
subset_meth_E <- meth[, colnames(meth)%in%list_E] #subset data from main m matrix for just enriched data

#test <- subset_meth_E %>% summarise_at(vars(E1:E9), mean, na.rm = TRUE)

mean_meth_E <- rowMeans(subset_meth_E)

# Mean gene meth control
subset_meth_C <- meth[, colnames(meth)%in%list_C] #subset data from main m matrix for just control data

mean_meth_C <- rowMeans(subset_meth_C)

# Coefficient of Variation for enriched samples
sd_meth_E <- rowSds(subset_meth_E)
cv_meth_E <- sd_meth_E/mean_meth_E
 
# Coefficient of Variation for control
sd_meth_C <- rowSds(subset_meth_C)
cv_meth_C <- sd_meth_C/mean_meth_C


# Coefficient of Variation among treatments
meanAmongTrt_meth <- cbind(mean_meth_E,mean_meth_C)
sd_meanAmongTrt_meth <- rowSds(meanAmongTrt_meth)
mean_meanAmongTrt_meth <- rowMeans(meanAmongTrt_meth)
cv_mean_AmongTrt_meth <- sd_meanAmongTrt_meth/mean_meanAmongTrt_meth

# Difference in gene methylation between treatments
diff_Trt_meth <- mean_meth_E-mean_meth_C

# Creat matrix of all CpG summary stats
meth_summary <- cbind(mean_meth,mean_meth_E,mean_meth_C,cv_meth_E,cv_meth_C,cv_mean_AmongTrt_meth,
                      diff_Trt_meth) # bind all data frames with summary stats
meth_summary[is.na(meth_summary)]<-0 #filter for all methylation data over 0
meth_summary <- data.frame(gene_id = rownames(meth_summary),meth_summary) #change gene ID column name
meth_annot <- left_join(cpg10xgene, meth_summary) #join Cpg and meth summary

# Saving full meth_annot data.frame
saveRDS(meth_annot,"RAnalysis/Data/GE_meth_compare/meth_annot_full.RData")

#meth_annot$label <- paste0(meth_annot$chr,"_",meth_annot$start,"_",meth_annot$stop)
#meth_reduce <- data.frame(label=meth_annot$label,meth_annot[,1:9])


#### Summarize the gene expression data ####
expression <- dge_gene$counts
subset_expression <- expression[, colnames(expression)%in%list_E] 
subset_expression <- subset_expression[rowSums(subset_expression[])>0,]
expression <- expression[rownames(expression) %in% rownames(subset_expression),]
expression_cpm_log <- log2(cpm(expression)+ 1) # log2 cpm transformation
expression <- expression_cpm_log

# Mean gene expression all samples
mean_express <- rowMeans(expression)

# mean gene expression enriched 
subset_express_E <- expression[, colnames(expression)%in%list_E] #subset data from main m matrix for just enriched data

mean_express_E <- rowMeans(subset_express_E) #take mean per gene ID

# Mean gene expression control
subset_express_C <- expression[, colnames(expression)%in%list_C] #subset data from main m matrix for just control data

mean_express_C <- rowMeans(subset_express_C) #take mean per gene ID

# Coefficient of Variation for control gene expression
sd_express_C <- rowSds(subset_express_C)
cv_express_C <- sd_express_C/mean_express_C

# Coefficient of Variation for enriched gene expression
sd_express_E <- rowSds(subset_express_E)
cv_express_E <- sd_express_E/mean_express_E

# Coefficient of Variation among treatments
meanAmongTrt_express <- cbind(mean_express_E,mean_express_C)
sd_meanAmongTrt_express <- rowSds(meanAmongTrt_express)
mean_meanAmongTrt_express <- rowMeans(meanAmongTrt_express)
cv_mean_AmongTrt_express <- sd_meanAmongTrt_express/mean_meanAmongTrt_express

# Difference in gene expression between Trt
diff_Trt_express <- mean_express_E-mean_express_C

# Create matrix of all gene expression summary stats
gene_summary <- cbind(mean_express,mean_express_E,mean_express_C,cv_express_E,cv_express_C,cv_mean_AmongTrt_express,
                      diff_Trt_express)
gene_summary[is.na(gene_summary)]<-0
colnames(gene_summary) <- paste0("gene_",colnames(gene_summary))
gene_summary <- data.frame(gene_id = rownames(gene_summary),gene_summary)

# transcriptomic data
transcriptomic <- read.csv("RAnalysis/Genome/GFF3.csv") #load annotation GFF3 file

#remove first three columns in data frame
transcriptomic <- transcriptomic[,-(1:3),drop=FALSE]
transcriptomic <- transcriptomic[,-(4:6),drop=FALSE]

#add gene between Pver_g to match gene_summary
transcriptomic$gene_id <-  gsub("Pver_g","Pver_gene_g", transcriptomic$gene_id)

gene_annot <- left_join(gene_summary, transcriptomic)

gene_annot <- gene_annot[,-(9:13),drop=FALSE]

# Saving full gene_annot data.frame
saveRDS(gene_annot,"RAnalysis/Data/GE_meth_compare/gene_annot_full.RData")

# Combine gene expression, exon count and CpG (DNA mehtylation data)
gSum <- left_join(meth_annot, gene_annot)

# Reducing the data.frame to include places were we have all the data.
gSum_red<-gSum[!is.na(gSum$cov10_count),]
gSum_red<-gSum_red[!is.na(gSum_red$gene_mean_express),]

# Saving reduced data.frame
saveRDS(gSum,"RAnalysis/Data/GE_meth_compare/Multi_geneSummaryComplete.RData")

# Saving the reduced data.frame
saveRDS(gSum_red,"RAnalysis/Data/GE_meth_compare/Multi_geneSummaryReduced.RData")
```

#compare global mean meth vs express overall
```{r}
mat <- readRDS("RAnalysis/Data/GE_meth_compare/Multi_geneSummaryReduced.RData")

## Mean Methylation vs Gene Expression following Liew et al
df <- mat

#log gene expression x % methylation
df$log_gene <- log(df$gene_mean_express+1)
df$per_meth <- df$mean_meth * 100

#plot of global express x meth
meth.express.compare <- ggplot(df, aes(x=per_meth, y=log_gene,)) + 
    geom_point() + ylim(0,2) + 
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
  stat_cor(method = "kendall", digits = 3, label.x = 60, label.y = 0.5 ) +
    theme_cowplot();meth.express.compare

### Figure Mean Methylation vs Gene Expression###
ggsave(meth.express.compare, filename="RAnalysis/Output/Final_Figures/Host/global_express_meth_correlation.png", height = 7, width = 10)

```


#compare global mean meth vs express for control and enriched
```{r}

#control log gene expression x % methylation
df$log_gene_C <- log(df$gene_mean_express_C+1)
df$per_meth_C <- df$mean_meth_C * 100

df$density <- get_density(df$gene_mean_express_C, df$mean_meth_C*100, n = 100)

#correlation for control
cor.C <- cor.test(df$per_meth_C, df$gene_mean_express_C, 
                    method = "kendall")
cor.C

control <- ggplot(df,aes(x=per_meth_C,y=log_gene_C)) + 
    geom_point() + ylim(0,2) + 
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
  stat_cor(method = "kendall", digits = 3, label.x = 60, label.y = 0.5 ) +
    theme_cowplot();control


#enriched log gene expression x % methylation
df$log_gene_E <- log(df$gene_mean_express_E+1)
df$per_meth_E <- df$mean_meth_E * 100

#correlation for control
cor.E <- cor.test(df$per_meth_E, df$gene_mean_express_E, 
                    method = "kendall")
cor.E

enriched <- ggplot(df,aes(x=per_meth_E,y=log_gene_E,)) + 
    geom_point() + ylim(0,2) + 
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
  stat_cor(method = "kendall", digits = 3, label.x = 60, label.y = 0.5 ) +
    theme_cowplot();enriched


### Figure Mean Methylation vs Gene Expression###

meth_meanexpress <- control + labs(title = "", subtitle = "Control") + enriched + labs(title = "", subtitle = "Enriched")

meth_meanexpress

ggsave(meth_meanexpress, filename="RAnalysis/Output/Final_Figures/Host/treatments_ME_meth_correlation.png", height = 7, width = 15)

```

#Multiple regression for mean expression and %meth for both treatments
```{r}
#make multiple regression dataframe
#first have to change data frames from long to wide, need to separate per.meth in one column and log_gene in one column and treatments into another
per.meth.C <- df %>% select(gene_id, per_meth_C) 
per.meth.C$treatment <- "control" #make column with treatment
names(per.meth.C)[2] ="per_meth" #rename per.meth.c to per.meth 

per.meth.E <- df %>% select(gene_id, per_meth_E) 
per.meth.E$treatment <- "enriched"
names(per.meth.E)[2] ="per_meth"

log.gene.C <- df %>% select(gene_id, log_gene_C) 
log.gene.C$treatment <- "control"
names(log.gene.C)[2] ="log_gene"

log.gene.E <- df %>% select(gene_id, log_gene_E) 
log.gene.E$treatment <- "enriched"
names(log.gene.E)[2] ="log_gene"

#rbind enriched and control data frames for per.meth and log.gene 
per.meth <- rbind(per.meth.C, per.meth.E)
log.gene <- rbind(log.gene.C, log.gene.E)

#combine treatment separated and per.meth and log.gene dataframes for ancova 
mult.reg <- left_join(per.meth, log.gene)

```

#Make multiple regression and analysis

```{r}
#make multiple regression plot with two lies for treatments
mult.reg <- ggplot(mult.reg, aes(x=per_meth, y=log_gene, color = treatment)) + 
      geom_point(color = "black") + geom_smooth(method="lm",se = FALSE) + 
      labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
      stat_poly_line() +
      stat_poly_eq(use_label(c("eq", "R2")), label.x = "right", label.y = "bottom", rr.digits = 3, size = 5) +
      theme_cowplot();mult.reg

ggsave(mult.reg, filename="RAnalysis/Output/Final_Figures/Host/mult.reg.correlation.png", height = 10, width = 15)
ggsave(mult.reg, filename="RAnalysis/Output/Final_Figures/Host/mult.reg.correlation.pdf", height = 10, width = 15)

#use cocor to compare correlations
require(cocor) # load package
cocor.indep.groups(r1.jk=+0.021, r2.hm=+0.022, n1=30, n2=30, alternative="two.sided", alpha=0.05, conf.level=0.95, null.value=0)

```

Results of a comparison of two correlations based on independent groups

Comparison between r1.jk = 0.021 and r2.hm = 0.022
Difference: r1.jk - r2.hm = -0.001
Group sizes: n1 = 30, n2 = 30
Null hypothesis: r1.jk is equal to r2.hm
Alternative hypothesis: r1.jk is not equal to r2.hm (two-sided)
Alpha: 0.05

fisher1925: Fisher's z (1925)
  z = -0.0037, p-value = 0.9971
  Null hypothesis retained

zou2007: Zou's (2007) confidence interval
  95% confidence interval for r1.jk - r2.hm: -0.5102 0.5084
  Null hypothesis retained (Interval includes 0)




```{r}
## Mean Methylation vs Gene Expression following Javi et al

#df$density <- get_density(df$gene_mean_express, df$mean_meth*100, n = 100)
mean_GbMvsGE <- ggplot(df,aes(x=gene_mean_express,y=mean_meth*100)) + 
    geom_point() + ylim(0,100) + 
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="Gene Expression (log2-cpm)",y="DNA Methylation (%)") +
  stat_cor(method = "kendall", digits = 3, label.x = 10, label.y = 90 ) +
    theme_cowplot();mean_GbMvsGE

df$density <- get_density(df$gene_mean_express_E, df$mean_meth_E*100, n = 100)
mean_GbMvsGE_enriched <- ggplot(df,aes(x=gene_mean_express_E,y=mean_meth_E*100)) + 
    geom_point(aes(color = density)) + ylim(0,100) + 
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="Gene Expression (log2-cpm)",y="DNA Methylation (%)") +
  stat_cor(method = "kendall", digits = 3, label.x = 10, label.y = 90 ) +
    theme_cowplot();mean_GbMvsGE_enriched

df$density <- get_density(df$gene_mean_express_C, df$mean_meth_C*100, n = 100)
mean_GbMvsGE_control <- ggplot(df,aes(x=gene_mean_express_C, y=mean_meth_C*100)) + 
    geom_point(aes(color = density)) + ylim(0,100) + 
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="Gene Expression (log2-cpm)",y="DNA Methylation (%)") +
  stat_cor(method = "kendall", digits = 3, label.x = 10, label.y = 90 ) +
    theme_cowplot();mean_GbMvsGE_control

## Mean Methylation vs Gene Expression CV (log)
#df$density <- get_density(log(df$gene_cv_express_E), df$mean_meth_E*100, n = 100)
meanGbMvsGEcv <- ggplot(df,aes(log(df$gene_cv_express_E),y=mean_meth*100)) + 
    geom_point() + ylim(0,100) + scale_color_viridis() +
    geom_smooth(method=lm,colour="orange",size=2) +
    labs(x=bquote("Gene Expression"~CV[Ind]~"(log)"),y="DNA Methylation (%)") +
  stat_cor(method = "kendall", digits = 3, label.x = -0.5, label.y = 90 ) +
    theme_cowplot();meanGbMvsGEcv

df$density <- get_density(log(df$gene_cv_express_E+1), df$mean_meth_E*100, n = 100)
meanGbMvsGEcv_enriched <- ggplot(df,aes(log(x=gene_cv_express_E),y=mean_meth_E*100)) + 
    geom_point(aes(color = density)) + ylim(0,100) + scale_color_viridis() +
    geom_smooth(method=lm,colour="orange",size=2) +
    labs(x=bquote("Gene Expression"~CV[Ind]~"(log)"),y="DNA Methylation (%)") +
  stat_cor(method = "kendall", digits = 3, label.x = -0.5, label.y = 90 ) +
    theme_cowplot();meanGbMvsGEcv_enriched

df$density <- get_density(log(df$gene_cv_express_C), df$mean_meth_C*100, n = 100)
meanGbMvsGEcv_control <- ggplot(df,aes(log(x=gene_cv_express_C+1),y=mean_meth_C*100)) + 
    geom_point(aes(color = density)) + ylim(0,100) + scale_color_viridis() +
    geom_smooth(method=lm,colour="orange",size=2) +
    labs(x=bquote("Gene Expression"~CV[Ind]~"(log)"),y="DNA Methylation (%)") +
  stat_cor(method = "kendall", digits = 3, label.x = 1, label.y = 90 ) +
    theme_cowplot();meanGbMvsGEcv_control

enriched_plots <- plot_grid(mean_GbMvsGE_enriched, meanGbMvsGEcv_enriched, labels=c("A. Enriched", "B. Enriched"), ncol=1)
control_plots <- plot_grid(mean_GbMvsGE_control, meanGbMvsGEcv_control, labels=c("C. Control", "D. Control"), ncol=1)

### Figure Mean Methylation vs Gene Expression###
fig_GE_meth <- plot_grid(enriched_plots, control_plots)
ggsave(fig_GE_meth, filename="RAnalysis/Output/Final_Figures/Host/GbM_GE_correlation.png", height = 10, width = 15)

#save global cpm correlation
ggsave(mean_GbMvsGE, filename="RAnalysis/Output/Final_Figures/Host/global_cpm_correlation.png", height = 10, width = 15)

#save global cv correlation
ggsave(meanGbMvsGEcv, filename="RAnalysis/Output/Final_Figures/Host/global_cv_correlation.png", height = 10, width = 15)
```

#make correlation plots comparing enriched and control for cpm 
```{r}
#make multiple regression dataframe
#first have to change data frames from long to wide, need to separate per.meth in one column and log_gene in one column and treatments into another
gene.mean.express.E <- df %>% select(gene_id, gene_mean_express_E) 
gene.mean.express.E$treatment <- "enriched" #make column with treatment
names(gene.mean.express.E)[2] ="mean_express" #rename per.meth.c to per.meth 

mean.meth.E <- df %>% select(gene_id, mean_meth_E) 
mean.meth.E$treatment <- "enriched"
names(mean.meth.E)[2] ="mean_meth"

gene.mean.express.C <- df %>% select(gene_id, gene_mean_express_C) 
gene.mean.express.C$treatment <- "control"
names(gene.mean.express.C)[2] ="mean_express"

mean.meth.C <- df %>% select(gene_id, mean_meth_C) 
mean.meth.C$treatment <- "control"
names(mean.meth.C)[2] ="mean_meth"

#rbind enriched and control data frames for per.meth and log.gene 
mean.meth <- rbind(mean.meth.C, mean.meth.E)
cpm.gene <- rbind(gene.mean.express.E, gene.mean.express.C)

#combine treatment separated and per.meth and log.gene dataframes for ancova 
mult.reg.cpm <- left_join(mean.meth, cpm.gene)

```

#Make multiple regression and analysis
```{r}

#make multiple regression plot with two lies for treatments
cpm.cor <- ggplot(mult.reg.cpm, aes(x=mean_express, y=mean_meth*100, color = treatment)) + 
      geom_point(color = "black") + geom_smooth(method="lm",se = FALSE) + 
      labs(x="Gene Expression (log2-cpm)",y="DNA Methylation (%)") +
      stat_poly_line() +
      stat_poly_eq(use_label(c("eq", "R2")), label.x = "right", label.y = "top", rr.digits = 3, size = 5) +
      theme_cowplot();cpm.cor

ggsave(mult.reg, filename="RAnalysis/Output/Final_Figures/Host/treatment.cpm.correlation.png", height = 10, width = 15)
ggsave(mult.reg, filename="RAnalysis/Output/Final_Figures/Host/treatment.cpm.correlation.pdf", height = 10, width = 15)

#use cocor to compare correlations
require(cocor) # load package
cocor.indep.groups(r1.jk=+0.030, r2.hm=+0.028, n1=30, n2=30, alternative="two.sided", alpha=0.05, conf.level=0.95, null.value=0)

```
  Results of a comparison of two correlations based on independent groups

Comparison between r1.jk = 0.03 and r2.hm = 0.028
Difference: r1.jk - r2.hm = 0.002
Group sizes: n1 = 30, n2 = 30
Null hypothesis: r1.jk is equal to r2.hm
Alternative hypothesis: r1.jk is not equal to r2.hm (two-sided)
Alpha: 0.05

fisher1925: Fisher's z (1925)
  z = 0.0074, p-value = 0.9941
  Null hypothesis retained


#make correlation plots comparing enriched and control for cv
```{r}
#make multiple regression dataframe
#first have to change data frames from long to wide, need to separate per.meth in one column and log_gene in one column and treatments into another
gene.cv.express.E <- df %>% select(gene_id, gene_cv_express_E) 
gene.cv.express.E$treatment <- "enriched" #make column with treatment
names(gene.cv.express.E)[2] ="cv_express" #rename per.meth.c to per.meth 

mean.meth.E <- df %>% select(gene_id, mean_meth_E) 
mean.meth.E$treatment <- "enriched"
names(mean.meth.E)[2] ="mean_meth"

gene.cv.express.C <- df %>% select(gene_id, gene_cv_express_C) 
gene.cv.express.C$treatment <- "control"
names(gene.cv.express.C)[2] ="cv_express"

mean.meth.C <- df %>% select(gene_id, mean_meth_C) 
mean.meth.C$treatment <- "control"
names(mean.meth.C)[2] ="mean_meth"

#rbind enriched and control data frames for per.meth and log.gene 
mean.meth.cv <- rbind(mean.meth.C, mean.meth.E)
cv.gene <- rbind(gene.cv.express.E, gene.cv.express.C)

#combine treatment separated and per.meth and log.gene dataframes for ancova 
mult.reg.cv <- left_join(mean.meth.cv, cv.gene)

```

#Make multiple regression and analysis cv
```{r}

#make multiple regression plot with two lies for treatments
cv.cor <- ggplot(mult.reg.cv, aes(log(x=cv_express), y=mean_meth*100, color = treatment)) + 
      geom_point(color = "black") + geom_smooth(method="lm",se = FALSE) + 
      labs(x=bquote("Gene Expression"~CV[Ind]~"(log)"),y="DNA Methylation (%)") +
      #xlim(-5,1.8) +
      stat_poly_line() +
      #scale_x_continuous(breaks =seq (-4.5, 1.5, by = 0.5)) +
      stat_poly_eq(use_label(c("eq", "R2")), label.x = "right", label.y = "top", rr.digits = 3, size = 5) +
      theme_cowplot();cv.cor

ggsave(cv.cor, filename="RAnalysis/Output/Final_Figures/Host/treatment.cv.correlation.png", height = 10, width = 15)
ggsave(cv.cor, filename="RAnalysis/Output/Final_Figures/Host/treatment.cv.correlation.pdf", height = 10, width = 15)

#use cocor to compare correlations
require(cocor) # load package
cocor.indep.groups(r1.jk=+0.037, r2.hm=+0.029, n1=30, n2=30, alternative="two.sided", alpha=0.05, conf.level=0.95, null.value=0)

```

Results of a comparison of two correlations based on independent groups

Comparison between r1.jk = 0.037 and r2.hm = 0.029
Difference: r1.jk - r2.hm = 0.008
Group sizes: n1 = 30, n2 = 30
Null hypothesis: r1.jk is equal to r2.hm
Alternative hypothesis: r1.jk is not equal to r2.hm (two-sided)
Alpha: 0.05

fisher1925: Fisher's z (1925)
  z = 0.0294, p-value = 0.9765
  Null hypothesis retained

zou2007: Zou's (2007) confidence interval
  95% confidence interval for r1.jk - r2.hm: -0.5018 0.5163
  Null hypothesis retained (Interval includes 0)











Differential methylation/gene expression correlation

```{r}

#Subset DM/GE dataset
mat_red <- subset(mat,select=c("gene_id","diff_Trt_meth","gene_diff_Trt_express","gene_cv_express_E","gene_cv_express_C"))
mat_red[,2] <- mat_red[,2]*100 

#Create a diff_cv variable for all contrasts 
mat_red$diff.cv <- c(mat_red$gene_cv_express_E-mat_red$gene_cv_express_C)
mat_red$methylDirection <- ifelse(mat_red$diff_Trt < 0,
                                      "Hypomethylation",
                                      "Hypermethylation")
```

```{r Plot GE vs meth diff}
#### Gene Expression vs. DMG change ####
p1<-ggplot(mat_red,
           aes(x=diff_Trt_meth,y=gene_diff_Trt_express,colour=methylDirection)) + 
  geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
  geom_point(size=3) + ylim(-2.5,2.5) + xlim(-11,11) +  
  theme_cowplot() +
  labs(y=bquote("Gene expression (FoldChange"~Log[2]~")"),x="Methylation change in DMGs (fraction difference)") +
  theme(plot.title = element_text(hjust=0.5), legend.position = "none")
p1
p2<-ggplot(mat_red,
           aes(x=diff_Trt_meth,y=diff.cv,colour=methylDirection)) + 
  geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
  geom_point(size=3) + ylim(-4,4) + xlim(-11,11) +  
  theme_cowplot() +
  labs(y=bquote("Gene Expression"~CV[ind]~"difference"),x="Methylation change in DMGs (fraction difference)") +
  theme(plot.title = element_text(hjust=0.5), legend.position = "none")
p2
fig <- plot_grid(p1, p2, labels = c("A", "B"))
ggsave(fig,filename="RAnalysis/Output/Final_Figures/Host/DMG_GE_correlation.png", height = 5, width = 10)
```

```{r}
## Statistical Model ##
# DMG Methylation diff vs. Mean diff Gene Expression
(dmg_ge <- summary(lm(mat_red$gene_diff_Trt_express~mat_red$diff_Trt_meth)))

# DMG Methylation diff vs. CV diff Gene Expression
(dmg_cv <- summary(lm(mat_red$diff.cv~mat_red$diff_Trt_meth)))
plot(lm(mat_red$gene_diff_Trt_express~mat_red$diff_Trt_meth))
plot(lm(mat_red$diff.cv~mat_red$diff_Trt_meth))
```

