---
title: "GO_Term_Enrichment_Analysis"
author: "daniellembecker"
edited by: "daniellembecker"
date: "9/20/2021"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Gene Ontology (GO) Enrichment Analysis 
## Previous steps include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DEG statistical data sheet
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
#if (!requireNamespace("DESeq2", quietly = TRUE)) { BiocManager::install("DESeq2") }
library("DESeq2")
library("tidyr")
library("flextable")
library("officer")
library("tidyverse")
library("dplyr")
library("pheatmap")
#if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")
library(rrvgo)
library("RColorBrewer")
if (!requireNamespace("genefilter", quietly = TRUE)) { BiocManager::install("genefilter") }
library("genefilter")
library("ggplot2")
library("gplots")
#if (!requireNamespace("limma", quietly = TRUE)) { BiocManager::install("limma") }
library("limma")
library("spdep") 
library("patchwork")
library("adegenet") 
#if (!requireNamespace("goseq", quietly = TRUE)) { BiocManager::install("goseq") }
library("goseq")
library("gridExtra")
#if (!requireNamespace("clusterProfiler", quietly = TRUE)) { BiocManager::install("clusterProfiler") }
library("clusterProfiler")
library("DataCombine")
library("VennDiagram")
#if (!requireNamespace("GSEABase", quietly = TRUE)) { BiocManager::install("GSEABase") }
library("GSEABase")
library("data.table")
#if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")
library("rrvgo")
#if ("org.Ce.eg.db" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("org.Ce.eg.db")
library(org.Ce.eg.db)
#if ("topGO" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("topGO")
library(topGO)
#if ("Rgraphviz" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("Rgraphviz")
library(Rgraphviz)
#if ("ggdendro" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("ddgendro")
library(ggdendro) # For extracting dendrogram data

```

# Import the data files 
```{r, echo=FALSE}
# Load go terms from functional annotation file
annot_GO <- read.csv("RAnalysis/Output/annot_GO.terms.host.csv")[-1]

# Remove rows where 'GO.ID' is 'unknown'
annot_GO <- subset(annot_GO, GO.ID != 'unknown')

# edit annot_GO to list GO terms all in one row per gene
annot_GO <- annot_GO %>%
  group_by(gene_id) %>%
  summarise(GO.ID = paste(GO.ID, collapse = ";")) %>%
  ungroup()

```

# Load in significant DEGs and gene count data
```{r}
#DEG results
DEG.res <- read.csv("RAnalysis/Output/DEGSeq2.sig.results.host.csv")[-1]

# Load filtered gene count file
gcount_filt <- read.csv("RAnalysis/Output/gene_count_matrix_filtered.csv")

# Rename first column X to gene_id
names(gcount_filt)[names(gcount_filt) == "X"] <- "gene_id"

```

Match up genes in gene list file to annotation file
```{r}
names(annot_GO)

gcount_filt2annot = match(gcount_filt$gene_id, annot_GO$gene_id) #match genes in gcount_filt to annot_GO

# The following is the number of probes without annotation 
sum(is.na(gcount_filt2annot))

row_nas<-which(is.na(gcount_filt2annot))

#view the genes that do not have a match in the annotation file
missing<-as.data.frame(gcount_filt2annot[row_nas])
#print(missing)
```
1,663 genes in gcount_filt do not have GO annotations

Reduce annotation file to only contain genes detected in our dataset.  
```{r}
filtered_Pverr.annot <- annot_GO[annot_GO$gene_id %in% gcount_filt$gene_id, ]
dim(filtered_Pverr.annot)
```
The annotation file now only contains genes that were detected in our dataset that have annotation information. This is 18158 out of the 27439 genes in our dataset. 

Biological Processes topGO

Use topGO for enrichment analysis on upregulated data
```{r}
### Generate vector with names in just the contrast we are analyzing
ID.vector.up <- DEG.res %>%
      filter(direction=="Upregulated") %>%
      pull(gene_id)
    
# Get a list of GO Terms for all genes detected in our dataset
GO.terms <- filtered_Pverr.annot %>%
  dplyr::select(gene_id, GO.ID) %>%
  separate_rows(GO.ID, sep = ";")

# Create a gene-to-GO mapping list
geneID2GO <- split(as.character(GO.terms$GO.ID), GO.terms$gene_id)
geneNames <- unique(GO.terms$gene_id)

# Create a named vector for allGenes
allGenes.up <- factor(as.integer(geneNames %in% ID.vector.up))
names(allGenes.up) <- geneNames

# Create a topGOdata object using the custom gene2GO mapping
GOdata.up_BP <- new("topGOdata",
              ontology = "BP",
              allGenes = allGenes.up,
              nodeSize = 10,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.up_BP

# Tells you how many signifcant genes there was in the analysis 
numSigGenes(GOdata.up_BP) #137
```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  BP 

 18158 available genes (all genes from the array):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 158  significant genes. 

 15385 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 137  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 8152 
   - number of edges = 17684 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled upregulated data
```{r}
# use default algorithm for fisher test 
fisher.up_BP <- runTest(GOdata.up_BP, algorithm = "weight", statistic = "fisher")

# Generate a results table
allRes_up_BP <- GenTable(GOdata.up_BP, 
                      classicFisher = fisher.up_BP, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.up_BP)))

# Convert character columns to numeric
allRes_up_BP$classicFisher <- as.numeric(allRes_up_BP$classicFisher)

# Apply p value adjustment
sig_Res_up_BP <- allRes_up_BP %>% filter(classicFisher < 0.01)



```

Create node figrues of top 5-10 sig terms for upregulated genes
```{r}
# Plot significant nodes
showSigOfNodes(GOdata.up_BP, score(fisher.up_BP), firstSigNodes = 5, useInfo = 'all')

# Print graph with specific parameters and save to directory
printGraph(GOdata.up_BP, fisher.up_BP, firstSigNodes = 5, fn.prefix = "RAnalysis/Output/tGO_graph_upregulated_BP", useInfo = "def", pdfSW = TRUE)

```

Use topGO for enrichment analysis on downregulated data
```{r}
### Generate vector with names in just the contrast we are analyzing
ID.vector.down <- DEG.res %>%
      filter(direction=="Downregulated") %>%
      pull(gene_id)
    
# Create a named vector for allGenes
allGenes.down <- factor(as.integer(geneNames %in% ID.vector.down))
names(allGenes.down) <- geneNames

# Create a topGOdata object using the custom gene2GO mapping
GOdata.down_BP <- new("topGOdata",
              ontology = "BP",
              allGenes = allGenes.down,
              nodeSize = 10,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.down_BP

# Tells you how many signifcant genes there was in the analysis 
numSigGenes(GOdata.down_BP) # 58
```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  BP 

 18158 available genes (all genes from the array):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 74  significant genes. 

 15385 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 58  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 8152 
   - number of edges = 17684 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled downregulated data
```{r}
# use default algorithm for fisher test 
fisher.down_BP <- runTest(GOdata.down_BP, algorithm = "weight", statistic = "fisher")

# Generate a results table
allRes_down_BP <- GenTable(GOdata.down_BP, 
                      classicFisher = fisher.down_BP, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.down_BP)))

# Convert character columns to numeric
allRes_down_BP$classicFisher <- as.numeric(allRes_down_BP$classicFisher)

# Apply p value adjustment
sig_Res_down_BP <- allRes_down_BP %>% filter(classicFisher < 0.01)

```

Create node figrues of top 5-10 sig terms for downregulated genes
```{r}
# Plot significant nodes
showSigOfNodes(GOdata.down_BP, score(fisher.down_BP), firstSigNodes = 5, useInfo = 'all')

# Print graph with specific parameters and save to directory
printGraph(GOdata.down_BP, fisher.down_BP, firstSigNodes = 5, fn.prefix = "RAnalysis/Output/tGO_graph_downregulated_BP", useInfo = "def", pdfSW = TRUE)
```




Molecular Functions topGO

Use topGO for enrichment analysis on upregulated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.up_MF <- new("topGOdata",
              ontology = "MF",
              allGenes = allGenes.up,
              nodeSize = 10,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.up_MF

# Tells you how many signifcant genes there was in the analysis 
numSigGenes(GOdata.up_MF) #133
```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  MF 

 18158 available genes (all genes from the array):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 158  significant genes. 

 16359 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 133  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 1564 
   - number of edges = 2054 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled upregulated data
```{r}
# use default algorithm for fisher test 
fisher.up_MF <- runTest(GOdata.up_MF, algorithm = "classic", statistic = "fisher")

# Generate a results table
allRes_up_MF <- GenTable(GOdata.up_MF, 
                      classicFisher = fisher.up_MF, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.up_MF)))

# Convert character columns to numeric
allRes_up_MF$classicFisher <- as.numeric(allRes_up_MF$classicFisher)

# Apply p value adjustment
sig_Res_up_MF <- allRes_up_MF %>% filter(classicFisher < 0.01)

```

Create node figrues of top 5-10 sig terms for upregulated genes
```{r}
# Plot significant nodes
showSigOfNodes(GOdata.up_MF, score(fisher.up_MF), firstSigNodes = 5, useInfo = 'all')

# Print graph with specific parameters and save to directory
printGraph(GOdata.up_MF, fisher.up_MF, firstSigNodes = 5, fn.prefix = "RAnalysis/Output/tGO_graph_upregulated_MF", useInfo = "def", pdfSW = TRUE)

```

Use topGO for enrichment analysis on downregulated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.down_MF <- new("topGOdata",
              ontology = "MF",
              allGenes = allGenes.down,
              nodeSize = 10,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.down_MF

# Tells you how many signifcant genes there was in the analysis 
numSigGenes(GOdata.down_MF) #64
```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  MF 

 18158 available genes (all genes from the array):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 74  significant genes. 

 16359 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 64  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 1564 
   - number of edges = 2054 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled downregulated data
```{r}
# use default algorithm for fisher test 
fisher.down_MF <- runTest(GOdata.down_MF, algorithm = "weight", statistic = "fisher")

# Generate a results table
allRes_down_MF <- GenTable(GOdata.down_MF, 
                      classicFisher = fisher.down_MF, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.down_MF)))

# Convert character columns to numeric
allRes_down_MF$classicFisher <- as.numeric(allRes_down_MF$classicFisher)

# Apply p value adjustment
sig_Res_down_MF <- allRes_down_MF %>% filter(classicFisher < 0.01)

```

Create node figrues of top 5-10 sig terms for downregulated genes
```{r}
# Plot significant nodes
showSigOfNodes(GOdata.down_MF, score(fisher.down_MF), firstSigNodes = 5, useInfo = 'all')

# Print graph with specific parameters
printGraph(GOdata.down_MF, fisher.down_MF, firstSigNodes = 5, fn.prefix = "RAnalysis/Output/tGO_graph_downregulated_MF", useInfo = "def", pdfSW = TRUE)

```


Cellular Components topGO

Use topGO for enrichment analysis on upregulated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.up_CC <- new("topGOdata",
              ontology = "CC",
              allGenes = allGenes.up,
              nodeSize = 10,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.up_CC

# Tells you how many signifcant genes there was in the analysis 
numSigGenes(GOdata.up_CC) #133
```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  CC 

 18158 available genes (all genes from the array):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 158  significant genes. 

 16359 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 133  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 1564 
   - number of edges = 2054 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled upregulated data
```{r}
# use default algorithm for fisher test 
fisher.up_CC <- runTest(GOdata.up_CC, algorithm = "classic", statistic = "fisher")

# Generate a results table
allRes_up_CC <- GenTable(GOdata.up_CC, 
                      classicFisher = fisher.up_CC, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.up_CC)))

# Convert character columns to numeric
allRes_up_CC$classicFisher <- as.numeric(allRes_up_CC$classicFisher)

# Apply p value adjustment
sig_Res_up_CC <- allRes_up_CC %>% filter(classicFisher < 0.01)

```

Create node figrues of top 5-10 sig terms for upregulated genes
```{r}
# Plot significant nodes
showSigOfNodes(GOdata.up_CC, score(fisher.up_CC), firstSigNodes = 5, useInfo = 'all')

# Print graph with specific parameters and save to directory
printGraph(GOdata.up_CC, fisher.up_CC, firstSigNodes = 5, fn.prefix = "RAnalysis/Output/tGO_graph_upregulated_CC", useInfo = "def", pdfSW = TRUE)

```

Use topGO for enrichment analysis on downregulated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.down_CC <- new("topGOdata",
              ontology = "CC",
              allGenes = allGenes.down,
              nodeSize = 10,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.down_CC

# Tells you how many signifcant genes there was in the analysis 
numSigGenes(GOdata.down_CC) #64
```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  CC 

 18158 available genes (all genes from the array):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 74  significant genes. 

 16359 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g1 Pver_g10 Pver_g100 Pver_g10000 Pver_g10001  ...
   - 64  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 1564 
   - number of edges = 2054 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled downregulated data
```{r}
# use default algorithm for fisher test 
fisher.down_CC <- runTest(GOdata.down_CC, algorithm = "weight", statistic = "fisher")

# Generate a results table
allRes_down_CC <- GenTable(GOdata.down_CC, 
                      classicFisher = fisher.down_CC, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.down_CC)))

# Convert character columns to numeric
allRes_down_CC$classicFisher <- as.numeric(allRes_down_CC$classicFisher)

# Apply p value adjustment
sig_Res_down_CC <- allRes_down_CC %>% filter(classicFisher < 0.01)

```

Create node figrues of top 5-10 sig terms for downregulated genes
```{r}
# Plot significant nodes
showSigOfNodes(GOdata.down_CC, score(fisher.down_CC), firstSigNodes = 5, useInfo = 'all')

# Print graph with specific parameters
printGraph(GOdata.down_CC, fisher.down_CC, firstSigNodes = 5, fn.prefix = "RAnalysis/Output/tGO_graph_downregulated_CC", useInfo = "def", pdfSW = TRUE)

```


Print Tables
```{r}

# Generate a results table
allRes_up_BP_10 <- GenTable(GOdata.up_BP, 
                      classicFisher = fisher.up_BP, 
                      ranksOf = "classicFisher", 
                      topNodes = 10)
allRes_up_BP_10$Direction <- "Up-Regulated"

# Generate a results table
allRes_down_BP_10 <- GenTable(GOdata.down_BP, 
                      classicFisher = fisher.down_BP, 
                      ranksOf = "classicFisher", 
                      topNodes = 10)
allRes_down_BP_10$Direction <- "Down-Regulated"

# Generate a results table
allRes_up_MF_10 <- GenTable(GOdata.up_MF, 
                      classicFisher = fisher.up_MF, 
                      ranksOf = "classicFisher", 
                      topNodes = 10)
allRes_up_MF_10$Direction <- "Up-Regulated"

# Generate a results table
allRes_down_MF_10 <- GenTable(GOdata.down_MF, 
                      classicFisher = fisher.down_MF, 
                      ranksOf = "classicFisher", 
                      topNodes = 10)
allRes_down_MF_10$Direction <- "Down-Regulated"

# Combine top results for down and upreagulated BP
Enrichment_Res_BP_10 <- rbind(allRes_up_BP_10,allRes_down_BP_10)

# Create publishable table
Enrichment_Res_BP_10_table <- Enrichment_Res_BP_10 %>%
  dplyr::select(-Expected)  # Remove the 'expected' column

# Create flextable object
flex_enrich <- flextable(Enrichment_Res_BP_10_table)

# Customize the flextable
flex_enrich <- flex_enrich %>%
  set_header_labels(
    GO.ID = "GO ID",
    Term = "Term",
    Annotated = "Annotated",
    Significant = "Significant",
    classicFisher = "Fisher p-value") %>%
  colformat_double(digits = 4) %>%
  bold(j = "classicFisher") %>%  # Bold all values in classicFisher column
  add_header_lines(values = "Top 10 Enriched GO Terms (Biological Process)") %>%
  autofit() %>%
  theme_box()

# Save flextable as PNG
png_file <- "RAnalysis/Output/DEG_top10_enrich_BP_table.png"
save_as_image(flex_enrich, path = png_file)

# Combine top results for down and upreagulated MF
Enrichment_Res_MF_10 <- rbind(allRes_up_MF_10,allRes_down_MF_10)

# Create publishable table
Enrichment_Res_MF_10_table <- Enrichment_Res_MF_10 %>%
  dplyr::select(-Expected)  # Remove the 'expected' column

# Create flextable object
flex_enrich <- flextable(Enrichment_Res_MF_10_table)

# Customize the flextable
flex_enrich <- flex_enrich %>%
  set_header_labels(
    GO.ID = "GO ID",
    Term = "Term",
    Annotated = "Annotated",
    Significant = "Significant",
    classicFisher = "Fisher p-value") %>%
  colformat_double(digits = 4) %>%
  bold(j = "classicFisher") %>%  # Bold all values in classicFisher column
  add_header_lines(values = "Top 10 Enriched GO Terms (Molecular Function)") %>%
  autofit() %>%
  theme_box()

# Save flextable as PNG
png_file <- "RAnalysis/Output/DEG_top10_enrich_MF_table.png"
save_as_image(flex_enrich, path = png_file)
```

