---
title: "GO_Term_Enrichment_Analysis"
author: "daniellembecker"
edited by: "daniellembecker"
date: "9/20/2021"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Gene Ontology (GO) Enrichment Analysis 
## Previous steps include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DMG statistical data sheet
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
if (!requireNamespace("DESeq2", quietly = TRUE)) { BiocManager::install("DESeq2") }
library("DESeq2")
library("tidyr")
library("tidyverse")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
if (!requireNamespace("genefilter", quietly = TRUE)) { BiocManager::install("genefilter") }
library("genefilter")
library("ggplot2")
library("gplots")
library("spdep") 
library("patchwork")
#library("aDMGenet") 
library("gridExtra")
if (!requireNamespace("clusterProfiler", quietly = TRUE)) { BiocManager::install("clusterProfiler") }
library("clusterProfiler")
library("DataCombine")
if (!requireNamespace("GSEABase", quietly = TRUE)) { BiocManager::install("GSEABase") }
library("GSEABase")
library("data.table")
if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")
library("rrvgo")
if ("org.Ce.eg.db" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("org.Ce.eg.db")
library(org.Ce.eg.db)
if ("topGO" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("topGO")
library(topGO)
if ("Rgraphviz" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("Rgraphviz")
library(Rgraphviz)
```

# Import the data files 
```{r, echo=FALSE}
# Load go terms from functional annotation file
annot_GO <- read.csv("RAnalysis/Output/annot_GO.terms.host.csv")[-1]

# Remove rows where 'GO.ID' is 'unknown'
annot_GO <- subset(annot_GO, GO.ID != 'unknown')

# edit annot_GO to list GO terms all in one row per gene
annot_GO <- annot_GO %>%
  group_by(gene_id) %>%
  summarise(GO.ID = paste(GO.ID, collapse = ";")) %>%
  ungroup()

```

# Load in DMG results
```{r}
#DMG results
DMG.res <- read.csv("RAnalysis/Output/DMGs_all.csv")[-1]

colnames(DMG.res)[1] <-"gene_id" # make colnames a true column called gene_id
#remove the extra "gene" to match the "Pver_gene_g#" format
DMG.res$gene_id <-  gsub("Pver_gene_novel_gene_","Pver_novel_gene_", DMG.res$gene_id)
DMG.res$gene_id <-  gsub("Pver_gene_split_gene_", "Pver_gene_", DMG.res$gene_id)
#remove extra characters to match annot
DMG.res$gene_id <-  gsub("_gene_g","_g",DMG.res$gene_id) #sub gene_g with _g
DMG.res$gene_id <-  gsub("_gene_","_",DMG.res$gene_id) #remove extra _
DMG.res$gene_id <-  gsub("_split_","_",DMG.res$gene_id) #remove split naming in gene id

# Use the new MD.10x file that has now been filtered for SNPs

meth_filt <- readRDS("RAnalysis/Data/filtered_bed_SNP_file.Rdata") #read in info

# Rename gene column to gene_id and select only necessary columns, summarize meth
meth_filt_select <- meth_filt  %>%
  dplyr::select(Sample.ID, meth, gene) %>%
  rename(gene_id = gene)

# Reshape the data
meth_wide <- meth_filt_select %>%
  group_by(gene_id, Sample.ID) %>%
  summarize(total_meth = sum(meth, na.rm = TRUE), .groups = "drop") %>%
  pivot_wider(names_from = Sample.ID, values_from = total_meth, values_fill = 0)

#remove the extra "gene" to match the "Pver_gene_g#" format
meth_wide$gene_id <-  gsub("Pver_gene_novel_gene_","Pver_novel_gene_", meth_wide$gene_id)
meth_wide$gene_id <-  gsub("Pver_gene_split_gene_", "Pver_gene_", meth_wide$gene_id)
#remove extra characters to match annot
meth_wide$gene_id <-  gsub("_gene_g","_g", meth_wide$gene_id) #sub gene_g with _g
meth_wide$gene_id <-  gsub("_gene_","_", meth_wide$gene_id) #remove extra _
meth_wide$gene_id <-  gsub("_split_","_", meth_wide$gene_id) #remove split naming in gene id
meth_wide$gene_id <-  gsub("-.*","",meth_wide$gene_id) #remove the extra g number after the dash

# Check if all gene IDs in GFF3 are included in  annot_GO
unmatched_genes <- setdiff(meth_wide$gene_id, annot_GO$gene_id)
# Print the unmatched gene IDs
print(unmatched_genes)

# Only genes that don't match after edited are the unknowns we filtered above
```

Match up genes in gene list file to annotation file
```{r}
names(annot_GO)

gcount_filt2annot = match(meth_wide$gene_id, annot_GO$gene_id) #match genes in gcount_filt to annot_GO

# The following is the number of probes without annotation 
sum(is.na(gcount_filt2annot))

row_nas<-which(is.na(gcount_filt2annot))

#view the genes that do not have a match in the annotation file
missing<-as.data.frame(gcount_filt2annot[row_nas])

```
364 genes in gcount_filt do not have GO annotations

Reduce annotation file to only contain genes detected in our dataset.  
```{r}
filtered_Pverr.annot <- annot_GO[annot_GO$gene_id %in% meth_wide$gene_id, ]
dim(filtered_Pverr.annot)

```

The annotation file now only contains genes that were detected in our dataset that have annotation information. This is 3,546 out of the 3,910 genes in our dataset at 10x coverage. 

Biological Processes topGO

Use topGO for enrichment analysis on hypermethylated data
```{r}
### Generate vector with names in just the contrast we are analyzing
ID.vector.hyper <- DMG.res %>%
      filter(methylation_status == "hypermethylated") %>%
      pull(gene_id)
    
# Get a list of GO Terms for all genes detected in our dataset
GO.terms <- filtered_Pverr.annot %>%
  dplyr::select(gene_id, GO.ID) %>%
  separate_rows(GO.ID, sep = ";")

# Create a gene-to-GO mapping list
geneID2GO <- split(as.character(GO.terms$GO.ID), GO.terms$gene_id)
geneNames <- unique(GO.terms$gene_id)

# Create a named vector for allGenes
allGenes.hyper <- factor(as.integer(geneNames %in% ID.vector.hyper))
names(allGenes.hyper) <- geneNames

# Create a topGOdata object using the custom gene2GO mapping
GOdata.hyper_BP <- new("topGOdata",
              ontology = "BP",
              allGenes = allGenes.hyper,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hyper_BP

# measures number sig genes
numSigGenes(GOdata.hyper_BP) #125

```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  BP 

 3546 available genes (all genes from the array):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10036  ...
   - 149  significant genes. 

 2953 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10041  ...
   - 125  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 2935 
   - number of edges = 6086 

------------------------- topGOdata object -------------------------


Run topGO statistical analyses on compiled hypermethylated data
```{r}
# fisher exact test from topgo manual
fisher.hyper_BP <- runTest(GOdata.hyper_BP, statistic = "fisher")

# Generate a results table
allRes_hyper_BP <- GenTable(GOdata.hyper_BP, 
                      classicFisher = fisher.hyper_BP, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.hyper_BP)))

# Convert character columns to numeric
allRes_hyper_BP$classicFisher <- as.numeric(allRes_hyper_BP$classicFisher)

```

Create node figrues of top 5-10 sig terms for hypermethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hyper_BP, score(fisher.hyper_BP), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypermethylated_BP.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hyper_BP, fisher.hyper_BP, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```

Use topGO for enrichment analysis on hypomethylated data
```{r}
### Generate vector with names in just the contrast we are analyzing
ID.vector.hypo <- DMG.res %>%
      filter(methylation_status =="hypomethylated") %>%
      pull(gene_id)
    
# Create a named vector for allGenes
allGenes.hypo <- factor(as.integer(geneNames %in% ID.vector.hypo))
names(allGenes.hypo) <- geneNames

# Create a topGOdata object using the custom gene2GO mapping
GOdata.hypo_BP <- new("topGOdata",
              ontology = "BP",
              allGenes = allGenes.hypo,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hypo_BP

# 
numSigGenes(GOdata.hypo_BP) #143

```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  BP 

 3546 available genes (all genes from the array):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10036  ...
   - 182  significant genes. 

 2953 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10041  ...
   - 143  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 2935 
   - number of edges = 6086 

------------------------- topGOdata object -------------------------


Run topGO statistical analyses on compiled hypomethylated data
```{r}
# fisher exact test from topgo manual
fisher.hypo_BP <- runTest(GOdata.hypo_BP, statistic = "fisher")

# Generate a results table
allRes_hypo_BP <- GenTable(GOdata.hypo_BP, 
                      classicFisher = fisher.hypo_BP, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.hypo_BP)))

# Convert character columns to numeric
allRes_hypo_BP$classicFisher <- as.numeric(allRes_hypo_BP$classicFisher)

```

Create node figrues of top 5-10 sig terms for hypomethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hypo_BP, score(fisher.hypo_BP), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypomethylated_BP.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hypo_BP, fisher.hypo_BP, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```



Molecular Functions topGO

Use topGO for enrichment analysis on hypermethylated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.hyper_MF <- new("topGOdata",
              ontology = "MF",
              allGenes = allGenes.hyper,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hyper_MF

numSigGenes(GOdata.hyper_MF) # 144
```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  MF 

 3546 available genes (all genes from the array):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10036  ...
   - 149  significant genes. 

 3262 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g10004 Pver_g1002 Pver_g10021 Pver_g10041 Pver_g10049  ...
   - 144  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 484 
   - number of edges = 620 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled hypermethylated data
```{r}
# fisher exact test from topgo manual
fisher.hyper_MF <- runTest(GOdata.hyper_MF, statistic = "fisher")

# Generate a results table
allRes_hyper_MF <- GenTable(GOdata.hyper_MF, 
                      classicFisher = fisher.hyper_MF, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.hyper_MF)))

# Convert character columns to numeric
allRes_hyper_MF$classicFisher <- as.numeric(allRes_hyper_MF$classicFisher)

```

Create node figrues of top 5-10 sig terms for hypermethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hyper_MF, score(fisher.hyper_MF), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypermethylated_MF.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hyper_MF, fisher.hyper_MF, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```

Use topGO for enrichment analysis on hypomethylated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.hypo_MF <- new("topGOdata",
              ontology = "MF",
              allGenes = allGenes.hypo,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hypo_MF

numSigGenes(GOdata.hypo_MF) #175
```

------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  MF 

 3546 available genes (all genes from the array):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10036  ...
   - 182  significant genes. 

 3262 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g10004 Pver_g1002 Pver_g10021 Pver_g10041 Pver_g10049  ...
   - 175  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 484 
   - number of edges = 620 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled hypomethylated data
```{r}
# fisher exact test from topgo manual
fisher.hypo_MF <- runTest(GOdata.hypo_MF, statistic = "fisher")

# Generate a results table
allRes_hypo_MF <- GenTable(GOdata.hypo_MF, 
                      classicFisher = fisher.hypo_MF, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.hypo_MF)))

# Convert character columns to numeric
allRes_hypo_MF$classicFisher <- as.numeric(allRes_hypo_MF$classicFisher)


```

Create node figrues of top 5-10 sig terms for hypomethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hypo_MF, score(fisher.hypo_MF), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypomethylated_MF.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hypo_MF, fisher.hypo_MF, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```


Molecular Functions topGO

Use topGO for enrichment analysis on hypermethylated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.hyper_MF <- new("topGOdata",
              ontology = "MF",
              allGenes = allGenes.hyper,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hyper_MF

numSigGenes(GOdata.hyper_MF) # 144
```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  MF 

 3546 available genes (all genes from the array):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10036  ...
   - 149  significant genes. 

 3262 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g10004 Pver_g1002 Pver_g10021 Pver_g10041 Pver_g10049  ...
   - 144  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 484 
   - number of edges = 620 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled hypermethylated data
```{r}
# fisher exact test from topgo manual
fisher.hyper_MF <- runTest(GOdata.hyper_MF, statistic = "fisher")

# Generate a results table
allRes_hyper_MF <- GenTable(GOdata.hyper_MF, 
                      classicFisher = fisher.hyper_MF, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.hyper_MF)))

# Convert character columns to numeric
allRes_hyper_MF$classicFisher <- as.numeric(allRes_hyper_MF$classicFisher)

```

Create node figrues of top 5-10 sig terms for hypermethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hyper_MF, score(fisher.hyper_MF), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypermethylated_MF.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hyper_MF, fisher.hyper_MF, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```

Use topGO for enrichment analysis on hypomethylated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.hypo_MF <- new("topGOdata",
              ontology = "MF",
              allGenes = allGenes.hypo,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hypo_MF

numSigGenes(GOdata.hypo_MF) #175
```

------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  MF 

 3546 available genes (all genes from the array):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10036  ...
   - 182  significant genes. 

 3262 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g10004 Pver_g1002 Pver_g10021 Pver_g10041 Pver_g10049  ...
   - 175  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 484 
   - number of edges = 620 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled hypomethylated data
```{r}
# fisher exact test from topgo manual
fisher.hypo_MF <- runTest(GOdata.hypo_MF, statistic = "fisher")

# Generate a results table
allRes_hypo_MF <- GenTable(GOdata.hypo_MF, 
                      classicFisher = fisher.hypo_MF, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.hypo_MF)))

# Convert character columns to numeric
allRes_hypo_MF$classicFisher <- as.numeric(allRes_hypo_MF$classicFisher)


```

Create node figrues of top 5-10 sig terms for hypomethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hypo_MF, score(fisher.hypo_MF), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypomethylated_MF.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hypo_MF, fisher.hypo_MF, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```



Cellular Components topGO

Use topGO for enrichment analysis on hypermethylated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.hyper_CC <- new("topGOdata",
              ontology = "CC",
              allGenes = allGenes.hyper,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hyper_CC

numSigGenes(GOdata.hyper_CC) # 144
```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  CC 

 3546 available genes (all genes from the array):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10036  ...
   - 149  significant genes. 

 2626 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10041  ...
   - 77  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 360 
   - number of edges = 594 

------------------------- topGOdata object -------------------------

Run topGO statistical analyses on compiled hypermethylated data
```{r}
# fisher exact test from topgo manual
fisher.hyper_CC <- runTest(GOdata.hyper_CC, statistic = "fisher")

# Generate a results table
allRes_hyper_CC <- GenTable(GOdata.hyper_CC, 
                      classicFisher = fisher.hyper_CC, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.hyper_CC)))

# Convert character columns to numeric
allRes_hyper_CC$classicFisher <- as.numeric(allRes_hyper_CC$classicFisher)

```

Create node figrues of top 5-10 sig terms for hypermethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hyper_CC, score(fisher.hyper_CC), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypermethylated_CC.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hyper_CC, fisher.hyper_CC, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```

Use topGO for enrichment analysis on hypomethylated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.hypo_CC <- new("topGOdata",
              ontology = "CC",
              allGenes = allGenes.hypo,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hypo_CC

numSigGenes(GOdata.hypo_CC) #81
```
------------------------- topGOdata object -------------------------

 Description:
   -   

 Ontology:
   -  CC 

 3546 available genes (all genes from the array):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10036  ...
   - 182  significant genes. 

 2626 feasible genes (genes that can be used in the analysis):
   - symbol:  Pver_g10004 Pver_g10011 Pver_g1002 Pver_g10021 Pver_g10041  ...
   - 81  significant genes. 

 GO graph (nodes with at least  5  genes):
   - a graph with directed edges
   - number of nodes = 360 
   - number of edges = 594 

------------------------- topGOdata object -------------------------


Run topGO statistical analyses on compiled hypomethylated data
```{r}
# fisher exact test from topgo manual
fisher.hypo_CC <- runTest(GOdata.hypo_CC, statistic = "fisher")

# Generate a results table
allRes_hypo_CC <- GenTable(GOdata.hypo_CC, 
                      classicFisher = fisher.hypo_CC, 
                      ranksOf = "classicFisher", 
                      topNodes = length(score(fisher.hypo_CC)))

# Convert character columns to numeric
allRes_hypo_CC$classicFisher <- as.numeric(allRes_hypo_CC$classicFisher)


```

Create node figrues of top 5-10 sig terms for hypomethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hypo_CC, score(fisher.hypo_CC), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypomethylated_CC.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hypo_CC, fisher.hypo_CC, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```



Filter all dataframes for pvalue < 0.05
Combine all dataframes to make dendograms next, add ontolology column and up vs. hypomethylated column
```{r}
# Filter and add columns in one line for each dataframe
filt.hyper.BP <- allRes_hyper_BP %>% 
  filter(classicFisher < 0.05) %>% 
  mutate(ontology = "BP", methylation_status = "hypermethylated")

filt.hypo.BP <- allRes_hypo_BP %>% 
  filter(classicFisher < 0.05) %>% 
  mutate(ontology = "BP", methylation_status = "hypomethylated")

filt.hyper.MF <- allRes_hyper_MF %>% 
  filter(classicFisher < 0.05) %>% 
  mutate(ontology = "MF", methylation_status = "hypermethylated")

filt.hypo.MF <- allRes_hypo_MF %>% 
  filter(classicFisher < 0.05) %>% 
  mutate(ontology = "MF", methylation_status = "hypomethylated")

filt.hyper.CC <- allRes_hyper_CC %>% 
  filter(classicFisher < 0.05) %>% 
  mutate(ontology = "CC", methylation_status = "hypermethylated")

filt.hypo.CC <- allRes_hypo_CC %>% 
  filter(classicFisher < 0.05) %>% 
  mutate(ontology = "CC", methylation_status = "hypomethylated")

#Save dataframe with all GO enrichment
GO.enrich.all <- rbind(filt.hyper.BP, filt.hypo.BP, filt.hyper.MF, filt.hypo.MF, filt.hyper.CC, filt.hypo.CC)

# Rename annotated column to numInCat and significant as numDEI 
GO.enrich.all <- GO.enrich.all %>%
  rename(numInCat = Annotated,  # Rename 'annotated' to 'numInCat'
         numDEI = Significant,
         num.expected.percent = Expected)  # Rename 'significant' to 'numDEI'
  
# write csv
write.csv(GO.enrich.all, "RAnalysis/Output/DMG_GO_enrichment_terms_all.csv")

```


Upload all go enrichment data for rrvgo so you do not have to load all steps above
```{r}
# Load in entire go terms dataframe here so that we don't need to run entire script above everytime
GO.enrich.all <- read.csv("RAnalysis/Output/DMG_GO_enrichment_terms_all.csv")[-1]

```

Run a loop to filter enrichment for up and downregulated using rrvgo for BP.
```{r}
# Initialize an empty dataframe to store results
methy_stat_go_results_BP <- data.frame()

# Make a list for methy_stat
methy_stat_list <- c("hypermethylated", "hypomethylated")

for (methy_stat_id in methy_stat_list) {
 
   # Filter for the current methy_stat
  go_results <- GO.enrich.all %>%
    filter(methylation_status == methy_stat_id) %>%
    filter(ontology == "BP") %>%
    filter(!is.na(classicFisher)) %>%
    arrange(classicFisher)

  # Print the current methy_stat and number of unique GO IDs
  print(methy_stat_id)
  print(length(unique(go_results$GO.ID)))

  # Reduce/collapse GO term set with the rrvgo package
  simMatrix <- calculateSimMatrix(go_results$GO.ID,
                                  orgdb = "org.Ce.eg.db", # c. elegans database
                                  ont = "BP",
                                  method = "Rel")

  # Calculate similarity scores
  scores <- setNames(-log10(go_results$classicFisher), go_results$GO.ID)
  reducedTerms <- reduceSimMatrix(simMatrix,
                                  scores,
                                  threshold = 0.8,
                                  orgdb = "org.Ce.eg.db")

  # Keep only the GO terms from the reduced list
  go_results <- go_results %>%
    filter(GO.ID %in% reducedTerms$go)

  # Add parent terms to the list of GO terms
  go_results$ParentTerm <- reducedTerms$parentTerm[match(go_results$GO.ID, reducedTerms$go)]

  # Add methy_stat to the results
  go_results$methylation_status <- methy_stat_id

  # Print the number of unique GO IDs after reduction
  print(length(unique(go_results$GO.ID)))

  # Append results to the dataframe
  methy_stat_go_results_BP <- rbind(methy_stat_go_results_BP, go_results)
}

write.csv(methy_stat_go_results_BP, "RAnalysis/Output/DMG_go_results_BP.csv")


```

Make dendrogram and dot plot for BP using semantic similarity

```{r}
# Load in BP methy_stat dataframe so you dont need to re run everything above
methy_stat_go_results_BP <- read.csv("RAnalysis/Output/DMG_go_results_BP.csv")[-1]

# There are duplicate GO.IDs in the data set for hypo and hyper methylation which is okay from literature, however rrvgo can only handle unique GO.IDs, filtering for more significant term to be used for the simatrix
# Select the more significant term for each GO.ID
filtered_methy_stat_go_results_BP <- methy_stat_go_results_BP %>%
  group_by(GO.ID) %>%
  filter(classicFisher == min(classicFisher)) %>%
  ungroup()

# calculated the sim matrix for go terms
combined_simMatrix_BP <- calculateSimMatrix(filtered_methy_stat_go_results_BP$GO.ID,
                                            orgdb = "org.Ce.eg.db",
                                            ont = "BP",
                                            method = "Rel")


# Convert similarity matrix to distance matrix
distMatrix_BP <- as.dist(1 - combined_simMatrix_BP)

# Create dendrogram
dendrogram_BP <- as.dendrogram(hclust(distMatrix_BP))

# Get dendrogram data
dendro_data_BP <- dendro_data(dendrogram_BP)

#Prepare data for plotting by grouping by GO.ID, direction, ParentTerm, and Term
plot_data_BP <- filtered_methy_stat_go_results_BP %>%
  group_by(GO.ID, methylation_status, ParentTerm, Term) %>%
  summarise(numInCat = sum(numInCat), numDEI = sum(numDEI)) %>%
  ungroup() %>%
  mutate(percentDMGs = (numDEI / numInCat) * 100)

# Add Term labels to dendrogram data
dendro_data_BP$labels <- dendro_data_BP$labels %>%
  left_join(filtered_methy_stat_go_results_BP %>% dplyr::select(GO.ID, Term, methylation_status) %>% distinct(), by = c("label" = "GO.ID")) %>% mutate(label = Term)

# Create the dendrogram plot
dendo.BP.plot <- ggplot() +
  # Add dendrogram segments
  geom_segment(data = dendro_data_BP$segments,
               aes(x = x, y = y, xend = xend, yend = yend)) +
  
  # Add labels for GO terms with colors based on direction
  geom_text(data = dendro_data_BP$labels,
            aes(x = x, y = y, label = label, hjust = 1, color = methylation_status),
            size = 8, nudge_y = -0.01) +
  
  # Add dots representing terms
  geom_point(data = plot_data_BP,
             aes(x = match(Term, dendro_data_BP$labels$label),
                 y = -3,
                 size = numInCat,
                 fill = percentDMGs),  # Use fill for color gradient
             shape = 21) +  # Use a filled shape
  
  # Customize the plot
  scale_size_continuous(range = c(1, 10)) +
  scale_color_manual(values = c("hypermethylated" = "red", "hypomethylated" = "blue"),
                     labels = c("hypermethylated" = "Hypermethylation", "hypomethylated" = "Hypomethylation"),
                     name = "Methylation Status",  # Change legend title here
                     guide = guide_legend(override.aes = list(size = 8))) +
  scale_fill_gradient(low = "coral", high = "purple3", 
                      name = "% DMGs", 
                      limits = c(0, 43),  # Set limits based on your data range
                      breaks = seq(0, 43, by = 10)) +  # Set custom breaks
  coord_flip() +
  ylim(-3, NA) +  # Extend the x-axis (effectively the y-axis) to -1.4
  
  # Apply minimal theme and customize legend and axis
  theme_minimal() +
  theme(axis.ticks.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        panel.grid = element_blank(),
        legend.text = element_text(size = 22),  # Increase legend text size
        legend.title = element_text(size = 24, face = "bold"),
        axis.title.x = element_text(size = 30, face = "bold")) + # Increase legend title size
  # Add labels for axes and legends
  labs(x = NULL, y = "Biological Processes", size = "Number of Genes", fill = "Percent DMGs");dendo.BP.plot

# Save the updated plot
ggsave("RAnalysis/Output/DMG_dendrogram_plot_BP.png", 
       plot = dendo.BP.plot, 
       width = 13,  
       height = 8, 
       units = "in", 
       dpi = 300)



```


Run a loop to filter enrichment for up and downregulated using rrvgo for MF.
```{r}
# Initialize an empty dataframe to store results
methy_stat_go_results_MF <- data.frame()

# Make a list for methy_stat
methy_stat_list <- c("hypermethylated", "hypomethylated")

for (methy_stat_id in methy_stat_list) {
 
   # Filter for the current methy_stat
  go_results <- GO.enrich.all %>%
    filter(methylation_status == methy_stat_id) %>%
    filter(ontology == "MF") %>%
    filter(!is.na(classicFisher)) %>%
    arrange(classicFisher)

  # Print the current methy_stat and number of unique GO IDs
  print(methy_stat_id)
  print(length(unique(go_results$GO.ID)))

  # Reduce/collapse GO term set with the rrvgo package
  simMatrix <- calculateSimMatrix(go_results$GO.ID,
                                  orgdb = "org.Ce.eg.db", # c. elegans database
                                  ont = "MF",
                                  method = "Rel")

  # Calculate similarity scores
  scores <- setNames(-log10(go_results$classicFisher), go_results$GO.ID)
  reducedTerms <- reduceSimMatrix(simMatrix,
                                  scores,
                                  threshold = 0.8,
                                  orgdb = "org.Ce.eg.db")

  # Keep only the GO terms from the reduced list
  go_results <- go_results %>%
    filter(GO.ID %in% reducedTerms$go)

  # Add parent terms to the list of GO terms
  go_results$ParentTerm <- reducedTerms$parentTerm[match(go_results$GO.ID, reducedTerms$go)]

  # Add methy_stat to the results
  go_results$methylation_status <- methy_stat_id

  # Print the number of unique GO IDs after reduction
  print(length(unique(go_results$GO.ID)))

  # Append results to the dataframe
  methy_stat_go_results_MF <- rbind(methy_stat_go_results_MF, go_results)
}

write.csv(methy_stat_go_results_MF, "RAnalysis/Output/DMG_go_results_MF.csv")


```


Make dendrogram and dot plot for MF using semantic similarity

```{r}
# Load in MF methy_stat dataframe so you dont need to re run everything above
methy_stat_go_results_MF <- read.csv("RAnalysis/Output/DMG_go_results_MF.csv")[-1]

# There are duplicate GO.IDs in the data set for hypo and hyper methylation which is okay from literature, however rrvgo can only handle unique GO.IDs, filtering for more significant term to be used for the simatrix
# Select the more significant term for each GO.ID
filtered_methy_stat_go_results_MF <- methy_stat_go_results_MF %>%
  group_by(GO.ID) %>%
  filter(classicFisher == min(classicFisher)) %>%
  ungroup()

# calculated the sim matrix for go terms
combined_simMatrix_MF <- calculateSimMatrix(filtered_methy_stat_go_results_MF$GO.ID,
                                            orgdb = "org.Ce.eg.db",
                                            ont = "MF",
                                            method = "Rel")


# Convert similarity matrix to distance matrix
distMatrix_MF <- as.dist(1 - combined_simMatrix_MF)

# Create dendrogram
dendrogram_MF <- as.dendrogram(hclust(distMatrix_MF))

# Get dendrogram data
dendro_data_MF <- dendro_data(dendrogram_MF)

#Prepare data for plotting by grouping by GO.ID, direction, ParentTerm, and Term
plot_data_MF <- filtered_methy_stat_go_results_MF %>%
  group_by(GO.ID, methylation_status, ParentTerm, Term) %>%
  summarise(numInCat = sum(numInCat), numDEI = sum(numDEI)) %>%
  ungroup() %>%
  mutate(percentDMGs = (numDEI / numInCat) * 100)

# Add Term labels to dendrogram data
dendro_data_MF$labels <- dendro_data_MF$labels %>%
  left_join(filtered_methy_stat_go_results_MF %>% dplyr::select(GO.ID, Term, methylation_status) %>% distinct(), by = c("label" = "GO.ID")) %>% mutate(label = Term)

# Create the dendrogram plot
dendo.MF.plot <- ggplot() +
  # Add dendrogram segments
  geom_segment(data = dendro_data_MF$segments,
               aes(x = x, y = y, xend = xend, yend = yend)) +
  
  # Add labels for GO terms with colors based on direction
  geom_text(data = dendro_data_MF$labels,
            aes(x = x, y = y, label = label, hjust = 1, color = methylation_status),
            size = 8, nudge_y = -0.01) +
  
  # Add dots representing terms
  geom_point(data = plot_data_MF,
             aes(x = match(Term, dendro_data_MF$labels$label),
                 y = -3,
                 size = numInCat,
                 fill = percentDMGs),  # Use fill for color gradient
             shape = 21) +  # Use a filled shape
  
  # Customize the plot
  scale_size_continuous(range = c(1, 10)) +
  scale_color_manual(values = c("hypermethylated" = "red", "hypomethylated" = "blue"),
                     labels = c("hypermethylated" = "Hypermethylation", "hypomethylated" = "Hypomethylation"),
                     name = "Methylation Status",  # Change legend title here
                     guide = guide_legend(override.aes = list(size = 8))) +
  scale_fill_gradient(low = "coral", high = "purple3", 
                      name = "% DMGs", 
                      limits = c(0, 43),  # Set limits based on your data range
                      breaks = seq(0, 43, by = 10)) +  # Set custom breaks
  coord_flip() +
  ylim(-3, NA) +  # Extend the x-axis (effectively the y-axis) to -1.4
  
  # Apply minimal theme and customize legend and axis
  theme_minimal() +
  theme(axis.ticks.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        panel.grid = element_blank(),
        legend.text = element_text(size = 22),  # Increase legend text size
        legend.title = element_text(size = 24, face = "bold"),
        axis.title.x = element_text(size = 30, face = "bold")) + # Increase legend title size
  # Add labels for axes and legends
  labs(x = NULL, y = "Molecular Functions", size = "Number of Genes", fill = "Percent DMGs");dendo.MF.plot

# Save the updated plot
ggsave("RAnalysis/Output/DMG_dendrogram_plot_MF.png", 
       plot = dendo.MF.plot, 
       width = 13,  
       height = 8, 
       units = "in", 
       dpi = 300)

```



Run a loop to filter enrichment for up and downregulated using rrvgo for CC.
```{r}
# Initialize an empty dataframe to store results
methy_stat_go_results_CC <- data.frame()

# Make a list for methy_stat
methy_stat_list <- c("Up", "Down")

for (methy_stat_id in methy_stat_list) {
 
   # Filter for the current methy_stat
  go_results <- GO.enrich.all %>%
    filter(methylation_status == methylation_status) %>%
    filter(ontology == "CC") %>%
    filter(!is.na(classicFisher)) %>%
    arrange(classicFisher)

  # Print the current methy_stat and number of unique GO IDs
  print(methy_stat_id)
  print(length(unique(go_results$GO.ID)))

  # Reduce/collapse GO term set with the rrvgo package
  simMatrix <- calculateSimMatrix(go_results$GO.ID,
                                  orgdb = "org.Ce.eg.db", # c. elegans database
                                  ont = "CC",
                                  method = "Rel")

  # Calculate similarity scores
  scores <- setNames(-log10(go_results$classicFisher), go_results$GO.ID)
  reducedTerms <- reduceSimMatrix(simMatrix,
                                  scores,
                                  threshold = 0.7,
                                  orgdb = "org.Ce.eg.db")

  # Keep only the GO terms from the reduced list
  go_results <- go_results %>%
    filter(GO.ID %in% reducedTerms$go)

  # Add parent terms to the list of GO terms
  go_results$ParentTerm <- reducedTerms$parentTerm[match(go_results$GO.ID, reducedTerms$go)]

  # Add methy_stat to the results
  go_results$methy_stat <- methy_stat_id

  # Print the number of unique GO IDs after reduction
  print(length(unique(go_results$GO.ID)))

  # Append results to the dataframe
  methy_stat_go_results_CC <- rbind(methy_stat_go_results_CC, go_results)
  
}

```

Make dendrogram and dot plot for CC using semantic similarity

```{r}
# Load in CC methy_stat dataframe so you dont need to re run everything above
methy_stat_go_results_CC <- read.csv("RAnalysis/Output/DMG_go_results_CC.csv")[-1]

# Filter the dataset based on numDEI cut off of at least two sig terms 
filtered_methy_stat_go_results_CC <- methy_stat_go_results_CC %>%
  filter(!numDEI >= 2)

# Combine similarity matrices for both methy_stats
combined_simMatrix_CC <- calculateSimMatrix(methy_stat_go_results_CC$GO.ID,
                                         orgdb = "org.Ce.eg.db",
                                         ont = "CC",
                                         method = "Rel")

# Convert similarity matrix to distance matrix
distMatrix_CC <- as.dist(1 - combined_simMatrix_CC)

# Create dendrogram
dendrogram_CC <- as.dendrogram(hclust(distMatrix_CC))

# Get dendrogram data
dendro_data_CC <- dendro_data(dendrogram_CC)

#Prepare data for plotting by grouping by GO.ID, methy_stat, ParentTerm, and Term
plot_data_CC <- methy_stat_go_results_CC %>%
  group_by(GO.ID, methy_stat, ParentTerm, Term) %>%
  summarise(numInCat = sum(numInCat)) %>%
  ungroup()

# Add Term labels to dendrogram data
dendro_data_CC$labels <- dendro_data_CC$labels %>%
  left_join(methy_stat_go_results_CC %>% dplyr::select(GO.ID, Term) %>% distinct(), by = c("label" = "GO.ID")) %>%
  mutate(label = Term)

# Create the dendrogram plot with numInCat overlay plot
dendo.CC.plot <- ggplot() +
  # Add dendrogram
  geom_segment(data = dendro_data_CC$segments,
               aes(x = x, y = y, xend = xend, yend = yend)) +
  # Add labels for GO terms
  geom_text(data = dendro_data_CC$labels,
            aes(x = x, y = y, label = label, hjust = 1),
            size = 3, nudge_y = -0.01) +
  # Add labels for ParentTerms higher up
#  geom_text(data = dendro_data_CC$labels,
#            aes(x = x, y = y + 1, label = ParentTerm, hjust = 1),
#            size = 3, nudge_y = 0.03, color = "blue") +
  # Add dots
  geom_point(data = plot_data_CC,
             aes(x = match(Term, dendro_data_CC$labels$label),
                 y = -0.5,
                 size = numInCat,
                 color = methy_stat)) +
  # Customize the plot
  scale_size_continuous(range = c(1, 10)) +
  scale_color_manual(values = c("Up" = "red", "Down" = "blue"),
                     labels = c("Up" = "hypermethylated", "Down" = "hypomethylated"),
                     guide = guide_legend(override.aes = list(size = 3))) + # Adjust legend point size here
  coord_flip() +
  ylim(-0.5, NA) +  # Extend the x-axis (effectively the y-axis) to -0.5
  theme_minimal() +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        panel.grid = element_blank()) +
  labs(x = NULL, y = NULL, size = "Number of Genes", color = "Methylation Status");dendo.CC.plot

# Save the plot with adjusted dimensions
ggsave("RAnalysis/Output/DMG_dendrogram_plot_CC.png", plot = dendo.CC.plot, width = 11, height = 8, units = "in")

```






