---
title: "DMG_DEG_pathway_analysis"
author: "Danielle M. Becker"
date: "20220720"
output: html_document
editor_options: 
  chunk_output_type: console
---

#### Gene Expression and DNA Methylation comparison ####
## Used to compare general relationship between DNA methylation and gene expression
## Code developed from Javier A Rodriguez-Casariego paper: https://onlinelibrary.wiley.com/doi/pdf/10.1111/mec.16246?casa_token=DGoNlwKZByAAAAAA:i6JDcylh98txNEMQnr6dUTFxKd1sX7xotbSMZJq-DDOyqZGUPHObc30PhaWOAqpEufHMcJJ_wlQbQvbG
## https://github.com/jarcasariego/ACER_clonal_divergence/blob/main/GE_GbM_association/rmd/GE_DNAm_control.Rmd

```{r load packages}

## load packages
if (!requireNamespace("ComplexHeatmap", quietly = TRUE)) { BiocManager::install("ComplexHeatmap") }
library(ComplexHeatmap)
library(clusterProfiler)
library(pheatmap)
library(magrittr)
library(rtracklayer)
library(GenomicRanges)
library(png)
library(stats)
library(GO.db)
library(AnnotationDbi)
library(viridis)
library(janitor)
library(rtracklayer) # on bioconducter
library(dplyr)
library(nortest)
library(edgeR)
library(factoextra)
library(cowplot)
library(adegenet)
library(ggplot2)
library(matrixStats)
library(grid)
library(gridExtra)
library(RColorBrewer)
library(patchwork)
library(stringr)
library(tidyverse)
library(ggpubr)
pal <- brewer.pal(n = 3, name = "PRGn")
```

Load metadata
```{r}
#### Data ####
# Read in metadata
meta.data <- read.csv("RAnalysis/Data/metadata.RNAseq.csv", header=T, sep=",", na.string="NA", stringsAsFactors = F) #read in info
colnames(meta.data) <- c("colonyID","treatment","block","Sample_ID")
ID <- meta.data$Sample_ID

# Load go terms from functional annotation file
annot_GO <- read.csv("RAnalysis/Output/annot_GO.terms.host.csv")[-1]

# edit annot_GO to list GO terms all in one row per gene
annot_GO <- annot_GO %>%
  group_by(gene_id) %>%
  summarise(GO.ID = paste(GO.ID, collapse = ";")) %>%
  ungroup()
```

Processing expression data
```{r}
counts <- read.csv("RAnalysis/Data/Poc_gene_count_matrix.csv", header = TRUE, row.names = 1) #read in genes count matrix
colnames(counts) <- str_sub(colnames(counts), 1, 3) #remove everything after the first _ to just have the sample names
# Order columns by sample name
colnames(counts) <- c("C17","C18","C19","C20","C21","C22","C23","C24","C25","C26","C27","C28","C29","C30","C31","C32","E10","E11","E12","E13","E14","E15","E16","E1","E2","E3", "E4", "E5", "E6", "E7", "E8", "E9")
counts <- counts[, colnames(counts) %in% ID] #intersect with ID to maintain only samples a
counts <- counts[, order(colnames(counts))]
dge_gene <- DGEList(counts,genes = rownames(counts))
test <- as.data.frame(dge_gene)
saveRDS(dge_gene, "RAnalysis/Data/RNA_gene_preNormalization_DGEListObj.Rdata")
```

Create Methylation Summary Table
```{r}
#SAMPLES E7, E8, AND C28 WERE EXCLUDED DUE TO DEDUPLICATION AND LOW COUNTS
# Read in gene methylation data
gene_dnam <- readRDS("RAnalysis/Data/filtered_bed_SNP_file.Rdata") #read in info

#rename genewiz sample.ID to include corresponding E and C in front for enriched and control libraries for later analysis
gene_dnam <- gene_dnam %>%
    mutate(Sample.ID = recode(Sample.ID, '1' = 'E3', '3' = 'E14', '4' = 'C26', '5' = 'C31', '6' = 'C29', '7' = 'E13','8' = 'C20', '9' = 'C18', '10' = 'C24','11' = 'E10', '12' = 'C27', '13' = 'C22', '14' = 'E11', '15' = 'E4', '17' = 'E6', '18' = 'C21', '20' = 'E12', '21' = 'E5', '22' = 'C30', '23' = 'E1', '24' = 'C23', '25' = 'E16', '26' = 'C19', '27' = 'C25', '28' = 'E2', '29' = 'E9', '30' = 'C32', '31' = 'E15', '32' = 'C17'))

#check to see all where renamed correctly
unique(gene_dnam$Sample.ID)

# Count unique loci before filtering
unique_loci_before <- gene_dnam %>%
  summarise(unique_loci = n_distinct(paste(scaffold, position, sep = "_")))

# Print the result
print(paste("Unique loci before filtering:", unique_loci_before$unique_loci))

# Change from long to wide format and calculate mean methylation per gene and sample
meth <- gene_dnam %>%
  group_by(Sample.ID, gene) %>%
  summarise(mean = mean(per.meth), n = n()) %>%
  spread(key = Sample.ID, value = mean)

# Count unique loci after filtering
# Note: After filtering, we need to re-evaluate the original dataset to see which loci remain
remaining_genes <- meth$gene

# Filter original dataset to include only remaining genes
gene_dnam_filtered <- gene_dnam %>%
  filter(gene %in% remaining_genes)

unique_loci_after <- gene_dnam_filtered %>%
  summarise(unique_loci = n_distinct(paste(scaffold, position, sep = "_")))

# Print the result
print(paste("Unique loci after filtering:", unique_loci_after$unique_loci))


#change from long to wide format for filtered ge
meth <- data.frame (gene_dnam_filtered %>%
                   group_by(Sample.ID, gene) %>%
                   summarise(mean = mean(per.meth), n = n()) %>%
                   spread(., key = Sample.ID, value = mean))

#calculate percent meth per gene
rownames(meth) <- meth$gene
meth <- meth[,-1]
meth <- as.matrix(meth)
meth <- meth/100
saveRDS(meth, "RAnalysis/Data/meth_dataset.RData")

meth.test <- data_frame(meth)

## Create a summary of methylation for each gene ##
mean_meth <- rowMeans(meth)
meta.data <- na.omit(meta.data)

# create list for sample IDs from metadata for treatment and select from that list with filtering, pipe, row means
list_E <- filter(meta.data, treatment == "enriched") #filter only enriched data from metadata
list_C <- filter(meta.data, treatment == "control") #filter only control data from metadata
list_E <- as.list(list_E$Sample_ID) #make list of just enriched IDs
list_C <- as.list(list_C$Sample_ID) #make list of just control IDs

# gene meth enriched 
subset_meth_E <- meth[, colnames(meth)%in%list_E] #subset data from main m matrix for just enriched data

mean_meth_E <- rowMeans(subset_meth_E)

# Mean gene meth control
subset_meth_C <- meth[, colnames(meth)%in%list_C] #subset data from main m matrix for just control data

mean_meth_C <- rowMeans(subset_meth_C)

# Coefficient of Variation for enriched samples
sd_meth_E <- rowSds(subset_meth_E)
cv_meth_E <- sd_meth_E/mean_meth_E
 
# Coefficient of Variation for control
sd_meth_C <- rowSds(subset_meth_C)
cv_meth_C <- sd_meth_C/mean_meth_C

# Coefficient of Variation among treatments
meanAmongTrt_meth <- cbind(mean_meth_E,mean_meth_C)
sd_meanAmongTrt_meth <- rowSds(meanAmongTrt_meth)
mean_meanAmongTrt_meth <- rowMeans(meanAmongTrt_meth)
cv_mean_AmongTrt_meth <- sd_meanAmongTrt_meth/mean_meanAmongTrt_meth

# Difference in gene methylation between treatments
diff_Trt_meth <- mean_meth_E-mean_meth_C

# Creat matrix of all CpG summary stats
meth_summary <- cbind(mean_meth,mean_meth_E,mean_meth_C,cv_meth_E,cv_meth_C,cv_mean_AmongTrt_meth,
                      diff_Trt_meth) # bind all data frames with summary stats
meth_summary[is.na(meth_summary)]<-0 #filter for all methylation data over 0
meth_summary <- data.frame(gene_id = rownames(meth_summary),meth_summary) #change gene ID column name

# Saving full meth_annot data.frame
saveRDS(meth_summary,"RAnalysis/Data/meth_annot_full.RData")

#### Summarize the gene expression data ####
expression <- dge_gene$counts
subset_expression <- expression[rowSums(expression[])>0,]
expression <- expression[rownames(expression) %in% rownames(subset_expression),]
expression_cpm_log <- log2(cpm(expression)+ 1) # log2 cpm transformation
expression <- expression_cpm_log

# Mean gene expression all samples
mean_express <- rowMeans(expression)

# mean gene expression enriched 
subset_express_E <- expression[, colnames(expression)%in%list_E] #subset data from main m matrix for just enriched data

mean_express_E <- rowMeans(subset_express_E) #take mean per gene ID

# Mean gene expression control
subset_express_C <- expression[, colnames(expression)%in%list_C] #subset data from main m matrix for just control data

mean_express_C <- rowMeans(subset_express_C) #take mean per gene ID

# Coefficient of Variation for control gene expression
sd_express_C <- rowSds(subset_express_C)
cv_express_C <- sd_express_C/mean_express_C

# Coefficient of Variation for enriched gene expression
sd_express_E <- rowSds(subset_express_E)
cv_express_E <- sd_express_E/mean_express_E

# Standard Error for control gene expression
se_express_C <- sd_express_C/sqrt(length((subset_express_C)))

# Standard Error for enriched gene expression
se_express_E <- sd_express_E/sqrt(length((subset_express_E)))

# Coefficient of Variation among treatments
meanAmongTrt_express <- cbind(mean_express_E,mean_express_C)
sd_meanAmongTrt_express <- rowSds(meanAmongTrt_express)
mean_meanAmongTrt_express <- rowMeans(meanAmongTrt_express)
cv_mean_AmongTrt_express <- sd_meanAmongTrt_express/mean_meanAmongTrt_express

# Difference in gene expression between Trt
diff_Trt_express <- mean_express_E-mean_express_C

# Create matrix of all gene expression summary stats
gene_summary <- cbind(mean_express,mean_express_E,mean_express_C, se_express_C, se_express_E, cv_express_E,cv_express_C,cv_mean_AmongTrt_express,
                      diff_Trt_express)
gene_summary[is.na(gene_summary)]<-0
colnames(gene_summary) <- paste0("gene_",colnames(gene_summary))
gene_summary <- data.frame(gene_id = rownames(gene_summary),gene_summary)

# transcriptomic data
transcriptomic <- read.csv("RAnalysis/Output/rename_structure_gff3.csv") #load annotation GFF3 file

#remove first three columns in data frame
transcriptomic <- transcriptomic[,-(1:3),drop=FALSE]
transcriptomic <- transcriptomic[,-(4:6),drop=FALSE]

#add gene between Pver_g to match gene_summary
transcriptomic$gene_id <-  gsub("Pver_g","Pver_gene_g", transcriptomic$gene_id)

#combine gene summmary and gene IDs from annot
gene_annot <- left_join(gene_summary, transcriptomic)

#delete length and gene information that is unnecessary
gene_annot <- gene_annot[,-(9:13),drop=FALSE]

# Saving full gene_annot data.frame
saveRDS(gene_annot,"RAnalysis/Data/gene_annot_full.RData")

# Combine gene expression, exon count and CpG (DNA mehtylation data)
gSum <- full_join(meth_summary, gene_annot, by = "gene_id")

# Reducing the data.frame to include places were we have all the data.
gSum_red<-gSum[!is.na(gSum$cov10_count),]
gSum_red<-gSum_red[!is.na(gSum_red$gene_mean_express),]

# Saving full data.frame
saveRDS(gSum,"RAnalysis/Data/Multi_geneSummaryComplete.RData")

# Saving the reduced data.frame
#saveRDS(gSum_red,"RAnalysis/Output/Multi_geneSummaryReduced.RData")
```

#compare gene level meth vs express overall
```{r}
mat <- readRDS("RAnalysis/Output/Multi_geneSummaryReduced.RData") 

## Mean Methylation vs Gene Expression following Liew et al
df <- mat

#log gene expression x % methylation
df$log_gene <- log(df$gene_mean_express+1.1)
df$per_meth <- df$mean_meth * 100

# Test for normality
ks.test(df$gene_mean_express, "pnorm", mean = mean(df$gene_mean_express), sd = sd(df$gene_mean_express))
ks.test(df$mean_meth, "pnorm", mean = mean(df$mean_meth), sd = sd(df$mean_meth))

#plot of gene level express x meth
meth.express.compare <- ggplot(df, aes(x=per_meth, y=log_gene,)) + 
    geom_point() + ylim(0.5, 2.7) + xlim(0, 98) +
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
  stat_cor(method = "kendall", digits = 2.8, label.x = 60, label.y = 0.5) +
    theme_cowplot();meth.express.compare

### Figure Mean Methylation vs Gene Expression###
ggsave(meth.express.compare, filename="RAnalysis/Output/gene_express_meth_correlation.png", height = 7, width = 10)

```

## Analysis of comparisons between methylation and expression data from differentially methylated genes and differentially expressed genes.
# Import the data files 
```{r}
# read in DEGs
DEG <- read.csv("RAnalysis/Output/DEGSeq2.sig.results.host.csv")[-1]
# Reorder columns to move gene_id to the first position
DEG <- DEG %>%
  dplyr::select(gene_id, everything())

# read in DMGs
DMG <- read.csv("RAnalysis/Output/DMGs_all.csv")[-1]
# Rename column from 'gene' to 'gene_id'
DMG <- DMG %>%
  dplyr::rename(gene_id = gene)

# Remove extra characters in df and DMGs to match DEG
df_rename <- mat
df_rename$gene_id <-  gsub("_gene_g","_g",df_rename$gene_id) #sub gene_g with _g
df_rename$gene_id <-  gsub("_gene_","_",df_rename$gene_id) #remove extra _
df_rename$gene_id <- gsub("-.*", "", df_rename$gene_id) # Modify the gene_id column to remove the dash and anything after it
df_rename$gene_id <-  gsub("Pver_split_", "Pver_",df_rename$gene_id) #remove the extra _gene_ after split

# Remove characters in DMG to match DEG
DMG_rename <- DMG
DMG_rename$gene_id <-  gsub("_gene_g","_g",DMG_rename$gene_id) #sub gene_g with _g
DMG_rename$gene_id <-  gsub("_gene_","_",DMG_rename$gene_id) #remove extra _
DMG_rename$gene_id <- gsub("-.*", "", DMG_rename$gene_id) # Modify the gene_id column to remove the dash and anything after 
DMG_rename$gene_id <-  gsub("Pver_split_", "Pver_",DMG_rename$gene_id) #remove the extra _gene_ after split

#log gene expression x % methylation
df_rename$log_gene <- log(df_rename$gene_mean_express+1.1)
df_rename$per_meth <- df_rename$mean_meth * 100

# Double check that all gene_ids found in DEG and DMGs are in df_rename
missing_in_df_rename_DEG <- setdiff(DEG$gene_id, df_rename$gene_id) 
missing_in_df_rename_DMG <- setdiff(DMG_rename$gene_id, df_rename$gene_id) 

#Check which are missing from DEG in DMG dataset
missing_in_DMG_rename_DEG <- setdiff(DEG$gene_id, DMG_rename$gene_id) # 249 without methylation info, only 3 genes in both DEG and DMG datasets

# Create a combined list of unique gene_ids from both DMG and DEG
combined_gene_ids <- unique(c(DMG_rename$gene_id, DEG$gene_id)) 

# Filter df_rename to keep only rows with gene_ids in combined_gene_ids
df_filtered <- df_rename[df_rename$gene_id %in% combined_gene_ids, ]

# Check how many rows have non-NA values for both log_gene and per_meth
valid_rows <- df_filtered[!is.na(df_filtered$log_gene) & !is.na(df_filtered$per_meth), ]

# Count the number of such rows (genes)
num_genes_with_data <- nrow(valid_rows)

# Display the result
num_genes_with_data #451

# Shapiro-Wilk Test for normality
shapiro.test(df_filtered$gene_mean_express) # not normal
shapiro.test(df_filtered$mean_meth) # not normal

# look at just degs in df rename with methylaiton data
# Filter df_filtered to include only rows with gene_id found in DEG
df_DEG_rename <- df_rename %>%
  dplyr::semi_join(DEG, by = "gene_id")

#log gene expression x % methylation
df_DEG_rename$log_gene <- log(df_DEG_rename$gene_mean_express+1.1)
df_DEG_rename$per_meth <- df_DEG_rename$mean_meth * 100

DEG.meth.express <- ggplot(df_DEG_rename, aes(x=per_meth, y=log_gene)) + 
    geom_point() +
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
  stat_cor(method = "kendall", digits = 2.8, label.x = 10, label.y = 0.5 ) +
    theme_cowplot();DEG.meth.express

### Figure Mean Methylation vs Gene Expression###
ggsave(DEG.meth.express, filename="RAnalysis/Output/DEG_correlation.png", height = 10, width = 10)

```

## Analysis of comparisons between methylation and expression data from hypothesis driven genes
# Import the data files 
```{r}
# Read in hypothesis informed gene list
gene_list <- read.csv("Bioinformatics/Output/targetgenelist_GOterms.csv")

# Remove duplicates based on both Gene.ID and GO.term, so if duplicate gene IDs by GO term its removed
cleaned_data <- gene_list %>%
  distinct(Gene.ID, GO.term, .keep_all = TRUE)

# Count the number of unique genes based on 'gene_id'
unique_genes_count <- cleaned_data %>%
  distinct(Gene.ID) %>%
  tally() #3519

# Remove extra characters in df to match gene_id in all genes dataframe
gene_list_cleaned <- cleaned_data %>%
  dplyr::rename(gene_id = Gene.ID) %>%  # Rename Gene.ID to gene_id
  mutate(gene_id = gsub("_gene_g", "_g", gene_id),          # Replace "_gene_g" with "_g"
         gene_id = gsub("_gene_", "_", gene_id),            # Remove extra "_gene_"
         gene_id = gsub("-.*", "", gene_id),                # Remove dash and anything after it
         gene_id = gsub("Pver_split_", "Pver_", gene_id),   # Remove "Pver_split_"
         gene_id = gsub("\\..*", "", gene_id),              # Remove anything after the last period
         gene_id = gsub("_model", "", gene_id))             # Remove "_model"

# Ensure gene_list_cleaned is a vector of gene IDs
gene_list_vector <- gene_list_cleaned$gene_id

# Check if any gene ids are missing from df_rename
missing_in_rename <- setdiff(gene_list_vector, df_rename$gene_id) # 39 gene_ids missing did not have >0 gene counts

# Check common gene IDs
common_gene_ids <- intersect(gene_list_vector, df_rename$gene_id)

# Number of common gene IDs
length(common_gene_ids) #unique gene_ids = 2,069

# Filter the df_rename dataframe for only the gene_id values that are found in the gene_list_cleaned
gene_filtered <- df_rename %>%
  filter(gene_id %in% gene_list_vector)

# Select only the 'gene_mean_express' and 'mean_meth' columns from the 'gene_filtered' dataframe
selected_columns <- gene_filtered %>%
  dplyr::select(gene_id, gene_mean_express, mean_meth)

# Combine with gene_list_cleaned to output for github with meth and express values for genes 
# Combine 'selected_columns' with 'gene_list_cleaned'
combined_data <- gene_list_cleaned %>%
  left_join(selected_columns, by = "gene_id")

#log gene expression x % methylation
combined_data$log_gene <- log(combined_data$gene_mean_express+1.1)
combined_data$per_meth <- combined_data$mean_meth * 100

# Count rows with non-missing values in both 'mean_meth' and 'gene_mean_express'
count_non_missing <- combined_data %>%
  filter(!is.na(mean_meth) & !is.na(gene_mean_express)) %>%
  tally() #2,580 genes with meth and express data

# Select only the 'gene_mean_express' and 'per_meth' columns from the 'gene_filtered' dataframe
final_df <- combined_data %>%
  dplyr::select(gene_id, GO.term, Function, gene_mean_express, per_meth)

# write dataframe
write.csv(final_df, "Bioinformatics/Output/gene_list_meth_express_cleaned.csv")

# Shapiro-Wilk Test for normality
shapiro.test(combined_data$gene_mean_express) # not normal
shapiro.test(combined_data$mean_meth) # not normal

# Check how many rows have non-NA values for both log_gene and per_meth
valid_rows <- combined_data[!is.na(combined_data$log_gene) & !is.na(combined_data$per_meth), ] #574

# Make three plots for each function category

# Filter data for 'Nutrient metabolism'
nutrient_metabolism_data <- combined_data %>%
  filter(Function == "Nutrient metabolism")

# Plot for 'Nutrient metabolism'
nutrient_metabolism_plot <- ggplot(nutrient_metabolism_data, aes(x=per_meth, y=log_gene,)) + 
    geom_point() + 
  geom_smooth(method=lm, colour="orange", size=2, na.rm = TRUE) +
  scale_color_viridis() +
  labs(x="DNA Methylation (%)", y = "log10 (mean expression + 1)") +
  ggtitle("A. Nutrient Metabolism") +  # Add title
  stat_cor(method = "kendall", digits = 2, label.x = 5, label.y = 0.5) +
  theme_cowplot();nutrient_metabolism_plot

# Filter data for 'Symbiotic maintenance'
symbiotic_maintenance_data <- combined_data %>%
  filter(Function == "Symbiotic maintenance")

# Plot for 'Symbiotic maintenance'
symbiotic_maintenance_plot <- ggplot(symbiotic_maintenance_data, aes(x=per_meth, y=log_gene,)) + 
    geom_point() + 
  geom_smooth(method=lm, colour="orange", size=2, na.rm = TRUE) +
  scale_color_viridis() +
  labs(x="DNA Methylation (%)", y = "log10 (mean expression + 1)") +
  ggtitle("B. Symbiotic Maintenance") +  # Add title
  stat_cor(method = "kendall", digits = 2, label.x = 5, label.y = 0.5) +
  theme_cowplot();symbiotic_maintenance_plot

# Filter data for 'Symbiotic nutrient exchange and signaling pathways'
symbiotic_exchange_data <- combined_data %>%
  filter(Function == "Symbiotic nutrient exchange and signaling pathways")

# Plot for 'Symbiotic nutrient exchange and signaling pathways'
symbiotic_exchange_plot <- ggplot(symbiotic_exchange_data, aes(x=per_meth, y=log_gene,)) + 
    geom_point() + 
  geom_smooth(method=lm, colour="orange", size=2, na.rm = TRUE) +
  scale_color_viridis() +
  labs(x="DNA Methylation (%)", y = "log10 (mean expression + 1)") +
  ggtitle("C. Symbiotic Nutrient Exchange and Signaling Pathways") +  # Add title
  stat_cor(method = "kendall", digits = 2, label.x = 5, label.y = 0.5) +
  theme_cowplot();symbiotic_exchange_plot

# Combine three figures 
corr.figures <- plot_grid(meth.express.compare, DEG.meth.express, , labels = c('A', 'B', 'C', 'D'), label_size = 16)

ggsave(filename=paste0("RAnalysis/Output/correlation_figures.png"), corr.figures, dpi=300, width=20, height=10, units="in", limitsize = FALSE)

```























