---
title: "GO_Term_Enrichment_Analysis"
author: "daniellembecker"
edited by: "daniellembecker"
date: "9/20/2021"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Gene Ontology (GO) Enrichment Analysis 
## Previous steps include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DMG statistical data sheet
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
if (!requireNamespace("DESeq2", quietly = TRUE)) { BiocManager::install("DESeq2") }
library("DESeq2")
library("tidyr")
library("tidyverse")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
if (!requireNamespace("genefilter", quietly = TRUE)) { BiocManager::install("genefilter") }
library("genefilter")
library("ggplot2")
library("gplots")
if (!requireNamespace("limma", quietly = TRUE)) { BiocManager::install("limma") }
library("limma")
library("spdep") 
library("patchwork")
library("aDMGenet") 
if (!requireNamespace("goseq", quietly = TRUE)) { BiocManager::install("goseq") }
library("goseq")
library("gridExtra")
if (!requireNamespace("clusterProfiler", quietly = TRUE)) { BiocManager::install("clusterProfiler") }
library("clusterProfiler")
library("DataCombine")
library("VennDiagram")
if (!requireNamespace("GSEABase", quietly = TRUE)) { BiocManager::install("GSEABase") }
library("GSEABase")
library("data.table")
if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")
library("rrvgo")
if ("org.Ce.eg.db" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("org.Ce.eg.db")
library(org.Ce.eg.db)
if ("topGO" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("topGO")
library(topGO)
if ("Rgraphviz" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("Rgraphviz")
library(Rgraphviz)
```

# Import the data files 
```{r, echo=FALSE}
# Load go terms from functional annotation file
annot_GO <- read.csv("RAnalysis/Output/annot_GO.terms.host.csv")[-1]

# Remove rows where 'GO.ID' is 'unknown'
annot_GO <- subset(annot_GO, GO.ID != 'unknown')

# edit annot_GO to list GO terms all in one row per gene
annot_GO <- annot_GO %>%
  group_by(gene_id) %>%
  summarise(GO.ID = paste(GO.ID, collapse = ";")) %>%
  ungroup()

```

# Load in DMG results
```{r}
#DMG results
DMG.res <- read.csv("RAnalysis/Output/DMGs_all.csv")[-1]

colnames(DMG.res)[1] <-"gene_id" # make colnames a true column called gene_id
#remove the extra "gene" to match the "Pver_gene_g#" format
DMG.res$gene_id <-  gsub("Pver_gene_novel_gene_","Pver_novel_gene_", DMG.res$gene_id)
DMG.res$gene_id <-  gsub("Pver_gene_split_gene_", "Pver_gene_", DMG.res$gene_id)
#remove extra characters to match annot
DMG.res$gene_id <-  gsub("_gene_g","_g",DMG.res$gene_id) #sub gene_g with _g
DMG.res$gene_id <-  gsub("_gene_","_",DMG.res$gene_id) #remove extra _
DMG.res$gene_id <-  gsub("_split_","_",DMG.res$gene_id) #remove split naming in gene id

# Use the new MD.10x file that has now been filtered for SNPs

meth_filt <- readRDS("RAnalysis/Data/filtered_bed_SNP_file.Rdata") #read in info


# Rename gene column to gene_id and select only necessary columns, summarize meth
meth_filt_select <- meth_filt  %>%
  select(Sample.ID, meth, gene) %>%
  rename(gene_id = gene)

# Reshape the data
meth_wide <- meth_filt_select %>%
  group_by(gene_id, Sample.ID) %>%
  summarize(total_meth = sum(meth, na.rm = TRUE), .groups = "drop") %>%
  pivot_wider(names_from = Sample.ID, values_from = total_meth, values_fill = 0)

#remove the extra "gene" to match the "Pver_gene_g#" format
meth_wide$gene_id <-  gsub("Pver_gene_novel_gene_","Pver_novel_gene_", meth_wide$gene_id)
meth_wide$gene_id <-  gsub("Pver_gene_split_gene_", "Pver_gene_", meth_wide$gene_id)
#remove extra characters to match annot
meth_wide$gene_id <-  gsub("_gene_g","_g", meth_wide$gene_id) #sub gene_g with _g
meth_wide$gene_id <-  gsub("_gene_","_", meth_wide$gene_id) #remove extra _
meth_wide$gene_id <-  gsub("_split_","_", meth_wide$gene_id) #remove split naming in gene id
meth_wide$gene_id <-  gsub("-.*","",meth_wide$gene_id) #remove the extra g number after the dash

# Check if all gene IDs in GFF3 are included in  annot_GO
unmatched_genes <- setdiff(meth_wide$gene_id, annot_GO$gene_id)
# Print the unmatched gene IDs
print(unmatched_genes)

# Only genes that don't match after edited are the unknowns we filtered above
```

Match up genes in gene list file to annotation file
```{r}
names(annot_GO)

gcount_filt2annot = match(meth_wide$gene_id, annot_GO$gene_id) #match genes in gcount_filt to annot_GO

# The following is the number of probes without annotation 
sum(is.na(gcount_filt2annot))

row_nas<-which(is.na(gcount_filt2annot))

#view the genes that do not have a match in the annotation file
missing<-as.data.frame(gcount_filt2annot[row_nas])

```
364 genes in gcount_filt do not have GO annotations

Reduce annotation file to only contain genes detected in our dataset.  
```{r}
filtered_Pverr.annot <- annot_GO[annot_GO$gene_id %in% meth_wide$gene_id, ]
dim(filtered_Pverr.annot)

```

The annotation file now only contains genes that were detected in our dataset that have annotation information. This is 3,546 out of the 3,910 genes in our dataset at 10x coverage. 

Biological Processes topGO

Use topGO for enrichment analysis on hypermethylated data
```{r}
### Generate vector with names in just the contrast we are analyzing
ID.vector.hyper <- DMG.res %>%
      filter(direction=="hypermethylated") %>%
      pull(gene_id)
    
# Get a list of GO Terms for all genes detected in our dataset
GO.terms <- filtered_Pverr.annot %>%
  dplyr::select(gene_id, GO.ID) %>%
  separate_rows(GO.ID, sep = ";")

# Create a gene-to-GO mapping list
geneID2GO <- split(as.character(GO.terms$GO.ID), GO.terms$gene_id)
geneNames <- unique(GO.terms$gene_id)

# Create a named vector for allGenes
allGenes.hyper <- factor(as.integer(geneNames %in% ID.vector.hyper))
names(allGenes.hyper) <- geneNames

# Create a topGOdata object using the custom gene2GO mapping
GOdata.hyper_BP <- new("topGOdata",
              ontology = "BP",
              allGenes = allGenes.hyper,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hyper_BP
numSigGenes(GOdata.hyper_BP) #7736
sum(feasible(GOdata.hyper_BP)) #15385
```



Run topGO statistical analyses on compiled hypermethylated data
```{r}
# weight01 is the default algorithm used in Alexa et al. (2006) and Li et al. 2018
weight01.fisher.hyper_BP <- runTest(GOdata.hyper_BP, statistic = "fisher")

# Generate a results table
allRes_hyper_BP <- GenTable(GOdata.hyper_BP, 
                      classicFisher.hyper = weight01.fisher.hyper_BP, 
                      ranksOf = "classicFisher.hyper", 
                      topNodes = length(score(weight01.fisher.hyper_BP)))

# Convert character columns to numeric
allRes_hyper_BP$classicFisher.hyper <- as.numeric(allRes_hyper_BP$classicFisher.hyper)

# only adjust p values if you know your data is independent, multiple testing correction was not applied to the resulting P values as the tests are considered to be nonindependent
#allRes.hyper$weight_bh_adjust <- p.adjust(allRes.hyper$classicFisher.hyper, method="BH") #add adjusted p-values
#allRes.hyper

```

Create node figrues of top 5-10 sig terms for hypermethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hyper_BP, score(weight01.fisher.hyper_BP), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypermethylated_BP.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hyper_BP, weight01.fisher.hyper_BP, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```

Use topGO for enrichment analysis on hypomethylated data
```{r}
### Generate vector with names in just the contrast we are analyzing
ID.vector.hypo <- DMG.res %>%
      filter(direction=="hypomethylated") %>%
      pull(gene_id)
    
# Create a named vector for allGenes
allGenes.hypo <- factor(as.integer(geneNames %in% ID.vector.hypo))
names(allGenes.hypo) <- geneNames

# Create a topGOdata object using the custom gene2GO mapping
GOdata.hypo_BP <- new("topGOdata",
              ontology = "BP",
              allGenes = allGenes.hypo,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hypo_BP
numSigGenes(GOdata.hypo_BP) #7736
sum(feasible(GOdata.hypo_BP)) #15385
```



Run topGO statistical analyses on compiled hypomethylated data
```{r}
# weight01 is the default algorithm used in Alexa et al. (2006) and Li et al. 2018
weight01.fisher.hypo_BP <- runTest(GOdata.hypo_BP, statistic = "fisher")

# Generate a results table
allRes_hypo_BP <- GenTable(GOdata.hypo_BP, 
                      classicFisher.hypo = weight01.fisher.hypo_BP, 
                      ranksOf = "classicFisher.hypo", 
                      topNodes = length(score(weight01.fisher.hypo_BP)))

# Convert character columns to numeric
allRes_hypo_BP$classicFisher.hypo <- as.numeric(allRes_hypo_BP$classicFisher.hypo)

# only adjust p values if you know your data is independent, multiple testing correction was not applied to the resulting P values as the tests are considered to be nonindependent
#allRes.hyper$weight_bh_adjust <- p.adjust(allRes.hyper$classicFisher.hyper, method="BH") #add adjusted p-values
#allRes.hyper

```

Create node figrues of top 5-10 sig terms for hypomethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hypo_BP, score(weight01.fisher.hypo_BP), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypomethylated_BP.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hypo_BP, weight01.fisher.hypo_BP, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```



Molecular Functions topGO

Use topGO for enrichment analysis on hypermethylated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.hyper_MF <- new("topGOdata",
              ontology = "MF",
              allGenes = allGenes.hyper,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hyper_MF
numSigGenes(GOdata.hyper_MF) #1564
sum(feasible(GOdata.hyper_MF)) #2054
```

Run topGO statistical analyses on compiled hypermethylated data
```{r}
# weight01 is the default algorithm used in Alexa et al. (2006) and Li et al. 2018
weight01.fisher.hyper_MF <- runTest(GOdata.hyper_MF, statistic = "fisher")

# Generate a results table
allRes_hyper_MF <- GenTable(GOdata.hyper_MF, 
                      classicFisher.hyper = weight01.fisher.hyper_MF, 
                      ranksOf = "classicFisher.hyper", 
                      topNodes = length(score(weight01.fisher.hyper_MF)))

# Convert character columns to numeric
allRes_hyper_MF$classicFisher.hyper <- as.numeric(allRes_hyper_MF$classicFisher.hyper)

# only adjust p values if you know your data is independent, multiple testing correction was not applied to the resulting P values as the tests are considered to be nonindependent
#allRes.hyper$weight_bh_adjust <- p.adjust(allRes.hyper$classicFisher.hyper, method="BH") #add adjusted p-values
#allRes.hyper

```

Create node figrues of top 5-10 sig terms for hypermethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hyper_MF, score(weight01.fisher.hyper_MF), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypermethylated_MF.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hyper_MF, weight01.fisher.hyper_MF, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```

Use topGO for enrichment analysis on hypomethylated data
```{r}
# Create a topGOdata object using the custom gene2GO mapping
GOdata.hypo_MF <- new("topGOdata",
              ontology = "MF",
              allGenes = allGenes.hypo,
              nodeSize = 5,
              annot = annFUN.gene2GO, 
              gene2GO = geneID2GO)

GOdata.hypo_MF
numSigGenes(GOdata.hypo_MF) #8165
sum(feasible(GOdata.hypo_MF)) #16359
```


Run topGO statistical analyses on compiled hypomethylated data
```{r}
# weight01 is the default algorithm used in Alexa et al. (2006) and Li et al. 2018
weight01.fisher.hypo_MF <- runTest(GOdata.hypo_MF, statistic = "fisher")

# Generate a results table
allRes_hypo_MF <- GenTable(GOdata.hypo_MF, 
                      classicFisher.hypo = weight01.fisher.hypo_MF, 
                      ranksOf = "classicFisher.hypo", 
                      topNodes = length(score(weight01.fisher.hypo_MF)))

# Convert character columns to numeric
allRes_hypo_MF$classicFisher.hypo <- as.numeric(allRes_hypo_MF$classicFisher.hypo)

# only adjust p values if you know your data is independent, multiple testing correction was not applied to the resulting P values as the tests are considered to be nonindependent
#allRes.hyper$weight_bh_adjust <- p.adjust(allRes.hyper$classicFisher.hyper, method="BH") #add adjusted p-values
#allRes.hyper

```

Create node figrues of top 5-10 sig terms for hypomethylated genes
```{r}
# view plot in line first
showSigOfNodes(GOdata.hypo_MF, score(weight01.fisher.hypo_MF), firstSigNodes = 5, useInfo = 'all')

# Set the file path and name
pdf_file_path <- "RAnalysis/Output/tGO_graph_hypomethylated_MF.pdf"
# Open a PDF device
pdf(pdf_file_path)
# Call printGraph
printGraph(GOdata.hypo_MF, weight01.fisher.hypo_MF, firstSigNodes = 10, fn.prefix = "tGO", useInfo = "all", pdfSW = FALSE)
# Close the PDF device
dev.off()
```

Filter all dataframes for pvalue < 0.05
Combine all dataframes to make dendograms next, add ontolology column and up vs. hypomethylated column
```{r}
# Filter and add columns in one line for each dataframe
filt.hyper.BP <- allRes_hyper_BP %>% 
  filter(classicFisher.hyper < 0.05) %>% 
  mutate(ontology = "BP", methylation status = "hypermethylated")

filt.hyper.MF <- allRes_hyper_MF %>% 
  filter(classicFisher.hyper < 0.05) %>% 
  mutate(ontology = "MF", methylation status = "hypermethylated")

filt.hypo.BP <- allRes_hypo_BP %>% 
  filter(classicFisher.hypo < 0.05) %>% 
  mutate(ontology = "BP", methylation status = "hypomethylated")

filt.hypo.MF <- allRes_hypo_MF %>% 
  filter(classicFisher.hypo < 0.05) %>% 
  mutate(ontology = "MF", methylation status = "hypomethylated")

```


