---
title: "DMG_DEG_pathway_analysis"
author: "Danielle M. Becker"
date: "20220720"
output: html_document
editor_options: 
  chunk_output_type: console
---

#### Gene Expression and DNA Methylation comparison ####
## Used to compare general relationship between DNA methylation and gene expression
## Code developed from Javier A Rodriguez-Casariego paper: https://onlinelibrary.wiley.com/doi/pdf/10.1111/mec.16246?casa_token=DGoNlwKZByAAAAAA:i6JDcylh98txNEMQnr6dUTFxKd1sX7xotbSMZJq-DDOyqZGUPHObc30PhaWOAqpEufHMcJJ_wlQbQvbG
## https://github.com/jarcasariego/ACER_clonal_divergence/blob/main/GE_GbM_association/rmd/GE_DNAm_control.Rmd

```{r load packages}

## load packages
if (!requireNamespace("ComplexHeatmap", quietly = TRUE)) { BiocManager::install("ComplexHeatmap") }
library(ComplexHeatmap)
library(clusterProfiler)
library(pheatmap)
library(magrittr)
library(rtracklayer)
library(GenomicRanges)
library(png)
library(stats)
library(GO.db)
library(AnnotationDbi)
library(viridis)
library(janitor)
library(rtracklayer) # on bioconducter
library(dplyr)
library(nortest)
library(edgeR)
library(factoextra)
library(cowplot)
library(adegenet)
library(ggplot2)
library(matrixStats)
library(grid)
library(gridExtra)
library(RColorBrewer)
library(patchwork)
library(stringr)
library(tidyverse)
library(ggpubr)
pal <- brewer.pal(n = 3, name = "PRGn")
```

Load metadata
```{r}
#### Data ####
# Read in metadata
meta.data <- read.csv("RAnalysis/Data/metadata.RNAseq.csv", header=T, sep=",", na.string="NA", stringsAsFactors = F) #read in info
colnames(meta.data) <- c("colonyID","treatment","block","Sample_ID")
ID <- meta.data$Sample_ID

# Load go terms from functional annotation file
annot_GO <- read.csv("RAnalysis/Output/annot_GO.terms.host.csv")[-1]

# edit annot_GO to list GO terms all in one row per gene
annot_GO <- annot_GO %>%
  group_by(gene_id) %>%
  summarise(GO.ID = paste(GO.ID, collapse = ";")) %>%
  ungroup()
```

Processing expression data
```{r}
counts <- read.csv("RAnalysis/Data/Poc_gene_count_matrix.csv", header = TRUE, row.names = 1) #read in genes count matrix
colnames(counts) <- str_sub(colnames(counts), 1, 3) #remove everything after the first _ to just have the sample names
# Order columns by sample name
colnames(counts) <- c("C17","C18","C19","C20","C21","C22","C23","C24","C25","C26","C27","C28","C29","C30","C31","C32","E10","E11","E12","E13","E14","E15","E16","E1","E2","E3", "E4", "E5", "E6", "E7", "E8", "E9")
counts <- counts[, colnames(counts) %in% ID] #intersect with ID to maintain only samples a
counts <- counts[, order(colnames(counts))]
dge_gene <- DGEList(counts,genes = rownames(counts))
test <- as.data.frame(dge_gene)
saveRDS(dge_gene, "RAnalysis/Data/RNA_gene_preNormalization_DGEListObj.Rdata")
```

Create Methylation Summary Table
```{r}
#SAMPLES E7, E8, AND C28 WERE EXCLUDED DUE TO DEDUPLICATION AND LOW COUNTS
# Read in gene methylation data
gene_dnam <- readRDS("RAnalysis/Data/filtered_bed_SNP_file.Rdata") #read in info

#rename sample.ID to include corresponding E and C in front for enriched and control for later analysis
gene_dnam <- gene_dnam %>%
    mutate(Sample.ID = recode(Sample.ID, '1' = 'E1', '2' = 'E2', '3' = 'E3', '4' = 'E4', '5' = 'E5', '6' = 'E6', '9' = 'E9','10' = 'E10', '11' = 'E11', '12' = 'E12','13' = 'E13', '14' = 'E14', '15' = 'E15', '16' = 'E16', '17' = 'C17', '18' = 'C18', '19' = 'C19', '20' = 'C20', '21' = 'C21', '22' = 'C22', '23' = 'C23', '24' = 'C24', '25' = 'C25', '26' = 'C26', '27' = 'C27', '29' = 'C29', '30' = 'C30', '31' = 'C31', '32' = 'C32'))

#check to see all where renamed correctly
unique(gene_dnam$Sample.ID)

# Count unique loci before filtering
unique_loci_before <- gene_dnam %>%
  summarise(unique_loci = n_distinct(paste(scaffold, position, sep = "_")))

# Print the result
print(paste("Unique loci before filtering:", unique_loci_before$unique_loci))

# Change from long to wide format and calculate mean methylation per gene and sample
meth <- gene_dnam %>%
  group_by(Sample.ID, gene) %>%
  summarise(mean = mean(per.meth), n = n()) %>%
  spread(key = Sample.ID, value = mean)

# Filter for a minimum of 5 CpG sites per gene
meth_filtered <- meth[which(meth$n >= 5), ]

# Count unique loci after filtering
# Note: After filtering, we need to re-evaluate the original dataset to see which loci remain
remaining_genes <- meth_filtered$gene

# Filter original dataset to include only remaining genes
gene_dnam_filtered <- gene_dnam %>%
  filter(gene %in% remaining_genes)

unique_loci_after <- gene_dnam_filtered %>%
  summarise(unique_loci = n_distinct(paste(scaffold, position, sep = "_")))

# Print the result
print(paste("Unique loci after filtering:", unique_loci_after$unique_loci))


#change from long to wide format
meth <- data.frame (gene_dnam %>%
                   group_by(Sample.ID, gene) %>%
                   summarise(mean = mean(per.meth), n = n()) %>%
                   spread(., key = Sample.ID, value = mean))

#filter for a min 5 Cpg per gene
meth <- meth[which(meth$n >= 5), ]
cpg10xgene <- meth[,1:2]  
colnames(cpg10xgene) <- c("gene_id", "cov10_count")
rownames(meth) <- meth$gene
meth <- meth[,-1]
meth <- as.matrix(meth)
meth <- meth/100
saveRDS(meth, "RAnalysis/Data/meth_dataset.RData")

meth.test <- data_frame(meth)

## Create a summary of methylation for each gene ##
mean_meth <- rowMeans(meth)
meta.data <- na.omit(meta.data)

# create list for sample IDs from metadata and select from that list with filtering, pipe, row means
list_E <- filter(meta.data, treatment == "enriched") #filter only enriched data from metadata
list_C <- filter(meta.data, treatment == "control") #filter only control data from metadata
list_E <- as.list(list_E$Sample_ID) #make list of just enriched IDs
list_C <- as.list(list_C$Sample_ID) #make list of just control IDs

# mean gene meth enriched 
subset_meth_E <- meth[, colnames(meth)%in%list_E] #subset data from main m matrix for just enriched data

mean_meth_E <- rowMeans(subset_meth_E)

# Mean gene meth control
subset_meth_C <- meth[, colnames(meth)%in%list_C] #subset data from main m matrix for just control data

mean_meth_C <- rowMeans(subset_meth_C)

# Coefficient of Variation for enriched samples
sd_meth_E <- rowSds(subset_meth_E)
cv_meth_E <- sd_meth_E/mean_meth_E
 
# Coefficient of Variation for control
sd_meth_C <- rowSds(subset_meth_C)
cv_meth_C <- sd_meth_C/mean_meth_C

# Coefficient of Variation among treatments
meanAmongTrt_meth <- cbind(mean_meth_E,mean_meth_C)
sd_meanAmongTrt_meth <- rowSds(meanAmongTrt_meth)
mean_meanAmongTrt_meth <- rowMeans(meanAmongTrt_meth)
cv_mean_AmongTrt_meth <- sd_meanAmongTrt_meth/mean_meanAmongTrt_meth

# Difference in gene methylation between treatments
diff_Trt_meth <- mean_meth_E-mean_meth_C

# Creat matrix of all CpG summary stats
meth_summary <- cbind(mean_meth,mean_meth_E,mean_meth_C,cv_meth_E,cv_meth_C,cv_mean_AmongTrt_meth,
                      diff_Trt_meth) # bind all data frames with summary stats
meth_summary[is.na(meth_summary)]<-0 #filter for all methylation data over 0
meth_summary <- data.frame(gene_id = rownames(meth_summary),meth_summary) #change gene ID column name
meth_annot <- left_join(cpg10xgene, meth_summary) #join Cpg and meth summary

# Saving full meth_annot data.frame
saveRDS(meth_annot,"RAnalysis/Data/meth_annot_full.RData")

#### Summarize the gene expression data ####
expression <- dge_gene$counts
subset_expression <- expression[rowSums(expression[])>0,]
expression <- expression[rownames(expression) %in% rownames(subset_expression),]
expression_cpm_log <- log2(cpm(expression)+ 1) # log2 cpm transformation
expression <- expression_cpm_log

# Mean gene expression all samples
mean_express <- rowMeans(expression)

# mean gene expression enriched 
subset_express_E <- expression[, colnames(expression)%in%list_E] #subset data from main m matrix for just enriched data

mean_express_E <- rowMeans(subset_express_E) #take mean per gene ID

# Mean gene expression control
subset_express_C <- expression[, colnames(expression)%in%list_C] #subset data from main m matrix for just control data

mean_express_C <- rowMeans(subset_express_C) #take mean per gene ID

# Coefficient of Variation for control gene expression
sd_express_C <- rowSds(subset_express_C)
cv_express_C <- sd_express_C/mean_express_C

# Coefficient of Variation for enriched gene expression
sd_express_E <- rowSds(subset_express_E)
cv_express_E <- sd_express_E/mean_express_E

# Standard Error for control gene expression
se_express_C <- sd_express_C/sqrt(length((subset_express_C)))

# Standard Error for enriched gene expression
se_express_E <- sd_express_E/sqrt(length((subset_express_E)))

# Coefficient of Variation among treatments
meanAmongTrt_express <- cbind(mean_express_E,mean_express_C)
sd_meanAmongTrt_express <- rowSds(meanAmongTrt_express)
mean_meanAmongTrt_express <- rowMeans(meanAmongTrt_express)
cv_mean_AmongTrt_express <- sd_meanAmongTrt_express/mean_meanAmongTrt_express

# Difference in gene expression between Trt
diff_Trt_express <- mean_express_E-mean_express_C

# Create matrix of all gene expression summary stats
gene_summary <- cbind(mean_express,mean_express_E,mean_express_C, se_express_C, se_express_E, cv_express_E,cv_express_C,cv_mean_AmongTrt_express,
                      diff_Trt_express)
gene_summary[is.na(gene_summary)]<-0
colnames(gene_summary) <- paste0("gene_",colnames(gene_summary))
gene_summary <- data.frame(gene_id = rownames(gene_summary),gene_summary)

# transcriptomic data
transcriptomic <- read.csv("RAnalysis/Output/rename_structure_gff3.csv") #load annotation GFF3 file

#remove first three columns in data frame
transcriptomic <- transcriptomic[,-(1:3),drop=FALSE]
transcriptomic <- transcriptomic[,-(4:6),drop=FALSE]

#add gene between Pver_g to match gene_summary
transcriptomic$gene_id <-  gsub("Pver_g","Pver_gene_g", transcriptomic$gene_id)

#combine gene summmary and gene IDs from annot
gene_annot <- left_join(gene_summary, transcriptomic)

#delete length and gene information that is unnecessary
gene_annot <- gene_annot[,-(9:13),drop=FALSE]

# Saving full gene_annot data.frame
saveRDS(gene_annot,"RAnalysis/Data/gene_annot_full.RData")

# Combine gene expression, exon count and CpG (DNA mehtylation data)
gSum <- full_join(meth_annot, gene_annot, by = "gene_id")

# Reducing the data.frame to include places were we have all the data.
gSum_red<-gSum[!is.na(gSum$cov10_count),]
gSum_red<-gSum_red[!is.na(gSum_red$gene_mean_express),]

# Saving reduced data.frame
saveRDS(gSum,"RAnalysis/Data/Multi_geneSummaryComplete.RData")

# Saving the reduced data.frame
saveRDS(gSum_red,"RAnalysis/Data/Multi_geneSummaryReduced.RData")
```

#compare global mean meth vs express overall
```{r}
mat <- readRDS("RAnalysis/Output/Multi_geneSummaryReduced_all.RData") 

## Mean Methylation vs Gene Expression following Liew et al
df <- mat

#log gene expression x % methylation
df$log_gene <- log(df$gene_mean_express+1.1)
df$per_meth <- df$mean_meth * 100

# Test for normality
ks.test(df$gene_mean_express, "pnorm", mean = mean(df$gene_mean_express), sd = sd(df$gene_mean_express))
ks.test(df$mean_meth, "pnorm", mean = mean(df$mean_meth), sd = sd(df$mean_meth))

#plot of global express x meth
meth.express.compare <- ggplot(df, aes(x=per_meth, y=log_gene,)) + 
    geom_point() + ylim(0.5, 2.7) + xlim(0, 98) +
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
  stat_cor(method = "kendall", digits = 2.8, label.x = 60, label.y = 0.5) +
    theme_cowplot();meth.express.compare

### Figure Mean Methylation vs Gene Expression###
ggsave(meth.express.compare, filename="RAnalysis/Output/global_express_meth_correlation.png", height = 7, width = 10)

```

## Analysis of comparisons between methylation and expression data from differentially methylated genes and differentially expressed genes.
# Import the data files 
```{r}
# read in DEGs
DEG <- read.csv("RAnalysis/Output/DEGSeq2.sig.results.host.csv")[-1]
# Reorder columns to move gene_id to the first position
DEG <- DEG %>%
  select(gene_id, everything())

# read in DMGs
DMG <- read.csv("RAnalysis/Output/DMGs_all.csv")[-1]
# Rename column from 'gene' to 'gene_id'
DMG <- DMG %>%
  rename(gene_id = gene)

# Remove extra characters in df and DMGs to match DEG
df_rename <- gSum
df_rename$gene_id <-  gsub("_gene_g","_g",df_rename$gene_id) #sub gene_g with _g
df_rename$gene_id <-  gsub("_gene_","_",df_rename$gene_id) #remove extra _
df_rename$gene_id <- gsub("-.*", "", df_rename$gene_id) # Modify the gene_id column to remove the dash and anything after it
df_rename$gene_id <-  gsub("Pver_split_", "Pver_",df_rename$gene_id) #remove the extra _gene_ after split

# Remove characters in DMG to match DEG
DMG_rename <- DMG
DMG_rename$gene_id <-  gsub("_gene_g","_g",DMG_rename$gene_id) #sub gene_g with _g
DMG_rename$gene_id <-  gsub("_gene_","_",DMG_rename$gene_id) #remove extra _
DMG_rename$gene_id <- gsub("-.*", "", DMG_rename$gene_id) # Modify the gene_id column to remove the dash and anything after it
DMG_rename$gene_id <-  gsub("Pver_split_", "Pver_",DMG_rename$gene_id) #remove the extra _gene_ after split

#log gene expression x % methylation
df_rename$log_gene <- log(df_rename$gene_mean_express+1.1)
df_rename$per_meth <- df_rename$mean_meth * 100

# Double check that all gene_ids found in DEG and DMGs are in df_rename
missing_in_df_rename_DEG <- setdiff(DEG$gene_id, df_rename$gene_id) 
missing_in_df_rename_DMG <- setdiff(DMG_rename$gene_id, df_rename$gene_id) 

#Check which are missing from DEG in DMG dataset
missing_in_DMG_rename_DEG <- setdiff(DEG$gene_id, DMG_rename$gene_id) # 249 without methylation info, only 3 genes in both DEG and DMG datasets

# Create a combined list of unique gene_ids from both DMG and DEG
combined_gene_ids <- unique(c(DMG_rename$gene_id, DEG$gene_id)) 

# Filter df_rename to keep only rows with gene_ids in combined_gene_ids
df_filtered <- df_rename[df_rename$gene_id %in% combined_gene_ids, ]

# Check how many rows have non-NA values for both log_gene and per_meth
valid_rows <- df_filtered[!is.na(df_filtered$log_gene) & !is.na(df_filtered$per_meth), ]

# Count the number of such rows (genes)
num_genes_with_data <- nrow(valid_rows)

# Display the result
num_genes_with_data #355

# Shapiro-Wilk Test for normality
shapiro.test(df_filtered$gene_mean_express) # not normal
shapiro.test(df_filtered$mean_meth) # not normal

#plot of DMG METH EXPRESS, 249 DEGs do not have methylation data due to filtering for methylation so really a DMG correlation
DMG.meth.express <- ggplot(df_filtered, aes(x=per_meth, y=log_gene,)) + 
    geom_point() + ylim(0, 2.4) + xlim(0,50) +
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
  stat_cor(method = "kendall", digits = 2.8, label.x = 40, label.y = 0.5 ) +
    theme_cowplot();DMG.meth.express

### Figure Mean Methylation vs Gene Expression###
ggsave(DMG.meth.express, filename="RAnalysis/Output/DMG_correlation.png", height = 10, width = 10)


```

## Analysis of comparisons between methylation and expression data from hypothesis driven genes
# Import the data files 
```{r}
# Read in hypothesis informed gene list
gene_list <- read.csv("Bioinformatics/Output/Nutrient_genesofinterest_Pverr.csv")[-1]

# Format gene list and make all individual Pver genes have an individual row and remove any spaces before or after gene id
gene_list_separated <- gene_list %>%
  separate_rows(Pverr.gene.ID, sep = "; |;|,") %>%
  mutate(Pverr.gene.ID = str_trim(Pverr.gene.ID)) %>%
  mutate(Pverr.gene.ID = str_replace(Pverr.gene.ID, "^Pver_novel_model_", "Pver_novel_")) %>%
  rename(gene_id = Pverr.gene.ID)

# Ensure gene_list_separated is a vector of gene IDs
gene_list_vector <- gene_list_separated$gene_id

# Check if any gene ids are missing from df_rename
missing_in_rename <- setdiff(gene_list_vector, df_rename$gene_id) # 39 gene_ids missing did not have >0 gene counts

# Check common gene IDs
common_gene_ids <- intersect(gene_list_vector, df_rename$gene_id)

# Number of common gene IDs
length(common_gene_ids) #unique gene_ids = 386

# Filter the df_rename dataframe for only the gene_id values that are found in the gene_list_separated
gene_filtered <- df_rename %>%
  filter(gene_id %in% gene_list_vector)

# Shapiro-Wilk Test for normality
shapiro.test(df_filtered$gene_mean_express) # not normal
shapiro.test(df_filtered$mean_meth) # not normal

#plot of DEG X DMG METH EXPRESS
hypoth_compare <- ggplot(gene_filtered, aes(x=per_meth, y=log_gene,)) + 
    geom_point() + #ylim(0, 2.1) + xlim(0,7.5) +
    geom_smooth(method=lm,colour="orange",size=2,na.rm = TRUE) +
    scale_color_viridis() +
    labs(x="DNA Methylation (%)", y = "log10 (mean expression +1)") +
  stat_cor(method = "spearman", digits = 2.8, label.x = 5, label.y = 0.5 ) +
    theme_cowplot();hypoth_compare

# Combine three figures 
corr.figures <- plot_grid(meth.express.compare, DMG.meth.express, hypoth_compare, labels = c('A', 'B', 'C'), label_size = 16)

ggsave(filename=paste0("RAnalysis/Output/correlation_figures.png"), corr.figures, dpi=300, width=20, height=10, units="in", limitsize = FALSE)

```



Show fold chnage and z scores for change between enriched and control treatments in our genes of interest for gene expression
```{r}
# next steps mean change in expression for fold change or z score in each gene that goes from control to enriched for target gene list, how much do they change between control and enriched and then show how they go up or down in control and enriched, does nutrient enrichment affect genes of interest or do genes of interest change with control and enriched 

# next steps mean change in methylation for fold change or z score in each gene that goes from control to enriched for target gene list, how much do they change between control and enriched and then show how they go up or down in control and enriched, does nutrient enrichment affect genes of interest or do genes of interest change with control and enriched 

# x axis treatment then calculating relative change between control and enriched gene expression and some may go down and some may go up and then same for methylation (relative change in control and enriched) show relative change in expression and methylation in our genes of interest that are represented in gene expression and x out of x in 

# Remove extra characters in gene_summary to match gene_list
express.change <- gene_summary
express.change$gene_id <-  gsub("_gene_g","_g",express.change$gene_id) #sub gene_g with _g
express.change$gene_id <-  gsub("_gene_","_",express.change$gene_id) #remove extra _
express.change$gene_id <- gsub("-.*", "", express.change$gene_id) # Modify the gene_id column to remove the dash and anything after it
express.change$gene_id <-  gsub("Pver_split_", "Pver_",express.change$gene_id) #remove the extra _gene_ after split
colnames(express.change)[colnames(express.change) == "gene_id"] <- "gene" # Rename gene_id to gene

# Calculate for gene expression data:
# Filter gene_summary for the genes of interest
target_genes_summary <- express.change[express.change$gene %in% gene_list_vector,]

# Calculate fold change for the target genes
target_genes_summary$fold_change <- mean_express_E[rownames(target_genes_summary)] / mean_express_C[rownames(target_genes_summary)]
target_genes_summary$fold_change[is.infinite(target_genes_summary$fold_change)] <- NA  # Handle infinite values

# Scatterplot of fold change with y-axis intervals and customized text sizes
ggplot(target_genes_summary, aes(x = gene, y = fold_change, color = fold_change)) +
  geom_point(size = 3) +
  geom_hline(yintercept = 1, linetype = "dashed", color = "red", size = 1) +  # Line at fold change = 1
  theme_bw() +
  labs(title = "Fold Change in Gene Expression Genes of Interest Between Enriched and Control Treatments",
       x = NULL,  # Remove x-axis label
       y = "Fold Change") +
  theme(axis.text.x = element_blank(),  # Remove x-axis labels
        axis.ticks.x = element_blank(),  # Remove x-axis ticks
        axis.text.y = element_text(size = 14),  # Increase y-axis text size
        axis.title.y = element_text(size = 16),  # Increase y-axis title size
        plot.title = element_text(size = 18, face = "bold"),  # Increase plot title size
        legend.text = element_text(size = 14),  # Increase legend text size
        legend.title = element_text(size = 16)) +  # Increase legend title size
  scale_color_viridis_c() +  # Optional: use a color scale for better visibility
  scale_y_continuous(breaks = seq(min(target_genes_summary$fold_change, na.rm = TRUE), 
                                  max(target_genes_summary$fold_change, na.rm = TRUE), 
                                  by = 1))  # Set y-axis intervals at 1

```


Show fold chnage and z scores for change between enriched and control treatments in our genes of interest for methylation
```{r}

# Remove extra characters in gene_summary to match gene_list
meth.change <- meth_summary
meth.change$gene_id <-  gsub("_gene_g","_g",meth.change$gene_id) #sub gene_g with _g
meth.change$gene_id <-  gsub("_gene_","_",meth.change$gene_id) #remove extra _
meth.change$gene_id <- gsub("-.*", "", meth.change$gene_id) # Modify the gene_id column to remove the dash and anything after it
meth.change$gene_id <-  gsub("Pver_split_", "Pver_",meth.change$gene_id) #remove the extra _gene_ after split
colnames(meth.change)[colnames(meth.change) == "gene_id"] <- "gene" # Rename gene_id to gene

# Calculate for gene expression data:
# Filter gene_summary for the genes of interest
target_genes_meth_summary <- meth.change[meth.change$gene %in% gene_list_vector,]

# Calculate fold change for the target genes
target_genes_meth_summary$fold_change <- mean_express_E[rownames(target_genes_meth_summary)] / mean_express_C[rownames(target_genes_meth_summary)]
target_genes_meth_summary$fold_change[is.infinite(target_genes_meth_summary$fold_change)] <- NA  # Handle infinite values

# Scatterplot of fold change with y-axis intervals and customized text sizes
ggplot(target_genes_meth_summary, aes(x = gene, y = fold_change, color = fold_change)) +
  geom_point(size = 3) +
  geom_hline(yintercept = 1, linetype = "dashed", color = "red", size = 1) +  # Line at fold change = 1
  theme_bw() +
  labs(title = "Fold Change in Methylation Genes of Interest Between Enriched and Control Treatments",
       x = NULL,  # Remove x-axis label
       y = "Fold Change") +
  theme(axis.text.x = element_blank(),  # Remove x-axis labels
        axis.ticks.x = element_blank(),  # Remove x-axis ticks
        axis.text.y = element_text(size = 14),  # Increase y-axis text size
        axis.title.y = element_text(size = 16),  # Increase y-axis title size
        plot.title = element_text(size = 18, face = "bold"),  # Increase plot title size
        legend.text = element_text(size = 14),  # Increase legend text size
        legend.title = element_text(size = 16)) +  # Increase legend title size
  scale_color_viridis_c()   # Optional: use a color scale for better visibility
```























