---
title: "dendogram"
author: "daniellembecker"
date: "12/01/23"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## Dendogram of clusters of GO and KEGG terms, enrichment output

#load libraries
```{r}
library(ggplot2)
library(dendextend)
library(plyr)
library(circlize)
library(gridExtra)
library(data.table)
library(stringdist)
library(tidytree)
library(ggpubr)
library(viridis)
library(treeio)
library(ggtree)
library(stringr)
library(readr)
```

#input data
```{r}
f_gene_groups=c('RAnalysis/Data/WGBS/dendogram/dendogram_DMG_BP.csv') 
```

#output directory + base name for the output graph
```{r}
outidx=c('RAnalysis/Output/Dendograms/DMG_clusters.graphs')

f_gene_groups_sep=","
```
 

#names of columns in the input csv
```{r}
gene_groups_ids='genes'
gene_groups_direction='percentDEInCat' #calculated as (number of DEGs in each GO/total number of genes in GO term)*100 _this is the percentage of differentially expressed 

#genes in each term. It will be used to set the color in the heatmap
gene_groups_term='term'
gene_groups_experiment='experiment'
gene_groups_size='set_size'


```

# functions:

```{r}
dist_funct=function(v,func){
  n=length(v)
  m=matrix(data=NA,nrow=n,ncol=n)
  for(j in 1:n){
    if(j<n){
      for(k in (j+1):n){
        dist = func(v[j],v[k])
        m[j,k]= dist
        m[k,j]= dist
      }
    }
  }
  return(m)
}

dist_overlap=function(x,y){
  x=strsplit(x,';')[[1]]
  y=strsplit(y,';')[[1]]
  z=length(intersect(x,y))
  return(1-(z/min(length(x),length(y))))
}

pairwise_dists=function(ids,dist_mat_){
  dist_mat=as.matrix(dist_mat_)
  ids_hash=list()
  rownames0=rownames(dist_mat_)
  for(j in 1:length(rownames0)){ids_hash[[rownames0[j]]]=j}
  dists=c()
  n=length(ids)
  for(j in 1:n){
    if(j<n){
      for(k in (j+1):n){
        dists=c(dists,dist_mat[ids_hash[[ids[j]]],ids_hash[[ids[k]]]])
      }
    }
  }
  return(dists)
}



colors1 <- c('lightskyblue', 'dodgerblue', 'steelblue', 'royalblue', 'limegreen', 'seagreen', 'darkolivegreen', 'indianred', 'slategray', 'darkred', 'mediumblue', 'forestgreen', 'turquoise', 'gold', 'orchid', 'deeppink', "darkkhaki", "coral", "red", "pink",   rainbow(n=100))
max1=20   #you can try different numbers and choose the best clustering 
```


##### run a for loop for each term to make connections for dendogram
```{r}
for(i in 1:length(f_gene_groups)){
  #geneGroups1 = read.csv(f_gene_groups[i],sep=f_gene_groups_sep,stringsAsFactors=F)
  geneGroups1 = fread(f_gene_groups[i],sep=f_gene_groups_sep,stringsAsFactors=F)
  cols=c(gene_groups_ids,gene_groups_direction,gene_groups_term,gene_groups_experiment, gene_groups_size)
  geneGroups1 = geneGroups1[,..cols,with=F]
  names(geneGroups1) = c('ids','score','term','experiment', 'size')
  geneGroups2 = geneGroups1[,.(nonredudnatIDs = paste(unique(strsplit(gsub("\\s+","",paste(ids,collapse=';'),perl=T)[[1]],";")[[1]]),collapse=";")),by=term]
  geneGroups3 = as.data.frame(geneGroups2)
  d = dist_funct(geneGroups3[,'nonredudnatIDs'],dist_overlap)
  colnames(d)=geneGroups3[,'term']
  rownames(d)=geneGroups3[,'term']
  hc = hclust(as.dist(d),method="ward.D2") #hclust_method)
  hcd = as.dendrogram(hc)
  hc[['tip.label']] = hc[['labels']] }
  
  
  cuts0=1:max1
  cutr1=list()
  cutr_clusters1=list()
  for(u in 1:length(cuts0)){
    cat(u,"\n")
    cutr1[[u]] = cutree(tree=hc,k=cuts0[u])
    cutr_freqs = as.data.frame(table(cutr1[[u]]))
    cutr_clusters1[[u]]=data.frame(term_name=names(cutr1[[u]]),cluster=cutr1[[u]])
    cutr_clusters1[[u]]=cutr_clusters1[[u]][order(cutr_clusters1[[u]]$cluster),]
  }
  
  idx0 = as.integer(max1)
  cutr=cutr1[[idx0]]
  cuts=cuts0[idx0]
  cutr_clusters=cutr_clusters1[[idx0]]
  cutr_tightClusters=cutr_clusters1[[length(cutr_clusters1)]]
  names(cutr_tightClusters) = c('term','cluster')
  
# Uppercase the first letter of row names in the data frame
cutr_tightClusters$label_upper <- tools::toTitleCase(rownames(cutr_tightClusters))

# Replace "Dna" with "DNA" in the label_upper column
cutr_tightClusters$label_upper <- str_replace(cutr_tightClusters$label_upper, "Dna", "DNA")

# Replace "Mrna" with "mRNA" in the label_upper column
cutr_tightClusters$label_upper <- str_replace(cutr_tightClusters$label_upper, "Mrna", "mRNA")

# make clusters on tree and organize
gg1 = ggtree(hc, layout = 'rectangular') %<+% cutr_tightClusters + #edits layout of tree
  geom_tippoint(aes(color = as.factor(cluster)), shape = "triangle", size = 8) + #adds triangles and adjusts size
  scale_colour_manual(values = colors1) + #distinguishes colors of clusters with data
  geom_tiplab(aes(label = label_upper), offset = 0.3, size = 6) + #references the label_upper column to name tree nodes
  xlim(-3, 3) #sets x limites on plot
  
  pdf(paste0(outidx[i],'_tree_BP.pdf'))
  print(gg1)
  dev.off()
  
  scores1 = as.data.frame(dcast(geneGroups1[,.(term,score,experiment)], term ~ experiment,value.var='score',mean),stringsAsFactors=F)
  scores2 = scores1[,c(1:ncol(scores1))]
  
# Convert scores1 to data frame
scores2 <- as.data.frame(scores1[, c(1:ncol(scores1))])

# Remove the "term" column
scores2 <- select(scores2, -c("term"))

# Set row names of scores2
rownames(scores2) <- scores1[, 1]

# Create the heatmap
gg2 <- gheatmap(gg1, scores2, offset = 2.5, width = 0.1, font.size = 8)

# Apply custom gradient to the heatmap
gg2 <- gg2 + viridis::scale_fill_viridis(na.value=NA, option = "inferno")

# Output to PDF
pdf(paste0(outidx[i], '_tree_scores_BP.pdf'), width = 21, height = 15)
print(gg2)
dev.off()

```



#make dot plot seperate from dendorgram
```{r}
#calculate means for set size and percentDEIncat for each term
# Calculate mean score for each term
mean_scores <- geneGroups1 %>%
  group_by(term) %>%
  summarise(mean_score = mean(score))

# Calculate sum of set size for each term
sum_set_sizes <- geneGroups1 %>%
  group_by(term) %>%
  summarise(sum_set_size = sum(size))

#combine data frames by term
dotplot <- left_join(mean_scores, sum_set_sizes)

#read in dotplot data
clusters <- read_csv("RAnalysis/Data/WGBS/dendogram/DMG_dotplot_BP.csv", col_names = TRUE)

# Convert term_name to factor
clusters$term <- factor(clusters$term)

# Define the desired order of term names
term_order <- c("Carboxylic acid metabolic process", "Carbohydrate derivative metabolic process", "Intracellular protein transport", "Lipid metabolic process", "Muscle system process", "Regulation of cellular process", "Organelle localization", "Cell adhesion", "Protein maturation", "Cellular localization", "Regulation of catalytic activity", "Carbohydrate metabolic process", "Cytoskeleton organization", "Nucleic acid metabolic process", "DNA metabolic process", "Nitrogen compound metabolic process", "DNA biosynthetic process", "DNA replication", "DNA repair", "Telomere organization", "Cellular homeostasis", "Cell differentiation", "Transmembrane transport", "Transport", "DNA-templated transcription", "Cellular nitrogen compound", "Biological process involved in interspecies interaction", "Microtubule-based movement", "Mitotic cell cycle", "Regulation of gene expression", "Cell cycle process", "Response to nutrient levels", "Cytoplasmic translation", "Response to nitrogen compound", "Viral process", "Response to stress", "Response to virus", "Immune system process", "Positive regulation of translation", "Cell motility", "Regulation of phospholipase activity", "Protein modification process", "Protein metabolic process", "Behavior", "Wound healing", "Cellular component organization", "Membrane organization", "Signaling", "Reproductive process", "Cellular metabolic process", "Positive regulation of cellular pH", "Anatomical structure development")

# Reorder the term_name factor levels
clusters <- clusters %>%
  mutate(term = factor(term, levels = rev(term_order)))

# Create plot with reordered terms
gg3 = ggplot (clusters, aes(x = comparison, y = term, size = set_size)) + 
            geom_count (aes(color = percentDEInCat)) + 
            scale_color_viridis(na.value=NA, option = "inferno") +
            scale_size(limits = c(1, 1527)) +  # Set the limits here
            labs(color = "%DMGs", size = "Number of Genes") +
            theme(panel.background = element_blank())

print(gg3)

ggsave("RAnalysis/Output/Dendograms/clusters_dotplot_DMG_BP.png",
       dpi = 300, 
       width = 10,
       height = 10)

```


