---
title: "GO Semantic Similarity"
author: "Danielle Becker"
date: "Last updated 2023/06/27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Prepare workspace
```{r, message=FALSE, warning=FALSE}
rm(list=ls()) # removes all prior objects
library(tidyverse)
library(org.Hs.eg.db)
library(simplifyEnrichment)
library(GOSemSim)
library(DOSE)
library(magick)
library(gridtext)
library(clusterProfiler)
library(enrichplot)
# we use ggplot2 to add x axis labels (ex: ridgeplot)
library(ggplot2)
```

#Use simplifyenrichment package (http://www.bioconductor.org/packages/release/bioc/vignettes/simplifyEnrichment/inst/doc/simplifyEnrichment.html) to calculate semantic similarity for GO terms in gene expression and methylation
#All dataframes were made in RAnalysis - Scripts - Host - Host_GO_KEGG_Enrichment_Analysis_Func_Annot

#Create semantic similarity matrices for differentially expressed gene categories

Load GO data and select only the GO_IDs associated with BP terms
```{r, message=FALSE, warning=FALSE}
  MatGO <- read_csv("RAnalysis/Data/RNA-seq/P.verrucosa/Host/DEG_GO_parent_terms_complete.csv", col_names = TRUE)
nrow(MatGO) # total #enriched terms 307
nrow(filter(MatGO, ontology=="BP")) #BP terms 210
nrow(filter(MatGO, ontology=="CC")) #CC terms 27
nrow(filter(MatGO, ontology=="MF")) #MF terms 70
MatGO.BP <- MatGO %>% filter(ontology == "BP")
MatBP <- MatGO.BP$category
MatGO.MF <- MatGO %>% filter(ontology == "MF")
MatMF <- MatGO.MF$category
MatGO.CC <- MatGO %>% filter(ontology == "CC")
MatCC <- MatGO.CC$category
```


Calculate a similarity matrix and save the output.
```{r}
Mat = GO_similarity(MatBP, ont = "BP", db = "org.Hs.eg.db")
pdf(file = "RAnalysis/Output/RNA-seq/Semantic_similarity/DEG_BP_plot.pdf", width = 7, height = 5)
simplifyGO(Mat, word_cloud_grob_param = list(max_width = 50), max_words=20)
dev.off()

MatMF = GO_similarity(MatMF, ont = "MF", db = "org.Hs.eg.db")
pdf(file = "RAnalysis/Output/RNA-seq/Semantic_similarity/DEG_MF_plot.pdf", width = 7, height = 5)
simplifyGO(MatMF, word_cloud_grob_param = list(max_width = 50), max_words=20)
dev.off()

MatCC = GO_similarity(MatCC, ont = "CC", db = "org.Hs.eg.db")
pdf(file = "RAnalysis/Output/RNA-seq/Semantic_similarity/DEG_CC_plot.pdf", width = 7, height = 5)
simplifyGO(MatCC, word_cloud_grob_param = list(max_width = 50), max_words=20)
dev.off()
```


#try using treeplot to show similarity and plot DEGs
```{r}
# load data
DEG <- read_csv("RAnalysis/Data/RNA-seq/P.verrucosa/Host/DEG_GO_parent_terms_complete.csv", col_names = TRUE)

#select gene IDs and GO terms
select <- DEG %>% select(slim_cat, genes)

#take from long to wide 
# Split genes into separate rows
geneList <- select %>%
  separate_rows(genes, sep = ',') %>%
  arrange(slim_cat)  # Optional: If you want to sort the result by GO term

unique_genes <- unique(geneList$genes)



library(topGO)
library(clusterProfiler)

# Create a named vector of genes
allGenes <- setNames(rep(1, length(geneList$genes)), geneList$genes)

# Load the topGO package
library(topGO)

# Gene selection function (modify based on your criteria)
geneSelectionFun <- function(allGenes) {
  return(allGenes > 0)  # Change this condition based on your significance criteria
}

# Create a TopGOdata object
godata <- new("topGOdata", ontology = "BP", allGenes = allGenes, geneSel = geneSelectionFun, nodeSize = 10, annot = annFUN.gene2GO, gene2GO = geneList)



# Run the topGO algorithm
result <- runTest(godata, algorithm = "classic", statistic = "fisher")

# Calculate pairwise similarities
edox2 <- pairwise_termsim(result)

# Create tree plots or visualize the results as needed
p1 <- treeplot(edox2)
p2 <- treeplot(edox2, hclust_method = "average")

# Display the plots side by side
aplot::plot_list(p1, p2, tag_levels = 'A')




```








