---
title: "GO Semantic Similarity"
author: "Danielle Becker"
date: "Last updated 2023/06/27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Prepare workspace
```{r, message=FALSE, warning=FALSE}
rm(list=ls()) # removes all prior objects
library(tidyverse)
library(org.Hs.eg.db)
library(simplifyEnrichment)
library(magick)
library(gridtext)
library(clusterProfiler)
library(enrichplot)
# we use ggplot2 to add x axis labels (ex: ridgeplot)
library(ggplot2)
```

#Use simplifyenrichment package (http://www.bioconductor.org/packages/release/bioc/vignettes/simplifyEnrichment/inst/doc/simplifyEnrichment.html) to calculate semantic similarity for GO terms in gene expression and methylation
#All dataframes were made in RAnalysis - Scripts - Host - Host_GO_KEGG_Enrichment_Analysis_Func_Annot

#Create semantic similarity matrices for differentially expressed gene categories

Load GO data and select only the GO_IDs associated with BP terms
```{r, message=FALSE, warning=FALSE}
  MatGO <- read_csv("RAnalysis/Data/RNA-seq/P.verrucosa/Host/DEG_GO_parent_terms_complete.csv", col_names = TRUE)
nrow(MatGO) # total #enriched terms 307
nrow(filter(MatGO, ontology=="BP")) #BP terms 210
nrow(filter(MatGO, ontology=="CC")) #CC terms 27
nrow(filter(MatGO, ontology=="MF")) #MF terms 70
MatGO.BP <- MatGO %>% filter(ontology == "BP")
MatBP <- MatGO.BP$category
MatGO.MF <- MatGO %>% filter(ontology == "MF")
MatMF <- MatGO.MF$category
MatGO.CC <- MatGO %>% filter(ontology == "CC")
MatCC <- MatGO.CC$category
```


Calculate a similarity matrix and save the output.
```{r}
Mat = GO_similarity(MatBP, ont = "BP", db = "org.Hs.eg.db")
pdf(file = "RAnalysis/Output/RNA-seq/Semantic_similarity/DEG_BP_plot.pdf", width = 7, height = 5)
simplifyGO(Mat, word_cloud_grob_param = list(max_width = 50), max_words=20)
dev.off()

MatMF = GO_similarity(MatMF, ont = "MF", db = "org.Hs.eg.db")
pdf(file = "RAnalysis/Output/RNA-seq/Semantic_similarity/DEG_MF_plot.pdf", width = 7, height = 5)
simplifyGO(MatMF, word_cloud_grob_param = list(max_width = 50), max_words=20)
dev.off()

MatCC = GO_similarity(MatCC, ont = "CC", db = "org.Hs.eg.db")
pdf(file = "RAnalysis/Output/RNA-seq/Semantic_similarity/DEG_CC_plot.pdf", width = 7, height = 5)
simplifyGO(MatCC, word_cloud_grob_param = list(max_width = 50), max_words=20)
dev.off()
```


#Create semantic similarity matrices for differentially methylated gene categories
Load GO data and select only the GO_IDs associated with BP terms
```{r, message=FALSE, warning=FALSE}
UEGO <- read_csv("2a-WGCNA/Output/GO.05.UE.csv", col_names = TRUE)
nrow(UEGO) # total #enriched terms
nrow(filter(UEGO, ontology=="BP")) #BP terms
nrow(filter(UEGO, ontology=="CC")) #CC terms
nrow(filter(UEGO, ontology=="MF")) #MF terms
UEGO.BP <- UEGO %>% filter(ontology == "BP")
UEBP <- UEGO.BP$category
# UEGO.MF <- UEGO %>% filter(ontology == "MF")
# UEMF <- UEGO.MF$category
```

Calculate a similarity matrix and save the output.
```{r}
UE = GO_similarity(UEBP, ont = "BP", db = "org.Hs.eg.db")
pdf(file = "3-Semantic-Similarity/Output/Biological_Process_Plots/UEBP_GOSEM.pdf", width = 7, height = 5)
simplifyGO(UE, word_cloud_grob_param = list(max_width = 50), max_words=20)
dev.off()
# UE = GO_similarity(UEMF, ont = "MF", db = "org.Hs.eg.db")
# simplifyGO(UE, word_cloud_grob_param = list(max_width = 50), max_words=20)
```

#try using clusterprofiler for the same thing

```{r}
#make annotation using gff and gaf for Pocillopora verrucosa
```


