---
title: "WGBS Markdown"
author: "daniellembecker"
date: "3/17/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load Libraries

```{r message = FALSE, warning = FALSE}
library(plotrix) 
library(ggplot2)
library(gridExtra)
library(seacarb) 
library(dplyr)
library(pheatmap)
library(tidyverse)
library(genefilter)
library(cowplot)
library(lsmeans)
library(ontologyIndex)
library(ontologySimilarity)
library(data.table)
library(RColorBrewer)
library(colorRamps)
library(GSEABase)
library(here)

```

#Downloaded all final .bed files for 5x and 10x from bluewaves to desktop: scp -r danielle_becker@bluewaves.uri.edu:/data/putnamlab/dbecks/Becker_E5/WGBS_Becker_E5/Becker_WGBS/CovtoCyto/*_enrichment.bed /Users/Danielle/Desktop/Putnam_Lab/Becker_E5/RAnalysis/Data/WGBS

###Samples 19, 16, and 2 have low data coverage (see WGBS workflow in Bioinformatics>WGBS>WGBS workflow for further information), not moving forward with these samples in statistical steps

Methylation Analysis
## Loading genomic and annotation

```{r}
#load sample information
sample.info <- read.csv("../Data/sample.info.csv", header=T, sep=",", na.string="NA", stringsAsFactors = F) #read in info
samp <- sample.info$Sample.ID # set sample info
samp <- gsub("[_]", "-", samp) #remove extra characters

#load genes gff 
Genes <- read_csv("../../RAnalysis/Data/Genome/Pver_genome_assembly_v1.0.gene.gff3", sep="\t") #read in data file
Genes <- Genes[,c(1,4,5,9)] #select desired columns only
colnames(Genes) <- c("scaffold", "start", "stop", "gene") #rename columns
Genes$gene <- gsub(";.*","",Genes$gene) #remove extra characters
Genes$gene <- gsub("ID=","",Genes$gene) #remove extra characters

#Load annotation file (ask Hollie about text file)
Annot <- read.csv("RAnalysis/Data/Genome/Panopea-generosa-genes-annotations.tab", header=FALSE, sep="\t", na.string="NA", stringsAsFactors = FALSE, skip=0)
colnames(Annot) <- c("gene","scaffold", "start", "end", "uniprot.id", "uniprot_taxa.id", "Uniprot.Name", "GO.IDs", "GO.Descript")
GO.data <- Annot[,c(1,8)] #select names and GOs
splitted <- strsplit(as.character(GO.data$GO.IDs), "; ") #split into multiple GO ids
Gene.GO.IDs <- data.frame(v1 = rep.int(GO.data$gene, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row 
colnames(Gene.GO.IDs) <- c("gene", "GO.IDs") #rename columns
Gene.GO.IDs$GO.IDs <- as.character(Gene.GO.IDs$GO.IDs)
#Gene.GO.IDs$GO.IDs[is.na(Gene.GO.IDs$GO.IDs)] <- "unknown" #replace NA with unknown
GO.IDs <-unique(Gene.GO.IDs$GO.IDs)
# downloaded 20210407: http://www.informatics.jax.org/gotools/data/input/map2MGIslim.txt
GoSlims <- read.csv("RAnalysis/Data/20210407_map2MGIslim.csv", header=TRUE, sep=",", na.string="NA", stringsAsFactors = FALSE, skip=0)
colnames(GoSlims) <- c("GO.IDs", "Cat","GO.Slim.Term")
Gene.GO.IDs.slims <- merge(Gene.GO.IDs, GoSlims, by="GO.IDs", all = TRUE)
Gene.GO.IDs.slims <- Gene.GO.IDs.slims[-1,]
GO.IDs.slims <- Gene.GO.IDs.slims[,-2]

```


#Load filtered methylation counts (see RAnalysis>Data>WGBS for all files)
```{r}
# Converge final enrichment .bed files, separate by 5x and 10x coverage

#Merge final .bed files at 5x coverage 
meth.data.5x <- list.files(path = "../Data/WGBS/5x_coverage/", pattern = ".bed$", full.names=TRUE) %>%
  set_names(.) %>% 
  map_dfr(read.csv,.id="Sample.ID", header=FALSE, sep="\t", na.string="NA", stringsAsFactors = FALSE) %>% 
  dplyr::select(-c(V3,V7:V14)) %>%
  group_by(Sample.ID)
colnames(meth.data.5x) <- c("Sample.ID", "scaffold", "position","per.meth","meth","unmeth","gene")
meth.data.5x$gene <- gsub(";.*","",meth.data.5x$gene) #remove extra characters
meth.data.5x$gene <- gsub("ID=","",meth.data.5x$gene) #remove extra characters
meth.data.5x$Sample.ID <- gsub("../Data/WGBS/5x_coverage/","",meth.data.5x$Sample.ID) #remove extra characters
meth.data.5x$Sample.ID <- gsub("_.*","",meth.data.5x$Sample.ID) #remove extra characters 
meth.data.5x$Sample.ID <- gsub("-","_",meth.data.5x$Sample.ID) #remove extra characters
meth.data.5x$Sample.ID <- gsub("/","",meth.data.5x$Sample.ID) #remove extra characters (backslash from sample.ID name)
MD.5x <- merge(meth.data.5x,sample.info, by="Sample.ID") #combine sample info and 5x coverage megafile
```

```{r}
#Merge final .bed files at 10x coverage
meth.data.10x <- list.files(path = "../Data/WGBS/10x_coverage/", pattern = ".bed$", full.names=TRUE) %>%
  set_names(.) %>% 
  map_dfr(read.csv,.id="Sample.ID", header=FALSE, sep="\t", na.string="NA", stringsAsFactors = FALSE) %>% 
  dplyr::select(-c(V3,V7:V14)) %>%
  group_by(Sample.ID)
colnames(meth.data.10x) <- c("Sample.ID", "scaffold", "position","per.meth","meth","unmeth","gene")
meth.data.10x$gene <- gsub(";.*","",meth.data.10x$gene) #remove extra characters
meth.data.10x$gene <- gsub("ID=","",meth.data.10x$gene) #remove extra characters
meth.data.10x$Sample.ID <- gsub("../Data/WGBS/10x_coverage/","",meth.data.10x$Sample.ID) #remove extra characters
meth.data.10x$Sample.ID <- gsub("_.*","",meth.data.10x$Sample.ID) #remove extra characters 
meth.data.10x$Sample.ID <- gsub("-","_",meth.data.10x$Sample.ID) #remove extra characters
meth.data.10x$Sample.ID <- gsub("/","",meth.data.10x$Sample.ID) #remove extra characters (backslash from sample.ID name)
MD.10x <- merge(meth.data.10x,sample.info, by="Sample.ID") #combine sample info and 10x coverage megafile
```

# Testing for Differentially Methylated Genes
```{r}
# Comparison of enriched vs control nutrient treatments
# Binomial GLM to test for differentially methylated genes
sub_meth_table.5x  <- MD.5x
sub_meth_table.5x$group <- paste0(sub_meth_table.5x$Sample.ID, sub_meth_table.5x$gene)

#filter for genes with >5 methylated positions
min.filt.5x <- count(sub_meth_table.5x, vars = c( group))
newdata.5x <- min.filt.5x[ which(min.filt.5x$n > 4), ]
sub_meth_table.5x <- sub_meth_table.5x[sub_meth_table.5x$group %in% newdata.5x $vars,]

# create data frame to stored results
results.5x <- data.frame()
gs.5x <- unique(sub_meth_table.5x$gene)

#make treatment a factor 

#first subset the unique dataframes and second run the GLMs
for(i in 1:length(sub_meth_table.5x$gene)){
  
  #subset the dataframe gene by gene
  sub_meth_table.5x1 <- subset(sub_meth_table.5x, gene ==gs.5x[i])
  
  # fit glm position model
  fit <- glm(matrix(c(meth, unmeth), ncol=2) ~ treatment, 
             data=sub_meth_table.5x1, family=binomial)
  a <- anova(fit, test="Chisq")
  
  # capture summary stats to data frame
  df <- data.frame(gene = sub_meth_table.5x1[1,7],
                   pval.treatment = a$`Pr(>Chi)`[2],
                   #pval.position = a$`Pr(>Chi)`[3], #uncomment if you want to include position of CpG within a gene
                   #pval.treatment_x_position = a$`Pr(>Chi)`[4], #uncomment if you want to include position of CpG within a gene interaction with treatment
                   stringsAsFactors = F)
  
  # bind rows of temporary data frame to the results data frame
  results.5x <- rbind(results.5x, df)
  
}
# An error will be generated here for contrasts. 
#This potential for contrasts (interactions) is included in the case one wants to examine the role of position of CpG within a gene
#Error in `contrasts<-`(`*tmp*`, value = contr.funs[1 + isOF[nn]]) : contrasts can be applied only to factors with 2 or more levels
#Continuing the analysis from results line will generate the results in the absence of the contrast (interaction).

results.5x[is.na(results.5x)] <- 0
results.5x$adj.pval.treatment <- p.adjust(results$pval.treatment, method='BH') #BH is a less conservative Benjamini & Hochberg (1995) correction in which p-values are multiplied by the number of comparisons

#results$adj.pval.position <- p.adjust(results$pval.position, method='BH') #uncomment if you want to include position of CpG within a gene
#result $adj.pval.treatment_x_position <- p.adjust(results$pval.treatment_x_position, method='BH') #uncomment if you want to include position of CpG within a gene interaction with treatment

Amb.sig <-results
Amb.sig <- Amb.sig[,c(1,3)]
Amb.sig <- Amb.sig[order(Time.Amb.sig$adj.pval.TimePoint),]
Amb.sig <- Amb.sig[which(Time.Amb.sig$adj.pval.TimePoint<0.05), ]

# Annotation of DMG under ambient conditions through time
Time.Amb.annot <- merge(Time.Amb.sig , Annot, by="gene", all.x=TRUE)
Time.Amb.annot <- Time.Amb.annot[order(Time.Amb.annot$adj.pval.TimePoint),]
Time.Amb.annot <- Time.Amb.annot[!duplicated(Time.Amb.annot$gene),]
write.table(Time.Amb.annot, 'RAnalysis/Output/Table_S1_Time_sig_annot.tsv', sep='\t', row.names=FALSE)

#Sanity check plotting of top DMGs by time
high <- MD.Time %>%	
  filter(gene== Time.Amb.sig$gene[1])
means <- aggregate(per.meth ~ TimePoint, data=high, FUN=mean)
ses <- aggregate(per.meth ~ TimePoint, data=high, FUN=std.error)
means$ses <- ses$per.meth
ggplot(means, aes(x=TimePoint, y=per.meth)) +
  geom_bar(stat="identity", color="black",
             position=position_dodge()) +
    geom_errorbar(aes(ymin=per.meth-ses, ymax=per.meth+ses), width=.2,
                  position=position_dodge(.9)) +
    xlab("TimePoint") + #plot x axis label
    ylab("Percent Methylation") + #plot y axis label
    ylim(0,100)+ #set y limits to 0 and 100%
    scale_fill_manual(values = c("blue", "purple", "red"))+
    theme_bw()

```


