---
title: "WGBS Markdown"
author: "daniellembecker"
date: "3/17/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## WGBS Gene Ontology (GO) Enrichment Analysis
## Previous steps for DNA methylation include WGBS workflow in Bioinformatics>WGBS>WGBS workflow 

## Load Libraries

```{r message = FALSE, warning = FALSE}
library(plotrix) 
library(ggplot2)
library(gridExtra)
library(factoextra)
library(seacarb) 
library(dplyr)
library(pheatmap)
library(tidyverse)
library(genefilter)
library(base)
library("DESeq2")
library(cowplot)
if ("lsmeans" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('lsmeans') 
library(lsmeans)
if ("ontologyIndex" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('ontologyIndex') 
library(ontologyIndex)
if ("ontologySimilarity" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('ontologySimilarity')
library(ontologySimilarity)
library(data.table)
library(RColorBrewer)
if ("colorRamps" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('colorRamps')
library(colorRamps)
if ("limma" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('limma')
# Load the package
library(limma)
library(GSEABase)
library(here)
library(lme4)
library(vegan)
library(goseq)
library(ggpubr)
library(patchwork)
library(rstatix)
library(stringr)
library(readr)
library(png)
library(grid)
library(strex)
library(cowplot)
library(parallel)
```

#Downloaded all final .bed files for and 10x from andromeda to desktop: scp -r danielle_becker@andromeda.uri.edu:/data/putnamlab/dbecks/Becker_E5/WGBS_Becker_E5/Becker_WGBS/CovtoCyto/*_enrichment.bed /Users/Danielle/Desktop/Putnam_Lab/Becker_E5/RAnalysis/Data/WGBS

**Libraries E7, E8, AND C28 had low data coverage, not moving forward with these samples in statistical steps.**

# Methylation Analysis

## Loading genomic and annotation information

```{r}
#load sample information
sample.info <- read.csv("RAnalysis/Data/metadata.WGBS.csv", 
                        header=T, sep=",", na.string="NA", stringsAsFactors = F) 

#load genes gff 
Genes <- read.csv("RAnalysis/Output/rename_structure_gff3.csv") %>% #read in data fill
  dplyr::select(gene_id, seqid, start, end, gene_id, length) %>% #select desired columns only 
  dplyr::rename(stop = end, gene = gene_id, scaffold = seqid) %>% #rename columns
  mutate(gene = gsub("_", "_gene_", gene))

#Load annotation file
Annot <- read.csv("Functional_Annotation/pver_annot_full.csv", row.names = 1) %>%
  dplyr::rename(gene = gene_id) %>% #rename columns
  mutate(gene = gsub("_", "_gene_", gene), #add _gene_g#
         gene = str_before_nth(gene, ".t", 1)) #remove the all after . "t#" number after the dash

#load all gene GO IDs organized and split
Gene.GO.IDs <- read.csv("Functional_Annotation/pver_GOterms_interprot_swissprot_blast_trembl_20211203.csv", row.names = 1) %>%
  dplyr::rename(gene = gene_id, GO.IDs = GO.ID) %>% 
  mutate(gene = gsub("_", "_gene_", gene), #add _gene_g#
         gene = str_before_nth(gene, ".t", 1)) #remove the all after . "t#" number after the dash

GO.IDs <-unique(Gene.GO.IDs$GO.IDs) # creates list of unique entries in the GO.IDs column 
```

## Load filtered methylation counts 10x (only need to run initially, redone on 20240814, commented out as we filtered SNPs from the output in the next section)

```{r}
# 
# #  #Merge final .bed files at 10x coverage
#   meth.data.10x <- list.files(path = "Bioinformatics/Data/10x_methcounts/", pattern = ".bed$", full.names=TRUE) %>%
#     purrr::set_names(.) %>% 
#     map_dfr(read.csv,.id="Sample.ID", header=FALSE, sep="\t", na.string="NA", stringsAsFactors = FALSE) %>% 
#     dplyr::select(-c(V3,V7:V14)) %>%
#     group_by(Sample.ID)
#   colnames(meth.data.10x) <- c("Sample.ID", "scaffold", "position","per.meth","meth","unmeth","gene")
#   meth.data.10x$gene <- gsub(";.*","",meth.data.10x$gene) #remove extra characters
#   meth.data.10x$gene <- gsub("ID=","",meth.data.10x$gene) #remove extra characters
#   meth.data.10x$Sample.ID <- gsub("Bioinformatics/Data/10x_methcounts//","",meth.data.10x$Sample.ID) #remove extra characters
#   meth.data.10x$Sample.ID <- gsub("_.*","",meth.data.10x$Sample.ID) #remove extra characters 
#   #meth.data.10x$Sample.ID <- gsub("-","_",meth.data.10x$Sample.ID) #remove extra characters
#   #meth.data.10x$Sample.ID <- gsub("/","",meth.data.10x$Sample.ID) #remove extra characters (backslash from sample.ID name)
#   MD.10x <- merge(meth.data.10x,sample.info, by="Sample.ID") #combine sample info and 10x coverage megafile
#  
#  # # # Filter genewiz sample IDs 2, 16, and 19 that had low coverage
#  # # # Define genewiz sample IDs with low coverage
#   low_coverage_samples <- c(2, 16, 19)
# # # # 
# # # # # Filter out these genewiz sample IDs from the combined dataset
#    MD.10x_filtered <- MD.10x[!MD.10x$Sample.ID %in% low_coverage_samples, ]
#    unique_sample_ids <- unique(MD.10x_filtered$Sample.ID)
#    print(unique_sample_ids)
# #  
# # # # #save sample info and percent methylation
#   saveRDS(MD.10x_filtered, "RAnalysis/Output/sample_all_methylated_data_10x.Rdata")

```

## Filter out any SNPs that could impact our methylation data and use updated and filtered data frame for next steps

```{r}
# Filtering SNPs found from BS-SNper, following this workflow: https://github.com/daniellembecker/DanielleBecker_Lab_Notebook/blob/master/_posts/2024-02-12-Testing-BS-SNPer-Molec-Underpinnings-WGBS.md and this script: https://github.com/hputnam/Becker_E5/blob/master/RAnalysis/Scripts/WGBS/BS-SNPer.filter.Rmd

# Use the new MD.10x file that has now been filtered for SNPs
MD.10x <- readRDS("RAnalysis/Data/filtered_bed_SNP_file.Rdata") #read in info
head(MD.10x)

#MD.10x_filtered <- MD.10x %>%
#  dplyr::filter(!fragment.ID %in% c("PV_26"))
```


## Testing for Differentially Methylated Genes at 10X coverage
Comparison of enriched vs control nutrient treatments at 10X coverage.
Data transformation steps 
```{r}
sub_meth_table.10x  <- MD.10x %>% 
  # create new column called 'group' based on loci position
  unite(Loc, scaffold, position, remove = FALSE, sep = " ") %>%
  
  #filtering out loci that have a median methylation of <1% across all samples
   group_by(Loc) %>%
   mutate(loci_median = median(per.meth)) %>%
   filter(loci_median > 0) %>% ungroup() %>%
  
  # filtering out genes with less than 5 positions of data 
  dplyr::group_by(gene) %>%
  mutate(n_positions = n_distinct(position)) %>% ungroup() %>% ## count the number of unique positions for each gene
  filter(n_positions >= 5) %>% dplyr::select(-n_positions)

gs.10x <- unique(sub_meth_table.10x$gene)

## Before loci filtering 1,118,733 rows for 3,910 genes
## After loci filtering 1,000,442 rows for 1,902 genes 
## After filtering for 5 **methylated** positions filtering 246,993 rows for 557 genes (length of gs.10x above)
```

Binomial GLM to test for differentially methylated genes. For 1,690 genes this took ~5-10 minutes. 
```{r}
# Split the sub_meth_table.10x dataframe by 'gene'
split_dfs <- split(sub_meth_table.10x, sub_meth_table.10x$gene)
# split_example <- sub_meth_table.10x %>% subset(gene == "Pver_gene_g97")

# Preparing objects for data collection
results.10x <- list()
results.10x$Stats <- data.frame(matrix(nrow=length(gs.10x), ncol=2), row.names=gs.10x)
colnames(results.10x$Stats) <- c("gene", "pvalue_treatment")

## Defining the run_glm function
run_glm <- function(data){
  
  #Bayesian model
  fit <- glm(matrix(c(meth, unmeth), ncol=2) ~ treatment, 
             data=data, family=binomial)
  
  a <- anova(fit, test="Chisq")
  
  Gene <- unique(data$gene)
  
  #Gathering convergence stats
  results.10x$Stats[Gene, ] <- c(Gene, a$`Pr(>Chi)`[2])
  
  write.table(results.10x$Stats[Gene, ], "RAnalysis/Output/GLM_output_with_extrafilter.csv", 
              col.names=F, row.names=T, quote=F, append=T)
  
}

### Lapply for the GLM 
### this will run through the run_glm function for multiple genes at a time and output
mclapply(split_dfs, run_glm)
```

Filtering to genes that are significantly different and justing p-values with 'BH' (a less conservative Benjamini & Hochberg (1995) correction in which p-values are multiplied by the number of comparisons). 
```{r}
glm_results <- read.delim2("RAnalysis/Output/GLM_output_with_extrafilter.csv", header=F, sep = " ") %>% 
  dplyr::select(-V1) %>% dplyr::rename(gene = V2, pvalue_treatment = V3)

## adjusting p-values 
glm_results$adj_pvalue_treatment <- p.adjust(glm_results$pvalue_treatment, method='BH')

## filtering to p-values above 0.05 
glm_results_significant <- glm_results %>%
  filter(adj_pvalue_treatment <= 0.050)

## Before filteringc 1,902 rows for 1,902 genes
## After filtering 376 rows for 376 significantly differentially methylated genes 
## With filtering for 5 methylated positions, 290 genes are significantly differentially methylated 
```

Adding gene annotation information and methylation data 

```{r}
glm_results_sig_annotated <- glm_results_significant %>% dplyr::select(gene, adj_pvalue_treatment) %>%
  left_join(., sub_meth_table.10x, by = "gene") %>%
  left_join(., Annot, by = "gene")
```

Calculating relative change 

```{r}
relative_change <- glm_results_sig_annotated %>%
  group_by(gene, treatment) %>%
  summarize(med_per_meth = median(per.meth), .groups = 'drop') %>%
  pivot_wider(names_from = treatment, values_from = med_per_meth) %>%
  mutate(relative_change = enriched - control)

ggplot(relative_change, aes(x = gene, y = relative_change, fill = relative_change)) +
  geom_point(size = 3, shape = 21, color = 'black') +
  theme_bw() +
  scale_fill_gradient2(na.value = "white", low = "#800f2f", 
                      mid = 'white', midpoint = 0, high = "#4f772d") + 
  labs(x = "Gene",
       y = "Change in Mean Methylation (%)",
       fill = "% change") +
  geom_hline(yintercept = 0, color = 'grey40', linetype = 'dashed') +
  theme(axis.text.x = element_blank(),
        legend.position = "right")  
```

Conduct PERMANOVA and PCoA on methylation matrix 
```{r}
## Using the filter step, suggest using median percent methylation for pcoa
methylation_matrix <- glm_results_sig_annotated %>%
  group_by(gene, Sample.ID) %>%
  mutate(med_per_meth = median(per.meth)) %>% 
  dplyr::select(gene, Sample.ID, med_per_meth) %>% distinct() %>%
  spread(gene, med_per_meth) %>%
  column_to_rownames(var = "Sample.ID")

# Select treatment data
meta_for_permanova <- glm_results_sig_annotated %>%
  dplyr::select(Sample.ID, treatment) %>% distinct()

# PERMANOVA 
adonis2(methylation_matrix ~ treatment, data = meta_for_permanova, method='bray')

# PERMDISP 
# Perform betadisper analysis
betadisper_results <- betadisper(vegdist(methylation_matrix, "eu"), meta_for_permanova$treatment)
anova(betadisper_results)
# Extract the eigenvalues
eigenvalues <- betadisper_results$eig

# Calculate the percentage of variance for PC1 and PC2
variance_percentages <- eigenvalues / sum(eigenvalues) * 100
pc1_variance <- round(variance_percentages[1], 2)
pc2_variance <- round(variance_percentages[2], 2)

# Extract PCoA coordinates and merge with metadata
pcoa_coordinates <- as.data.frame(scores(betadisper_results, display = "sites")) %>%
  rownames_to_column(var = "Sample.ID") %>%
  left_join(., meta_for_permanova, by = "Sample.ID")

# Create the plot with percentage of variance on axes
pcoa_coordinates %>%
  ggplot(aes(x = PCoA1, y = PCoA2, fill = treatment)) +
  geom_point(size = 3, shape = 21, color = "black") +
  theme_bw() +
  stat_ellipse(alpha = 0.2, size = 1, level = 0.95) +  # Add ellipses with transparency 
  scale_fill_manual(values = c("control" = "grey", "enriched" = "black")) +
  labs(fill = "Treatment",
    x = paste0("PCoA1 (", pc1_variance, "%)"),
    y = paste0("PCoA2 (", pc2_variance, "%)")
  ) +
  annotate("text", x = -100, y = -97, label = "PERMANOVA p=0.007", 
           hjust = 0, #vjust = 1.1, 
           size = 3, color = "grey20");pcoa_coordinates


# Run PCA for comparison for DMGs
# Perform PCA on the methylation matrix
pca_results <- prcomp(methylation_matrix, scale. = TRUE)

# Extract PCA variance
pca_variance <- pca_results$sdev^2
pca_variance_percent <- pca_variance / sum(pca_variance) * 100
pc1_variance <- round(pca_variance_percent[1], 2)
pc2_variance <- round(pca_variance_percent[2], 2)

# Extract PCA coordinates and merge with metadata
pca_coordinates <- as.data.frame(pca_results$x) %>%
  rownames_to_column(var = "Sample.ID") %>%
  left_join(., meta_for_permanova, by = "Sample.ID")

# Create the PCA plot with percentage of variance on axes
pca_coordinates %>%
  ggplot(aes(x = PC1, y = PC2, fill = treatment)) +
  geom_point(size = 3, shape = 21, color = "black") +
  theme_bw() +
  stat_ellipse(alpha = 0.2, size = 1) +  # Add ellipses with transparency 
  scale_fill_manual(values = c("control" = "grey", "enriched" = "black")) +
  labs(fill = "Treatment",
       x = paste0("PC1 (", pc1_variance, "%)"),
       y = paste0("PC2 (", pc2_variance, "%)")
  ) +
  annotate("text", x = -11, y = -11, label = "PERMANOVA p=0.006", 
           hjust = 0, size = 3, color = "grey20")

ggsave("RAnalysis/Output/DMG_PCA.png", width = 6.5, height = 5)

# Run PERMANOVA using the adonis2 function
permanova_results <- adonis2(methylation_matrix ~ treatment, data = meta_for_permanova, method = 'bray')

# View the results of the PERMANOVA
print(permanova_results)

```

Make heatmap for DMGs 
```{r}
# Filter metadata for the two treatments
meta_for_heatmap <- meta_for_permanova %>%
  filter(treatment %in% c("control", "enriched"))

# Ensure that the row names of the methylation_matrix are in the same order as the metadata
methylation_matrix_filtered <- methylation_matrix[rownames(methylation_matrix) %in% meta_for_heatmap$Sample.ID, ]

# Subset methylation_matrix to include only samples from these treatments
methylation_matrix_filtered <- methylation_matrix_filtered[meta_for_heatmap$Sample.ID, ]

# Define colors for treatments
annotation_col <- meta_for_heatmap %>%
  select(Sample.ID, treatment) %>%
  column_to_rownames(var = "Sample.ID")

ann_colors <- list(
  treatment = c(
    "control" = "grey", 
    "enriched" = "black"
  )
)

# Create the heatmap
# Generate the heatmap
DMG_heatmap <- pheatmap(
  methylation_matrix_filtered,
  cluster_rows = TRUE,  # Cluster genes
  cluster_cols = TRUE,  # Cluster samples
  scale = "row",        # Scale by row (genes)
  show_rownames = TRUE, # Show gene names
  show_colnames = TRUE, # Show sample IDs
  annotation_col = annotation_col, # Add treatment as annotation
  annotation_colors = ann_colors,   # Apply annotation colors
  color = colorRampPalette(c("blue", "white", "red"))(100), # Customize color palette
  main = "Heatmap of DMGs between Treatments")

DMG_heatmap

png(file = "RAnalysis/Output/DMG_heatmap_host.png")
DMG_heatmap
dev.off()
```


### Make Volcano plot of hyper and hypo methylated DMGs
```{r}
#make meth count matrix, select gene, frag ID, and per meth from dataframe
matrix <- MD.10x %>%
  dplyr::select(gene, fragment.ID, per.meth)

#change from long to wide, set fun.aggregate for mean % meth per gene as it includes each position for each sample in the calculation
countdata <- reshape2::dcast(matrix, gene ~ fragment.ID, value.var = "per.meth", fun.aggregate = mean)

#make first row removed and rownames
countData <- countdata[,-1]
rownames(countData) <- countdata[,1]

#make all column values integers 
countData <- countData %>% mutate_if(is.numeric, as.integer)

#filter out sample IDs 2, 19 and 16 from sample_info and colData data frame because they are not being used due to deduplication errors
samp.info <- sample.info[!row.names(sample.info) %in% c(2,16,19),]

# Define the columns for libraries to remove
columns_to_remove <- c("PV_7", "PV_8", "PV_28")  # Adjust with the column names you want to remove

# Remove columns from colData
countData <- countData[, !colnames(countData) %in% columns_to_remove]

##Merge the treatment columns into a new column , group. Set group as a factor.
samp.info$treatment <- factor(samp.info$treatment, levels = c("control","enriched"))

# Update countData column names to match samp.info
colnames(countData) <- samp.info$fragment.ID

# Reorder countData columns to match the row names in samp.info
countData <- countData[, samp.info$fragment.ID]

# Update row names of samp.info to match countData columns
rownames(samp.info) <- samp.info$fragment.ID

# Check if colnames(countData) and rownames(samp.info) match
identical(sort(colnames(countData)), sort(rownames(samp.info)))  # Should be TRUE

#Set DESeq2 design
gdds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = samp.info,
                              design = ~treatment)

# Run DMG analysis
dds <- DESeq(gdds)

# Extract log2fold change values
results <- results(dds)
log2fold_change <- results$log2FoldChange
gene_names <- rownames(results)
log2fold_change_df <- data.frame(Gene = gene_names, Log2FoldChange = log2fold_change)

#rename gene column
log2fold_change_df <- log2fold_change_df %>%
  dplyr::rename(gene = Gene)

# Merge with the methylation data frame 
final_df <- left_join(glm_results, log2fold_change_df, by = "gene")

#rename gene column
final_df <- final_df %>%
 dplyr::rename(log2FoldChange = Log2FoldChange)

final_df <- final_df %>%
  dplyr::rename(padj = adj_pvalue_treatment)

final_df <- final_df %>%
  dplyr::rename(pvalue = pvalue_treatment)

# Create a new column for -log10(padj), noticed outliers in data
final_df$log_padj <- -log10(final_df$padj)

# Remove rows where log_padj is greater than 10
final_df <- final_df[final_df$log_padj <= 100, ]

# Define criteria for hypermethylation, hypomethylation
final_df$methylation_status <- ifelse(final_df$padj > 0.05, "Not Significant",
                                       ifelse(final_df$log2FoldChange > 0, "Hypermethylated",
                                              ifelse(final_df$log2FoldChange < 0, "Hypomethylated", "Not Classified")))

# Set up the color palette for hyper, hypo, and non-significant genes
colors <- c("Hypermethylated" = "red", "Hypomethylated" = "blue", "Not Significant" = "black")

# Count specific statuses
count_hypermethylated <- sum(final_df$methylation_status == 'Hypermethylated') #130
count_hypomethylated <- sum(final_df$methylation_status == 'Hypomethylated') #158

# Create the volcano plot using ggplot2
DMG.volcano <- ggplot(final_df, aes(x = log2FoldChange, y = -log10(padj), color = methylation_status)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(values = colors, name = "Methylation Status", 
                     breaks = c("Hypermethylated", "Hypomethylated", "Not Significant"),
                     labels = c("Hypermethylated (130)", "Hypomethylated (158)", "Not Significant"),
                     drop = FALSE) +  # Include all specified breaks in the legend
  labs(title = "", x = "log2FoldChange", y = "-log10(adjusted p-value)") +
  theme_bw() +
  geom_hline(yintercept = 0, linetype = "dashed", color = "black");DMG.volcano

ggsave("RAnalysis/Output/DMG.volcano.png", DMG.volcano, width = 8, height = 6, units = c("in"))


# Make volcano and PCoA plot together
#load png image of DMG PCA
DMG.PCA <- readPNG("RAnalysis/Output/DMG_PCA.png")
PCA.DMG <- rasterGrob(DMG.PCA, interpolate=TRUE)

Figure2 <- plot_grid( DMG.volcano, PCA.DMG, labels = c('A', 'B'), label_size = 14)

ggsave(filename=paste0("RAnalysis/Output/DMG.volcano_pca.png"), plot=Figure2, dpi=300, width=10, height=5, units="in", limitsize = FALSE)

# Add methylation status assignments to the significant results and ouput for topGO enrichment analysis
merge_df <- final_df %>%
  dplyr::rename(adj_pvalue_treatment = padj)

merge_df <- merge_df %>%
  dplyr::rename(pvalue_treatment = pvalue)

# Merge for just significant genes for enrichment analysis
DMGs_all <- merge(glm_results_significant, merge_df)

write.csv(DMGs_all, "RAnalysis/Output/DMGs_all.csv")

```

