---
title: "WGBS Markdown"
author: "daniellembecker"
date: "3/17/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load Libraries

```{r message = FALSE, warning = FALSE}
library(plotrix) 
library(ggplot2)
library(gridExtra)
library(seacarb) 
library(dplyr)
library(pheatmap)
library(tidyverse)
library(genefilter)
library(cowplot)
library(lsmeans)
library(ontologyIndex)
library(ontologySimilarity)
library(data.table)
library(RColorBrewer)
library(colorRamps)
library(GSEABase)
library(here)

```

#Downloaded all final .bed files for 5x and 10x from bluewaves to desktop: scp -r danielle_becker@bluewaves.uri.edu:/data/putnamlab/dbecks/Becker_E5/WGBS_Becker_E5/Becker_WGBS/CovtoCyto/*_enrichment.bed /Users/Danielle/Desktop/Putnam_Lab/Becker_E5/RAnalysis/Data/WGBS

###Samples 19, 16, and 2 have low data coverage (see WGBS workflow in Bioinformatics>WGBS>WGBS workflow for further information), not moving forward with these samples in statistical steps

Methylation Analysis
## Loading genomic and annotation

```{r}
#load sample information
sample.info <- read.csv("../Data/sample.info.csv", header=T, sep=",", na.string="NA", stringsAsFactors = F) #read in info
samp <- sample.info$Sample.ID # set sample info
samp <- gsub("[_]", "-", samp) #remove extra characters

#load genes gff 
Genes <- read_csv("../../RAnalysis/Data/Genome/Pver_genome_assembly_v1.0.gene.gff3", sep="\t") #read in data file
Genes <- Genes[,c(1,4,5,9)] #select desired columns only
colnames(Genes) <- c("scaffold", "start", "stop", "gene") #rename columns
Genes$gene <- gsub(";.*","",Genes$gene) #remove extra characters
Genes$gene <- gsub("ID=","",Genes$gene) #remove extra characters
```


#Load filtered methylation counts (see RAnalysis>Data>WGBS for all files)
```{r}
# Converge final enrichment .bed files, separate by 5x and 10x coverage

#Merge final .bed files at 5x coverage 
meth.data.5x <- list.files(path = "../Data/WGBS/5x_coverage/", pattern = ".bed$", full.names=TRUE) %>%
  set_names(.) %>% 
  map_dfr(read.csv,.id="Sample.ID", header=FALSE, sep="\t", na.string="NA", stringsAsFactors = FALSE) %>% 
  dplyr::select(-c(V3,V7:V14)) %>%
  group_by(Sample.ID)
colnames(meth.data.5x) <- c("Sample.ID", "scaffold", "position","per.meth","meth","unmeth","gene")
meth.data.5x$gene <- gsub(";.*","",meth.data.5x$gene) #remove extra characters
meth.data.5x$gene <- gsub("ID=","",meth.data.5x$gene) #remove extra characters
meth.data.5x$Sample.ID <- gsub("../Data/WGBS/5x_coverage/","",meth.data.5x$Sample.ID) #remove extra characters
meth.data.5x$Sample.ID <- gsub("_.*","",meth.data.5x$Sample.ID) #remove extra characters 
meth.data.5x$Sample.ID <- gsub("-","_",meth.data.5x$Sample.ID) #remove extra characters
meth.data.5x$Sample.ID <- gsub("/","",meth.data.5x$Sample.ID) #remove extra characters (backslash from sample.ID name)
MD.5x <- merge(meth.data.5x,sample.info, by="Sample.ID") #combine sample info and 5x coverage megafile
```

```{r}
#Merge final .bed files at 10x coverage
meth.data.10x <- list.files(path = "../Data/WGBS/10x_coverage/", pattern = ".bed$", full.names=TRUE) %>%
  set_names(.) %>% 
  map_dfr(read.csv,.id="Sample.ID", header=FALSE, sep="\t", na.string="NA", stringsAsFactors = FALSE) %>% 
  dplyr::select(-c(V3,V7:V14)) %>%
  group_by(Sample.ID)
colnames(meth.data.10x) <- c("Sample.ID", "scaffold", "position","per.meth","meth","unmeth","gene")
meth.data.10x$gene <- gsub(";.*","",meth.data.10x$gene) #remove extra characters
meth.data.10x$gene <- gsub("ID=","",meth.data.10x$gene) #remove extra characters
meth.data.10x$Sample.ID <- gsub("../Data/WGBS/10x_coverage/","",meth.data.10x$Sample.ID) #remove extra characters
meth.data.10x$Sample.ID <- gsub("_.*","",meth.data.10x$Sample.ID) #remove extra characters 
meth.data.10x$Sample.ID <- gsub("-","_",meth.data.10x$Sample.ID) #remove extra characters
meth.data.10x$Sample.ID <- gsub("/","",meth.data.10x$Sample.ID) #remove extra characters (backslash from sample.ID name)
MD.10x <- merge(meth.data.10x,sample.info, by="Sample.ID") #combine sample info and 10x coverage megafile
```




