---
title: "GO MWU input"
author: "Danielle Becker"
date: "2024-07-22"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r install packages}
#if ("rtracklayer" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rtracklayer")
#BiocManager::install("rrvgo")

library(dplyr)
library(ggplot2)
library(rtracklayer)
library(rrvgo)
library(GO.db)
library(tidyr)
library(forcats)
library(scales)
sessionInfo() #list of packages after library-ing these packages
```


# Import the data files 
```{r, echo=FALSE}
# Structural annotation GFF3 file
GFF3 <- read.csv("RAnalysis/Output/RNA-seq/DEG/Host/rename_structure_gff3.csv")

# Load go terms from functional annotation file
annot_GO <- read.csv("RAnalysis/Output/RNA-seq/DEG/Host/annot_GO.terms.host.csv")[-1]

# Remove rows where 'GO.ID' is 'unknown'
# in go mwu analysis, genes with unknown are in the annotation
#annot_GO <- subset(annot_GO, GO.ID != 'unknown')

# edit annot_GO to list GO terms all in one row per gene
annot_GO <- annot_GO %>%
  group_by(gene_id) %>%
  summarise(GO.ID = paste(GO.ID, collapse = ";")) %>%
  ungroup()

```

# Identifying treatment and all expressed pver genes (poverA = 0.50,10)
```{r}
#treatment information
treatmentinfo <- read.csv("RAnalysis/Data/RNA-seq/metadata.RNAseq.csv", header = TRUE, sep = ",")
str(treatmentinfo)
head(treatmentinfo)

# load all DEGs
DEG.all <- read.csv("RAnalysis/Output/RNA-seq/DEG/Host/DEGs.all.20240530.csv")
nrow(DEG.all)

#rename x column to gene_id
colnames(DEG.all)[1] <-"gene_id" # make colnames a true column called gene_id
head(DEG.all)

#load DEG significant results
DEG.sig <- read.csv("RAnalysis/Output/RNA-seq/DEG/Host/DEGs.sig.20240530.csv")[,-1]
nrow(DEG.sig)

# Load filtered gene count file
gcount_filt <- read.csv("RAnalysis/Output/RNA-seq/DEG/gene_count_matrix_filtered.csv")

# Rename first column X to gene_id
names(gcount_filt)[names(gcount_filt) == "X"] <- "gene_id"

```

### Isolating comparisons and using fold change 

I'm going to isolate comparisons and then use fold change as input to GO MWU rather than a binary of 1 (DE) or 0 (not DE). 

Match up genes in gene list file to annotation file
```{r}
names(annot_GO)

gcount_filt2annot = match(gcount_filt$gene_id, annot_GO$gene_id) #match genes in gcount_filt to annot_GO

# The following is the number of probes without annotation 
sum(is.na(gcount_filt2annot))

row_nas<-which(is.na(gcount_filt2annot))

#view the genes that do not have a match in the annotation file
missing<-as.data.frame(gcount_filt2annot[row_nas])
#print(missing)
```
0 genes in gcount_filt do not have GO annotations because unknown is allowed to stay in go MWU analysis

Reduce annotation file to only contain genes detected in our dataset.  
```{r}
filtered_Pverr.annot <- annot_GO[annot_GO$gene_id %in% gcount_filt$gene_id, ]
dim(filtered_Pverr.annot)

```

The annotation file now only contains genes that were detected in our dataset that have annotation information. This is 19,821 out of the 27439 genes in our dataset. 

19,821 genes were retained in the filtered GO dataset. Write table for GO MWU analysis 
```{r}
write.table(filtered_Pverr.annot, "RAnalysis/Scripts/RNA-seq/goMWU/gene_to_go_DB.tab",row.names = FALSE, sep = "\t", quote = FALSE)
```

Filter the counts by the remaining annotations 
```{r}
# Counts 
filt_counts_annot <- filtered_Pverr.annot %>%
  left_join(DEG.all, by = "gene_id") %>%
  dplyr::select(c("gene_id", "log2FoldChange")) %>% na.omit
```
19,821 DEGs in filt_counts_annot

Make GO_MWU input csv file with fold changes 
```{r}
write.csv(filt_counts_annot, "RAnalysis/Scripts/RNA-seq/goMWU/DEGs_all.csv", row.names = FALSE, quote = FALSE)
```


