---
title: "pver_annot_compile"
author: "daniellembecker"
date: "10/20/2021"
output: html_document
---

Load libraries
```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(dplyr)
```

#This script takes the results of functional annotation services and combines the results. Nucleotide sequences from ReefGenomics (http://pver.reefgenomics.org/download/) were annotated using DIAMONDSEARCH BLASTX, resulting in 26436 hits. These hits were used as input into:
#  - Uniprot
#  - Blast2GO

#Additional annotation was provided by
#  - InterProScan



## Diamond BLAST

# Load DIAMOND BLAST results

```{r}
pver_blast <- read_tsv("Functional_Annotation/Diamond/Pver_annot.tab", col_names = FALSE)
colnames(pver_blast) <- c("seqName", "top_hit", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore",  "qlen", "slen")
head(pver_blast)
dim(pver_blast) #26436 x 14
```



## Uniprot

# Uniprot mapping occurred on 20211015.
#EMBL/GenBank/DDBJ full transcripts  generated from Diamond BLAST were mapped to UniProtKB database IDs.

#Because there were many Diamond BLAST hits for Pver, I had to break up the tab file into chunks that Uniprot could handle. So there are 14 Pver_Uniprot files to be read and then rbind() together

```{r}
u1 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabaa.tab", col_names = TRUE)
colnames(u1) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u1)
dim(u1) # 415 x 10
```

```{r}
u2 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabab.tab", col_names = TRUE)
colnames(u2) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u2)
dim(u2) # 430 x 10
```

```{r}
u3 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabac.tab", col_names = TRUE)
colnames(u3) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u3)
dim(u3) # 433 x 10
```

```{r}
u4 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabad.tab", col_names = TRUE)
colnames(u4) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u4)
dim(u4) # 434 x 10
```

```{r}
u5 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabae.tab", col_names = TRUE)
colnames(u5) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u5)
dim(u5) # 439 x 10
```

```{r}
u6 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabaf.tab", col_names = TRUE)
colnames(u6) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u6)
dim(u6) # 444 x 10
```

```{r}
u7 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabag.tab", col_names = TRUE)
colnames(u7) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u7)
dim(u7) # 460 x 10
```

```{r}
u8 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabah.tab", col_names = TRUE)
colnames(u8) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u8)
dim(u8) # 474 x 10
```

```{r}
u9 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabai.tab", col_names = TRUE)
colnames(u9) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u9)
dim(u9) # 523 x 10
```

```{r}
u10 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabaj.tab", col_names = TRUE)
colnames(u10) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u10)
dim(u10) # 591 x 10
```

```{r}
u11 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabak.tab", col_names = TRUE)
colnames(u11) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u11)
dim(u11) # 444 x 10
```

```{r}
u12 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabal.tab", col_names = TRUE)
colnames(u12) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u12)
dim(u12) # 373 x 10
```

```{r}
u13 <- read_tsv("Functional_Annotation/Uniprot/uniprot_tabam.tab", col_names = TRUE)
colnames(u13) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u13)
dim(u13) # 442 x 10
```

```{r}
u14 <- read_tsv("Functional_Annotation/Uniprot/uniprot_taban.tab", col_names = TRUE)
colnames(u14) <- c("top_hit", "uniprotkb_entry", "status", "gene_names", "organism", "protein_names", "length", "gene_ontology", "go_ids", "my_list")
head(u14)
dim(u14) # 80 x 10
```


#Compile the Uniprot files 
```{r}
uniprot_results <- bind_rows(u1,u2,u3,u4,u5,u6,u7,u8,u9,u10,u11,u12,u13,u14)
uniprot_results <- unique(uniprot_results)
head(uniprot_results)
dim(uniprot_results) # 4738 x 10
uniprot_results <- filter(uniprot_results, grepl("GO",go_ids)) #Select only gnes with GO terms
dim(uniprot_results) # 2857 x 10
```

# Generate a list of GO terms - UniProt
```{r}
uniprot_GO <- select(uniprot_results, my_list, go_ids)
splitted <- strsplit(as.character(uniprot_GO$go_ids), ";") #split into multiple GO ids
uniprot_GO <- data.frame(v1 = rep.int(uniprot_GO$my_list, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
uniprot_GO <- unique(uniprot_GO)
colnames(uniprot_GO) <- c("gene_id", "GO.ID")
uniprot_GO$GO.ID <- gsub(" ", "", uniprot_GO$GO.ID)
nrow(uniprot_GO) # 6680 total GO terms from Uniprot
length(unique(uniprot_GO$GO.ID)) # 1106 unique GO terms from Uniprot
length(unique(uniprot_GO$gene_id)) # 2857 unique gene ids from Uniprot
```
 
# Not technically gene ids. Uniprot has no gene id info--actually ids from Uniprot. When I combine the Uniport and B2G files, the uniprot ids will then be associated with gene ids 


## Blast2GO

#Nucleotide CDS sequences were annotated using DIAMONDSEARCH BLASTX, resulting in 29515. These hits were used as input into Blast2GO to obtain GO terms using the 12/05/2020 obo database.

```{r}
B2G_results <- read.csv("Functional_Annotation/Blast2GO/pver_blast2go_table.csv")
B2G_results <- select(B2G_results, c("SeqName", "Description", "Length", "e.Value", "sim.mean", "GO.IDs", "GO.Names"))
colnames(B2G_results) <- c("seqName", "top_hit", "length", "eValue", "simMean", "GO.ID", "GO_names")
head(B2G_results)
dim(B2G_results) # 26436 x 7
B2G_results <- filter(B2G_results, grepl("GO",GO.ID)) #Genes with GO terms - 136
dim(B2G_results) # 4303 x 7
```


# Generate a list of GO terms - B2G
```{r}
B2G_results_GO <- select(B2G_results, top_hit, GO.ID)
splitted <- strsplit(as.character(B2G_results_GO$GO.ID), ";") #split into multiple GO ids
B2G_results_GO <- data.frame(v1 = rep.int(B2G_results_GO$top_hit, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
B2G_results_GO <- unique(B2G_results_GO)
colnames(B2G_results_GO) <- c("gene_id", "GO.ID")
B2G_results_GO$GO.ID <- gsub("F:", "", B2G_results_GO$GO.ID)
B2G_results_GO$GO.ID <- gsub("C:", "", B2G_results_GO$GO.ID)
B2G_results_GO$GO.ID <- gsub("P:", "", B2G_results_GO$GO.ID)
B2G_results_GO$GO.ID <- gsub(" ", "", B2G_results_GO$GO.ID)
head(B2G_results_GO)
nrow(B2G_results_GO) # 8686 total GO terms from B2G
length(unique(B2G_results_GO$GO.ID)) # 1246 unique GO terms from B2G
length(unique(B2G_results_GO$gene_id)) # 3068 unique gene ids from B2G
```



# Find intersections and unique results for each method (Uniprot and B2G)

## GO
# Intersection between GO terms for B2G and Uniprot

```{r}
BU_GO <- intersect(B2G_results_GO$GO.ID, uniprot_GO$GO.ID)
length(unique(BU_GO)) # 975 similar GO terms between B2G and Uniprot
```

# Difference in GO terms for B2G and Uniprot - B2G
```{r}
BUunique_GO <- setdiff(B2G_results_GO$GO.ID, uniprot_GO$GO.ID)
length(unique(BUunique_GO)) # 271 GO terms unique to B2G
```


# Difference in GO terms for B2G and Uniprot - Uniprot
```{r}
UBunique <- setdiff(uniprot_GO$GO.ID, B2G_results_GO$GO.ID)
length(unique(UBunique)) # 131 GO terms unique to Uniprot
```


## gene id
# Intersection between gene id for B2G and Uniprot
```{r}
BU_gene <- intersect(B2G_results_GO$gene_id, uniprot_GO$gene_id)
length(unique(BU_gene)) # 2852 similar gene ids between B2G and Uniprot
```


# Difference in gene ids for B2G and Uniprot
```{r}
BUunique_gene <- setdiff(B2G_results_GO$gene_id, uniprot_GO$gene_id)
length(unique(BUunique_gene)) # 216 gene ids unique to B2G
```


# Difference in gene ids for B2G and Uniprot
```{r}
UBunique_gene <- setdiff(uniprot_GO$gene_id, B2G_results_GO$gene_id)
length(unique(UBunique_gene)) # 5 gene ids unique to uniprot
```



## Merge Annotations - uniprot + b2g
```{r}
pver_annot <- left_join(pver_blast, B2G_results, by="seqName")
pver_annot <- select(pver_annot, seqName, top_hit.x, length.x, evalue, bitscore, simMean, GO.ID, GO_names)
colnames(pver_annot) <- c("gene_id", "top_hit", "length", "evalue", "bitscore", "simMean", "GO.ID", "GO_names")
uniprot_results_merge <- select(uniprot_results, -top_hit)
uniprot_results_merge <- rename(uniprot_results_merge, "top_hit"="my_list")
pver_annot <- left_join(pver_annot, uniprot_results_merge, by="top_hit")
pver_annot$GO <- paste(pver_annot$GO.ID, pver_annot$go_ids, sep=';') #generate new column with concatenated GO IDs
pver_annot$GO_terms <- paste(pver_annot$GO_names, pver_annot$gene_ontology, sep=';') #generate new column with concatenated GO IDs
pver_annot <- select(pver_annot, c("top_hit", "gene_id", "length.x", "evalue", "bitscore", "simMean", "uniprotkb_entry", "status", "protein_names", "gene_names", "organism", "GO", "GO_terms"))
pver_annot <- rename(pver_annot, "GO.IDs"="GO")
colnames(pver_annot) <- c("top_hit", "gene_id", "length", "eValue", "bitscore","simMean", "UniProtKB_entry", "status", "protein_names", "gene_names","organism", "GO_IDs","GO_terms")
names(pver_annot)
head(pver_annot)
tail(pver_annot)
dim(pver_annot)
write.csv(pver_annot, "Functional_Annotation/Output/pver_FuncAnn_UniP_B2G.csv")
```




# IPS
```{r}
IPS_GO <- read.csv("~/Desktop/PutnamLab/Repositories/SedimentStress/SedimentStress/Output/GOseq/acerv_GOterms.csv", header=TRUE)
IPS_GO <- select(IPS_GO, -X)
colnames(IPS_GO)[1] <-"gene_id"
#IPS_GO$gene_id <- gsub(".m1", "", IPS_GO$gene_id)
#IPS_GO$gene_id <- gsub("model", "TU", IPS_GO$gene_id)
length(unique(IPS_GO$gene_id)) # 12828
IPS_GO <- select(IPS_GO, c("gene_id", "GO_term")) # taking out source and score tho I want to leave them in--just for comparison though
splitted <- strsplit(as.character(IPS_GO$GO), ",") #split into multiple GO ids
IPS_GO <- data.frame(v1 = rep.int(IPS_GO$gene_id, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their 
colnames(IPS_GO) <- c("gene_id", "GO.ID")
IPS_GO <- unique(IPS_GO)
head(IPS_GO)
nrow(IPS_GO) # 30085 total GO terms from IPS
length(unique(IPS_GO$GO.ID)) # 1945 unique GO terms from IPS
length(unique(IPS_GO$gene_id)) # 12828 unique gene ids from IPS
```


# From B2G and Uniprot
```{r}
FuncAnn_UniP_B2G <- read.csv("~/Desktop/PutnamLab/Repositories/FunctionalAnnotation/FunctionalAnnotation/final_Annotations/acerv_FuncAnn_UniP_B2G.csv")
length(unique(FuncAnn_UniP_B2G$gene_id)) # 29515
FuncAnn_UniP_B2G_GO <- filter(FuncAnn_UniP_B2G, grepl("GO",GO.ID)) #Select only gnes with GO terms
length(unique(FuncAnn_UniP_B2G_GO$gene_id)) # 2172
FuncAnn_UniP_B2G_GO <- select(FuncAnn_UniP_B2G_GO, c("gene_id", "GO.ID")) # taking out source and score tho I want to leave them in--just for comparison though
splitted <- strsplit(as.character(FuncAnn_UniP_B2G_GO$GO.ID), ";") #split into multiple GO ids
FuncAnn_UniP_B2G_GO <- data.frame(v1 = rep.int(FuncAnn_UniP_B2G_GO$gene_id, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their 
colnames(FuncAnn_UniP_B2G_GO) <- c("gene_id", "GO.ID")
FuncAnn_UniP_B2G_GO$GO.ID <- gsub("F:", "", FuncAnn_UniP_B2G_GO$GO.ID)
FuncAnn_UniP_B2G_GO$GO.ID <- gsub("C:", "", FuncAnn_UniP_B2G_GO$GO.ID)
FuncAnn_UniP_B2G_GO$GO.ID <- gsub("P:", "", FuncAnn_UniP_B2G_GO$GO.ID)
FuncAnn_UniP_B2G_GO$GO.ID <- gsub(" ", "", FuncAnn_UniP_B2G_GO$GO.ID)
FuncAnn_UniP_B2G_GO <- unique(FuncAnn_UniP_B2G_GO)
head(FuncAnn_UniP_B2G_GO)
nrow(FuncAnn_UniP_B2G_GO) # 8390 total GO terms from B2G + Uniprot
length(unique(FuncAnn_UniP_B2G_GO$GO.ID)) # 1329 unique GO terms from B2G + Uniprot
length(unique(FuncAnn_UniP_B2G_GO$gene_id)) # 2172 unique gene ids from B2G + Uniprot
```


# Find intersections and unique results for each method (B2G + Uniprot and IPS)
## GO
# Intersection between GO terms for B2G + Uniprot and IPS
```{r}
IF_GO <- intersect(IPS_GO$GO.ID, FuncAnn_UniP_B2G_GO$GO.ID)
length(unique(IF_GO)) # 747 similar GO terms between B2G + Uniprot and IPS
```


# Difference in GO terms for B2G + Uniprot and IPS - FuncAnn first
```{r}
FIunique_GO <- setdiff(FuncAnn_UniP_B2G_GO$GO.ID, IPS_GO$GO.ID)
length(unique(FIunique_GO)) # 582 GO terms unique to B2G + Uniprot
```


# Difference in GO terms for B2G + Uniprot and IPS - IPS
```{r}
IFunique_GO <- setdiff(IPS_GO$GO.ID, FuncAnn_UniP_B2G_GO$GO.ID)
length(unique(IFunique_GO)) # 1198 GO terms unique to Uniprot
```


## gene id
# Intersection between GO terms for B2G + Uniprot and IPS
```{r}
IF_gene<- intersect(IPS_GO$gene_id, FuncAnn_UniP_B2G_GO$gene_id)
length(unique(IF_gene)) # 1140 similar gene ids between B2G + Uniprot and IPS
```

# Difference in gene ids for B2G + Uniprot and IPS - B2G + Uniprot
```{r}
FIunique_gene <- setdiff(FuncAnn_UniP_B2G_GO$gene_id, IPS_GO$gene_id)
length(unique(FIunique_gene)) # 1032 gene ids unique to B2G + Uniprot
```


# Difference in gene ids for B2G + Uniprot and IPS - IPS
```{r}
IFunique_gene <- setdiff(IPS_GO$gene_id, FuncAnn_UniP_B2G_GO$gene_id)
length(unique(IFunique_gene)) # 11688 gene ids unique to IPS
```



# Aggregate results 
```{r}
agg <- aggregate(IPS_GO$GO.ID, list(IPS_GO$gene_id), paste, collapse = ",")
colnames(agg) <- c("gene_id", "GO.ID")
```


```{r}
full_annot <- merge(agg, FuncAnn_UniP_B2G, by = "gene_id", all.x = T)
full_annot$GO.IDs <- paste(full_annot$GO.ID.x, full_annot$GO.ID.y, sep=';') #generate new column with concatenated GO IDs
full_annot <- select(full_annot, -c(X, GO.ID.x, GO.ID.y))
write.csv(full_annot, file = "~/Desktop/PutnamLab/Repositories/FunctionalAnnotation/FunctionalAnnotation/final_Annotations/acerv_fullAnnot.csv")
```





