---
title: "pver_GO_annot_compare"
author: "daniellembecker"
date: "11/5/2021"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries 
library("tidyverse")
library("dplyr")
```

#This script takes the results of functional annotation services and combines the results. Protein sequences from ReefGenomics (http://pver.reefgenomics.org/download/) were annotated using SWISSPROT BLASTP, TREMBL BLASTP, and DIAMONDSEARCH BLASTP. These hits were used as input into:

#  - Blast2GO

#Additional annotation was provided by

#  - InterProScan

## Load Swissprot/Blast2Go results
```{r}
# Load pver full annot results (diamond blast, swissprot, InterProScan; B2G)
swissprot <- read.csv("Functional_Annotation/Blast2GO/Pver_swissprot_blast2go_table.csv", header=T, na.strings=c("","NA", "---NA---", "no IPS match", "no GO terms", "n/a"))
dim(swissprot) # 27,439
head(swissprot)

#remove description NA rows 
#swissprot <- swissprot %>% drop_na("Description")


#remove X row, rename SeqName to gene_ID
swissprot <- swissprot[-c(1:2)]
swissprot <- rename(swissprot, gene_id = SeqName)

#remove empty unnecessary rows dont know if I need yet, will know after combining
#swissprot <- swissprot[, -c(13:15)] 

head(swissprot)

```

## Load Trembl/Blast2Go results
```{r}
# Load pver full annot results (diamond blast, swissprot, InterProScan; B2G)
trembl <- read.csv("Functional_Annotation/Blast2GO/Pver_trembl_blast2go_table.csv", header=T, na.strings=c("","NA", "---NA---", "no IPS match", "no GO terms", "n/a"))
dim(trembl) # 7,463
head(trembl)

#remove X row, rename SeqName to gene_ID
trembl <- trembl[-c(1:2)]
trembl <- rename(trembl, gene_id = SeqName)

#remove empty unnecessary rows 
#trembl <- trembl[, -c(13:15)] 

head(trembl)
```

## Load Diamond/Blast2Go results
```{r}
# Load pver full annot results (diamond blast, swissprot, InterProScan; B2G)
diamond <- read.csv("Functional_Annotation/Blast2GO/Pver_diamond_blast2go_table.csv", header=T, na.strings=c("","NA", "---NA---", "no IPS match", "no GO terms", "n/a"))
dim(diamond) # 26,436
head(diamond)

#remove X row, rename SeqName to gene_ID
diamond <- diamond[-c(1:2)]
diamond <- rename(diamond, gene_id = SeqName)

#remove empty unnecessary rows #dont know if I need yet, will know after combining
#diamond <- diamond[, -c(13:15)] 

#head(diamond)
```

## Load Interprot/Blast2Go results
```{r}
# Load pver full annot results (diamond blast, swissprot, InterProScan; B2G)
interprotscan <- read.csv("Functional_Annotation/Blast2GO/Pver_interprotscan_blast2go_table.csv", header=T, na.strings=c("","NA", "---NA---", "no IPS match", "no GO terms", "n/a"))
dim(interprotscan) # 27,439
head(interprotscan)

#remove X row, rename SeqName to gene_ID
interprotscan <- interprotscan[-c(1:2)]
interprotscan <- rename(interprotscan, gene_id = SeqName)

#remove empty unnecessary rows #dont know if I need yet, will know after combining
#interprotscan <- interprotscan[, -c(2:11)] 

head(interprotscan)
```

## Merge Annotations - trembl, swissprot, and interprotscan
```{r}
#Jill made two files due to the different GO terms, csv that had 
swiss_trem <- left_join(swissprot, trembl, by= c("gene_id"))
swiss_trem_inter <- left_join(swiss_trem, interprotscan, by = c("gene_id"))
swiss_trem_inter_diamond <- left_join(swiss_trem_inter, diamond, by= c("gene_id"))

#swiss_trem_diamond <- rbind(swiss_trem, diamond, by= c("gene_id"))

#swiss_trem_inter <- left_join(swiss_trem, interprotscan, by = c("gene_id"))

#sum(is.na(swiss_trem_inter$Description)) #720 NAs in description

pver_annot <- select(pver_annot, seqName, top_hit.x, length.x, evalue, bitscore, simMean, GO.ID, GO_names)
colnames(pver_annot) <- c("gene_id", "top_hit", "length", "evalue", "bitscore", "simMean", "GO.ID", "GO_names")
uniprot_results_merge <- select(uniprot_results, -top_hit)
uniprot_results_merge <- rename(uniprot_results_merge, "top_hit"="my_list")
pver_annot <- left_join(pver_annot, uniprot_results_merge, by="top_hit")
pver_annot$GO <- paste(pver_annot$GO.ID, pver_annot$go_ids, sep=';') #generate new column with concatenated GO IDs
pver_annot$GO_terms <- paste(pver_annot$GO_names, pver_annot$gene_ontology, sep=';') #generate new column with concatenated GO IDs
pver_annot <- select(pver_annot, c("top_hit", "gene_id", "length.x", "evalue", "bitscore", "simMean", "uniprotkb_entry", "status", "protein_names", "gene_names", "organism", "GO", "GO_terms"))
pver_annot <- rename(pver_annot, "GO.IDs"="GO")
colnames(pver_annot) <- c("top_hit", "gene_id", "length", "eValue", "bitscore","simMean", "UniProtKB_entry", "status", "protein_names", "gene_names","organism", "GO_IDs","GO_terms")
names(pver_annot)
head(pver_annot)
tail(pver_annot)
dim(pver_annot)
write.csv(pver_annot, "Functional_Annotation/Output/pver_FuncAnn_UniP_B2G.csv")
```



```{r}
# Count #GO terms, and enzyme codes 

length(na.omit(pver_annot$Swissprot_Hits)) # 19,536 sequences had hits against SwissProt database -- said 19,540 in our output file
length(na.omit(pver_annot$SwissProt_GO_IDs)) # 13,211 sequences were assigned GO terms
length(na.omit(pver_annot$SwissProt_Enzyme_Codes)) # 5,274 sequences were assigned enzyme codes 

# Count # InterProScan hits and GO terms
length(na.omit(pver_annot$InterPro.IDs)) # 25,679 sequences had hits against IPS database 
length(na.omit(pver_annot$InterPro.GO.IDs)) # 15,946 sequences were assigned GO terms 
```

```{r}
mean(na.omit(pver_annot$SwissProt_Num_GO)) # on average, ~12 GO terms assigned to a sequence for Swiss Prot

```

```{r}
# dfs separating GO IDs so each GO ID has its own row 


# SwissProt
GO_swiss <- select(pver_annot, SeqName, SwissProt_GO_IDs)
GO_swiss$SwissProt_GO_IDs <- gsub(",", ";", GO_swiss$SwissProt_GO_IDs)
splitted <- strsplit(as.character(GO_swiss$SwissProt_GO_IDs), ";") #split into multiple GO ids
GO_swiss <- data.frame(v1 = rep.int(GO_swiss$SeqName, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
colnames(GO_swiss) <- c("SeqName", "GO.ID")
GO_swiss$GO.ID <- gsub("F:", "", GO_swiss$GO.ID)
GO_swiss$GO.ID <- gsub("P:", "", GO_swiss$GO.ID)
GO_swiss$GO.ID <- gsub("C:", "", GO_swiss$GO.ID)
GO_swiss$GO.ID <- gsub(" ", "", GO_swiss$GO.ID)
GO_swiss[GO_swiss == "NA"] <- NA
GO_swiss <- na.omit(GO_swiss) # remove rows with NAs
dim(GO_swiss) # 165053 total GO terms

# InterProScan
GO_IPS <- select(pver_annot, SeqName, InterPro.GO.IDs)
GO_IPS$InterPro.GO.IDs <- gsub(",", ";", GO_IPS$InterPro.GO.IDs)
splitted <- strsplit(as.character(GO_IPS$InterPro.GO.IDs), ";") #split into multiple GO ids
GO_IPS <- data.frame(v1 = rep.int(GO_IPS$SeqName, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
colnames(GO_IPS) <- c("SeqName", "GO.ID")
GO_IPS$GO.ID <- gsub("F:", "", GO_IPS$GO.ID)
GO_IPS$GO.ID <- gsub("P:", "", GO_IPS$GO.ID)
GO_IPS$GO.ID <- gsub("C:", "", GO_IPS$GO.ID)
GO_IPS$GO.ID <- gsub(" ", "", GO_IPS$GO.ID)
GO_IPS[GO_IPS == "NA"] <- NA
GO_IPS <- na.omit(GO_IPS) # remove rows with NAs
dim(GO_IPS) # 42,128 total GO terms 
```

```{r}
# Join all GO dfs and remove duplicate rows 
GO_all <- rbind(GO_swiss, GO_IPS)
GO_all <- unique(GO_all)
write.csv(GO_all, "Functional_Annotation/Final_Annotations/pver_GOterms_interprot_swissprot_20211105.csv")

# Identify how many unique seqNames and GO IDs are in the df 
length(unique(GO_all$SeqName)) # 19035 seqs have one or more GO terms 
length(unique(GO_all$GO.ID)) # 16121 unique GO terms 
```


## Merge Annotations - trembl, swissprot, and interprotscan
```{r}
swiss_trem_inter <- left_join(swissprot, trembl, by="gene_id")
pver_annot <- select(pver_annot, seqName, top_hit.x, length.x, evalue, bitscore, simMean, GO.ID, GO_names)
colnames(pver_annot) <- c("gene_id", "top_hit", "length", "evalue", "bitscore", "simMean", "GO.ID", "GO_names")
uniprot_results_merge <- select(uniprot_results, -top_hit)
uniprot_results_merge <- rename(uniprot_results_merge, "top_hit"="my_list")
pver_annot <- left_join(pver_annot, uniprot_results_merge, by="top_hit")
pver_annot$GO <- paste(pver_annot$GO.ID, pver_annot$go_ids, sep=';') #generate new column with concatenated GO IDs
pver_annot$GO_terms <- paste(pver_annot$GO_names, pver_annot$gene_ontology, sep=';') #generate new column with concatenated GO IDs
pver_annot <- select(pver_annot, c("top_hit", "gene_id", "length.x", "evalue", "bitscore", "simMean", "uniprotkb_entry", "status", "protein_names", "gene_names", "organism", "GO", "GO_terms"))
pver_annot <- rename(pver_annot, "GO.IDs"="GO")
colnames(pver_annot) <- c("top_hit", "gene_id", "length", "eValue", "bitscore","simMean", "UniProtKB_entry", "status", "protein_names", "gene_names","organism", "GO_IDs","GO_terms")
names(pver_annot)
head(pver_annot)
tail(pver_annot)
dim(pver_annot)
write.csv(pver_annot, "Functional_Annotation/Output/pver_FuncAnn_UniP_B2G.csv")
```


# Aggregate results 
```{r}
agg <- aggregate(IPS_GO$GO.ID, list(IPS_GO$gene_id), paste, collapse = ",")
colnames(agg) <- c("gene_id", "GO.ID")
```


```{r}
full_annot <- merge(agg, FuncAnn_UniP_B2G, by = "gene_id", all.x = T)
full_annot$GO.IDs <- paste(full_annot$GO.ID.x, full_annot$GO.ID.y, sep=';') #generate new column with concatenated GO IDs
full_annot <- select(full_annot, -c(X, GO.ID.x, GO.ID.y))
write.csv(full_annot, file = "~/Desktop/PutnamLab/Repositories/FunctionalAnnotation/FunctionalAnnotation/final_Annotations/acerv_fullAnnot.csv")
```






