---
title: "pver_GO_annot_compare"
author: "daniellembecker"
date: "11/5/2021"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries 
library("tidyverse")
library("dplyr")
```

#The first part of this script takes the results of functional annotation services and combines the results. Protein sequences from ReefGenomics (http://pver.reefgenomics.org/download/) were annotated using SWISSPROT BLASTP, TREMBL BLASTP, and DIAMONDSEARCH BLASTP. These hits were used as input into:

#  - Blast2GO

#Additional annotation was provided by

#  - InterProScan

## Load Swissprot/Blast2Go results
```{r}
# Load pver full annot results (diamond blast, swissprot, InterProScan; B2G)
swissprot <- read.csv("Functional_Annotation/Blast2GO/Pver_swissprot_blast2go_table.csv", header=T, na.strings=c("","NA", "---NA---", "no IPS match", "no GO terms", "n/a"))
dim(swissprot) # 27,439
head(swissprot)

#remove X row, rename SeqName to gene_ID
swissprot <- swissprot[-c(1:2)]
swissprot <- rename(swissprot, gene_id = SeqName)

#remove empty unnecessary rows dont know if I need yet, will know after combining
swissprot <- swissprot[, -c(12:14)] 

head(swissprot)

```

## Load Trembl/Blast2Go results
```{r}
# Load pver full annot results (diamond blast, swissprot, InterProScan; B2G)
trembl <- read.csv("Functional_Annotation/Blast2GO/Pver_trembl_blast2go_table.csv", header=T, na.strings=c("","NA", "---NA---", "no IPS match", "no GO terms", "n/a"))
dim(trembl) # 7,463
head(trembl)

#remove X row, rename SeqName to gene_ID
trembl <- trembl[-c(1:2)]
trembl <- rename(trembl, gene_id = SeqName)

#remove empty unnecessary rows 
trembl <- trembl[, -c(12:14)] 


head(trembl)
```

## Load Diamond/Blast2Go results
```{r}
# Load pver full annot results (diamond blast, swissprot, InterProScan; B2G)
diamond <- read.csv("Functional_Annotation/Blast2GO/Pver_diamond_blast2go_table.csv", header=T, na.strings=c("","NA", "---NA---", "no IPS match", "no GO terms", "n/a"))
dim(diamond) # 26,436
head(diamond)

#remove X row, rename SeqName to gene_ID
diamond <- diamond[-c(1:2)]
diamond <- rename(diamond, gene_id = SeqName)

#remove empty unnecessary rows #dont know if I need yet, will know after combining
diamond <- diamond[, -c(12:14)] 

head(diamond)
```

## Load Interprot/Blast2Go results
```{r}
# Load pver full annot results (diamond blast, swissprot, InterProScan; B2G)
interprotscan <- read.csv("Functional_Annotation/Blast2GO/Pver_interprotscan_blast2go_table.csv", header=T, na.strings=c("","NA", "---NA---", "no IPS match", "no GO terms", "n/a"))
dim(interprotscan) # 27,439
head(interprotscan)

#remove X row, rename SeqName to gene_ID
interprotscan <- interprotscan[-c(1:2)]
interprotscan <- rename(interprotscan, gene_id = SeqName)

#remove empty unnecessary rows 
interprotscan <- interprotscan[, -2] 
interprotscan <- interprotscan[-c(3:10)]

head(interprotscan)
```

## Merge Annotations - trembl, swissprot, and interprotscan
```{r}
#combining all blast2go and diamond results
swiss_trem <- left_join(swissprot, trembl, by= c("gene_id"))
swiss_trem_inter <- left_join(swiss_trem, interprotscan, by = c("gene_id"))
pver_annot <- left_join(swiss_trem_inter, diamond, by= c("gene_id"))

write.csv(pver_annot, "Functional_Annotation/Final_Annotations/pver_annot_full.csv")
```



```{r}
# Count #GO terms, and enzyme codes 

# Count SwissProt hits and GO terms
length(na.omit(pver_annot$SWISS_.Hits)) # 19,536 sequences had hits against SwissProt database 
length(na.omit(pver_annot$SWISS_GO.IDs)) # 19,450 sequences were assigned GO terms
length(na.omit(pver_annot$SWISS_Enzyme.Codes)) # 7,627 sequences were assigned enzyme codes 

# Count Trembl hits and GO terms
length(na.omit(pver_annot$TREMBL_.Hits)) # 6,743 sequences had hits against Trembl database
length(na.omit(pver_annot$TREMBL_GO.IDs)) # 4,399 sequences were assigned GO terms
length(na.omit(pver_annot$TREMBL_Enzyme.Codes)) # 787 sequences were assigned enzyme codes

# Count InterProScan hits and GO terms
length(na.omit(pver_annot$InterPro.IDs)) # 25,679 sequences had hits against IPS database 
length(na.omit(pver_annot$InterPro.GO.IDs)) # 15,946 sequences were assigned GO terms 

# Count Diamond Blast hits and GO terms
length(na.omit(pver_annot$BLAST_.Hits)) # 26,426 sequences had hits against NCBI nr database
length(na.omit(pver_annot$BLAST_GO.IDs)) # 4,303 sequences were assigned GO terms
length(na.omit(pver_annot$BLAST_Enzyme.Codes)) # 1,098 sequences were assigned enzyme codes


```

```{r}
# check on average how many GO terms assigned to a sequence for each software

mean(na.omit(pver_annot$SWISS_.GO)) # on average, ~12 GO terms assigned to a sequence for Swiss Prot

mean(na.omit(pver_annot$TREMBL_.GO)) # on average, ~3 GO terms assigned to a sequence for Trembl

mean(na.omit(pver_annot$BLAST_.GO)) # on average, ~3 GO terms assigned to a sequence for Diamond Blast

```

############################################################################################

#The second part of this script takes the combined functional annotation results and organizes a list of GO terms per each gene model.

```{r}
# make individual dfs separating GO IDs so each GO ID has its own row for each software

# SwissProt
GO_swiss <- select(pver_annot, gene_id, SWISS_GO.IDs)
GO_swiss$SWISS_GO.IDs <- gsub(",", ";", GO_swiss$SWISS_GO.IDs)
splitted <- strsplit(as.character(GO_swiss$SWISS_GO.IDs), ";") #split into multiple GO ids
GO_swiss <- data.frame(v1 = rep.int(GO_swiss$gene_id, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
colnames(GO_swiss) <- c("gene_id", "GO.ID")
GO_swiss$GO.ID <- gsub("F:", "", GO_swiss$GO.ID)
GO_swiss$GO.ID <- gsub("P:", "", GO_swiss$GO.ID)
GO_swiss$GO.ID <- gsub("C:", "", GO_swiss$GO.ID)
GO_swiss$GO.ID <- gsub(" ", "", GO_swiss$GO.ID)
GO_swiss[GO_swiss == "NA"] <- NA
#GO_swiss <- na.omit(GO_swiss) # remove rows with NAs
dim(GO_swiss) # 235,923 total GO terms

# Trembl
GO_trembl <- select(pver_annot, gene_id, TREMBL_GO.IDs)
GO_trembl$TREMBL_GO.IDs <- gsub(",", ";", GO_trembl$TREMBL_GO.IDs)
splitted <- strsplit(as.character(GO_trembl$TREMBL_GO.IDs), ";") #split into multiple GO ids
GO_trembl <- data.frame(v1 = rep.int(GO_trembl$gene_id, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
colnames(GO_trembl) <- c("gene_id", "GO.ID")
GO_trembl$GO.ID <- gsub("F:", "", GO_trembl$GO.ID)
GO_trembl$GO.ID <- gsub("P:", "", GO_trembl$GO.ID)
GO_trembl$GO.ID <- gsub("C:", "", GO_trembl$GO.ID)
GO_trembl$GO.ID <- gsub(" ", "", GO_trembl$GO.ID)
GO_trembl[GO_trembl == "NA"] <- NA
#GO_trembl <- na.omit(GO_trembl) # remove rows with NAs
dim(GO_trembl) # 13,222 total GO terms


# InterProScan
GO_IPS <- select(pver_annot, gene_id, InterPro.GO.IDs)
GO_IPS$InterPro.GO.IDs <- gsub(",", ";", GO_IPS$InterPro.GO.IDs)
splitted <- strsplit(as.character(GO_IPS$InterPro.GO.IDs), ";") #split into multiple GO ids
GO_IPS <- data.frame(v1 = rep.int(GO_IPS$gene_id, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
colnames(GO_IPS) <- c("gene_id", "GO.ID")
GO_IPS$GO.ID <- gsub("F:", "", GO_IPS$GO.ID)
GO_IPS$GO.ID <- gsub("P:", "", GO_IPS$GO.ID)
GO_IPS$GO.ID <- gsub("C:", "", GO_IPS$GO.ID)
GO_IPS$GO.ID <- gsub(" ", "", GO_IPS$GO.ID)
GO_IPS[GO_IPS == "NA"] <- NA
#GO_IPS <- na.omit(GO_IPS) # remove rows with NAs
dim(GO_IPS) # 42,128 total GO terms 


# Diamond Blast
GO_blast <- select(pver_annot, gene_id, BLAST_GO.IDs)
GO_blast$BLAST_GO.IDs <- gsub(",", ";", GO_blast$BLAST_GO.IDs)
splitted <- strsplit(as.character(GO_blast$BLAST_GO.IDs), ";") #split into multiple GO ids
GO_blast <- data.frame(v1 = rep.int(GO_blast$gene_id, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
colnames(GO_blast) <- c("gene_id", "GO.ID")
GO_blast$GO.ID <- gsub("F:", "", GO_blast$GO.ID)
GO_blast$GO.ID <- gsub("P:", "", GO_blast$GO.ID)
GO_blast$GO.ID <- gsub("C:", "", GO_blast$GO.ID)
GO_blast$GO.ID <- gsub(" ", "", GO_blast$GO.ID)
GO_blast[GO_blast == "NA"] <- NA
#GO_blast <- na.omit(GO_blast) # remove rows with NAs
dim(GO_blast) # 11,288 total GO terms

```

```{r}

# Compare GO terms across databases
# Make db identifier column
GO_blast$database <- "blast"
GO_swiss$database <- "swissprot"
GO_IPS$database <- "IPS"
GO_trembl$database <- "trembl"

# Merge dfs by gene_id 
# First, merge by blast and swissprot
db_compare <- merge(GO_blast, GO_swiss, by = "gene_id", all = T)

# Then merge w/ IPS
db_compare <- merge(db_compare, GO_IPS, by = "gene_id", all = T)
colnames(db_compare) <- c("gene_id", "GO.ID.blast", "database.blast", "GO.ID.swissprot", "database.swissprot", "GO.ID.IPS", "database.IPS")

# Then merge w/ trembl
db_compare <- merge(db_compare, GO_trembl, by = "gene_id", all = T)
colnames(db_compare) <- c("gene_id", "GO.ID.blast", "database.blast", "GO.ID.swissprot", "database.swissprot", "GO.ID.IPS", "database.IPS", "GO.ID.trembl", "database.trembl")

# Find intersections among dbs
all <- na.omit(as.data.frame(intersect(intersect(intersect(db_compare$GO.ID.blast, db_compare$GO.ID.swissprot), db_compare$GO.ID.IPS), db_compare$GO.ID.trembl)))
colnames(all) <- "Blast_SP_IPS_blast_trembl"
db1 <- na.omit(as.data.frame(intersect(db_compare$GO.ID.blast, db_compare$GO.ID.swissprot)))
colnames(db1) <- "Blast_SP"
db2 <- na.omit(as.data.frame(intersect(db_compare$GO.ID.blast, db_compare$GO.ID.IPS)))
colnames(db2) <- "Blast_IPS"
db3 <- na.omit(as.data.frame(intersect(db_compare$GO.ID.IPS, db_compare$GO.ID.swissprot)))
colnames(db3) <- "SP_IPS"
db4 <- na.omit(as.data.frame(intersect(db_compare$GO.ID.swissprot, db_compare$GO.ID.trembl)))
colnames(db4) <- "SP_trembl"
db5 <- na.omit(as.data.frame(intersect(db_compare$GO.ID.blast, db_compare$GO.ID.trembl)))
colnames(db5) <- "blast_trembl"
db6 <- na.omit(as.data.frame(intersect(db_compare$GO.ID.IPS, db_compare$GO.ID.trembl)))
colnames(db6) <- "IPS_trembl"


dim(all) # 520 GO terms shared between blast, swissprot, IPS, and trembl
dim(db1) # 1185 GO terms shared between blast & swissprot
dim(db2) # 1056 GO terms shared between blast & IPS
dim(db3) # 2805 GO terms shared between swissprot & IPS
dim(db4) # 1100 GO terms shared between swissprot & trembl
dim(db5) # 593 GO terms shared between blast & trembl
dim(db6) # 794 GO terms shared between IPS & trembl

# Check unique values 
length(unique(all$Blast_SP_IPS_blast_trembl)) # 520
length(unique(db1$Blast_SP)) # 618
length(unique(db2$Blast_IPS)) # 429
length(unique(db3$SP_IPS)) # 1588
length(unique(db4$SP_trembl)) # 1100
length(unique(db5$blast_trembl)) # 593
length(unique(db6$IPS_trembl)) # 794

```


```{r}
# Join all GO dfs and remove duplicate rows 
GO_all <- rbind(GO_swiss, GO_IPS, GO_blast, GO_trembl)
GO_all <- unique(GO_all)
write.csv(GO_all, "Functional_Annotation/Final_Annotations/pver_GOterms_interprot_swissprot_blast_trembl_20211203.csv")

# Identify how many unique gene_ids and GO IDs are in the df 
length(unique(GO_all$gene_id)) # 24,037 gene_ids have one or more GO terms, means that not all gene models got assigned GO terms (3,402 did not), but kept NAs for downstream analysis
length(unique(GO_all$GO.ID)) # 17,024 unique GO terms 
```


