---
title: "Host RNA_seq_DESeq2"
author: "echille"
edited by: "daniellembecker"
date: "20230126"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Differential Expression Analysis 
## Follow RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow before statistical analysis to make gene counts and transcript counts matrices

### Set up workspace

Load libraries
```{r, message=FALSE, warning=FALSE}
library("genefilter")
library("car")
library("devtools")
library("DESeq2")
library("factoextra")
library("MuMIn")
library("tidyverse")
library("tidyr")
library("lme4")
library("RColorBrewer")
library("magrittr")
library("ggplot2")
library("goseq")
library("pheatmap")
library("stats")
library("gridExtra")
library("VennDiagram")
library("here")
library("plotrix")
library("Glimma")
library("edgeR")
library("tibble")
library("ggpubr")
```

Import the data files 
```{r}
#treatment information
treatmentinfo <- read.csv("RAnalysis/Data/RNA-seq/P.verrucosa/metadata.RNAseq.csv", header = TRUE, sep = ",")

treatmentinfo%>%
  select(sample_id, block, fragment.ID, treatment)

head(treatmentinfo)

#gene count matrix
gcount <- as.data.frame(read.csv("RAnalysis/Data/RNA-seq/P.verrucosa/Host/Poc_gene_count_matrix.csv", row.names="gene_id"), colClasses = double)

#gcount$gene_id <-  gsub("_gene_","_",gcount$gene_id)
#gcount$gene_id <-  gsub("_split_","_",gcount$gene_id)
#gcount$gene_id <-  gsub("-.*","",gcount$gene_id) #remove the extra g number after the dash
#row.names(gcount) <- gcount$gene_id

dim(gcount)
head(gcount)


#remove extra characters from multiple column header names for sample ids, skip first column labeled gene counts so specify 1:ncol
for ( col in 1:ncol(gcount)){
  colnames(gcount)[col] <-  sub("_R1.fastp.trim.20230215.fq.gz.sam.sorted.bam.merge.gtf", "", colnames(gcount)[col])
}

#subset the gcount matrix and reorder it so it matches the treatment info
gcount <- gcount[ ,treatmentinfo$sample_id]
head(gcount)

# Make sure treatment info ='s the rows of count data
all(rownames(treatmentinfo) %in% colnames(gcount)) # must come out TRUE

```


##### Pre-filter gene counts
## Pre-filtering our dataset to reduce the memory size dataframe, increase the speed of the transformation and testing functions, and improve quality of statistical analysis by removing low-coverage counts. Removed counts could represent outliers in the data and removing these improves sensitivity of statistical tests. We will filter out low coverage samples. Here, we will keep P=50% percent of the samples have counts over A=10, allowing only half our samples to have a value less than 10 per gene.

```{r}
#pOverA: Specifying the minimum count for a proportion of samples for each gene. Here, we are using a pOverA of 0.5, 10. This is because we have 32 samples with a minimum of n=1 sample per fragment and half in enriched and half in control. Therefore, we will accept genes that are present in 16/32 = 0.5 of the samples because we may expect different expression by treatment We are further setting the minimum count of genes to 10, such that 50% of the samples must have a gene count of >10 in order for the gene to remain in the data set.  

filt <- filterfun(pOverA(0.90,10))

#create filter for the counts data
gfilt <- genefilter(gcount, filt)

#identify genes to keep by count filter
gkeep <- gcount[gfilt,]

#identify gene lists
gn.keep <- rownames(gkeep)

#gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt <- as.data.frame(gcount[which(rownames(gcount) %in% gn.keep),])
head(gcount_filt)
dim(gcount_filt)

```
Display current order of metadata and gene count matrix.  
```{r}
treatmentinfo$sample_id
colnames(gcount_filt)
```

These are not in the same order, re order them.  
```{r}
list<-colnames(gcount_filt)
list<-as.factor(list)

treatmentinfo$sample_id<-as.factor(treatmentinfo$sample_id)

# Re-order the levels
treatmentinfo$sample_id <- factor(as.character(treatmentinfo$sample_id), levels=list)

# Re-order the data.frame
treatmentinfo_ordered <- treatmentinfo[order(treatmentinfo$sample_id),]
treatmentinfo_ordered$sample_id

treatmentinfo_ordered<-treatmentinfo_ordered%>%
  dplyr::select(sample_id, block, treatment)
```

Check the order now that it has been corrected. 
```{r}
treatmentinfo_ordered$sample_id
colnames(gcount_filt)
```

##Merge the treatment columns into a new column , group. Set group as a factor.
```{r}
treatmentinfo_ordered$treatment <- factor(treatmentinfo_ordered$treatment, levels = c("control","enriched"))
```

#Create a DESeqDataSet design from gene count matrix and labels. Here we set the design to look at treatment to test for any differences in gene expression across timepoints attributed to treatment.
#not using DESeq due to random factors, but can still use this to visualize clusters
```{r}
#Set DESeq2 design
gdds <- DESeqDataSetFromMatrix(countData = gcount_filt,
                              colData = treatmentinfo_ordered,
                              design = ~treatment)
```

#### Visualize gene count data

We're looking to see if the samples of the same treatments cluster

##### Log-transform the count data
First we are going to log-transform the data using a variance stabilizing transforamtion (vst). This is only for visualization purposes. Essentially, this is roughly similar to putting the data on the log2 scale. It will deal with the sampling variability of low counts by calculating within-group variability (if blind=FALSE). Importantly, it does not use the design to remove variation in the data, and so can be used to examine if there may be any variability do to technical factors such as extraction batch effects.

To do this we first need to calculate the size factors of our samples. This is a rough estimate of how many reads each sample contains compared to the others. In order to use VST (the faster log2 transforming process) to log-transform our data, the size factors need to be less than 4. Otherwise, there could be artefacts in our results.
```{r}
SF.gdds <- estimateSizeFactors(gdds) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than four to use vst
print(sizeFactors(SF.gdds)) #View size factors
```

Our size factors are all less than 4, so we can use VST!
```{r}
gvst <- vst(gdds, blind=FALSE) #apply a variance stabilizing transformation to minimize effects of small counts and normalize library size
```

# Conduct PERMANOVA and PCA on global expression and DEG

Export data for PERMANOVA test.  
```{r}
test<-t(assay(gvst)) #export as matrix
test<-as.data.frame(test)

#add category columns
test$sample_id <- rownames(test)
test$treatment <- treatmentinfo_ordered$treatment[match(test$sample_id, treatmentinfo_ordered$sample_id)]
```

Build PERMANOVA model.  
```{r}
#last column with treatment metadata test[19823] and test[19822], need to start at third to last column for the numberic error issue

scaled_test <-prcomp(test[c(1:19821)], scale=TRUE, center=TRUE)
fviz_eig(scaled_test)

# scale data
vegan <- scale(test[c(1:19821)])

# PerMANOVA 
permanova <- vegan::adonis2(vegan ~ treatment, data = test, method='eu')
permanova
```

Gene expression is significantly different between Treatments (p=0.023).  

Permutation test for adonis under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 999

adonis2(formula = vegan ~ treatment, data = test, method = "eu")
          Df SumOfSqs      R2      F Pr(>F)  
treatment  1    32572 0.05301 1.6793  0.016 *
Residual  30   581879 0.94699                
Total     31   614451 1.00000                
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1  

#### Heatmap of sample to sample distances
```{r}
# Using vst object, plot heat-map of sample-to-sample distances
sampleDists <- dist(t(assay(gvst))) # calculate distance matrix
sampleDistMatrix <- as.matrix(sampleDists) # create distance matrix
rownames(sampleDistMatrix) <- colnames(vst) # assign row names
colnames(sampleDistMatrix) <- NULL # assign col names 
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255) # assign colors 
heatmap <- pheatmap(sampleDistMatrix, # plot distance matrix
         clustering_distance_rows = sampleDists, # cluster rows
         clustering_distance_cols = sampleDists, # cluster cols
         col=colors) # set colors
heatmap
pdf(file = "RAnalysis/Output/RNA-seq/DEG/Host/sampletosample_heatmap_host.pdf")
heatmap
dev.off()

```


#### Principal component plot of samples
```{r}
gPCAdata <- plotPCA(gvst, intgroup = c("treatment"), returnData=TRUE)
percentVar <- round(100*attr(gPCAdata, "percentVar")) #plot PCA of samples with all data
ggplot(gPCAdata, aes(PC1, PC2, color=treatment)) + 
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  scale_color_manual(values = c(control="black", enriched="cadetblue3")) +
  coord_fixed() +
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
        #panel.grid.major = element_blank(), #Set major gridlines
        #panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) + #Set the plot background +
  stat_ellipse()

ggsave(path = "RAnalysis/Output/Final_Figures/Host/", filename = "DEGSeq2_general_PCA_plot_host.pdf")
ggsave(path = "RAnalysis/Output/Final_Figures/Host/", filename = "DEGSeq2_general_PCA_plot_host.png")
```

##### Run DE analysis for overall example 

#Run differential expression test using a Wald model. 
```{r, message = FALSE}
DEGSeq2_wald <- DESeq(gdds) # run differential expression test by treatment (?) using wald test 
# estimating size factors
# estimating dispersions
# gene-wise dispersion estimates
# mean-dispersion relationship
# final dispersion estimates
# fitting model and testing
```

#Explore significant p-values for treatment
```{r, message = FALSE}
DEGSeq2.results <- results(DEGSeq2_wald, contrast=c("treatment","control","enriched")) #run wald model
head(DEGSeq2.results)
sum(DEGSeq2.results$padj < 0.05, na.rm=TRUE) #How many adjusted p-values were less than 0.05?

#256

DEGSeq2.results <- as.data.frame(DEGSeq2.results)
#create data frame with only DESeq2 DEG's with < 0.05 significance 

DEGSeq2.sig.results <- subset(DEGSeq2.results, padj < 0.05)
DEGSeq2.sig.results <- as.data.frame(DEGSeq2.sig.results)
DEGSeq2.sig.results$gene_id <-rownames(DEGSeq2.sig.results)
write.csv(DEGSeq2.sig.results, 'RAnalysis/Output/RNA-seq/DEG/Host/DEGSeq2.sig.results.host.csv')  

```

# Make MA plot from Glimma
```{r}
#### The MA plot is a visualisation that plots the log-fold-change between experimental groups (M) against the mean expression across all the samples (A) for each gene. The Glimma MA plot contains two main components: a plot of summary statistics across all genes that have been tested, and a plot of gene expression from individual samples for a given gene

MA_plot <- ggmaplot(DEGSeq2.results, main = expression("Control" %->% "Enriched"),
         fdr = 0.05, fc = 1, size = 1,
            palette = c("#B31B21", "#1465AC", "darkgray"),
         genenames = as.vector(DEGSeq2.results$names),
   font.label = c("bold", 16),
   font.legend = c("bold", 12),
   font.main = c("bold", 14),
         ggtheme = ggplot2::theme_minimal(),
         top = 0, label.select = c("BUB1", "CD83")); MA_plot

```


# Use ggvolc to visualize DEG up and down regulated genes
```{r}

# Define criteria for upregulation and downregulation

#rename for moving forward
volcano <-DEGSeq2.results

# Create a new column for -log10(padj), noticed outliers in data
volcano$log_padj <- -log10(volcano$padj)

# Remove rows where log_padj is greater than 15
volcano <- volcano[volcano$log_padj <= 15, ]

volcano$gene_expression_status <- ifelse(volcano$padj > 0.05, "Not Significant",
                                       ifelse(volcano$log2FoldChange > 0, "Upregulated",
                                              ifelse(volcano$log2FoldChange < 0, "Downregulated", "Not Classified")))

# Set up the color palette for hyper, hypo, and non-significant genes
colors <- c("Upregulated" = "red", "Downregulated" = "blue", "Not Significant" = "black")

# Create the volcano plot using ggplot2
DEG.volcano <- ggplot(volcano, aes(x = log2FoldChange, y = -log10(padj), color = gene_expression_status)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(values = colors, name = "Gene Expression Status (padj < 0.05)", 
                     breaks = c("Upregulated", "Downregulated", "Not Significant"),
                     labels = c("Upregulated (84)", "Downregulated (172)", "Not Significant"),
                     drop = FALSE) +  # Include all specified breaks in the legend
  labs(title = "", x = "log2FoldChange", y = "-log10(adjusted p-value)") +
  theme_bw() +
  geom_hline(yintercept = 0, linetype = "dashed", color = "black"); DEG.volcano

ggsave("RAnalysis/Output/RNA-seq/DEG/DEG.volcano.pdf", DEG.volcano, width = 8, height = 5, units = c("in"))

#load png image of DMG PCA

DEG.PCA <- readPNG("RAnalysis/Output/Final_Figures/Host/DEGSeq2_DEG_PCA_plot_host.png")
PCA.DEG <- rasterGrob(DEG.PCA, interpolate=TRUE)

Figure3 <- plot_grid(PCA.DEG, DEG.volcano, labels = c('A', 'B'), label_size = 14)

ggsave(filename=paste0("RAnalysis/Output/Final_Figures/Host/manuscript/DEG.PCA.volcano.png"), plot=Figure3, dpi=300, width=12, height=5, units="in", limitsize = FALSE)
ggsave(filename=paste0("RAnalysis/Output/Final_Figures/Host/manuscript/DEG.PCA.volcano.pdf"), plot=Figure3, dpi=300, width=12, height=5, units="in", limitsize = FALSE)


```




## Identify unique genes from DEGs from treatments 
```{r}
dim(DEGSeq2.sig.results) # 256 x 7
length(unique(DEGSeq2.sig.results$gene_id)) # 256 unique genes between all treatments 
write.csv(DEGSeq2.sig.results, file = "RAnalysis/Output/RNA-seq/DEG/Host/DEGs.all_treatment_20230126_host.csv")
```

## Visualize differentially expressed genes from DESeq2
```{r}
# Subset list of genes by those which padj>0.
DEGSeq2.output <- as.data.frame(DEGSeq2.sig.results)
dim(DEGSeq2.output)
DEGSeq2.results.all <- DEGSeq2.sig.results
DEGSeq2 <- rownames(DEGSeq2.results.all) # list all gene names 
DEGSeq2 <- unique(DEGSeq2) # select only unique gene names 
DEGSeq2_list <- gdds[which(rownames(gdds) %in% DEGSeq2)] # filter gdds DESeq2 object by unique gene names
dim(DEGSeq2_list) # 15 x 15
print(counts(DEGSeq2_list))

# As determined above, size factors all less than 4, so proceed with VST
# apply a variance stabilizing transformation to minimize effects of small counts and normalize library size

DEGvst <- vst(DEGSeq2_list, blind = FALSE, nsub = nrow(counts(DEGSeq2_list)))

dim(DEGvst) # 15 x 15
print(assay(DEGvst)) # look at vst-transformed gene count data 

# Plot heat map with diff expressed genes
# Make a matrix for computing similarity
mat <- assay(DEGvst)#[DEG.results.all, ] # make an expression object
mat <- mat - rowMeans(mat) #difference in expression compared to average across all samples
dim(mat)
ann_colors <- list(treatment= c(control="black", enriched="cadetblue3"))
df_DEGSeq2 <- as.data.frame(colData(DEGvst)[c("treatment")]) #make dataframe for column naming and associated treatment
DEGSeq2_heatmap <- pheatmap(mat, scale= "row", legend=TRUE, annotation_legend=TRUE, annotation_col=df_DEGSeq2, annotation_colors = ann_colors,
                            clustering_distance_rows="euclidean", clustering_method = "average",
                            show_rownames =FALSE,
                            show_colnames =TRUE,
                            cluster_cols = TRUE)


pdf(file = "RAnalysis/Output/Final_Figures/Host/DEGSeq2_DEG_heatmap_host.pdf")
DEGSeq2_heatmap
dev.off()


# PCA plot of diff-expressed genes 
DEGSeq2_PCAdata <- plotPCA(DEGvst, intgroup = c("treatment"), returnData=TRUE)
percentVar_pca <- round(100*attr(DEGSeq2_PCAdata, "percentVar")) #plot PCA of samples with all data
DEGSeq2_PCA_plot <- ggplot(DEGSeq2_PCAdata, aes(PC1, PC2, color=treatment)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar_pca[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_pca[2],"% variance")) +
  scale_color_manual(values = c(control="black", enriched="cadetblue3")) +
  coord_fixed() +
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
        #panel.grid.major = element_blank(), #Set major gridlines
        #panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) + #Set the plot background +
  stat_ellipse()


DEGSeq2_PCA_plot

ggsave(path = "RAnalysis/Output/Final_Figures/Host/", filename = "DEGSeq2_DEG_PCA_plot_host.pdf")
ggsave(path = "RAnalysis/Output/Final_Figures/Host/", filename = "DEGSeq2_DEG_PCA_plot_host.png")

# PCA plot is of deferentially expressed genes only

# Save results

write.csv(counts(DEGSeq2_list), file="RAnalysis/Output/RNA-seq/DEG/Host/DEGSeq2_list_unique_host.csv")

```

Export data for PERMANOVA test.  
```{r}
test1<-t(assay(DEGvst)) #export as matrix
test1<-as.data.frame(test1)

#add category columns
test1$sample_id <- rownames(test1)
test1$treatment <- treatmentinfo_ordered$treatment[match(test1$sample_id, treatmentinfo_ordered$sample_id)]
```

Build PERMANOVA model.  
```{r}
#last column with treatment metadata test[254] and test[253], need to start at third to last column for the numberic error issue

scaled_test1 <-prcomp(test1[c(1:252)], scale=TRUE, center=TRUE)
fviz_eig(scaled_test1)

# scale data
vegan <- scale(test1[c(1:252)])

# PerMANOVA 
permanova<-adonis2(vegan ~ treatment, data = test1, method='eu')
permanova
```

Gene expression between DEGs also signiicant

Permutation test for adonis under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 999

adonis2(formula = vegan ~ treatment, data = test1, method = "eu")
          Df SumOfSqs      R2      F Pr(>F)    
treatment  1   1867.5 0.23905 9.4246  0.001 ***
Residual  30   5944.5 0.76095                  
Total     31   7812.0 1.00000  

## DEG summary info 
## Run to calculate the DEGs, and upregulated vs. downregulated genes
```{r}
length(unique(DEGSeq2.sig.results$gene_id)) # 256 unique DEGs
down <- DEGSeq2.sig.results %>% filter(log2FoldChange < 0) 
length(unique(down$gene_id)) # 172 unique DEGs downregulated
up <- DEGSeq2.sig.results %>% filter(log2FoldChange > 0)
length(unique(up$gene_id)) # 84 unique DEGs upregulated

```

```{r}
#make gct expression data set for GSEA analysis

norm_counts <- counts(SF.gdds, normalized = T)

fid <- "norm_counts.gct" 
writeLines(c("#1.2", paste(nrow(norm_counts), ncol(norm_counts) - 2, collapse="\t")), fid, sep="\n")
write.table(norm_counts, "RAnalysis/Data/norm_counts.txt", file=fid, quote=FALSE, row.names=FALSE, col.names=TRUE, sep="\t", append = TRUE)

#write to .txt or tab deliniated file

```






