---
title: "Host RNA_seq_DESeq2"
author: "echille"
edited by: "daniellembecker"
date: "20230126"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Differential Expression Analysis 
## Follow RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow before statistical analysis to make gene counts and transcript counts matrices

### Set up workspace

Load libraries
```{r, message=FALSE, warning=FALSE}
library("genefilter")
library("DESeq2")
library("factoextra")
library("tidyverse")
library("RColorBrewer")
library("pheatmap")
library("ggpubr")
library("png")
library("grid")
library("cowplot")
```

Import the data files 
```{r}
#treatment information
treatmentinfo <- read.csv("RAnalysis/Data/RNA-seq/P.verrucosa/metadata.RNAseq.csv", header = TRUE, sep = ",")

#view treatment information
head(treatmentinfo)

#load gene count matrix
gcount <- as.data.frame(read.csv("RAnalysis/Data/RNA-seq/P.verrucosa/Host/Poc_gene_count_matrix.csv", row.names="gene_id"), colClasses = double)

#confirm expectations of 32 samples by 27439 genes
dim(gcount)
#view data
head(gcount)

#remove extra characters from multiple column header names for sample ids
for ( col in 1:ncol(gcount)){
  colnames(gcount)[col] <-  sub("_R1.fastp.trim.20230215.fq.gz.sam.sorted.bam.merge.gtf", "", colnames(gcount)[col])
}

#set the metadata row names as the sample names
row.names(treatmentinfo) <- treatmentinfo$sample_id

# Make sure treatment info ='s the rows of count data
all(rownames(treatmentinfo) %in% colnames(gcount)) # must come out TRUE

#check that the order of the samples in the gene count columns matches the order in the treatment info rows

#Display current order of metadata and gene count matrix.  
row.names(treatmentinfo) 
colnames(gcount)
#These are not in the same order, re order them.  

#identify the genecount order
list<-colnames(gcount)
list<-as.factor(list)

# Re-order the levels of the treatment information
treatmentinfo$sample_id <- factor(as.character(treatmentinfo$sample_id), levels=list)

# Re-order the data.frame
treatmentinfo_ordered <- treatmentinfo[order(treatmentinfo$sample_id),]

#Check the order of metadata and gene count matrix.  
row.names(treatmentinfo_ordered)
treatmentinfo_ordered$sample_id
colnames(gcount)

#all are now in the same order 
```

##### Filter gene counts
## Filtering our dataset to reduce the memory size dataframe, increase the speed of the transformation and testing functions, and improve quality of statistical analysis by removing low-coverage counts. Removed counts could represent outliers in the data and removing these improves sensitivity of statistical tests. We will filter out low coverage samples. Here, we will keep P=50% percent of the samples have counts over A=10, allowing only half our samples to have a value less than 10 per gene.

```{r}
#pOverA: Specifying the minimum count for a proportion of samples for each gene. Here, we are using a pOverA of 0.5, 10. This is because we have 32 samples, with half in enriched (n=16) and half in control (n=16). Therefore, we will accept genes that are present in 16/32 = 0.5 of the samples because we may expect different expression by treatment. We are further setting the minimum count of genes to 10, such that 50% of the samples must have a gene count of >10 in order for the gene to remain in the data set.  

filt <- filterfun(pOverA(0.50,10))

#create filter for the counts data
gfilt <- genefilter(gcount, filt)

#identify genes to keep by count filter
gkeep <- gcount[gfilt,]

#identify gene lists
gn.keep <- rownames(gkeep)

#gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt <- as.data.frame(gcount[which(rownames(gcount) %in% gn.keep),])
head(gcount_filt)
dim(gcount_filt)
dim(gcount)

```
We now have 19,821 genes after filtering, reduced from 27,439 genes before filtering. 


Check the order now that the gene count matrix has been filtered 
```{r}
row.names(treatmentinfo_ordered)
treatmentinfo_ordered$sample_id
colnames(gcount_filt)
```

##Set the treatment columns as a factor and identify the levels.
```{r}
unique(treatmentinfo_ordered$treatment)
treatmentinfo_ordered$treatment <- factor(treatmentinfo_ordered$treatment, levels = c("control","enriched"))
```

#Create a DESeqDataSet design from gene count matrix and labels. Here we set the design to look at treatment to test for any differences in gene expression across timepoints attributed to treatment.
#not using DESeq due to random factors, but can still use this to visualize clusters
```{r}
#Set DESeq2 design
gdds <- DESeqDataSetFromMatrix(countData = gcount_filt,
                              colData = treatmentinfo_ordered,
                              design = ~treatment)
```

#### Visualize gene count data

We're looking to see if the samples of the same treatments cluster

##### Log-transform the count data
First we are going to transform the data using a variance stabilizing transforamtion (vst). This is only for visualization purposes.  It will deal with the sampling variability of low counts by calculating within-group variability (if blind=FALSE). Importantly, it does not use the design to remove variation in the data, and so can be used to examine if there may be any variability do to technical factors such as extraction batch effects.

To do this we first need to calculate the size factors of our samples. This is a rough estimate of how many reads each sample contains compared to the others. In order to use VST transform our data, the size factors need to be less than 4. Otherwise, there could be artifacts in our results.
```{r}
SF.gdds <- estimateSizeFactors(gdds) #estimate size factors to determine if we can use vst to transform our data. 
print(sizeFactors(SF.gdds)) #View size factors
range(sizeFactors(SF.gdds))
```

Our size factors are all less than 4, so we can use VST
```{r}
gvst <- vst(gdds, blind=FALSE) #apply a variance stabilizing transformation to minimize effects of small counts and normalize library size
```

#### Heatmap of sample to sample distances
```{r}
# Using vst object, plot heat-map of sample-to-sample distances
sampleDists <- dist(t(assay(gvst))) # calculate distance matrix
sampleDistMatrix <- as.matrix(sampleDists) # create distance matrix
rownames(sampleDistMatrix) <- colnames(vst) # assign row names
colnames(sampleDistMatrix) <- NULL # assign col names 
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255) # assign colors 
heatmap <- pheatmap(sampleDistMatrix, # plot distance matrix
         clustering_distance_rows = sampleDists, # cluster rows
         clustering_distance_cols = sampleDists, # cluster cols
         col=colors) # set colors
heatmap
pdf(file = "RAnalysis/Output/RNA-seq/DEG/Host/sampletosample_heatmap_host.pdf")
heatmap
dev.off()

```


#### Principal component plot of samples
```{r}
gPCAdata <- plotPCA(gvst, intgroup = c("treatment"), returnData=TRUE)
percentVar <- round(100*attr(gPCAdata, "percentVar")) #plot PCA of samples with all data
ggplot(gPCAdata, aes(PC1, PC2, color=treatment)) + 
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  scale_color_manual(values = c(control="black", enriched="grey")) +
  coord_fixed() +
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
        #panel.grid.major = element_blank(), #Set major gridlines
        #panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) + #Set the plot background +
  stat_ellipse()

ggsave(path = "RAnalysis/Output/Final_Figures/Host/", filename = "DEGSeq2_general_PCA_plot_host.pdf")
ggsave(path = "RAnalysis/Output/Final_Figures/Host/", filename = "DEGSeq2_general_PCA_plot_host.png")
```

##### Run DE analysis 

#Run differential expression test using a Wald model. 
```{r, message = FALSE}
DEGSeq2_wald <- DESeq(gdds) # run differential expression test by treatment using wald test 
```

#Explore significant adjusted p-values, considering FDR, for treatment

When running results, the contrast command should be in this order: factor being tested, numerator (the treatment), and denominator (the reference treatment). 

Contrast is: "a character vector with exactly three elements: the name of a factor in the design formula, the name of the numerator level for the fold change, and the name of the denominator level for the fold change (simplest case)"  

Use resultsNames to show you the model outputs from DESeq2. 
```{r, message = FALSE}
resultsNames(DEGSeq2_wald)

DEGSeq2.results <- results(DEGSeq2_wald, contrast=c("treatment","enriched","control"))
#view DEG analysis results 

head(DEGSeq2.results)
dim(DEGSeq2.results)
sum(DEGSeq2.results$padj < 0.05, na.rm=TRUE) #p-values were less than 0.05

DEGSeq2.results <- as.data.frame(DEGSeq2.results)
#create data frame with only DESeq2 DEG's with < 0.05 significance 

DEGSeq2.sig.results <- subset(DEGSeq2.results, padj < 0.05)
DEGSeq2.sig.results <- as.data.frame(DEGSeq2.sig.results)
DEGSeq2.sig.results$gene_id <-rownames(DEGSeq2.sig.results)
write.csv(DEGSeq2.sig.results, 'RAnalysis/Output/RNA-seq/DEG/Host/DEGSeq2.sig.results.host.csv')  

```
252 genes are differential at padj<0.05 with no fold change filtering. 

# Use ggvolc to visualize DEG up and down regulated genes

AH removed the filtering for pvalue - it was taking out the highly significant genes.  

```{r}

# Define criteria for upregulation and downregulation

#rename for moving forward
volcano <-DEGSeq2.results

# Create a new column for -log10(padj), noticed outliers in data
volcano$log_padj <- -log10(volcano$padj)

# Remove rows where log_padj is greater than 15
#volcano <- volcano[volcano$log_padj <= 15, ]

volcano$gene_expression_status <- ifelse(volcano$padj > 0.05, "Not Significant",
                                       ifelse(volcano$log2FoldChange > 0, "Upregulated",
                                              ifelse(volcano$log2FoldChange < 0, "Downregulated", "Not Classified")))

# Set up the color palette for hyper, hypo, and non-significant genes
colors <- c("Upregulated" = "red", "Downregulated" = "blue", "Not Significant" = "black")

# Create the volcano plot using ggplot2
DEG.volcano <- ggplot(volcano, aes(x = log2FoldChange, y = -log10(padj), color = gene_expression_status)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(
    values = colors, 
    name = "Gene Expression Status",  
    breaks = c("Upregulated", "Downregulated", "Not Significant"),
    labels = c("Upregulated (174)", "Downregulated (78)", "Not Significant"),
    drop = FALSE
  ) +  # Include all specified breaks in the legend
  labs(title = "", x = "log2FoldChange", y = "-log10(adjusted p-value)") +
  theme_bw() +
  geom_hline(yintercept = 0, linetype = "dashed", color = "black"); DEG.volcano

ggsave("RAnalysis/Output/RNA-seq/DEG/DEG.volcano.pdf", DEG.volcano, width = 8, height = 5, units = c("in"))

#load png image of DEG PCA
DEG.PCA <- readPNG("RAnalysis/Output/Final_Figures/Host/DEGSeq2_DEG_PCA_plot_host.png")
PCA.DEG <- rasterGrob(DEG.PCA, interpolate=TRUE)
Figure3 <- plot_grid(DEG.volcano, PCA.DEG, labels = c('A', 'B'), label_size = 14)

ggsave(filename=paste0("RAnalysis/Output/Final_Figures/Host/manuscript/DEG.PCA.volcano.png"), plot=Figure3, dpi=300, width=12, height=5, units="in", limitsize = FALSE)
ggsave(filename=paste0("RAnalysis/Output/Final_Figures/Host/manuscript/DEG.PCA.volcano.pdf"), plot=Figure3, dpi=300, width=12, height=5, units="in", limitsize = FALSE)
```


## Identify unique genes from DEGs from treatments 
```{r}
dim(DEGSeq2.sig.results) # 
length(unique(DEGSeq2.sig.results$gene_id)) # 
write.csv(DEGSeq2.sig.results, file = "RAnalysis/Output/RNA-seq/DEG/Host/DEGs.all_treatment_20240530_host.csv")
```

## Visualize differentially expressed genes from DESeq2
```{r}

DEGSeq2.results.all <- DEGSeq2.sig.results
DEGSeq2 <- rownames(DEGSeq2.results.all) # list all gene names 
DEGSeq2_list <- gdds[which(rownames(gdds) %in% DEGSeq2)] # filter gdds DESeq2 object by gene names
dim(DEGSeq2_list) 
print(counts(DEGSeq2_list))

# As determined above, size factors all less than 4, so proceed with VST
# apply a variance stabilizing transformation to DEGs to minimize effects of small counts and normalize library size
DEGvst <- vst(DEGSeq2_list, blind = FALSE, nsub = nrow(counts(DEGSeq2_list)))

dim(DEGvst) 
print(assay(DEGvst)) # look at vst-transformed gene count data 

# Plot heat map with diff expressed genes
# Make a matrix for computing similarity
mat <- assay(DEGvst) # make an expression object
mat <- mat - rowMeans(mat) #difference in expression compared to average across all samples
dim(mat)
ann_colors <- list(treatment= c(control="black", enriched="grey"))
df_DEGSeq2 <- as.data.frame(colData(DEGvst)[c("treatment")]) #make dataframe for column naming and associated treatment
DEGSeq2_heatmap <- pheatmap(mat, scale= "row", legend=TRUE, annotation_legend=TRUE, annotation_col=df_DEGSeq2, annotation_colors = ann_colors,
                            clustering_distance_rows="euclidean", clustering_method = "average",
                            show_rownames =FALSE,
                            show_colnames =TRUE,
                            cluster_cols = TRUE)
DEGSeq2_heatmap

pdf(file = "RAnalysis/Output/Final_Figures/Host/DEGSeq2_DEG_heatmap_host.pdf")
DEGSeq2_heatmap
dev.off()

#make label size for plot below
label_size <- 16

# PCA plot of diff-expressed genes 
DEGSeq2_PCAdata <- plotPCA(DEGvst, intgroup = c("treatment"), returnData=TRUE)
percentVar_pca <- round(100*attr(DEGSeq2_PCAdata, "percentVar")) #plot PCA of samples with all data
#make label size for plot below
label_size <- 16

DEGSeq2_PCA_plot <- ggplot(DEGSeq2_PCAdata, aes(PC1, PC2, color=treatment)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar_pca[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_pca[2],"% variance")) +
  scale_color_manual(values = c(control="black", enriched="grey")) +
  coord_fixed() +
  theme_bw() + #Set background color
   theme(
    panel.border = element_blank(),
    axis.line = element_line(colour = "black"),
    plot.background = element_blank(),
    # Increase the size of axis text
    axis.text = element_text(size = label_size),
    # Increase the size of x and y axis titles
    axis.title = element_text(size = label_size),
    # Increase the size of legend text
    legend.text = element_text(size = label_size),
    # Increase the size of legend title
    legend.title = element_text(size = label_size)
  ) +
  stat_ellipse()


DEGSeq2_PCA_plot

ggsave("RAnalysis/Output/Final_Figures/Host/DEGSeq2_DEG_PCA_plot_host.pdf", DEGSeq2_PCA_plot, width = 8, height = 5, units = c("in"))
ggsave("RAnalysis/Output/Final_Figures/Host/DEGSeq2_DEG_PCA_plot_host.png", DEGSeq2_PCA_plot, width = 8, height = 5, units = c("in"))
# PCA plot is of DEGs only

# Save results
write.csv(counts(DEGSeq2_list), file="RAnalysis/Output/RNA-seq/DEG/Host/DEGSeq2_list_unique_host.csv")
```

## DEG summary info 
## Run to calculate the DEGs, and upregulated vs. downregulated genes
```{r}
length(unique(DEGSeq2.sig.results$gene_id)) # 
down <- DEGSeq2.sig.results %>% filter(log2FoldChange < 0) 
length(unique(down$gene_id)) # DEGs downregulated
up <- DEGSeq2.sig.results %>% filter(log2FoldChange > 0)
length(unique(up$gene_id)) # DEGs upregulated

```
78 downregulated; 174 upregulated 

## Visualize fold change of DEGs between treatments 

```{r}
plotCounts(DEGSeq2_wald, gene=rownames(up)[1], intgroup="treatment")

plotCounts(DEGSeq2_wald, gene=rownames(down)[1], intgroup="treatment")
```





