---
title: "GO_Term_Enrichment_Analysis"
author: "daniellembecker"
edited by: "daniellembecker"
date: "9/20/2021"
output: html_document
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## RNAseq Gene Ontology (GO) Enrichment Analysis 
## Previous steps include RNAseq workflow in Bioinformatics>RNAseq>RNAseq workflow and Differential Gene Expression statistical analysis in RAnalysis>Scripts>RNAseq_Differential_Gene_Expression to make DEG statistical data sheet
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
library("DESeq2")
library("tidyverse")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
library("genefilter")
library("ggplot2")
library("gplots")
library("limma")
library("spdep") 
library("patchwork")
library("adegenet") 
library("goseq")
library("gridExtra")
library("clusterProfiler")
library("DataCombine")
library("VennDiagram")
library("GSEABase")
library("data.table")
if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")
library("rrvgo")
if ("org.Ce.eg.db" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("org.Ce.eg.db")
library(org.Ce.eg.db)
```

# Import the data files 
```{r, echo=FALSE}
# Structural annotation GFF3 file
GFF3 <- read.csv("RAnalysis/Output/RNA-seq/GOSeq/Host/rename_structure_gff3.csv")

# Load go terms from functional annotation file
annot_GO <- read.csv("RAnalysis/Output/RNA-seq/GOSeq/Host/annot_GO.terms.host.csv")[-1]

# Check if all gene IDs in GFF3 are in annot_GO
unmatched_genes <- setdiff(GFF3$gene_id, annot_GO$gene_id)
# Print the unmatched gene IDs
print(unmatched_genes)

```

# Identifying treatment and all expressed pver genes (poverA = 0.50,10)
```{r}
#treatment information
treatmentinfo <- read.csv("RAnalysis/Data/RNA-seq/metadata.RNAseq.csv", header = TRUE, sep = ",")
str(treatmentinfo)
head(treatmentinfo)

#DEG significant results
DEG.res <- read.csv("RAnalysis/Output/RNA-seq/DEG/Host/DEGSeq2.sig.results.host.csv")[,-1]
nrow(DEG.res)
#DEG.res$gene_id <-  gsub("_gene","",DEG.res$gene_id) #remove extra characters

colnames(DEG.res)[7] <-"gene_id" # make colnames a true column called gene_id
head(DEG.res)

# Load filtered gene count file
gcount_filt <- read.csv("RAnalysis/Output/RNA-seq/DEG/gene_count_matrix_filtered.csv")

```

# Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for each set of DEGs.
#Build GOSEQ vector 
#GOseq requires a vector of all genes, all differentially expressed genes, and gene lengths
```{r}
#Make ID and length vectors for all genes, UPREG and DOWNREG
#all genes
DEG <- GFF3[GFF3$gene_id %in% DEG.res$gene_id, ][-1]
dim(DEG) # 252 x 10
DEG_names <- as.vector(DEG$gene_id)
gene_vector <- as.integer(GFF3$gene_id%in%DEG_names)
names(gene_vector) <- GFF3$gene_id

# Filter for upregulated genes
DEG.up <- DEG.res[DEG.res$direction == "Upregulated", ]

# Upregulated ID and length vectors
DEG.up <- GFF3[GFF3$gene_id %in% DEG.up$gene_id, ][-1]
dim(DEG.up) # 174 x 10
DEG_names.up <- as.vector(DEG.up$gene_id)
gene_vector.up <- as.integer(GFF3$gene_id%in%DEG_names.up)
names(gene_vector.up) <- GFF3$gene_id

# Filter for downregulated genes
DEG.down <- DEG.res[DEG.res$direction == "Downregulated", ]

# Downregulated ID and length vectors
DEG.down <- GFF3[GFF3$gene_id %in% DEG.down$gene_id, ][-1]
dim(DEG.down) # 78 x 10
DEG_names.down <- as.vector(DEG.down$gene_id)
gene_vector.down <- as.integer(GFF3$gene_id%in%DEG_names.down)
names(gene_vector.down) <- GFF3$gene_id

# Make ID vector 
IDvector <- GFF3$gene_id

# Make length vector 
lengthVector <- GFF3$length

```

#Calculate Probability Weighting Function
```{r}
pwf<-nullp(gene_vector, IDvector, bias.data=lengthVector) #weight vector by length of gene
pwf.up<-nullp(gene_vector.up, IDvector, bias.data=lengthVector) #weight vector by length of gene for upreg
pwf.down<-nullp(gene_vector.down, IDvector, bias.data=lengthVector) #weight vector by length of gene for downreg
```


# Perform GOseq
## Find enriched GO terms, "selection-unbiased testing for category enrichment amongst differentially expressed (DE) genes for RNA-seq data"
```{r}
### Perform GOseq for all, upreg and down reg
## Perform GOseq for all genes
# Find enriched GO terms, "selection-unbiased testing for category enrichment amongst differentially expressed (DE) genes for RNA-seq data"
GO.wall<-goseq(pwf, IDvector, gene2cat=annot_GO, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
#adjust p-values 
GO.wall$bh_adjust <-  p.adjust(GO.wall$over_represented_pvalue, method="BH") 
  
## Perform GOseq for upreg genes  
GO.wall.up<-goseq(pwf.up, IDvector, gene2cat=annot_GO, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
#adjust p-values 
GO.wall.up$bh_adjust <-  p.adjust(GO.wall.up$over_represented_pvalue, method="BH") 
   
## Perform GOseq for downreg genes   
GO.wall.down<-goseq(pwf.down, IDvector, gene2cat=annot_GO, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO.wall.down$bh_adjust <-  p.adjust(GO.wall.down$over_represented_pvalue, method="BH") 
    
# Using manually entered categories.
# Calculating the p-values...
# 'select()' returned 1:1 mapping between keys and columns
write.csv(GO.wall, file = "RAnalysis/Output/RNA-seq/GOSeq/Host/pver_GO_ALL_host.csv")
write.csv(GO.wall.up, file = "RAnalysis/Output/RNA-seq/GOSeq/Host/pver_GO_UPREG_host.csv")
write.csv(GO.wall.down, file = "RAnalysis/Output/RNA-seq/GOSeq/Host/pver_GO_DOWNREG_host.csv")

# Find significantly enriched GO terms in all genes
enriched.GO.05 <-GO.wall$category[GO.wall$over_represented_pvalue<.05]
enriched.GO.05 <-data.frame(enriched.GO.05)
colnames(enriched.GO.05) <- c("category")
enriched.GO.05 <- merge(enriched.GO.05, GO.wall, by="category")
enriched.GO.05 <- enriched.GO.05[order(-enriched.GO.05$numDEInCat),]
enriched.GO.05$term <- as.factor(enriched.GO.05$term)
head(enriched.GO.05)

# Find significantly enriched GO terms in upregulated genes
enriched.GO.05.up <-GO.wall.up$category[GO.wall.up$over_represented_pvalue<.05]
enriched.GO.05.up <-data.frame(enriched.GO.05.up)
colnames(enriched.GO.05.up) <- c("category")
enriched.GO.05.up <- merge(enriched.GO.05.up, GO.wall.up, by="category")
enriched.GO.05.up <- enriched.GO.05.up[order(-enriched.GO.05.up$numDEInCat),]
enriched.GO.05.up$term <- as.factor(enriched.GO.05.up$term)
head(enriched.GO.05.up)

# Find significantly enriched GO terms in downregulated genes
enriched.GO.05.down <-GO.wall.down$category[GO.wall.down$over_represented_pvalue<.05]
enriched.GO.05.down <-data.frame(enriched.GO.05.down)
colnames(enriched.GO.05.down) <- c("category")
enriched.GO.05.down <- merge(enriched.GO.05.down, GO.wall.down, by="category")
enriched.GO.05.down <- enriched.GO.05.down[order(-enriched.GO.05.down$numDEInCat),]
enriched.GO.05.down$term <- as.factor(enriched.GO.05.down$term)
head(enriched.GO.05.down)

# Subset enriched GO terms by ontology (BP, CC, MF) and save as csv. 
MF <- subset(enriched.GO.05, ontology=="MF")
MF <- MF[order(-MF$numDEInCat),]
CC <- subset(enriched.GO.05, ontology=="CC") 
CC <- CC[order(-CC$numDEInCat),] 
BP <- subset(enriched.GO.05, ontology=="BP")
BP <- BP[order(-BP$numDEInCat),]
write.csv(MF, file = "RAnalysis/Output/RNA-seq/GOSeq/Host/pver_MF_Sig_Enriched_GO.05.host.csv")
write.csv(CC, file = "RAnalysis/Output/RNA-seq/GOSeq/Host/pver_CC_Sig_Enriched_GO.05.host.csv") 
write.csv(BP, file = "RAnalysis/Output/RNA-seq/GOSeq/Host/pver_BP_Sig_Enriched_GO.05.host.csv")
write.csv(enriched.GO.05, file = "RAnalysis/Output/RNA-seq/GOSeq/Host/pver_Sig_Enriched_GO.05_ALL_host.csv")

nrow(enriched.GO.05) # 312
nrow(filter(enriched.GO.05, ontology=="BP")) #number sig BP terms 211
nrow(filter(enriched.GO.05, ontology=="MF")) #number sig MF terms 76
nrow(filter(enriched.GO.05, ontology=="CC")) #number sig CC terms 22
```

## Merge DEG gene ids and GO info
```{r}
# #include enrichment information for each category (GO Term)
colnames(annot_GO) <- c("gene_id", "category")
merge <- merge(enriched.GO.05, annot_GO, by.x = "category") # contains gene ids and GO info 

# Merge DEG with merge 
merge_again <- merge(merge, DEG, by = "gene_id") # contains GO info and gene counts for DEGs
```


## Find GOslim terms
# Run GOslim to get broader categories
```{r}
#load in generic GO database, has all of the upper level categories for GO terms. Ex: regulation of cell division would be under cellular regulation
slim <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") #get GO database

## filtering all of BP (do MF and CC seperately)
BP_GO <- enriched.GO.05 %>%
  filter(ontology=="BP")
BPGO_collection <- GOCollection(BP_GO$category) #Make library of query terms
slims_bp <- data.frame(goSlim(BPGO_collection, slim, "BP")) #Find common parent terms to slim down our list
slims_bp$category <- row.names(slims_bp) #save rownames as category

## filtering all of MF
MF_GO <- enriched.GO.05 %>%
  filter(ontology=="MF")
MFGO_collection <- GOCollection(MF_GO$category) #Make library of query terms
slims_mf <- data.frame(goSlim(MFGO_collection, slim, "MF")) #Find common parent terms to slim down our list
slims_mf$category <- row.names(slims_mf) #save rownames as category

## filtering all of CC
CC_GO <- enriched.GO.05 %>%
  filter(ontology=="CC")
CCGO_collection <- GOCollection(CC_GO$category) #Make library of query terms
slims_cc <- data.frame(goSlim(CCGO_collection, slim, "CC")) #Find common parent terms to slim down our list
slims_cc$category <- row.names(slims_cc) #save rownames as category
```

# Get mapped terms, using functions from Sam White's Biostars [post](https://support.bioconductor.org/p/128407/#128409).
```{r}
#custom function from Sam White's, gets mapped ids for all of your query terms 
#Write function mappedIds to get the query terms that mapped to the slim categories
mappedIds <-
  function(df, collection, OFFSPRING) #the command to run requires a dataframe of slim terms, like slims_MF above, your list of query terms, and the offspring from the GOCollection by goSlim
  {
    map <- as.list(OFFSPRING[rownames(df)]) # Subset GOcollection offspring by the rownames of your dataframe
    mapped <- lapply(map, intersect, ids(collection)) #Find the terms that intersect between the subset made above of your query terms and the GOids from the GO collection
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L)) #Add column "go_terms" with matching terms 
    df #show resulting dataframe
  }
#Run function for MF and BP terms
BPslim <- mappedIds(slims_bp, BPGO_collection, GOBPOFFSPRING)
MFslim <- mappedIds(slims_mf, MFGO_collection, GOMFOFFSPRING)
CCslim <- mappedIds(slims_cc, CCGO_collection, GOCCOFFSPRING)
```

# Remove duplicate matches, keeping the broader umbrella term
```{r}
#filtering out duplicates
#BP
BPslim <- filter(BPslim, Count>0 & Term!="biological_process") #filter out empty slims and term "biological process"
BPsplitted <- strsplit(as.character(BPslim$go_terms), ";") #split into multiple GO ids
BPslimX <- data.frame(Term = rep.int(BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) #list all
BPslimX <- merge(BPslimX, BPslim[,c(1,3:4)], by="Term") #Add back counts, term, and category info
BPslimX <- unique(setDT(BPslimX)[order(go_term, -Count)], by = "go_term") #remove duplicate offspring terms, keeping only those that appear in the larger umbrella term (larger Count number)
BPslim <- data.frame(slim_term=BPslimX$Term, slim_cat=BPslimX$category, category=BPslimX$go_term) #rename columns
head(BPslim)

#MF
MFslim <- filter(MFslim, Count>0 & Term!="molecular_function") #filter out empty slims and term "molecular function"
MFsplitted <- strsplit(as.character(MFslim$go_terms), ";") #split into multiple GO ids
MFslimX <- data.frame(Term = rep.int(MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) #list all
MFslimX <- merge(MFslimX, MFslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
MFslimX <- unique(setDT(MFslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
MFslim <- data.frame(slim_term=MFslimX$Term, slim_cat=MFslimX$category, category=MFslimX$go_term) #rename columns
head(MFslim)

#CC
CCslim <- filter(CCslim, Count>0 & Term!="cellular_component") #filter out empty slims and term "molecular function"
CCsplitted <- strsplit(as.character(CCslim$go_terms), ";") #split into multiple GO ids
CCslimX <- data.frame(Term = rep.int(CCslim$Term, sapply(CCsplitted, length)), go_term = unlist(CCsplitted)) #list all
CCslimX <- merge(CCslimX, CCslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
CCslimX <- unique(setDT(CCslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
CCslim <- data.frame(slim_term=CCslimX$Term, slim_cat=CCslimX$category, category=CCslimX$go_term) #rename columns
head(CCslim)

```

# Save slim info with GO enrichment info for heatmap dataframes.
```{r}
GO.BP <- right_join(BPslim, filter(enriched.GO.05, ontology=="BP"), by="category") #add back GO enrichment info for each offspring term
GO.MF <- right_join(MFslim, filter(enriched.GO.05, ontology=="MF"), by="category") #add back GO enrichment info for each offspring term
GO.CC <- right_join(CCslim, filter(enriched.GO.05, ontology=="CC"), by="category") #add back GO enrichment info for each offspring term
```


## Make heatmap
```{r}
BPplot <- GO.BP %>% mutate(term = fct_reorder(term, -over_represented_pvalue)) %>%
    ggplot(aes(x = ontology, y = term)) + 
    geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
      scale_y_discrete(position = "right") +
      facet_grid(slim_term~ ., scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
      theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
      strip.text.y.left = element_text(angle=0, size = 20, face = "bold"),
      strip.text.x = element_text(size = 30, face = "bold"),
      axis.title = element_blank(),
      axis.text.x = element_text(size = 30),
      axis.text = element_text(size = 14), legend.title = element_text(size = 20), legend.text =     
      element_text(size = 20))

MFplot <- GO.MF %>% mutate(term = fct_reorder(term, -over_represented_pvalue)) %>%
    ggplot(aes(x = ontology, y = term)) + 
    geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
      scale_y_discrete(position = "right") +
      facet_grid(slim_term~ ., scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
      theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
      strip.text.y.left = element_text(angle=0, size = 20, face = "bold"),
      strip.text.x = element_text(size = 30, face = "bold"),
      axis.title = element_blank(),
      axis.text.x = element_text(size = 30),
      axis.text = element_text(size = 14), legend.title = element_text(size = 20), legend.text =     
      element_text(size = 20))
fig5 <- BPplot + MFplot

ggsave("RAnalysis/Output/Final_Figures/Host/GO_enrichment_DEG_heatmap_host.pdf", fig5, width = 30, height = 30, units = c("in"))
```

## Make supplemental table summarizing GO enrichment
```{r}
GO.enrichment.dat <- bind_rows(GO.BP, GO.MF, GO.CC)
head(GO.enrichment.dat)

#Make dataframe of GO results for clustering and heatmap. 
#add gene_IDs. To get gene_IDs we will merge with the GO.terms DF.
GOgenes <- data.frame(gene_id=annot_GO$gene_id, category=annot_GO$category) 
GOgenes$gene_id <- as.character(GOgenes$gene_id) #make gene ID a character so we can collapse our many near-identical columns

GO.enrichment.summary  <- left_join(GO.enrichment.dat, GOgenes, by="category" ) #join the DFs

GO.enrichment.summary <- GO.enrichment.summary %>% #collapse and have gene IDs for a particular term in a single row as a comma-sep list. 
  group_by(slim_term, slim_cat, category, over_represented_pvalue, under_represented_pvalue, numDEInCat, numInCat, term, ontology, bh_adjust) %>%
  summarise(genes = toString(gene_id)) %>% #rename collapsed gene_ID column "gene"
  ungroup()


write.csv(GO.enrichment.summary, 'RAnalysis/Output/RNA-seq/GOSeq/Host/GO.enrichment.summary.table.DEG.host.csv') 
```


## Exploring unique go_slim terms
```{r}
#number of unique go slim terms per category
length(unique(GO.enrichment.summary$slim_term)) #44 unique go terms

#view slim terms per ontology category
## filtering go slim category of BP 
BP_slim <- GO.enrichment.summary %>%
  filter(ontology=="BP")
view(BP_slim)
#number of BP go slim terms
length(unique(BP_slim$slim_term)) #32

## filtering go slim category of MF 
MF_slim <- GO.enrichment.summary %>%
  filter(ontology=="MF")
view(MF_slim)
#number of BP go slim terms
length(unique(MF_slim$slim_term)) #8

## filtering go slim category of CC 
CC_slim <- GO.enrichment.summary %>%
  filter(ontology=="CC")
view(CC_slim)

#number of CC go slim terms
length(unique(CC_slim$slim_term)) #4

```

# Make plots of ontology seperated DEGs
```{r}
MF_plot_GO <- GO.MF  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                   #mutate(term = fct_reorder(term, Day)) %>%
                   ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom"
                   ) +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("Molecular Function") +
                   theme_bw() + #Set background color 
   #facet_wrap(~slim_term) +
   theme(axis.text.y = element_text(hjust = 1))   +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes

 MF_plot_GO

 ggsave("RAnalysis/Output/Final_Figures/Host/GO.plot.MF.terms.host.pdf", MF_plot_GO , width = 10, height = 10, units = c("in"))

 BP_plot_GO <- GO.BP  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                   ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom") +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("Biological Processes") +
   facet_grid(slim_term ~ ., scales = "free", space='free') + 
   theme_bw() +
   theme(axis.text.y = element_text(hjust = 1))   +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes



BP_plot_GO

 ggsave("RAnalysis/Output/Final_Figures/Host/GO.plot.BP.terms.host.pdf", BP_plot_GO , width = 10, height = 49, units = c("in"))
 

 CC_plot_GO <- GO.CC  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                   #mutate(term = fct_reorder(term, Day)) %>%
                   ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom"
                   ) +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("Cellular Components") +
                   theme_bw() + #Set background color 
   #facet_wrap(~slim_term) +
   theme(axis.text.y = element_text(hjust = 1))   +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes

 CC_plot_GO

 ggsave("RAnalysis/Output/Final_Figures/Host/GO.plot.CC.terms.host.pdf", CC_plot_GO , width = 10, height = 10, units = c("in"))

```

#trying different approach to heatmaps using Ariana approach
```{r}
GO.plot.BP <-  ggplot(GO.BP, aes(x = ontology, y = term)) + 
    geom_point(aes(size=numInCat, fill= -log10(over_represented_pvalue), colour=-log10(over_represented_pvalue))) + 
    scale_size(name="Number of Genes", range=c(5,12), limits=c(1,500))+
    scale_fill_gradient(name = "Adjusted p-value", low="black", high="gray")+
    scale_colour_gradient(name = "Adjusted p-value", low="black", high="gray")+
    facet_grid(slim_term ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 60, multi_line = TRUE))+
    theme_bw() + 
    theme(panel.border = element_blank(), 
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
          strip.background =element_rect(fill="white"),
          strip.text.y = element_text(angle=0, size = 18, face="bold"),
          strip.text.x = element_text(size = 16),
          axis.text.x = element_text(size = 20, angle=45, hjust=1),
          axis.text.y = element_text(size = 16),
          axis.title.x = element_text(size = 22, face="bold"),
          legend.text=element_text(size=16),
          title=element_text(size=22, face="bold"),
          plot.title = element_text(hjust = 0.1),
          axis.title.y = element_blank())

ggsave(filename=paste0("RAnalysis/Output/RNA-seq/GOSeq/Host/final.BP.plot.host.png"), plot=GO.plot.BP, dpi=300, width=30, height=49, units="in")

```


# Make upregulated and downregulated plots
```{r}
#remove NA from enriched upregulated data frame
enriched.GO.05.up <- na.omit(enriched.GO.05.up)

#select for 30 lowest p values for each group 
BP.filt.u <- enriched.GO.05.up %>% filter(ontology == "BP")

CC.filt.u <- enriched.GO.05.up %>% filter(ontology == "CC")

MF.filt.u <- enriched.GO.05.up %>% filter(ontology == "MF")

#select only 30 lowest pvalues and rejoin 
BP.filt2.u <- BP.filt.u %>% slice_min(order_by = over_represented_pvalue, n = 30)

MF.filt2.u <- MF.filt.u %>% slice_min(order_by = over_represented_pvalue, n = 30)

#rejoin dataframes

enriched.up <- rbind(MF.filt2.u, BP.filt2.u, CC.filt.u)

# rename ontology terms
enriched.up <- enriched.up %>%
    mutate(ontology = recode(ontology, BP = 'Biological Process', MF = 'Molecular Function', CC =  'Cellular Component' ))

#create upreg plot
upreg.DEG.plot <- enriched.up  %>%  
  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>% 
  mutate(ontology = factor(ontology, levels = c("Biological Process", "Molecular Function", "Cellular Component"))) %>% ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom"
                   ) +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("Upregulated Differentially Expressed Genes") +
                   theme_bw() + #Set background color 
   facet_grid(ontology ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 55, multi_line = TRUE)) +
   theme(axis.text.y = element_text(hjust = 1))   +
  theme(plot.title = element_text(size = 25, face = "bold", hjust = 0.95)) +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes


 ggsave("RAnalysis/Output/Final_Figures/Host/upreg.DEG.plot.pdf", upreg.DEG.plot , width = 12, height = 20, units = c("in"))
 
 
##make downregulated plot
#remove NA from enriched downregulated data frame
enriched.GO.05.down <- na.omit(enriched.GO.05.down)

#select for 30 lowest p values for each group, first filter by ontology
BP.filt.d <- enriched.GO.05.down %>% filter(ontology == "BP")

CC.filt.d <- enriched.GO.05.down %>% filter(ontology == "CC")

MF.filt.d <- enriched.GO.05.down %>% filter(ontology == "MF")

#select only 30 lowest pvalues and rejoin 
BP.filt2.d <- BP.filt.d %>% slice_min(order_by = over_represented_pvalue, n = 30)

MF.filt2.d <- MF.filt.d %>% slice_min(order_by = over_represented_pvalue, n = 30)

#rejoin dataframes

enriched.down <- rbind(MF.filt2.d, BP.filt2.d, CC.filt.d)


# rename ontology terms
enriched.down <- enriched.down %>%
    mutate(ontology = recode(ontology, BP = 'Biological Process', MF = 'Molecular Function', CC =  'Cellular Component' ))


#make downreg plot
downreg.DEG.plot <- enriched.down  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>% 
  mutate(ontology = factor(ontology, levels = c("Biological Process", "Molecular Function", "Cellular Component"))) %>% ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom"
                   ) +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("Downregulated Differentially Expressed Genes") +
                   theme_bw() + #Set background color 
   facet_grid(ontology ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 55, multi_line = TRUE))+
   theme(axis.text.y = element_text(hjust = 1))   +
  theme(plot.title = element_text(size = 25, face = "bold", hjust = 0.95)) +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes


ggsave("RAnalysis/Output/Final_Figures/Host/downreg.DEG.plot.pdf", downreg.DEG.plot , width = 12, height = 20, units = c("in"))


#use patchwork to combine figures and add labels
Figure1 <- upreg.DEG.plot + downreg.DEG.plot + plot_annotation(tag_levels = 'A')
 
ggsave(filename=paste0("RAnalysis/Output/Final_Figures/Host/manuscript/down.upreg.DEG.png"), plot=Figure1, dpi=300, width=18, height=20, units="in")
ggsave(filename=paste0("RAnalysis/Output/Final_Figures/Host/manuscript/down.upreg.DEG.pdf"), plot=Figure1, dpi=300, width=18, height=20, units="in")

```


#######################################################################################

## KEGG enrichment analysis

## Obtained KEGG ontology terms by using KofamScan in command-line; see step 3 in E. Chille [lab notebook](https://github.com/echille/E.-Chille-Open-Lab-Notebook/blob/master/_posts/2020-10-08-M-capitata-functional-annotation-pipeline.md) for instructions. 
## If run into issues with downloading KofamScan software, etc.; see https://github.com/Putnam-Lab/Lab_Management/issues/21

# INFORMATION FOR KEGG ENRICHMENT ANALYSIS FOR R FOUND HERE:
# https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html 
#http://yulab-smu.top/clusterProfiler-book/chapter6.html#kegg-over-representation-test

# Load KO -- gene mapping info
```{r}

#Load in KO term functional annotation and edit to match GFF3 and annotation file
KFS.KO.orig <- read_tsv("../../../Desktop/Pver_KO_annot.tsv.zip", col_names = TRUE)

KFS.KO <- KFS.KO.orig[-1,c(1,2,3)]
colnames(KFS.KO) <- c("sig", "gene_id", "ko")
KFS.KO <- filter(KFS.KO, sig=="*")
KFS.KO <- KFS.KO[,c(2:3)]

#need to apply the same changes from Pver_annot file to the KFS.KO for gene_id
KFS.KO$gene_id <-  gsub(".t[1-67890900]","", KFS.KO$gene_id) #remove everything after g#
KFS.KO$gene_id <- gsub("Pver_split_", "Pver_", KFS.KO$gene_id)
KFS.KO$gene_id <- gsub("Pver_", "Pver_gene_", KFS.KO$gene_id)
KFS.KO$gene_id <-  gsub("_gene_g","_g", KFS.KO$gene_id) #sub gene_g with _g
KFS.KO$gene_id <-  gsub("_gene_","_", KFS.KO$gene_id) #remove extra _
KFS.KO$gene_id <-  gsub("\\..*","",KFS.KO$gene_id) #remove extra characters after the first .
head(KFS.KO)
tail(KFS.KO)

#remove eval and rename to KO.terms all
KO.terms <- KFS.KO
KO.terms <- unique(KO.terms)
str(KO.terms)
head(KO.terms)
colnames(KO.terms) <- c("gene_id", "GO.ID")

#Bind KO and GO references
GO.terms <- read_csv("RAnalysis/Output/RNA-seq/GOSeq/Host/annot_GO.terms.host.csv")[-1]

GOKO.terms <- bind_rows(GO.terms, KO.terms)

```


## Perform Kegg enrichment with goseq package
```{r}
#used str() to see that GOKO.terms was a tibble not a dataframe, convert to dataframe
GOKO.terms <- as.data.frame(GOKO.terms)

#Perform goseq
KOwall.up <- goseq(pwf.up, GFF3$gene_id, gene2cat=GOKO.terms, test.cats=c("KEGG"), method="Wallenius", use_genes_without_cat=TRUE)
KOwall.down <- goseq(pwf.down, GFF3$gene_id, gene2cat=GOKO.terms, test.cats=c("KEGG"), method="Wallenius", use_genes_without_cat=TRUE)
```

Replace NAs with "KEGG" and extract KO terms from results
```{r}
#Upregulated
up.KO.05<-KOwall.up$category[KOwall.up$over_represented_pvalue<.05]
up.KO.05<-data.frame(up.KO.05)
colnames(up.KO.05) <- c("category")
up.KO.05 <- merge(up.KO.05, KOwall.up, by="category")
up.KO.05$ontology <- replace_na(up.KO.05$ontology, "KEGG")
up.KO.05 <- filter(up.KO.05, ontology=="KEGG")
up.KO.05 <- up.KO.05[order(up.KO.05$ontology, up.KO.05$over_represented_pvalue, -up.KO.05$numDEInCat),]
up.KO.05$term <- as.factor(up.KO.05$term)
nrow(up.KO.05)

#Downregulated
down.KO.05<-KOwall.down$category[KOwall.down$over_represented_pvalue<.05]
down.KO.05<-data.frame(down.KO.05)
colnames(down.KO.05) <- c("category")
down.KO.05 <- merge(down.KO.05, KOwall.down, by="category")
down.KO.05$ontology <- replace_na(down.KO.05$ontology, "KEGG")
down.KO.05 <- filter(down.KO.05, ontology=="KEGG")
down.KO.05 <- down.KO.05[order(down.KO.05$ontology, down.KO.05$over_represented_pvalue, -down.KO.05$numDEInCat),]
down.KO.05$term <- as.factor(down.KO.05$term)
nrow(down.KO.05)
```

Add KO definitions
```{r}
#Prep definition data
KFS.KO.def <- subset(KFS.KO.orig, select=c("#","KO", "KO definition"))
colnames(KFS.KO.def) <- c("sig","category", "term")
KFS.KO.def <- filter(KFS.KO.def, sig=="*")
KFS.KO.def <- KFS.KO.def[,c(2:3)]
#Merge with KEGG output
up.KO.05 <- unique(left_join(up.KO.05[,-6], KFS.KO.def, by=c("category")))
down.KO.05 <- unique(left_join(down.KO.05[,-6], KFS.KO.def, by=c("category")))
```

Write output KEGG enrichment files
```{r}
write.csv(up.KO.05, file = "RAnalysis/Output/RNA-seq/KEGG_ontology/KO.05.up.csv")
write.csv(down.KO.05, file = "RAnalysis/Output/RNA-seq/KEGG_ontology/KO.05.down.csv")
```

## Make supplemental table summarizing KO enrichment
## Must do separately for up and down so we don't get duplicate gene entries. Add dir info
```{r}
#Add dir info
up.KO.05$dir <- "up"
down.KO.05$dir <- "down"
```

Make dataframe of GO results for clustering and heatmap. 
```{r}
#add gene_IDs. To get gene_IDs we will merge with the GO.terms DF.
GOKOgenes <- data.frame(gene_id=GOKO.terms$gene_id, category=GOKO.terms$GO.ID) #First have to make the "by" column the same for both
```

Using R script from Mass Lab. Needs columns "experiment" where we will put cluster information, "term", "percentDEInCat", and "gene" with all the DE genes associated with that GO term.
```{r}
#DO UP (up)
KOgenes_U <- filter(GOKOgenes)

KOdf_U <- left_join(up.KO.05, KOgenes_U, by="category" ) #join the DFs
KOdf_U$gene_id <- as.character(KOdf_U$gene_id) #make gene ID a character so we can collapse our many near-identical columns

```

```{r}
#DO DOWN (down)
KOgenes_D <- filter(GOKOgenes)
KOdf_D <- left_join(down.KO.05, KOgenes_D, by="category" ) #join the DFs
KOdf_D$gene_id <- as.character(KOdf_D$gene_id) #make gene ID a character so we can collapse our many near-identical columns

```

Bind rows for up and down and save!
```{r}
KOdf <- bind_rows(KOdf_U, KOdf_D)
KOdf <- na.omit(KOdf)
head(KOdf)
str(KOdf)
write.csv(KOdf, file = "RAnalysis/Output/RNA-seq/KEGG_ontology/KOenrichmentsummary.csv", row.names = FALSE)
```

```{r}
#KEGG plot
KEGGplot <- KOdf  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                   #mutate(term = fct_reorder(term, Day)) %>%
                   ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                   geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                   geom_point(size=3, shape = 15, colour = "grey70") +
                   coord_flip() +
                   theme(
                     panel.grid.minor.y = element_blank(),
                     panel.grid.major.y = element_blank(),
                     legend.position="bottom"
                   ) +
                   xlab("") +
                   ylab("-Log10 (pvalue)") +
                   ggtitle("KEGG terms") +
                   theme_bw() + #Set background color 
   theme(axis.text.y = element_text(hjust = 1))   +
                   theme(panel.border = element_blank(), # Set border
                         panel.grid.major = element_blank(), #Set major gridlines
                         panel.grid.minor = element_blank(), #Set minor gridlines
                         axis.line = element_line(colour = "black"), #Set axes color
                         plot.background=element_blank())  #Set the plot background #set title attributes

KEGGplot


ggsave("RAnalysis/Output/RNA-seq/KEGG_ontology/KEGGplot.png", KEGGplot, width = 10, height = 10, units = c("in"))
 
```





