---
title: "BS-SNPer.filter"
author: "daniellembecker"
date: "2024-03-11"
output: html_document
---

# Filtering SNPs found from BS-SNper, following this workflow: https://github.com/daniellembecker/DanielleBecker_Lab_Notebook/blob/master/_posts/2024-02-12-Testing-BS-SNPer-Molec-Underpinnings-WGBS.md

# Using .tab output file from BS-SNPer to filter SNPs found in 10x .bed files made following this workflow: https://github.com/daniellembecker/DanielleBecker_Lab_Notebook/blob/master/_posts/2022-12-10-P.verrucosa-WGBS-Workflow-Host.md and combined with this script: https://github.com/hputnam/Becker_E5/blob/master/RAnalysis/Scripts/WGBS/WGBS.Rmd

# Load libraries 

```{r}
library(dplyr)
```

# Load data

```{r}
# load merged .bed files at 10x coverage
# this file has sampleID and methylation information

bed_data <- readRDS("RAnalysis/Output/WGBS/sample_all_methylated_data_10x.Rdata") #read in info

# Read SNP positions from the .tab file
snp_data <- read.table("RAnalysis/Data/WGBS/BS-SNPer/CT-SNPs.tab", header = FALSE)

# Rename columns
colnames(snp_data) <- c("scaffold", "position", "placeholder", "ref.allele", "alter.allele", "quality.score", "filter.status", "add.info", "genotype", "related.info")

# Filter SNP positions from the .bed file
filtered_bed_data <- bed_data[bed_data$scaffold %in% snp_data$scaffold & bed_data$position %in% snp_data$position, ]

# Write the filtered data to a new .bed file
write.table(filtered_bed_data, "RAnalysis/Data/WGBS/BS-SNPer/filtered_bed_SNP_file.bed", sep = "\t", quote = FALSE, row.names = FALSE, col.names = TRUE)



```



