---
title: "BS-SNPer.filter"
author: "daniellembecker"
date: "2024-03-11"
output: html_document
---

# Filtering SNPs found from BS-SNper, following this workflow: https://github.com/daniellembecker/DanielleBecker_Lab_Notebook/blob/master/_posts/2024-02-12-Testing-BS-SNPer-Molec-Underpinnings-WGBS.md

# Using .tab output file from BS-SNPer to filter SNPs found in 10x .bed files made following this workflow: https://github.com/daniellembecker/DanielleBecker_Lab_Notebook/blob/master/_posts/2022-12-10-P.verrucosa-WGBS-Workflow-Host.md and combined with this script: https://github.com/hputnam/Becker_E5/blob/master/RAnalysis/Scripts/WGBS/WGBS.Rmd

# Load libraries 

```{r}
library(dplyr)
library(tidyverse)
```

# Load data

```{r}
# load merged .bed files at 10x coverage
# this file has sampleID and methylation information

bed_data <- readRDS("RAnalysis/Output/WGBS/sample_all_methylated_data_10x.Rdata") #read in info

# Read SNP positions from the .tab file
snp_data <- read.table("RAnalysis/Data/WGBS/BS-SNPer/CT-SNPs.tab", header = FALSE)

# Rename columns
colnames(snp_data) <- c("scaffold", "position", "placeholder", "ref.allele", "alter.allele", "quality.score", "filter.status", "add.info", "genotype", "related.info")

# Make new column with unique scaffold and position identifier for both data frames to merge and  filter SNPs from snp_data

bed_data$scaff.pos <- paste(bed_data$scaffold, bed_data$position, sep = "-")

snp_data$scaff.pos <- paste(snp_data$scaffold, snp_data$position, sep = "-")

# Using various approaches to see if there is data in either data frame that matches, which would elude to there being SNPs to filter in the bed_data frame

# Try merge function to merge any data found in scaff.pos colummn in another dataframe
merge_data <- merge(bed_data, snp_data, by = "scaff.pos")

# Try anti join in each data frame to see which data was not included or would not be filtered, each data frame is unique
anti_data <- snp_data %>%
  anti_join(bed_data, by = "scaff.pos")

test <- anti_join(bed_data, snp_data, by=c("scaffold", "position"))

# Check for shared values
shared_values <- intersect(snp_data$scaff.pos, bed_data$scaff.pos)

# Check for unique values in df1
unique_df1 <- setdiff(snp_data$scaff.pos, bed_data$scaff.pos)

# There seems to be no SNPs in my current bed file data frame........this may be due to the fact that the original .bam files used in BS-SNPer were not yet filtered, unlike our current .bed files that went through filtering steps in our bioinformatic workflow. This could mean any possible SNPs were filtered out.


```


# Try filtering approach with original .vcf output file from BS-SNPer
```{r}

# read in CT-SNP.vcf following Stevens approach: https://github.com/hputnam/Geoduck_Meth/blob/master/code/15-filtering-ct-snps.Rmd

ct <- read.csv("../../CT-SNP.vcf", header = FALSE, sep = "\t")

# make new column combining scaffold and position like above 
ct2 <- ct %>% 
  mutate(loci = paste0(V1, "_", V2))

# make new column combining scaffold and position like above 
bed2 <- bed_data %>%
  mutate(loci = paste0(scaffold, "_", position))

# join data frames based on matching values in loci column
inner.join.dat <- inner_join(bed2, ct2, by = "loci")

#512 SNPs that match my bed files dataframe

#  find rows in the first data frame that do not have a match in the second data frame based on loci column

anti.join.dat <- bed2 %>%
  anti_join(ct2, by = "loci")

# 1155895 unique data in my bed2 files, this matches that there are 512 SNPs in my bed data frame

# Use anti.join.dat as the filtered bed data file moving forward 

filtered.bed
```




