---
title: "WGBS Markdown"
author: "daniellembecker"
date: "3/17/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Molecular Underpinnings Chronic Nutrient Enrichment Project

## WGBS Gene Ontology (GO) Enrichment Analysis
## Previous steps for DNA methylation include WGBS workflow in Bioinformatics>WGBS>WGBS workflow 

## Load Libraries

```{r message = FALSE, warning = FALSE}
library(plotrix) 
library(ggplot2)
library(gridExtra)
library(seacarb) 
library(dplyr)
library(pheatmap)
library(tidyverse)
library(genefilter)
library(base)
library("DESeq2")
library(cowplot)
if ("lsmeans" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('lsmeans') 
library(lsmeans)
if ("ontologyIndex" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('ontologyIndex') 
library(ontologyIndex)
if ("ontologySimilarity" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('ontologySimilarity')
library(ontologySimilarity)
library(data.table)
library(RColorBrewer)
if ("colorRamps" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('colorRamps')
library(colorRamps)
if ("limma" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('limma')
library(limma)
library(GSEABase)
library(here)
library(lme4)
library(vegan)
library(goseq)
library(ggpubr)
library(patchwork)
library(rstatix)
library(stringr)
library(readr)
library(png)
library(grid)
library(cowplot)
```

#Downloaded all final .bed files for 5x and 10x from bluewaves to desktop: scp -r danielle_becker@bluewaves.uri.edu:/data/putnamlab/dbecks/Becker_E5/WGBS_Becker_E5/Becker_WGBS/CovtoCyto/*_enrichment.bed /Users/Danielle/Desktop/Putnam_Lab/Becker_E5/RAnalysis/Data/WGBS

###Samples E8 had low data coverage (see WGBS workflow in Bioinformatics>WGBS>WGBS workflow for further information), not moving forward with these samples in statistical steps

Methylation Analysis
## Loading genomic and annotation

```{r}
#load sample information
sample.info <- read.csv("RAnalysis/Data/WGBS/metadata.WGBS.csv", header=T, sep=",", na.string="NA", stringsAsFactors = F) #read in info
#samp <- sample.info$Sample.ID # set sample info
#samp <- gsub("[_]", "-", samp) #remove extra characters

#load genes gff 
Genes <- read.csv("RAnalysis/Genome/Pver_genome_assembly_v1.0.gene.gff3", head = TRUE, sep= " ") #read in data fill
Genes <- Genes[,c(9,1,4,5)] #select desired columns only
colnames(Genes) <- c("gene","scaffold", "start", "stop" ) #rename columns
Genes$gene <- gsub(";.*","",Genes$gene) #remove extra characters
Genes$gene <- gsub("ID=","",Genes$gene) #remove extra characters
Genes$length <- Genes$stop-Genes$start # calculated lengths from start and stop

#Load annotation file
Annot <- read.csv("Functional_Annotation/Final_Annotations/pver_annot_full.csv", row.names = 1)
colnames(Annot)[1] <- "gene"

#edit gene ID to match GFF3
Annot$gene <-  gsub("\\..*","",Annot$gene) #remove the all after . "t#" number after the dash
Annot$gene <-  gsub("_","_gene_",Annot$gene) #add _gene_g#


#load all gene GO IDs organized and split

Gene.GO.IDs <- read.csv("Functional_Annotation/Final_Annotations/pver_GOterms_interprot_swissprot_blast_trembl_20211203.csv", row.names = 1)
colnames(Gene.GO.IDs) <- c("gene", "GO.IDs", "database") #rename columns

#edit gene ID to match GFF3
Gene.GO.IDs$gene <-  gsub("\\..*","",Gene.GO.IDs$gene) #remove the all after . "t#" number after the dash
Gene.GO.IDs$gene <-  gsub("_","_gene_",Gene.GO.IDs$gene) #add _gene_g#
Gene.GO.IDs$GO.IDs <- as.character(Gene.GO.IDs$GO.IDs) #make IDs into characters
GO.IDs <-unique(Gene.GO.IDs$GO.IDs) #removes duplicates from columns, compiles both lists to get all GO terms without duplicates
```

#Load filtered methylation counts 10x 
```{r}
#Nyssa-stringr package, nice and clean so you can pipe and not have to rename, etc., use expression that does it all in one line regex 
#Merge final .bed files at 10x coverage
meth.data.10x <- list.files(path = "RAnalysis/Data/WGBS/10x_methcounts/", pattern = ".bed$", full.names=TRUE) %>%
  purrr::set_names(.) %>% 
  map_dfr(read.csv,.id="Sample.ID", header=FALSE, sep="\t", na.string="NA", stringsAsFactors = FALSE) %>% 
  dplyr::select(-c(V3,V7:V14)) %>%
  group_by(Sample.ID)
colnames(meth.data.10x) <- c("Sample.ID", "scaffold", "position","per.meth","meth","unmeth","gene")
meth.data.10x$gene <- gsub(";.*","",meth.data.10x$gene) #remove extra characters
meth.data.10x$gene <- gsub("ID=","",meth.data.10x$gene) #remove extra characters
meth.data.10x$Sample.ID <- gsub("RAnalysis/Data/WGBS/10x_methcounts//","",meth.data.10x$Sample.ID) #remove extra characters
meth.data.10x$Sample.ID <- gsub("_.*","",meth.data.10x$Sample.ID) #remove extra characters 
#meth.data.10x$Sample.ID <- gsub("-","_",meth.data.10x$Sample.ID) #remove extra characters
#meth.data.10x$Sample.ID <- gsub("/","",meth.data.10x$Sample.ID) #remove extra characters (backslash from sample.ID name)
MD.10x <- merge(meth.data.10x,sample.info, by="Sample.ID") #combine sample info and 10x coverage megafile

#save sample info and percent methylation
saveRDS(MD.10x, "RAnalysis/Output/WGBS/sample_all_methylated_data_10x.Rdata")

MD.10x <- readRDS("RAnalysis/Output/WGBS/sample_all_methylated_data_10x.Rdata") #read in info

head(MD.10x)
```


# Testing for Differentially Methylated Genes at 10X coverage
```{r}
# Comparison of enriched vs control nutrient treatments
# At 10x coverage
# Binomial GLM to test for differentially methylated genes
sub_meth_table.10x  <- MD.10x
sub_meth_table.10x$group <- paste0(sub_meth_table.10x$Sample.ID, sub_meth_table.10x$gene)

#filter for genes with >5 methylated positions
min.filt.10x <- dplyr::count(sub_meth_table.10x, vars = c( group))
newdata.10x <- min.filt.10x[ which(min.filt.10x$n > 4), ]
sub_meth_table.10x <- sub_meth_table.10x[sub_meth_table.10x$group %in% newdata.10x$vars,]

# create data frame to stored results
results.10x <- data.frame()
gs.10x <- unique(sub_meth_table.10x$gene)


#all code below is commented out because only need to run this for the initial step
# #first subset the unique dataframes and second run the GLMs
# #purr package function called map, much cleaner for for loops (how to loop over a model in the tidyverse, can have a tidy output for your anova) computational time is faster with purr and it is cleaner
# for(i in 1:length(sub_meth_table.10x$gene)){
# 
#   #subset the dataframe gene by gene
#   sub_meth_table.10x1 <- subset(sub_meth_table.10x, gene ==gs.10x[i])
# 
#   # fit glm position model
#   #think about this as just a logistic regression, linear model on binomial distribution is binomial distribution, just adding a random effect for block. P-values based off of maximum likelihood approach, not gonna be like an anova table.
#   fit <- glm(matrix(c(meth, unmeth), ncol=2) ~ treatment,
#              data=sub_meth_table.10x1, family=binomial)
#   a <- anova(fit, test="Chisq")
# 
#   # capture summary stats to data frame
#   df <- data.frame(gene = sub_meth_table.10x1[1,7],
#                    pval.treatment = a$`Pr(>Chi)`[2],
#                    #pval.position = a$`Pr(>Chi)`[3], #uncomment if you want to include position of CpG within a gene
#                    #pval.treatment_x_position = a$`Pr(>Chi)`[4], #uncomment if you want to include position of CpG within a gene interaction with treatment
#                    stringsAsFactors = F)
# 
#   # bind rows of temporary data frame to the results data frame
#   results.10x <- rbind(results.10x, df)
# 
# }
# 
# write.csv(results.10x, "RAnalysis/Output/WGBS/meth.results.10x.csv")

 # An error will be generated here for contrasts. 
#This potential for contrasts (interactions) is included in the case one wants to examine the role of position of CpG within a gene
#Error in `contrasts<-`(`*tmp*`, value = contr.funs[1 + isOF[nn]]) : contrasts can be applied only to factors with 2 or more levels
#Continuing the analysis from results line will generate the results in the absence of the contrast (interaction).

results.10x <- read.csv("RAnalysis/Output/WGBS/meth.results.10x.csv", row.names = 1)


results.10x[is.na(results.10x)] <- 0
results.10x$adj.pval.treatment <- p.adjust(results.10x$pval.treatment, method='BH') #BH is a less conservative Benjamini & Hochberg (1995) correction in which p-values are multiplied by the number of comparisons

#results$adj.pval.position <- p.adjust(results$pval.position, method='BH') #uncomment if you want to include position of CpG within a gene
#result $adj.pval.treatment_x_position <- p.adjust(results$pval.treatment_x_position, method='BH') #uncomment if you want to include position of CpG within a gene interaction with treatment

# Identifying DMG with significant main effect 
DMG.sig.10x <-results.10x
DMG.sig.10x <- DMG.sig.10x[order(DMG.sig.10x$adj.pval.treatment),]
DMG.sig.10x <- DMG.sig.10x[which(DMG.sig.10x$adj.pval.treatment<0.05), ]
sum(DMG.sig.10x$adj.pval.treatment < 0.05, na.rm=TRUE) #How many adjusted p-values were less than 0.05?

#368

# Annotation of DMG under between treatments 
DMG.sig.annot.10x <- left_join(DMG.sig.10x , Annot, by="gene")
#DMG.sig.annot <- DMG.sig.annot[!duplicated(DMG.sig.annot$gene),]
write.table(DMG.sig.annot.10x, 'RAnalysis/Output/WGBS/Treatment_sig_annot.10x.tsv', sep='\t', row.names=FALSE)


```

#plot global, DMG, and DMG per DEG for bar plots between enriched and control
```{r}
#Sanity check plotting of DMGs by treatment
subdf.DMG <- MD.10x  %>%	
  filter(gene == DMG.sig.annot.10x$gene[2])
means.DMG <- aggregate(per.meth ~ treatment, data=subdf.DMG, FUN=mean)
ses.DMG <- aggregate(per.meth ~ treatment, data=subdf.DMG, FUN=std.error)
means.DMG$ses <- ses.DMG$per.meth
means.DMG$per.meth <- means.DMG$per.meth

# build the linear model to see if tretaments differ and check assumptions
model  <- lm(per.meth ~ treatment, data = subdf.DMG)

# Create a QQ plot of residuals
ggqqplot(residuals(model))
#not normal, use kruskall willis test

# use Kruskal-Wallis test by rank which is a non-parametric alternative to one-way ANOVA test, which extends the two-samples Wilcoxon test in the situation where there are more than two groups. It’s recommended when the assumptions of one-way ANOVA test are not met.

DMG.mod <- kruskal.test(per.meth ~ treatment, data = subdf.DMG)

DMG.mod

#results: Kruskal-Wallis chi-squared = 118.91, df = 1, p-value < 2.2e-16, significantly higher percent meth in DMGs in enriched treatment

#make bar plot of means DMG
plot.DMG <- ggplot(means.DMG, aes(x=treatment, y=per.meth, fill = treatment)) +
  geom_bar(stat="identity",
             position=position_dodge(), show.legend = FALSE) +
    geom_errorbar(aes(ymin=per.meth-ses, ymax=per.meth+ses), width=.2,
                  position=position_dodge(.9)) +
    xlab("Treatment") + #plot x axis label
    ylab("Percent Methylation in DMG") + #plot y axis label
    scale_fill_manual(values = c(control="black", enriched="cadetblue3")) +
    scale_y_continuous(limits = c(0,12.5), expand = c(0, 0)) +
    ggtitle("Kruskal-Wallis, p < 2.2e-16") +
    theme_bw()

# #try the boxplot option here
# plot.DMG <- ggboxplot(subdf.DMG, x = "treatment", y = "per.meth",
#           color = "treatment")+
#   scale_color_manual(values = c(control="black", enriched="cadetblue3")) +
#   stat_compare_means(method = "kruskal.test") +
#   xlab("Treatment") + #plot x axis label
#     ylab("Percent Methylation in DMG") + #plot y axis label
#   scale_y_continuous(limits = c(0,70), expand = c(0, 0)) +
#   theme_bw()

ggsave("RAnalysis/Output/WGBS/per.meth.10x.DMG.pdf", plot.DMG)
ggsave("RAnalysis/Output/WGBS/per.meth.10x.DMG.png", plot.DMG)



#Sanity check plotting of DMGs by treatment
subdf.global <- MD.10x # %>%	
  #filter(gene== DMG.sig.annot.10x$gene[2])
means.global <- aggregate(per.meth ~ treatment, data=subdf.global, FUN=mean)
ses.global <- aggregate(per.meth ~ treatment, data=subdf.global, FUN=std.error)
means.global$ses <- ses.global$per.meth
means.global$per.meth <- means.global$per.meth

plot.global <- ggplot(means.global, aes(x=treatment, y=per.meth, fill = treatment)) +
  geom_bar(stat="identity",
             position=position_dodge(), show.legend = FALSE) +
    geom_errorbar(aes(ymin=per.meth-ses, ymax=per.meth+ses), width=.2,
                  position=position_dodge(.9)) +
    xlab("Treatment") + #plot x axis label
    ylab("Percent Methylation in Global") + #plot y axis label
    scale_fill_manual(values = c(control="black", enriched="cadetblue3")) +
    scale_y_continuous(expand = expansion(mult = c(0, .1)))+
    theme_bw()

ggsave("RAnalysis/Output/WGBS/per.meth.10x.global.pdf", plot.global)
ggsave("RAnalysis/Output/WGBS/per.meth.10x.global.png", plot.global)

#compare dmg and deg, create plot compared %dmgs in DEGs, load deg data
deg <- read.csv("RAnalysis/Output/RNA-seq/DEG/Host/DEGSeq2.sig.results.host.csv")

#make gene list for gene_id in deg dataframe
gene_list <-c(deg$gene_id)

#remove _gene_ in gene name from MD.10x
#MD.10x$gene <- gsub("_gene","", MD.10x$gene) #remove extra characters

#looking at percent methylation in DEG
subdf.dmg.deg <- MD.10x %>%	
  filter(gene %in% gene_list)
means.dmg.deg <- aggregate(per.meth ~ treatment, data=subdf.dmg.deg, FUN=mean)
ses.dmg.deg <- aggregate(per.meth ~ treatment, data=subdf.dmg.deg, FUN=std.error)
means.dmg.deg$ses <- ses.dmg.deg$per.meth
means.dmg.deg$per.meth <- means.dmg.deg$per.meth

plot.dmg.deg <- ggplot(means.dmg.deg, aes(x=treatment, y=per.meth, fill = treatment)) +
  geom_bar(stat="identity",
             position=position_dodge(), show.legend = FALSE) +
    geom_errorbar(aes(ymin=per.meth-ses, ymax=per.meth+ses), width=.2,
                  position=position_dodge(.9)) +
    xlab("Treatment") + #plot x axis label
    ylab("Percent Methylation in DEG") + #plot y axis label
    scale_fill_manual(values = c(control="black", enriched="cadetblue3")) +
    scale_y_continuous(expand = expansion(mult = c(0, .1)))+
    ggtitle("Kruskal-Wallis, p = 0.04") +
    theme_bw()

ggsave("RAnalysis/Output/WGBS/per.meth.in.DEG.pdf", plot.dmg.deg)
ggsave("RAnalysis/Output/WGBS/per.meth.in.DEG.png", plot.dmg.deg)


# build the linear model to see if tretaments differ and check assumptions
model  <- lm(per.meth ~ treatment, data = subdf.dmg.deg)

# Create a QQ plot of residuals
ggqqplot(residuals(model))
#not normal, use kruskall willis test

# use Kruskal-Wallis test by rank which is a non-parametric alternative to one-way ANOVA test, which extends the two-samples Wilcoxon test in the situation where there are more than two groups. It’s recommended when the assumptions of one-way ANOVA test are not met.

dmg.deg.mod <- kruskal.test(per.meth ~ treatment, data = subdf.dmg.deg)

dmg.deg.mod


```

#Calculate hyper- and hypo- methylation from data using limma package in significant DMGs

```{r}
#set up dataframe for limma analysis
# Assume you have a data frame 'methylation_data' with columns: Gene, Condition1, Condition2
# 'Gene' column contains gene names, 'Condition1' and 'Condition2' contain methylation values

#select certain columns
methylation_data <- MD.10x %>%
  filter(gene %in% DMG.sig.annot.10x$gene)

# Data wrangling to calculate the mean percent methylation for each gene
aggregated_data <- methylation_data %>%
  group_by(gene, treatment) %>%
  summarize(percent.meth = mean(per.meth, na.rm = TRUE))

# Specify the name of the control treatment group
control_group <- "control"

# Filter data for the control group
control_data <- aggregated_data %>%
  filter(treatment == control_group) %>%
  select(gene, ControlMean = percent.meth)

# Merge with the original data to get the mean percent methylation for each gene and treatment
merged_data <- merge(aggregated_data, control_data, by = "gene", all.x = TRUE)

# Calculate the mean difference for each gene and treatment compared to the control
merged_data$MeanDifference <- merged_data$percent.meth - merged_data$ControlMean

# Identify hyper and hypomethylated genes
# Adjust the threshold based on your specific criteria
hyper_genes <- merged_data$gene[merged_data$MeanDifference > 0]
hypo_genes <- merged_data$gene[merged_data$MeanDifference < 0]

# Count the number of hyper and hypomethylated genes
num_hyper_genes <- length(hyper_genes) #203
num_hypo_genes <- length(hypo_genes) #165

```



# Conduct PERMANOVA and PCA on global methylation and DMG
#### Principal component plot of samples global DMG
```{r}
#First need to make a matrix with percent meth count data for each sample and gene ID
#make meth count matrix, select gene, frag ID, and per meth from dataframe
matrix <- MD.10x %>%
  dplyr::select(gene, fragment.ID, per.meth)

#change from long to wide, set fun.aggregate for mean % meth per gene as it includes each position for each sample in the calculation
countdata <- reshape2::dcast(matrix, gene ~ fragment.ID, value.var = "per.meth", fun.aggregate = mean)

#make first row removed and rownames
countData <- countdata[,-1]
rownames(countData) <- countdata[,1]

#make all column values integers 
countData <- countData %>% mutate_if(is.numeric, as.integer)

#filter out sample 2, 19 and 16 from sample_info data frame becuase they are not being used due to deduplication errors
samp.info <- sample.info[!row.names(sample.info) %in% c(8,11,12),]

##Merge the treatment columns into a new column , group. Set group as a factor.
samp.info$treatment <- factor(samp.info$treatment, levels = c("control","enriched"))

#Set DESeq2 design
gdds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = samp.info,
                              design = ~treatment)

#estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than four to use vst
SF.gdds <- estimateSizeFactors(gdds) 
print(sizeFactors(SF.gdds))

#Our size factors are all less than 4, so we can use VST!
#apply a variance stabilizing transformation to minimize effects of small counts and normalize library size
gvst <- varianceStabilizingTransformation(gdds, blind=FALSE) 

gPCAdata <- plotPCA(gvst, intgroup = c("treatment"), returnData=TRUE)
percentVar <- round(100*attr(gPCAdata, "percentVar")) #plot PCA of samples with all data
ggplot(gPCAdata, aes(PC1, PC2, color=treatment)) + 
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  scale_color_manual(values = c(control="black", enriched="grey")) +
  coord_fixed() +
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
        #panel.grid.major = element_blank(), #Set major gridlines
        #panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) + #Set the plot background +
  stat_ellipse()

ggsave(path = "RAnalysis/Output/Final_Figures/Host/", filename = "meth_general_PCA_plot_host.pdf")
ggsave(path = "RAnalysis/Output/Final_Figures/Host/", filename = "meth_general_PCA_plot_host.png")

```


### Make Volcano plot of hyper and hypo methylated DMGs
```{r}
# Run DESeq2 analysis
dds <- DESeq(gdds)

# Extract log2fold change values
results <- results(dds)
log2fold_change <- results$log2FoldChange
gene_names <- rownames(results)
log2fold_change_df <- data.frame(Gene = gene_names, Log2FoldChange = log2fold_change)

#rename gene column
log2fold_change_df <- log2fold_change_df %>%
  dplyr::rename(gene = Gene)

# Merge with the methylation data frame
final_df <- left_join(results.10x, log2fold_change_df, by = "gene")

#rename gene column
final_df <- final_df %>%
 dplyr::rename(log2FoldChange = Log2FoldChange)

final_df <- final_df %>%
  dplyr::rename(padj = adj.pval.treatment)

final_df <- final_df %>%
  dplyr::rename(pvalue = pval.treatment)

# Create a new column for -log10(padj), noticed outliers in data
final_df $log_padj <- -log10(final_df $padj)

# Remove rows where log_padj is greater than 10
final_df <- final_df[final_df$log_padj <= 100, ]

# Define criteria for hypermethylation, hypomethylation

final_df$methylation_status <- ifelse(final_df$padj > 0.05, "Not Significant",
                                       ifelse(final_df$log2FoldChange > 0, "Hypermethylated",
                                              ifelse(final_df$log2FoldChange < 0, "Hypomethylated", "Not Classified")))

# Set up the color palette for hyper, hypo, and non-significant genes
colors <- c("Hypermethylated" = "red", "Hypomethylated" = "blue", "Not Significant" = "black")

# Create the volcano plot using ggplot2
DMG.volcano <- ggplot(final_df, aes(x = log2FoldChange, y = -log10(padj), color = methylation_status)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(values = colors, name = "Methylation Status", 
                     breaks = c("Hypermethylated", "Hypomethylated", "Not Significant"),
                     labels = c("Hypermethylated (203)", "Hypomethylated (165)", "Not Significant"),
                     drop = FALSE) +  # Include all specified breaks in the legend
  labs(title = "", x = "log2FoldChange", y = "-log10(adjusted p-value)") +
  theme_bw() +
  geom_hline(yintercept = 0, linetype = "dashed", color = "black"); DMG.volcano

ggsave("RAnalysis/Output/WGBS/DMG.volcano.pdf", DMG.volcano, width = 8, height = 6, units = c("in"))
  
#load png image of DMG PCA

DMG.PCA <- readPNG("RAnalysis/Output/Final_Figures/Host/DMG_PCA_plot_host.png")

PCA.DMG <- rasterGrob(DMG.PCA, interpolate=TRUE)

Figure2 <- plot_grid( DMG.volcano, DMG_PCA_plot, labels = c('A', 'B'), label_size = 14)

ggsave(filename=paste0("RAnalysis/Output/Final_Figures/Host/manuscript/DMG.PCA.volcano.png"), plot=Figure2, dpi=300, width=12, height=5, units="in", limitsize = FALSE)

ggsave(filename=paste0("RAnalysis/Output/Final_Figures/Host/manuscript/DMG.PCA.volcano.pdf"), plot=Figure2, dpi=300, width=12, height=5, units="in", limitsize = FALSE)

```



# Conduct PERMANOVA and PCA on differential expression 

Export data for PERMANOVA test.  
```{r}
test<-t(assay(gvst)) #export as matrix
test<-as.data.frame(test)

#add category columns
test$fragment.ID <- rownames(test)
test$treatment <- samp.info$treatment[match(test$fragment.ID, samp.info$fragment.ID)]
```

Build PERMANOVA model.  
```{r}
#last column with treatment metadata test[3912] and test[3911], need to start at third to last column for the numeric error issue
#also getting Error in prcomp.default(test1[c(1:3910)], center = TRUE, scale. = TRUE) : cannot rescale a constant/zero column to unit variance, so had to remove constant data values
test1 <- test %>% purrr::keep(~length(unique(.x)) != 1)

scaled_test <-prcomp(test1[c(1:3899)], center = TRUE, scale. = TRUE)
fviz_eig(scaled_test)

# scale data
vegan <- scale(test1[c(1:3899)])

# PerMANOVA 
permanova <- adonis2(vegan ~ treatment, data = test1, method='eu')
permanova
```

There are significant differences between treatments for control and enriched sites in global methylation

adonis2(formula = vegan ~ treatment, data = test1, method = "eu")
          Df SumOfSqs      R2      F Pr(>F)   
treatment  1     4191 0.03839 1.0778  0.003 **
Residual  27   104981 0.96161                 
Total     28   109172 1.00000                 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

## Visualize differentially methylated genes
```{r}
# Subset list of genes by those which padj>0.
DMG.sig.result <- as.data.frame(DMG.sig.10x)

#make first row removed and add gene as rownames
DMG.sig.results <- DMG.sig.result[,-1]
rownames(DMG.sig.results) <- DMG.sig.result[,1]

DMG.output <- as.data.frame(DMG.sig.results)
dim(DMG.output)

DMG.results.all <- (DMG.sig.results)
DMG <- rownames(DMG.results.all) # list all gene names 
DMG <- unique(DMG) # select only unique gene names 
DMG_list <- gdds[which(rownames(gdds) %in% DMG)] # filter gdds DMG object by unique gene names
dim(DMG_list) # 15 x 15
print(counts(DMG_list))


# As determined above, size factors all less than 4, so proceed with VST
# apply a variance stabilizing transformation to minimize effects of small counts and normalize library size
DMGvst <- varianceStabilizingTransformation(DMG_list, blind=FALSE) 

dim(DMGvst) # 15 x 15
print(assay(DMGvst)) # look at vst-transformed gene count data 

# Plot heat map with diff expressed genes
# Make a matrix for computing similarity
mat <- assay(DMGvst)#[DEG.results.all, ] # make an expression object
mat <- mat - rowMeans(mat) #difference in expression compared to average across all samples
dim(mat)
ann_colors <- list(treatment= c(control="black", enriched="grey"))
df_DMG <- as.data.frame(colData(DMGvst)[c("treatment")]) #make dataframe for column naming and associated treatment
DMG_heatmap <- pheatmap(mat, scale= "row", legend=TRUE, annotation_legend=TRUE, annotation_col=df_DMG, annotation_colors = ann_colors,
                            clustering_distance_rows="euclidean", clustering_method = "average",
                            show_rownames =FALSE,
                            show_colnames =TRUE,
                            cluster_cols = TRUE)

pdf(file = "RAnalysis/Output/Final_Figures/Host/DMG_heatmap_host.pdf")
DMG_heatmap
dev.off()

png(file = "RAnalysis/Output/Final_Figures/Host/DMG_heatmap_host.png")
DMG_heatmap
dev.off()

#make label size for plot below
label_size <- 16

# PCA plot of diff-expressed genes 
DMG_PCAdata <- plotPCA(DMGvst, intgroup = c("treatment"), returnData=TRUE)
percentVar_pca <- round(100*attr(DMG_PCAdata, "percentVar")) #plot PCA of samples with all data
DMG_PCA_plot <- ggplot(DMG_PCAdata, aes(PC1, PC2, color=treatment)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar_pca[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_pca[2],"% variance")) +
  scale_color_manual(values = c(control="black", enriched="grey")) +
  coord_fixed() +
  theme_bw() + #Set background color
  theme(
    panel.border = element_blank(),
    axis.line = element_line(colour = "black"),
    plot.background = element_blank(),
    # Increase the size of axis text
    axis.text = element_text(size = label_size),
    # Increase the size of x and y axis titles
    axis.title = element_text(size = label_size),
    # Increase the size of legend text
    legend.text = element_text(size = label_size),
    # Increase the size of legend title
    legend.title = element_text(size = label_size)
  ) +#Set the plot background +
  stat_ellipse()

DMG_PCA_plot

ggsave("RAnalysis/Output/Final_Figures/Host/DMG_PCA_plot_plot_host.png", DMG_PCA_plot, width = 10, height = 10, units = c("in"))
ggsave("RAnalysis/Output/Final_Figures/Host/DMG_PCA_plot_host.pdf", DMG_PCA_plot, width = 10, height = 10, units = c("in"))

```

# Conduct PERMANOVA and DMG

Export data for PERMANOVA test.  
```{r}
test2<-t(assay(DMGvst)) #export as matrix
test2<-as.data.frame(test2)

#add category columns
test2$fragment.ID <- rownames(test2)
test2$treatment <- samp.info$treatment[match(test2$fragment.ID, samp.info$fragment.ID)]
```

Build PERMANOVA model.  
```{r}
#last column with treatment metadata test[370] and test[369], need to start at third to last column for the numeric error issue
scaled_test2 <-prcomp(test2[c(1:368)], center = TRUE, scale. = TRUE)
fviz_eig(scaled_test2)

# scale data
vegan <- scale(test2[c(1:368)])

# PerMANOVA 
permanova <- adonis2(vegan ~ treatment, data = test2, method='eu')
permanova
```

There are significant differences between treatments for control and enriched sites in DMG

adonis2(formula = vegan ~ treatment, data = test2, method = "eu")
          Df SumOfSqs      R2     F Pr(>F)    
treatment  1    711.3 0.06903 2.002  0.001 ***
Residual  27   9592.7 0.93097                 
Total     28  10304.0 1.00000                 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## 10x GO enrichment
```{r}
# GO Enrichment Analysis of Treatment for 10x data
##### GO enrichment of DMGs #####

DMG.trt <- as.character(DMG.sig.10x$gene) #set the enrichment test list

#change to vectors
ALL.vector <-c(t(Genes$gene))
DMG.vector <-c(t(DMG.trt))
ID.vector <- Genes$gene
LENGTH.vector <-Genes$length

#DMG.trt.vector <-c(t(DMG.trt)) #change to vectors
gene.vector=as.integer(ALL.vector%in%DMG.vector) #Construct new vector with 1 for DEG and 0 for others
names(gene.vector)=ALL.vector #set names
DMG.pwf<-nullp(gene.vector, ID.vector, bias.data=LENGTH.vector) #weight vector by length of gene

#Find enriched GO terms, 
GO.wall<-goseq(DMG.pwf, ID.vector, gene2cat=Gene.GO.IDs, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)

GO.sig.trt <- GO.wall %>%
  filter(over_represented_pvalue <0.05) %>%
  arrange(., ontology)
write.csv(GO.sig.trt , file = "RAnalysis/Output/WGBS/GO.sig.trt.10x.csv")

#GO.sig.trt$term <- as.factor(GO.sig.trt$term)
GO.sig.trt[is.na(GO.sig.trt)] <- "UN"

nrow(GO.sig.trt) #118
nrow(filter(GO.sig.trt, ontology=="BP")) #number sig BP terms (75)
nrow(filter(GO.sig.trt, ontology=="MF")) #number sig MF terms (37)
nrow(filter(GO.sig.trt, ontology=="CC")) #number sig CC terms (6)

#check any missing GO terms manually and add as needed
which(GO.sig.trt$ontology=="UN")

GO.sig.trt[117,6] <- "innate immune response-activating signal transduction"
GO.sig.trt[117,7] <- "BP"

GO.sig.trt[118,6] <- "cell-cell adherens junction"
GO.sig.trt[118,7] <- "CC"

```

## Find GOslim terms
## This is where you put GO terms to larger umbrella terms

# Run GOslim to get broader categories
```{r}
#load in generic GO database, has all of the upperlevel categories for GO terms. Ex: regulation of cell division would be under cellular regulation
slim <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") #get GO database

## filtering all of BP (do MF and CC seperately)
BP_GO <- GO.sig.trt %>%
  filter(ontology=="BP")
BPGO_collection <- GOCollection(BP_GO$category) #Make library of query terms
slims_bp <- data.frame(goSlim(BPGO_collection, slim, "BP")) #Find common parent terms to slim down our list
slims_bp$category <- row.names(slims_bp) #save rownames as category

## filtering all of MF
MF_GO <- GO.sig.trt %>%
  filter(ontology=="MF")
MFGO_collection <- GOCollection(MF_GO$category) #Make library of query terms
slims_mf <- data.frame(goSlim(MFGO_collection, slim, "MF")) #Find common parent terms to slim down our list
slims_mf$category <- row.names(slims_mf) #save rownames as category

## filtering all of CC
CC_GO <- GO.sig.trt %>%
  filter(ontology=="CC")
CCGO_collection <- GOCollection(CC_GO$category) #Make library of query terms
slims_cc <- data.frame(goSlim(CCGO_collection, slim, "CC")) #Find common parent terms to slim down our list
slims_cc$category <- row.names(slims_cc) #save rownames as category
```

# Get mapped terms, using functions from Sam White's Biostars [post](https://support.bioconductor.org/p/128407/#128409).
```{r}
#custom function from Sam White's, gets mapped ids for all of your query terms 
#Write function mappedIds to get the query terms that mapped to the slim categories
mappedIds <-
  function(df, collection, OFFSPRING) #the command to run requires a dataframe of slim terms, like slims_MF above, your list of query terms, and the offspring from the GOCollection by goSlim
  {
    map <- as.list(OFFSPRING[rownames(df)]) # Subset GOcollection offspring by the rownames of your dataframe
    mapped <- lapply(map, intersect, ids(collection)) #Find the terms that intersect between the subset made above of your query terms and the GOids from the GO collection
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L)) #Add column "go_terms" with matching terms 
    df #show resulting dataframe
  }
#Run function for MF and BP terms
BPslim <- mappedIds(slims_bp, BPGO_collection, GOBPOFFSPRING)
MFslim <- mappedIds(slims_mf, MFGO_collection, GOMFOFFSPRING)
CCslim <- mappedIds(slims_cc, CCGO_collection, GOCCOFFSPRING)
```

# Remove duplicate matches, keeping the broader umbrella term
```{r}
#filtering out duplicates, keep the largest branch that has the most GO terms

#BP
BPslim <- filter(BPslim, Count>0 & Term!="biological_process") #filter out empty slims and term "biological process"
BPsplitted <- strsplit(as.character(BPslim$go_terms), ";") #split into multiple GO ids
BPslimX <- data.frame(Term = rep.int(BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) #list all
BPslimX <- merge(BPslimX, BPslim[,c(1,3:4)], by="Term") #Add back counts, term, and category info
BPslimX <- unique(setDT(BPslimX)[order(go_term, -Count)], by = "go_term") #remove duplicate offspring terms, keeping only those that appear in the larger umbrella term (larger Count number)
BPslim <- data.frame(slim_term=BPslimX$Term, slim_cat=BPslimX$category, category=BPslimX$go_term) #rename columns
head(BPslim)

#MF
MFslim <- filter(MFslim, Count>0 & Term!="molecular_function") #filter out empty slims and term "molecular function"
MFsplitted <- strsplit(as.character(MFslim$go_terms), ";") #split into multiple GO ids
MFslimX <- data.frame(Term = rep.int(MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) #list all
MFslimX <- merge(MFslimX, MFslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
MFslimX <- unique(setDT(MFslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
MFslim <- data.frame(slim_term=MFslimX$Term, slim_cat=MFslimX$category, category=MFslimX$go_term) #rename columns
head(MFslim)

#CC
CCslim <- filter(CCslim, Count>0 & Term!="cellular_component") #filter out empty slims and term "molecular function"
CCsplitted <- strsplit(as.character(CCslim$go_terms), ";") #split into multiple GO ids
CCslimX <- data.frame(Term = rep.int(CCslim$Term, sapply(CCsplitted, length)), go_term = unlist(CCsplitted)) #list all
CCslimX <- merge(CCslimX, CCslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
CCslimX <- unique(setDT(CCslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
CCslim <- data.frame(slim_term=CCslimX$Term, slim_cat=CCslimX$category, category=CCslimX$go_term) #rename columns
head(CCslim)

```

# Save slim info with GO enrichment info for heatmap dataframes.
```{r}
GO.BP <- right_join(BPslim, filter(GO.sig.trt, ontology=="BP"), by="category") #add back GO enrichment info for each offspring term
GO.MF <- right_join(MFslim, filter(GO.sig.trt, ontology=="MF"), by="category") #add back GO enrichment info for each offspring term
GO.CC <- right_join(CCslim, filter(GO.sig.trt, ontology=="CC"), by="category") #add back GO enrichment info for each offspring term
```


## Make heatmap
```{r}
BPplot <- GO.BP %>% mutate(term = fct_reorder(term, -over_represented_pvalue)) %>%
    ggplot(aes(x = ontology, y = term)) + 
    geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
      scale_y_discrete(position = "right") +
      facet_grid(slim_term~ ., scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
      theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
      strip.text.y.left = element_text(angle=0, size = 11, face = "bold"),
      strip.text.x = element_text(size = 12, face = "bold"),
      axis.title = element_blank(),
      axis.text = element_text(size = 12), legend.title = element_text(size = 12), legend.text =     
      element_text(size = 11))
MFplot <- GO.MF %>% mutate(term = fct_reorder(term, -over_represented_pvalue)) %>% 
    ggplot(aes(x = ontology, y = term)) + 
    geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
      scale_y_discrete(position = "right") +
      facet_grid(slim_term~ ., scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
      theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
      strip.text.y.left = element_text(angle=0, size = 11, face = "bold"),
      strip.text.x = element_text(size = 12, face = "bold"),
      axis.title = element_blank(),
      axis.text = element_text(size = 12), legend.title = element_text(size = 12), legend.text =
      element_text(size = 11))
fig5 <- BPplot + MFplot
ggsave("RAnalysis/Output/WGBS/DMG_heatmap_GO_terms.10x.pdf", fig5, width = 20, height = 20, units = c("in"))

```

## Make supplemental table summarizing GO enrichment
```{r}
GO.enrichment.summary.10x <- bind_rows(GO.BP, GO.MF, GO.CC)

#add gene_IDs. To get gene_IDs we will merge with the GO.terms DF.
GOgenes <- data.frame(gene_id=Gene.GO.IDs$gene, category=Gene.GO.IDs$GO.IDs) 
GOgenes$gene_id <- as.character(GOgenes$gene_id) #make gene ID a character so we can collapse our many near-identical columns

GO.enrichment.summary.10x  <- left_join(GO.enrichment.summary.10x, GOgenes, by="category" ) #join the DFs

#adjust p-values 
GO.enrichment.summary.10x$bh_adjust <-  p.adjust(GO.enrichment.summary.10x$over_represented_pvalue, method="BH")

GO.enrichment.summary.10x <- GO.enrichment.summary.10x %>% #collapse and have gene IDs for a particular term in a single row as a comma-sep list. 
  group_by(slim_term, slim_cat, category, over_represented_pvalue, under_represented_pvalue, numDEInCat, numInCat, term, ontology, bh_adjust) %>%
  summarise(genes = toString(gene_id)) %>% #rename collapsed gene_ID column "gene"
  ungroup()

head(GO.enrichment.summary.10x)
write.csv(GO.enrichment.summary.10x, 'RAnalysis/Output/WGBS/GO.enrichment.summary.table.ontology.10x.csv') 
```

